<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Modeller of Allelic Gene Expression (MAGE): Detection of
        allele-specific expression and allelic imbalance • MAGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Modeller of Allelic Gene Expression (MAGE): Detection of
        allele-specific expression and allelic imbalance">
<meta property="og:description" content="Detection of imprinting, loss of 
    imprinting, cis-eQTLS and induced allelic imbalance.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">MAGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/MAGE_expanded_tutorial.html">MAGE expanded tutorial</a>
    </li>
    <li>
      <a href="articles/MAGE_tutorial.html">MAGE tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="mage">MAGE<a class="anchor" aria-label="anchor" href="#mage"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>Welcome to MAGE’s readme!</p>
<p>MAGE is the Modeller of Allelic Gene Expression, an R package providing extensive functions for various RNAseq-based allelic analyses. This ranges from basic tasks such as (solely) RNAseq-based genotyping, to the analysis of more complex population-level phenomena such as (differential) allelic bias, allellic divergence and (loss of) imprinting analyses. More information about what these are and how to model them using MAGE can be found in the package vignette.</p>
<div class="section level2">
<h2 id="system-requirements">System requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h2>
<div class="section level3">
<h3 id="hardware-requirements">Hardware requirements<a class="anchor" aria-label="anchor" href="#hardware-requirements"></a>
</h3>
<p>MAGE requires a standard computer (server) with sufficient RAM for in-memory operations.</p>
</div>
<div class="section level3">
<h3 id="os-requirements">OS requirements<a class="anchor" aria-label="anchor" href="#os-requirements"></a>
</h3>
<p>This package is supported for Windows, Linux and MacOS. The package has been tested on the following systems:</p>
<ul>
<li>Linux: Ubuntu 18.04 (bionic), Ubuntu 16.04 (Xenial)</li>
<li>Windows: Windows 10</li>
</ul>
</div>
<div class="section level3">
<h3 id="other-software-requirements">Other software requirements<a class="anchor" aria-label="anchor" href="#other-software-requirements"></a>
</h3>
<p>MAGE is an R software package with additional C/C++ code under-the-hood, and as such relies on (versions listed are those used for MAGE’s latest test, though more recent versions should work fine as well):</p>
<ul>
<li>R (v4.0.2)</li>
<li>GNU multiple precision arithmetic library (gmp; v6.1.2)</li>
<li>Boost Multiple Precision Floating-Point Reliable Library (mpfr; v4.0.1)</li>
</ul>
<p>Furthermore, MAGE relies on other R packages for its operations, which are listed in the tutorial: <a href="https://biobix.github.io/MAGE/articles/MAGE_tutorial.html#session-info" class="external-link uri">https://biobix.github.io/MAGE/articles/MAGE_tutorial.html#session-info</a></p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>If you want to install MAGE in a conda environment including an R-installation, the following bash commands sets up the previously listed software dependencies, then opens the environment, and launches R:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> <span class="op">&lt;</span>envname<span class="op">&gt;</span> r-essentials r-base r-devtools gmp mpfr</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate <span class="op">&lt;</span>envname<span class="op">&gt;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>Setting up this environment takes about 9 minutes on a linux-based server (Ubuntu 18.04; bionic)</p>
<p>MAGE source code can be found on its associated Github page: <a href="https://github.com/Biobix/MAGE" class="external-link uri">https://github.com/Biobix/MAGE</a>. As such, MAGE can be installed using the <code>install_github</code> function from R’s <code>devtools</code> package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"BioBix/MAGE"</span><span class="op">)</span></span></code></pre></div>
<p>From a clean R installation (e.g. as in the conda environment set up previously), this installation takes about 2 minutes.</p>
<p>For a local installation, rtools will be required. In case the code chunk above throws the error <code>ERROR: loading failed for 'i386'</code>, running the following instead potentially solves the issue:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"BioBix/MAGE"</span>, INSTALL_opts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"--no-multiarch"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>MAGE contains a vignette going over its entire anaylis pipeline, which can be found here: <a href="https://biobix.github.io/MAGE/articles/MAGE_tutorial.html" class="external-link uri">https://biobix.github.io/MAGE/articles/MAGE_tutorial.html</a></p>
<p>Besides this “regular” vignette an “expanded” one is included as well (<a href="https://biobix.github.io/MAGE/articles/MAGE_expanded_tutorial.html" class="external-link uri">https://biobix.github.io/MAGE/articles/MAGE_expanded_tutorial.html</a>), containing more in-depth code using MAGE’s base functions while the regular one uses wrapper functions that handle many analyses and intermediary steps under-the hood. As such, the regular vignette is recommended for first-time users or users that just want a plug-and-play pipeline, while the expanded vignette provides more insight into MAGE’s analyses and is recommended when setting out to create your own custom/specialized analysis pipeline.</p>
<p>Running the entire vignette takes about 2.5 hours on a standard computer (local installation; Windows 10) if parallellization is not used (single-core). Enabling parallellization when running MAGE on large-scale datasets on server is advised.</p>
<p>MAGE has an associated paper pending publication. Coming soon…</p>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>For theoretical/technical questions or issues while using the MAGE package, please contact me at <a href="mailto:cedric.stroobandt@ugent.be" class="email">cedric.stroobandt@ugent.be</a>; for everything else, contact Tim De Meyer at <a href="mailto:tim.demeyer@ugent.be" class="email">tim.demeyer@ugent.be</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/AGPL-3" class="external-link">AGPL-3</a>.0 license</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MAGE</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Cedric Stroobandt <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
