<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Cedric Stroobandt" />

<meta name="date" content="2022-08-28" />

<title>MAGE tutorial</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">MAGE tutorial</h1>
<h4 class="author">Cedric Stroobandt</h4>
<h4 class="date">2022-08-28</h4>



<div id="mage-modeller-of-allelic-gene-expression" class="section level1">
<h1>MAGE: Modeller of Allelic Gene Expression</h1>
<p>The MAGE package provides extensive functions for RNAseq-based
allelic analyses, ranging from basic tasks such as genotyping - albeit
allele-specific-expression (ASE) aware and on RNAseq data only - to the
analysis of more complex ASE-related phenomena such as allelic bias (AB)
detection, differential allelic divergence (AD) detection in case
(tumor) populations compared to controls, and detection of imprinted
loci with subsequent detection of loss of this imprinting (LOI) in
cases. More information on these phenomena is provided throughout this
vignette. As MAGE relies on population-level models, large numbers of
samples are extremely beneficial.</p>
</div>
<div id="a-toy-dataset" class="section level1">
<h1>A toy dataset</h1>
<p>MAGE’s main input consists of per-locus per-sample nucleotide counts
(A/T/C/G). These are relatively straightforward to extract from BAM/SAM
files via custom scripting or established pipelines such as GATK’s
CollectAllelicCounts functionality (<a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037594071-CollectAllelicCounts" class="uri">https://gatk.broadinstitute.org/hc/en-us/articles/360037594071-CollectAllelicCounts</a>),
though the latter already filters the data down to one reference- and
one variant-allele count.</p>
<p>MAGE includes a small example of such data, including both a control
(e.g. healthy tissue) and case (e.g. tumor tissue) dataset:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;MAGE&quot;</span>, <span class="at">package =</span> <span class="st">&quot;MAGE&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(ControlCounts))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">locus_id</th>
<th align="left">reference alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">sample_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">397</td>
<td align="right">0</td>
<td align="left">sample1</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">398</td>
<td align="right">0</td>
<td align="left">sample2</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">184</td>
<td align="right">0</td>
<td align="right">217</td>
<td align="right">0</td>
<td align="left">sample3</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">111</td>
<td align="right">0</td>
<td align="right">230</td>
<td align="right">0</td>
<td align="left">sample4</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">437</td>
<td align="right">0</td>
<td align="right">577</td>
<td align="right">0</td>
<td align="left">sample5</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">280</td>
<td align="right">0</td>
<td align="left">sample6</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(CaseCounts))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">locus_id</th>
<th align="left">reference alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">sample_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">582</td>
<td align="right">0</td>
<td align="left">sample1</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">229</td>
<td align="right">5</td>
<td align="right">309</td>
<td align="right">0</td>
<td align="left">sample2</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">847</td>
<td align="right">0</td>
<td align="left">sample3</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">134</td>
<td align="right">0</td>
<td align="left">sample4</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">8</td>
<td align="right">4</td>
<td align="right">4751</td>
<td align="right">5</td>
<td align="left">sample5</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">722</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">4</td>
<td align="left">sample6</td>
</tr>
</tbody>
</table>
<p>These toy datasets contain nucleotide counts for 269 loci, every
locus being covered by a varying number of samples (e.g. 127 samples
covering loc1 in the control dataset):</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(ControlCounts<span class="sc">$</span>locus_id))</span></code></pre></div>
<p>[1] 269</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(ControlCounts<span class="sc">$</span>locus_id <span class="sc">==</span> <span class="st">&quot;loc1&quot;</span>)</span></code></pre></div>
<p>[1] 127</p>
<p>Only a control dataset is required for RNAseq-based genotyping and
allelic bias detection, but differential analyses studying e.g. the
effect of disease on allelic expression (differential allelic
divergence, loss of imprinting) require a comparison of cases against
controls.</p>
</div>
<div id="input-pre-processing" class="section level1">
<h1>Input pre-processing</h1>
<p>MAGE expects its input data to be formatted as <em>lists of
dataframes, every dataframe containing the per-sample nucleotide counts
of one particular locus</em>, and has some requirements about naming as
well. The first function we’ll be using is
<code>standard_alleles()</code> which expects the columns
<code>ref_alleles</code>, <code>A</code>, <code>T</code>, <code>C</code>
and <code>G</code>. As such, the <code>reference alleles</code> column
in the toy dataset should be renamed to <code>ref_alleles</code>; each
function’s input requirements are listed on their respective help
pages.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MAGE expects lists:</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>controlList <span class="ot">&lt;-</span> <span class="fu">list</span>() </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>caseList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">unique</span>(ControlCounts<span class="sc">$</span>locus_id)){              <span class="co"># For every locus...</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  interDF <span class="ot">&lt;-</span> ControlCounts[ControlCounts<span class="sc">$</span>locus_id<span class="sc">==</span>n,] <span class="co"># extract per-sample nucleotide counts</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(interDF)[<span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">&quot;ref_alleles&quot;</span>                <span class="co"># re-name the reference_allele column </span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  controlList[[n]] <span class="ot">&lt;-</span> interDF                          <span class="co"># put it into the list</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">unique</span>(CaseCounts<span class="sc">$</span>locus_id)){</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  interDF <span class="ot">&lt;-</span> CaseCounts[CaseCounts<span class="sc">$</span>locus_id<span class="sc">==</span>n,]</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(interDF)[<span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">&quot;ref_alleles&quot;</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  caseList[[n]] <span class="ot">&lt;-</span> interDF</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="determine-one-reference--and-one-variant-allele" class="section level1">
<h1>Determine one reference- and one variant allele</h1>
<p>MAGE employs beta-binomial modelling of allele counts, which implies
only two alleles are possible (which is often the case in human
populations anyway). Accounting for more alleles would require a
different approach, possibilities being multiple beta-binomial models
for all alleles against one set reference or even all pairwise
combinations, or (beta-)multinomial models; the latter is, however, not
supported by MAGE at the moment.</p>
<p>So as it stands, we need to pick one reference and one variant allele
and retain only their count data. This can be done by hand by a
researcher based on biological knowledge or through some automated
procedure, e.g. retaining only the most common alleles. MAGE’s
<code>standard_alleles()</code> function combines both approaches by
taking a look at the standard alleles for every locus according to dbSNP
which are supplied by the <code>ref_alleles</code> column, then picking
the two most common alleles from among these suggested standard alleles
only. Should one wish to solely rely on RNAseq count files without
external annotation, it’s possible to supply
<code>standard_alleles()</code> with <code>A/T/C/G</code> strings in all
of its <code>ref_alleles</code> columns, which will result in the most
abundant alleles being picked.</p>
<p><code>standard_alleles()</code> updates each dataframe in our lists,
retaining only two alleles in the <code>ref_alleles</code> column and
including additional columns specifying which allele is chosen as
reference and which as variant allele, alongside their respective counts
(the reference will be the most prevalent of the two in terms of total
count in all RNAseq data combined):</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">names</span>(controlList)){</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  controlList[[n]] <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">standard_alleles</span>(controlList[[n]])</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(controlList[[<span class="st">&quot;loc27&quot;</span>]]))</span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="6%" />
<col width="11%" />
<col width="15%" />
<col width="3%" />
<col width="5%" />
<col width="3%" />
<col width="3%" />
<col width="12%" />
<col width="5%" />
<col width="5%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">locus_id</th>
<th align="left">ref_alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">sample_id</th>
<th align="left">ref</th>
<th align="left">var</th>
<th align="right">ref_count</th>
<th align="right">var_count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1043</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">67</td>
<td align="right">43</td>
<td align="right">0</td>
<td align="left">sample1</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">67</td>
<td align="right">43</td>
</tr>
<tr class="even">
<td align="left">1067</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">2</td>
<td align="right">101</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="left">sample2</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">101</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">1091</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">80</td>
<td align="right">0</td>
<td align="left">sample3</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0</td>
<td align="right">80</td>
</tr>
<tr class="even">
<td align="left">1115</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">90</td>
<td align="right">0</td>
<td align="left">sample4</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0</td>
<td align="right">90</td>
</tr>
<tr class="odd">
<td align="left">1139</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">40</td>
<td align="right">63</td>
<td align="right">0</td>
<td align="left">sample5</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">40</td>
<td align="right">63</td>
</tr>
<tr class="even">
<td align="left">1163</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">94</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">sample6</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">94</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>We could run <code>standard_alleles()</code> on the case data as
well, but to make comparisons between the two datasets possible we
enforce these to be the same for corresponding loci in the control- and
case-datasets (this makes sense from a biological perspective as
well):</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">names</span>(caseList)){</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  interDF <span class="ot">&lt;-</span> caseList[[n]]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  interDF<span class="sc">$</span>ref_alleles <span class="ot">&lt;-</span> controlList[[n]]<span class="sc">$</span>ref_alleles[<span class="dv">1</span>]</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  interDF<span class="sc">$</span>ref <span class="ot">&lt;-</span> controlList[[n]]<span class="sc">$</span>ref[<span class="dv">1</span>]</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  interDF<span class="sc">$</span>var <span class="ot">&lt;-</span> controlList[[n]]<span class="sc">$</span>var[<span class="dv">1</span>]</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  interDF<span class="sc">$</span>ref_count <span class="ot">&lt;-</span> interDF[,<span class="fu">which</span>(<span class="fu">colnames</span>(interDF)<span class="sc">==</span>interDF<span class="sc">$</span>ref[<span class="dv">1</span>])]</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  interDF<span class="sc">$</span>var_count <span class="ot">&lt;-</span> interDF[,<span class="fu">which</span>(<span class="fu">colnames</span>(interDF)<span class="sc">==</span>interDF<span class="sc">$</span>var[<span class="dv">1</span>])]</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  caseList[[n]] <span class="ot">&lt;-</span> interDF</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="prior-filtering" class="section level1">
<h1>Prior filtering</h1>
<p>Before doing any analyses, it might be beneficial (in terms of
runtime) to filter out low-quality loci. MAGE’s
<code>prior_filter()</code> offers this functionality, leaving out loci
based on a required minimal median coverage across all samples, a
minimal number of samples, or a crudely estimated minimal minor allele
frequency (just the % abundance of the variant allele over all RNAseq
counts). After all, if the intent of a researcher is to study ASE effect
which only occur in heterozygotes, then loci with an extremely low minor
allele frequency will have very little of those to base analyses on.
Also, if its argument is set to TRUE, it checks whether the total
nucleotide counts truly suggest there’s only two occuring alleles or
whether there’s heuristic evidence to the contrary, in which case the
locus is removed (details are found on this function’s help page).</p>
<p>For now, we’ll restrict ourselves to just a minimal number of samples
filter of 20; the first analysis is genotyping after all, and
researchers might be interested in getting the most likely genotype for
every sample regardless of median coverage or minor allele frequency
across the entire locus (as MAGE is a population-level modeller a
minimal number of samples of at least 20 is always advised). We apply no
actual filtering on the case dataset but rather retain its data if it’s
retained in the control set as well, as comparison against controls is
its main purpose. We do run with no actual filter setting applied
though, which reduces the function to removing samples having a
zero-count for both reference- and variant-alleles (not removing these
would give NA-results during several calculations in the pipeline). This
reduces the toy datasets from 269 to 264 loci:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">names</span>(controlList)){</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  controlList[[n]] <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">prior_filter</span>(controlList[[n]], <span class="at">min_median_cov =</span> <span class="dv">0</span>, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_nr_samples =</span> <span class="dv">20</span>, <span class="at">checkref_filter =</span> <span class="cn">TRUE</span>, <span class="at">prior_allelefreq_filter =</span> <span class="cn">FALSE</span>, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_PriorAlleleFreq =</span> <span class="dv">0</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  caseList[[n]] <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">prior_filter</span>(caseList[[n]], <span class="at">min_median_cov =</span> <span class="dv">0</span>, </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_nr_samples =</span> <span class="dv">0</span>, <span class="at">checkref_filter =</span> <span class="cn">FALSE</span>, <span class="at">prior_allelefreq_filter =</span> <span class="cn">FALSE</span>, </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_PriorAlleleFreq =</span> <span class="dv">0</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">is.null</span>(controlList[[n]])){</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    caseList[[n]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="metaparameter-estimation" class="section level1">
<h1>Metaparameter estimation</h1>
<p>Before getting to model fitting, we need to determine the
population-level metaparameters shared across all loci, those being the
inbreeding coefficient of the population and the sequencing error rate.
These values can be set based on prior knowledge about the population
and the sequencing protocol of choice, but MAGE’s can estimate them
based on observed <em>control</em> data, which is recommended in all
cases for a good model fit; This function fits a simple binomial mixture
model assuming no allelic bias (i.e. perfectly balanced expression of
the reference- and variant allele in heterozygotes) to our read count
data via expectation maximization:</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(counts_{ref}|counts_{total}) = &amp;\ P_{rr} * {\tt
pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=1-SE) \ + \\
&amp;\ P_{rv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total},
p=0.5) \ + \\
&amp;\ P_{vv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=SE)
\\
\end{aligned}
\]</span></p>
<p>With <span class="math inline">\((P_{rr}, P_{rv}, P_{vv})\)</span>
to-be-fitted genotype frequencies from which the inbreeding coefficient
can be estimated, and SE a to-be-fitted sequencing error rate estimate.
The function also returns a number of outputs that can be used as
reliability measures, including an updated minor allele frequency
estimate, median coverage and number of samples. The SE-estimate itself
shouldn’t be absurdly high either. Default cutoff values for these
metrics are suggested below. We <em>don’t</em> filter these loci out
however; a suboptimal fit by may be remedied by the later beta-binomial
fit allowing unequal heterozygous allele expression for actual
genotyping and allelic bias detection.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>SE_vec <span class="ot">&lt;-</span> <span class="fu">c</span>()   <span class="co"># Vector for saving reliable sequencing error rate estimates</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>F_vec <span class="ot">&lt;-</span> <span class="fu">c</span>()    <span class="co"># Vector for saving reliable inbreeding coefficient estimates</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>pA_filt <span class="ot">&lt;-</span> <span class="fl">0.15</span> <span class="co"># Loci with a low minor allele frequency are less reliable</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>SE_filt <span class="ot">&lt;-</span> <span class="fl">0.035</span> <span class="co"># Loci which return a high estimated sequencing error are suspicious</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>NumSamp_filt <span class="ot">&lt;-</span> <span class="dv">20</span> <span class="co"># Loci with a lot of samples are reliable for estimating inbreeding</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>MedianCov_filt <span class="ot">&lt;-</span> <span class="dv">4</span> <span class="co"># Loci with low coverage are unreliable</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">names</span>(controlList)){</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  MetaEst_res <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">AllelicMeta_est</span>(<span class="at">ref_counts =</span> controlList[[n]]<span class="sc">$</span>ref_count, </span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>                  <span class="at">var_counts =</span> controlList[[n]]<span class="sc">$</span>var_count)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># You can store each locus&#39; Sequencing Error (SE) and inbreeding coefficient (F) estimate </span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># in its dataframe, if you want:</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  controlList[[n]]<span class="sc">$</span>est_SE <span class="ot">&lt;-</span> MetaEst_res<span class="sc">$</span>SE</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  controlList[[n]]<span class="sc">$</span>est_inbr <span class="ot">&lt;-</span> MetaEst_res<span class="sc">$</span>F_inbr</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  controlList[[n]]<span class="sc">$</span>allelefreq_0 <span class="ot">&lt;-</span> MetaEst_res<span class="sc">$</span>allelefreq</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Only take a locus&#39; estimates into account if the locus is high-quality:</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span>(MetaEst_res<span class="sc">$</span>allelefreq <span class="sc">&lt;=</span> pA_filt <span class="sc">||</span> MetaEst_res<span class="sc">$</span>allelefreq <span class="sc">&gt;=</span> (<span class="dv">1</span> <span class="sc">-</span> pA_filt) </span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># allelefreq returns allele frequency of the ref-allele,</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>        <span class="co"># so just in case this is the minor allele on population level </span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>        <span class="co"># (even though it is the most expresse one across all RNAseq data),</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>        <span class="co"># we perform the filtering like this</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>      <span class="sc">||</span> MetaEst_res<span class="sc">$</span>SE <span class="sc">&gt;</span> SE_filt) <span class="sc">&amp;</span> <span class="fu">nrow</span>(controlList[[n]])<span class="sc">&gt;=</span>NumSamp_filt <span class="sc">&amp;</span> </span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>      <span class="fu">median</span>(controlList[[n]]<span class="sc">$</span>ref_count <span class="sc">+</span> controlList[[n]]<span class="sc">$</span>var_count) <span class="sc">&gt;=</span> MedianCov_filt) {</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>      SE_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(SE_vec, MetaEst_res<span class="sc">$</span>SE)</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>      F_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(F_vec, MetaEst_res<span class="sc">$</span>F_inbr)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>SEmedian <span class="ot">&lt;-</span> <span class="fu">median</span>(SE_vec)</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>Fmedian <span class="ot">&lt;-</span> <span class="fu">median</span>(F_vec)</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">c</span>(SEmedian, Fmedian))</span></code></pre></div>
<p>[1] 0.0033650786 [2] -0.0004878278</p>
<p>These metaparameter estimates seem reasonable. Note that the
sequencing error estimate may be quite a bit higher than expected from
modern sequencing machines, but this is a good thing as an extremely low
sequencing error metaparameter would make fitting of homozygous peaks
later down the pipeline way too strict, failing to accomodate for even a
single faulty read (Setting SE lower than 0.002 is not recommended). An
inbreeding coefficient close to zero is a sign of a normal population
and commonplace when working on human cancer data, but it could be more
extreme in specific experimental setups.</p>
</div>
<div id="genotyping-rnaseq-data-allelic-bias-detection" class="section level1">
<h1>Genotyping RNAseq data &amp; allelic bias detection</h1>
<p>For both genotyping and cis-eqtl detection, we can use MAGE’s
<code>EMfit_betabinom_robust</code> function, which fits a beta-binomial
mixture model using expectation maximization while allowing for unequal
allelic expression in heterozygotes:</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(counts_{ref}|counts_{total}) = &amp;\ P_{rr} * {\tt
pBetaBinom}(x=counts_{ref} \ |\ n=counts_{total}, \pi=1-SE,
\theta=\theta_{hom}) \ + \\
&amp;\ P_{rv} * {\tt pBetaBinom}(x=counts_{ref} \ |\ n=counts_{total},
\pi=\pi_{het}, \theta=\theta_{het}) \ + \\
&amp;\ P_{vv} * {\tt pBetaBinom}(x=counts_{ref} \ |\ n=counts_{total},
\pi=SE, \theta=\theta_{hom}) \\
\end{aligned}
\]</span></p>
<p>With <span class="math inline">\((P_{rr}, P_{rv}, P_{vv})\)</span>
fitted genotype frequencies, <span class="math inline">\((\theta_{hom},
\theta_{het})\)</span> fitted overdispersion parameters in homozygotes
and heterozygotes respectively, and <span class="math inline">\(\pi_{het}\)</span> the heterozygous <span class="math inline">\(\pi\)</span> parameter indicating the expected
reference allele fraction, thus allowing for unequal allelic expression
in heterozygotes. is a function included in the MAGE package for the
beta-binomial PMF, using a <span class="math inline">\(\pi,
\theta\)</span> parameterisation (probability of success i.e. reference
read, and overdispersion, respectively) instead of the classical <span class="math inline">\(\alpha, \beta\)</span> parameterisation as given
on e.g. <a href="https://en.wikipedia.org/wiki/Beta-binomial_distribution" class="uri">https://en.wikipedia.org/wiki/Beta-binomial_distribution</a>.
The relationship between them is <span class="math inline">\(\pi =
\alpha/(\alpha+\beta)\)</span> and <span class="math inline">\(\theta =
1/(\alpha+\beta)\)</span>, resulting in a <span class="math inline">\(\pi\)</span> between 0 and 1 and a <span class="math inline">\(\theta\)</span> between 0 and infinity.</p>
<p>We term unequal heterozygous allele expression capture by <span class="math inline">\(\pi_{het}\)</span> <em>Allelic Bias (AB)</em>. AB
can be caused by both biological (e.g. cis-eqtl loci) and technical
(e.g. alignment bias) mechanisms. The fit is robust in the sense that
samples with a disproportionately high impact on the heterozygous
distributional parameters (determined using parameter-MLEs leaving out
that particular sample) are not used for the final fit, though they are
retained in the output data and genotyped using the final model (while
being marked as outlier). The reason for focussing on heterozygous
distributional parameters is that these are still being fitted (the
homozygous <span class="math inline">\(\pi\)</span> parameters are
already fixed by the metaparameter SE) and are actually affected by
allele-specific expression effects (<span class="math inline">\(\pi_{het}\)</span> can indicate AB and <span class="math inline">\(\theta{het}\)</span> can indicate allelic
divergence, see later; meanwhile <span class="math inline">\(\theta_hom\)</span> is a nuisance parameter at
best). Do note, however, that the robust fit is significantly slower
than the non-robust one provided by the function (a factor times the
number of samples), and its merit for genotyping and AB-detection is
limited, so one might opt for the latter instead. A robust fit is,
however, very valuable for statistical inference on the overdispersion
parameter <span class="math inline">\(\theta\)</span> (see the next
section).</p>
<p>MAGE’s functions have many optional input parameters to customize the
fitting procedure, but throughout this tutorial and for most
applications the defaults should suffice. Feel free to read up on all
options on each function’s help page. After performing the fit and
writing results of interest into dataframes, we also perform a
chi-square test assessing Hardy-Weinberg Equilibrium using our final
genotypes and the previously determined inbreeding coefficient
hyperparameter, to be used for filtering later. Finally, the code below
has been adapted to running MAGE in parallel across multiple cores on a
server. For this, just change the value or to the desired number of
cores in the code below (for now it’s at 1 as it will otherwise not work
on a local Windows-installation). Parallellisation using the package
requires input data as a list with length equal to and all commands to
be combined in a function, but if parallellization is not required the
code below can be written much simpler using a for-loop.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>NC <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="co"># Number of Cores</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>NS <span class="ot">&lt;-</span> <span class="fu">length</span>(controlList) </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>spl <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">cumsum</span>(<span class="fu">rep</span>(<span class="fu">floor</span>(NS<span class="sc">/</span>NC),NC)<span class="sc">+</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>,NS<span class="sc">-</span><span class="fu">floor</span>(NS<span class="sc">/</span>NC)<span class="sc">*</span>NC),</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>         <span class="fu">rep</span>(<span class="dv">0</span>,NC<span class="sc">-</span>NS<span class="sc">+</span><span class="fu">floor</span>(NS<span class="sc">/</span>NC)<span class="sc">*</span>NC)))) <span class="co"># Helps in splitting input data</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>ParCTRL <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">&quot;list&quot;</span>, <span class="at">length=</span>NC)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>NC){ <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  ParCTRL[[i]] <span class="ot">&lt;-</span> controlList[(spl[i]<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span>(spl[i<span class="sc">+</span><span class="dv">1</span>])]</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>BetaBinomGenotyping <span class="ot">&lt;-</span> <span class="cf">function</span>(data){</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  positions <span class="ot">&lt;-</span> <span class="fu">names</span>(data)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  results <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (z <span class="cf">in</span> positions) {</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    MAGEres <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">EMfit_betabinom_robust</span>(<span class="at">data_counts =</span> data[[z]], </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">SE =</span> SEmedian, <span class="at">inbr =</span> Fmedian)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>    data[[z]] <span class="ot">&lt;-</span> MAGEres<span class="sc">$</span>data_hash</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># median_AB also calculates a robust median AB,</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># besides the AB as determined during EMfit_betabinom_robust&#39;s fitting procedure.</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># This can be used as additional filter when detecting significant AB:</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    med_AB <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">median_AB</span>(data[[z]]<span class="sc">$</span>ref_count, data[[z]]<span class="sc">$</span>var_count, </span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>                               data[[z]]<span class="sc">$</span>allelefreq[<span class="dv">1</span>], Fmedian) </span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>    res_loc <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">&quot;position&quot;</span> <span class="ot">=</span> z, <span class="st">&quot;probshift&quot;</span> <span class="ot">=</span> <span class="fu">as.numeric</span>(MAGEres<span class="sc">$</span>AB), </span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;LRT&quot;</span> <span class="ot">=</span> <span class="fu">as.numeric</span>(MAGEres<span class="sc">$</span>AB_lrt), <span class="st">&quot;p&quot;</span> <span class="ot">=</span> <span class="fu">as.numeric</span>(MAGEres<span class="sc">$</span>AB_p),</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;quality&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>quality, <span class="st">&quot;allele.frequency&quot;</span> <span class="ot">=</span> data[[z]]<span class="sc">$</span>allelefreq[<span class="dv">1</span>], </span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;reference&quot;</span> <span class="ot">=</span> data[[z]]<span class="sc">$</span>ref[<span class="dv">1</span>], <span class="st">&quot;variant&quot;</span> <span class="ot">=</span> data[[z]]<span class="sc">$</span>var[<span class="dv">1</span>], </span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;est_SE&quot;</span> <span class="ot">=</span> data[[z]]<span class="sc">$</span>est_SE[<span class="dv">1</span>], <span class="st">&quot;coverage&quot;</span> <span class="ot">=</span> data[[z]]<span class="sc">$</span>coverage[<span class="dv">1</span>],</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;nr_samples&quot;</span> <span class="ot">=</span> <span class="fu">nrow</span>(data[[z]]), <span class="st">&quot;median_AB&quot;</span> <span class="ot">=</span> med_AB, </span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;rho_rr&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>rho_rr, <span class="st">&quot;rho_rv&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>rho_rv, <span class="st">&quot;rho_vv&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>rho_vv, </span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;theta_hom&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>theta_hom, <span class="st">&quot;theta_het&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>theta_het,</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;theta_hom_NoShift&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>theta_hom_NoShift, </span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;theta_het_NoShift&quot;</span> <span class="ot">=</span> MAGEres<span class="sc">$</span>theta_het_NoShift, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>    results <span class="ot">&lt;-</span> <span class="fu">rbind</span>(results, res_loc) <span class="co"># results; one position per line</span></span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>  results <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">HWE_chisquared</span>(<span class="at">data =</span> data, Fmedian, <span class="at">results =</span> results)</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>  results<span class="sc">$</span>Chi2PVAL[<span class="fu">is.na</span>(results<span class="sc">$</span>Chi2PVAL)] <span class="ot">&lt;-</span> </span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>  results<span class="sc">$</span>Chi2STAT[<span class="fu">is.na</span>(results<span class="sc">$</span>Chi2STAT)] <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(data, results))</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>cl <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">makeCluster</span>(<span class="fu">getOption</span>(<span class="st">&quot;cl.cores&quot;</span>, NC))</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>parallel<span class="sc">::</span><span class="fu">clusterExport</span>(cl, <span class="fu">c</span>(<span class="st">&quot;Fmedian&quot;</span>, <span class="st">&quot;SEmedian&quot;</span>))</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a>GenoFinData <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">parLapply</span>(cl, <span class="at">X =</span> ParCTRL, <span class="at">fun =</span> BetaBinomGenotyping)</span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a>parallel<span class="sc">::</span><span class="fu">stopCluster</span>(cl)</span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a>ParCTRL <span class="ot">&lt;-</span> <span class="fu">lapply</span>(GenoFinData, <span class="st">`</span><span class="at">[[</span><span class="st">`</span>, <span class="dv">1</span>)</span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a>controlList <span class="ot">&lt;-</span> <span class="fu">do.call</span>(c, <span class="fu">lapply</span>(GenoFinData, <span class="st">`</span><span class="at">[[</span><span class="st">`</span>, <span class="dv">1</span>))</span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a>Geno_AB_res <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">&quot;rbind&quot;</span>, <span class="fu">lapply</span>(GenoFinData, <span class="st">`</span><span class="at">[[</span><span class="st">`</span>, <span class="dv">2</span>))</span></code></pre></div>
<p>Calling upon MAGE’s function can results in a memory limit warning
which can usually be ignored; most of the time it signifies that a
certain locus’ beta-binomial fit was so close to a regular binomial fit
that the overdispersion parameter is extremely close to zero, leading to
precision errors in beta-binomial calculations which will trigger MAGE
to use regular binomial functions instead.</p>
<p>The genotyping results can now be retrieved from the
<code>controlList</code> object, being included as the “genotypeN”
column in each locus’ dataframe:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(controlList[[<span class="st">&quot;loc27&quot;</span>]][,<span class="fu">c</span>(<span class="st">&quot;locus_id&quot;</span>, <span class="st">&quot;sample_id&quot;</span>, <span class="st">&quot;ref&quot;</span>, </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                                            <span class="st">&quot;var&quot;</span>, <span class="st">&quot;genotypeN&quot;</span>)]))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">locus_id</th>
<th align="left">sample_id</th>
<th align="left">ref</th>
<th align="left">var</th>
<th align="left">genotypeN</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1043</td>
<td align="left">loc27</td>
<td align="left">sample1</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rv</td>
</tr>
<tr class="even">
<td align="left">1067</td>
<td align="left">loc27</td>
<td align="left">sample2</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rr</td>
</tr>
<tr class="odd">
<td align="left">1091</td>
<td align="left">loc27</td>
<td align="left">sample3</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">vv</td>
</tr>
<tr class="even">
<td align="left">1115</td>
<td align="left">loc27</td>
<td align="left">sample4</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">vv</td>
</tr>
<tr class="odd">
<td align="left">1139</td>
<td align="left">loc27</td>
<td align="left">sample5</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rv</td>
</tr>
<tr class="even">
<td align="left">1163</td>
<td align="left">loc27</td>
<td align="left">sample6</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rr</td>
</tr>
</tbody>
</table>
<p>The <code>Geno_AB_res</code> dataframe contains the results of
AB-detection, i.e. testing whether the reference allele expression ratio
differs from a “perfectly balanced” ratio of 0.5 (i.e. is there a
preference to express one of both alleles?). The “probshift” column
contains the actual fitted allelic ratio while the “p” column contains
the p-value of a likelihood-ratio-test assessing significant deviance
from 0.5; both can be filtered on to retain only significantly
allelically biased loci with a big enough effect size.</p>
<p>Other columns contain additional filter criteria, like</p>
<ul>
<li>“median_AB”: a median (robust) reference allele ratio estimate in
case you want to play it extra safe in retaining significantly biased
loci</li>
<li>“quality”: will contain “!” if a locus contained no heterozygous
samples to detect AB on</li>
<li>“allele.frequency”: referring to the reference allele frequency
across genotypes; you may not want to trust loci with extreme allele
frequencies as these probably contain very little heterozygous
individuals</li>
<li>“coverage”: contains the median coverage of all samples covering the
locus)</li>
<li>“nr_samples”: number of samples covering the locus</li>
</ul>
<p>The code below first retains loci showcasing statistically
significant AB (controlling the false discovery rate at 0.05) which are
of good quality. Then, we also construct a more high-fidelity set of
loci, retaining only those of good quality with a fitted allelic ratio
greater than 0.6 or smaller than 0.4, yet not more extreme than 0.9 or
0.1 because those could be the product of poorly fitted loci. The median
ratio should also be larger than 0.6 or smaller than 0.4, allele
frequency should not be more extreme then 0.9 or 0.1, median coverage
should be over 10 and number of samples should be over 80. These are
reasonable filter criteria, though their specific values are dependent
on each end user’s desires en the specific dataset used (e.g. requiring
over 80 samples is perhaps too stringent for smaller datasets and too
liberal for larger ones).</p>
<p>In BOTH CASES, we <em>only retain loci that are in Hardy-Weinberg
Equilibrium (HWE)</em> according to the chi square test performed
earlier. Since we retain only those loci for which we can’t find
sufficient evidence against the null hypothesis of HWE, this is a
statistically weak conclusion, yet in scientific research a cut-off of
the p-value of this test of 0.001 is often used to filter out loci with
which something seems wrong (see e.g. , , ); such extreme deviations
from HWE should not occur in a normal population.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Statistical evidence for significant AB at the 5% FDR level:</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(Geno_AB_res<span class="sc">$</span>position[<span class="fu">p.adjust</span>(Geno_AB_res<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>) <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>      Geno_AB_res<span class="sc">$</span>quality <span class="sc">!=</span> <span class="st">&quot;!&quot;</span> <span class="sc">&amp;</span> Geno_AB_res<span class="sc">$</span>Chi2PVAL <span class="sc">&gt;=</span> <span class="fl">0.001</span>], <span class="at">collapse =</span> <span class="st">&quot;, &quot;</span>))</span></code></pre></div>
<p>[1] “loc0, loc5, loc8, loc9, loc11, loc12, loc14, loc16, loc18,
loc20, loc24, loc25, loc26, loc27, loc28, loc29, loc30, loc31, loc34,
loc35, loc39, loc41, loc44, loc45, loc46, loc49, loc54, loc55, loc56,
loc62, loc63, loc67, loc71, loc72, loc73, loc74, loc75, loc78, loc80,
loc86, loc90, loc92, loc94, loc95, loc97, loc99, loc101, loc103, loc105,
loc111, loc114, loc119, loc120, loc121, loc123, loc124, loc126, loc131,
loc132, loc145, loc147, loc149, loc151, loc153, loc154, loc155, loc158,
loc161, loc164, loc171, loc174, loc176, loc179, loc182, loc184, loc185,
loc186, loc187, loc189, loc194, loc198, loc199, loc204, loc206, loc218,
loc220, loc221, loc225, loc232, loc236, loc238, loc240, loc241, loc242,
loc244, loc245, loc247, loc248, loc254, loc259, loc260, loc261, loc262,
loc263, loc264, loc265, loc266, loc267”</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Statistical evidence for significant AB at the 5% FDR level,</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># only retaining reliable (high-quality) loci with a large enough effect size:</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(Geno_AB_res<span class="sc">$</span>position[<span class="fu">p.adjust</span>(Geno_AB_res<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>) <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> </span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  Geno_AB_res<span class="sc">$</span>quality <span class="sc">!=</span> <span class="st">&quot;!&quot;</span> <span class="sc">&amp;</span> (Geno_AB_res<span class="sc">$</span>probshift <span class="sc">&gt;</span> <span class="fl">0.6</span> <span class="sc">|</span> Geno_AB_res<span class="sc">$</span>probshift <span class="sc">&lt;</span> <span class="fl">0.4</span>) </span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&amp;</span> Geno_AB_res<span class="sc">$</span>probshift <span class="sc">&lt;</span> <span class="fl">0.9</span> <span class="sc">&amp;</span>  Geno_AB_res<span class="sc">$</span>probshift <span class="sc">&gt;</span> <span class="fl">0.1</span> <span class="sc">&amp;</span> (Geno_AB_res<span class="sc">$</span>median_AB <span class="sc">&gt;</span> <span class="fl">0.6</span> <span class="sc">|</span> </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  Geno_AB_res<span class="sc">$</span>median_AB <span class="sc">&lt;</span> <span class="fl">0.4</span>) <span class="sc">&amp;</span> Geno_AB_res<span class="sc">$</span>allele.frequency <span class="sc">&lt;</span> <span class="fl">0.9</span> <span class="sc">&amp;</span> </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  Geno_AB_res<span class="sc">$</span>allele.frequency <span class="sc">&gt;</span> <span class="fl">0.2</span> <span class="sc">&amp;</span> Geno_AB_res<span class="sc">$</span>coverage <span class="sc">&gt;</span> <span class="dv">10</span> <span class="sc">&amp;</span> </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  Geno_AB_res<span class="sc">$</span>nr_samples <span class="sc">&gt;</span> <span class="dv">80</span> <span class="sc">&amp;</span> Geno_AB_res<span class="sc">$</span>Chi2PVAL <span class="sc">&gt;=</span> <span class="fl">0.001</span>], <span class="at">collapse =</span> <span class="st">&quot;, &quot;</span>))</span></code></pre></div>
<p>[1] “loc0, loc9, loc11, loc16, loc24, loc41, loc62, loc67, loc86,
loc92, loc94, loc111, loc119, loc124, loc131, loc149, loc155, loc179,
loc186, loc245, loc254, loc260, loc263, loc264”</p>
<p>MAGE contains a function to plot histograms of the reference allele
fraction across samples for a specific locus, while also including PMF
plots of both the unshifted (assuming no AB) and shifted models. The
theta-parameters for the unshifted model are included in the previous
output as they are calculated anyway to test for a significant shift
using a likelihood ratio test. This plot can illustrate just how much
better the shifted model fits to the data, as illustrated below.</p>
<p>Keep in mind, however, that for a beta-binomial PMF, its shape
(variance) depends on the coverage it is calculated on even when
normalized from the reference allele count to a 0-to-1 reference allele
fraction, such as on the x-axis in the plot below. As such, we have to
pick a coverage on which to base the PMF plots ( argument). The median
coverage across samples is a good pick for this, but the PMF plots may
not fit certain data points of the histogram bars all that well if these
correspond to samples with outlying coverage, even if the actual fitted
beta-binomial model fits them extremely well. This is because the fitted
model allows for a flexible per-sample coverage parameter; for our PMF
plot we have to pick just one.</p>
<p>To accomodate for this, one might opt to set the argument to TRUE,
which will re-calculate observed per-sample reference allele fraction
that are to be plotted in the histogram using the MAGE’s beta-binomial
quantile function given the observed per-sample data probability as
input, but while using total counts (i.e. we transform observed
reference allele fractions to the fractions we “would have observed” -
i.e. with the same cumulative distribution function value - given
observed reads, assuming the fitted beta-binomial model is correct).
It’s up to the end user to use this argument, though for a comfortable
interpretation of the plots it’s recommended if the per-sample coverage
is very variable.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>PlotData <span class="ot">&lt;-</span> controlList[[<span class="st">&quot;loc11&quot;</span>]]</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>PlotData_eqtl <span class="ot">&lt;-</span> Geno_AB_res[Geno_AB_res<span class="sc">$</span>position<span class="sc">==</span><span class="st">&quot;loc11&quot;</span>,]</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>loc_plot <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">MAGE_EMfitplot</span>(<span class="at">ref_counts=</span>PlotData<span class="sc">$</span>ref_count, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">var_counts=</span>PlotData<span class="sc">$</span>var_count, <span class="at">pr=</span>PlotData_eqtl<span class="sc">$</span>rho_rr, <span class="at">prv=</span>PlotData_eqtl<span class="sc">$</span>rho_rv,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pv=</span>PlotData_eqtl<span class="sc">$</span>rho_vv, <span class="at">theta_hom=</span>PlotData_eqtl<span class="sc">$</span>theta_hom, </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_het=</span>PlotData_eqtl<span class="sc">$</span>theta_het, <span class="at">pr_NoShift=</span><span class="fu">mean</span>(PlotData<span class="sc">$</span>prr_H0), </span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">prv_NoShift=</span><span class="fu">mean</span>(PlotData<span class="sc">$</span>prv_H0), <span class="at">pv_NoShift=</span><span class="fu">mean</span>(PlotData<span class="sc">$</span>pvv_H0), </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_hom_NoShift=</span>PlotData_eqtl<span class="sc">$</span>theta_hom_NoShift, </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_het_NoShift=</span>PlotData_eqtl<span class="sc">$</span>theta_het_NoShift, </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">probshift=</span><span class="fu">as.numeric</span>(PlotData_eqtl<span class="sc">$</span>probshift), <span class="at">SE=</span>SEmedian, </span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">ScaleCount =</span> PlotData_eqtl<span class="sc">$</span>coverage, <span class="at">ScaleHist =</span> <span class="cn">TRUE</span>, <span class="at">plot_NoShift =</span> <span class="cn">TRUE</span>, <span class="at">nbins =</span> <span class="dv">30</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>loc_plot</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAzAAAAHgCAMAAABNd3gvAAABAlBMVEUAAAAAADoAAGYAOmYAOpAAZrYAoIcqr7IxeW4zMzM6AAA6ADo6AGY6OpA6ZmY6kJA6kNtNTU1NTW5NTY5NbqtNjshNu9VUhZFYWsFZWVlmAABmADpmAGZmOgBmZgBmkJBmtv9uTU1uTW5uTY5ubqtuq+SByb6M1MmOTU2OTW6OTY6OyP+QOgCQZgCQkDqQkGaQtpCQ27aQ2/+ZU0mgIPCk1eGrbk2rbm6rbo6ryKur5P+2ZgC22/+2/7a2/9u2///Ijk3I///bkDrb/7bb///kq27k///mSzXpo5nr6+vy8vL/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///8IJMgIAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAazklEQVR4nO2dC3vbVnZFaddyE8dtrcSxx560VTJjjzNSO1OLedSa1rYaW5nKmsqW8P//SnFBgAQkkTwXOADOuVj7cyI+wK3Nw7t0XyQ4yxBCYs3GDoCQJwEMQhECGIQiBDAIRQhgEIoQwCAUoUhg/rpdkmPi1Y+rL1vCjmgLMP5sCQswYqVW/xaaz1Vs7t69cgMvGMAkaQswY9oCjD9bwgKMWKnV344rYUW2AOPPlrAAI1Zq9bfjSliRLcD4syUswIiVWv3tuBJWZAsw7mxZVh7TFmDc2QLMmLYA48+WsAAjVmr1t+NKWJEtwPizJSzAiJVa/e24ElZkCzD+bAkLMGKlVn87roQV2QKMO1uWlce07Q2YR4/EIXuWKIM0bJwAxlVlx+1hxMRoA9Iiq6+XlbAA06dEGYRZI0Ub9BUWYIRZfb2shAWYPiXKIMwaKdqgr7AAI8zq62UlLMD0KVEGYdZI0QZ9hQUYYVZXLyvLymPaVq0KYPoQwLiqLMBIs/p6WQkLMH1KlEGYNVK0QV9hfQFzcTArtHf2xcvsb39Z3Fj9BJi+XQkrsjUFzIOlf2Cm/rOLRBmEWSNFG/QVFmCEWX29rIQFmKAlMGdf/PnebBaunIWfRzv5peOdsy/+NJvd+WUxdLv9GmB82LoK6xaYl/Ue5jSn4+Jg7+ze7dcXBztZ+C87DuRMFBiWlce0NQVMMed/cBWYj0/3srP7r8/u7RXXAz/FbQDTSQDTytYUMDf3MPlwLPwrLuagHM9KrqYKDGHHtPUAzNn9/z7YWwETMRoDmNFtXYVNBJiLg3+9/zpbDMnuvz69FblwJsogzBop2qCvsGMCM593Aaaaoyx+Hs92wopZNenPu5gYakQZhFkjRRv0FdYvMNlRYCTX0YKVYr7/p1lxY1gciOllRBmEWSNFG/QV1hcw65UPxDrsYYoyCLNGijboK2x/wGxXDoya13Gxiamw6Z+E5nMVm7t3VWymKsM9zNm9Yl2MHqYU+zBj2latyjAwXSXKIMwaKdqgr7AjA/NkJUmEvqRTqFaiDfoKCzCyrM5eVsKmCMxfHwEMbdBZWIARZvX1shI2SWAe14mRRPjfmwUww7gSVmQLMMJCtRLLyq4qCzAAs14A08oWYISFaiXaoK+wAAMwY9u6CgswADO2rauwowLT2LmURACYUV0JK7IFGGGhWok26CtsgsAUH40pflwczP7xfv30ZBcHN59JRqdQrUQb9BU2XWAWl84mCQzLymPamgJmdk01YE6L812e3f+X2Sycp+y/nob7b7/++HS2OFHZ7O/+GWAiBDCtbN0A8/Gb18XZye49yE7v/LLoXkIPc/Qg3Bx+nM6mAQxhx7Q1BcymIdnH3xYfuAyMLGnJ/4Wbc5YCTlMZkhF2TFs3wIQTk996eR2YMDK79bKYzRwBjD1bV2FT27hcjMWawHxTzPvpYazaugqbEjCn4fR914EpJi/5XcxhjNq6CpsSMNlRuUpWA+bioFglu1WcTnYyq2SEHdPWDzCtpVOoVmJZ2VVlAQZg1gtgWtkCjLBQrUQb9BUWYABmbFtXYQEGYMa2dRU2qY1LgOnXlbAiW4ARFqqVaIO+wgIMwIxt6yoswKQIDMvKY9oCjLBQrQQwrioLMEkCQ9gxbQFGWKhWog36Cjs2MI8Axo8rYUW2/QHz+DHA0AZ9hU0eGNFJl3QK1Uq0QV9hpwJMtvGkSzqFaiXaoK+w3oCpnSumUP2sMdUnx/7h38JplmJOuqRTqFZiWdlVZZMEZnGapZiTLukUqpUAxlVl3QGzaUi2+mxyVp5dKROedEmnUK1EG/QVNl1gYk66pFOoVqIN+gqbMDCZ/KRLOoVqJdqgr7DjAlPf6pdE2AxMGG0d366AiTnpkk6hWok26CtsSsBkx7PZP32z7GEiTrqkU6hWog36CpsUMK2lU6hWog36CgswKQLDsvKYtgAjLFQrAYyryuoA8243aL/4+dVbgIkRbdBXWLUe5sPD99mb/bhGCDC0wd5cjQPz6ftX2eVPrwAmVrRBX2G1gHn3dQ7Ni2JgluvzXIJ22ti47NbkEbKmTcCEDiY7/67eywhgjd7p70s6f1laiT/avsIq9TBhBrPQch4j+N0AQxvszdU2MG+eLS8BTJRYVnZVWSVgFgOx0M1c/hyzrAwwAOOrskrAFFOYYj/my+VCmeB3AwxtsDdX08C0bIQAQxvszXXSwEQptfrbcSWsyLZHYKQbl1FKrf52XAkrsu2zh5nPAcaPK2FFtgDjz5awACNWavVvIZaVx7TtF5hHAKMvgBnTFmD82RIWYMRKrf52XAkrsgUYf7aETRSY2s5l96dS5VVz8mtLWIARK7X623ElrMgWYPzZEhZgxEqt/i3EsvKYtgDjzhZgxrQFGH+2hAUYsVKrvx1XwopsAcafLWETBYaNS9qgr7AAk6QtYQFGrNTqb8eVsCJbgHFny7LymLYA484WYMa0BRh/toQFGLFSq78dV8KKbAHGny1hEwWGjUvaoK+wAJOkLWEBRqzU6m/HlbAiW4BxZ8uy8pi2AOPOFmDGtAUYf7aEBRixUqu/HVfCimwBxp8tYRMFho1L2qCvsGP3MCtiuj+VKq+ak19bwgKMWKnV344rYUW2AOPOlmXlMW0Bxp0twIxpCzD+bAkLMGKlVn87roQV2QKMP1vCpgrMauey+1Op8qo5+bUlLMCIlVr97bgSVmQLMP5sCQswYqVW/xZiWXlM25bACNQARt19wprPVWzu3lWxmaroYfzYEtZfDyP43QBDG+zNFWCkSq3+dlwJK7LtFRg2Lh25ElZkCzD+bAkLMGKlVn87roQV2QKMO1v2Yca0BRh3tgAzpi3A+LMlLMCIlVr97bgSVmQLMP5sCZsqMGxcOnIlrMgWYPzZEhZgxEqt/nZcCSuyBRh3tiwrj2kLMO5sAWZMW4DxZ0vYZIFZEtP9qVR51Zz82hIWYMRKrf52XAkrsu0XmOXOZfenUuVVc/JrS1iAESu1+ttxJazIFmD82RIWYMRKrf4txLLymLYA484WYMa0BRh/toQFGLFSq78dV8KKbAHGny1hkwWGjUs/roQV2QKMP1vCAoxYqdXfjithRbYA486WZeUxbQHGnS3AjGl7EzAfn956CTB2bQlrDJgsO5rN7vwCMEZtCWsOmNDLzGYPAMakLWENArNA5vbrbsCwcenHdaJhn9R1crL9F68B5ng228mHZusGZoIkjxfEAIwT14mGVQHm4mA22wsXTtd1MYIkADPVNjiAqy1gPj5dOxQDmPFtWVbWtNUakgGMWVuA0bRVGpI9yP/bARibtoS1BszRTkHNBmIESQBmqm1wAFdbwHx8unnGDzDj2hI2XWDYuHTjOtGwKkOy44DKx6fs9Nu0JWxfwAjWU24EJjudzTa+MwZgxrQlrDlgtkqQ5HGdmO5PrMqr5uTWlmVlTVuASd4WYDRtVYA5uxeGZNVbL9/t7u5+9Tb79GL34XuAMWBLWGPAXNmBebMf/n/543727muAMWBLWGPAlMvKpS5/ehV+fPrj2+z8d28BZnxbwhoD5uKgDkw+FNvd3c/Of/8++/R9YOfzXNl2Pa5/IEZwPEIjqLENM5+LH9eYwzS2LM+/exV6mQ8PK2CCBOg2gOn+l6ACXM3Jry1he+phTk62294ETPisZW3SH/Rmf9XDAMzItoQ1BswNerPfbg4DMPpiWVnTVh+YMBa7/Pnt5Y/PWqySAYy+AEbTVgeY49ls77j6PP+73d0vX2Xt9mEAxo3rRMOqAHN053+e7il9HgZgnLhONKzOpH8vbMXofB4GYJy4TjSsMWDYuHTjOtGwKkOy4zAkU/k8DMC4cZ1o2Dow83nbSb/a52EAxo3rRMPqALNVgiQAw7Jyb64AI5X1+g9gCzCathrA3PDWGIAxZEtYY8CU2Px2w7cqCZIAzFTb4ACuFoHJTjd8p5IgCcBMtQ0O4GoTGDYubdoSth9gTjoCs/bLYWJ6mPLUl92fWJVXzcmvLWF7AubksNOkf9M3wwqSAMxU2+AArraAEUiQBGBYVu7NFWCksl7/AWwBRtNWcUi2aStGkARgptoGB3C1BUx2GqYvOm++BBgvrhMNq9PDqH3dBcC4cZ1oWGvAsHHpxXWiYVWGZHrfD1MR0/2JVXnVnPzaErYfYObz1qtkx2HCv3cNE4AxYUtYc8BslSAJwLCs3JsrwEhlvf4D2AKMpq3akGx1XjKAMWZLWGvA6J2XDGDcuE40rAYwmqdZAhgvrhMNW1tVNgFM+YGY7k+syqvm5NeWsMaA0TwvGcB4cZ1o2Ma+ZetJv955yQDGi+tEw+oAs1WCJADDsnJvrraAaX7HJcAYswUYTVsNYJrfogww1mwJawyYbOOeJcCMbUtYY8BonvkSYLy4TjSstUk/G5deXCcaFmDStyVsL8DM562AEcz4AWZcW8KaA2bTu2IAZmRblpU1bQEmeVuA0bQFmPRtCQswYlmv/xC2hAUYsazXfwhbwvYBzElrYLaeJzYKmMWZ/Lo/sSqvmpNfW8JaAkYmQRKAmWobHMC1D2BODgEmTVvCAoxY1us/gC3Lypq2AJO8LcBo2g4EjEANYNTdEVLRFWDE6rGHeUQP48F1omENDskAxoXrRMOaA2axc9n9iVV51Zz82hIWYMSyXv8hbAnbBzBh3xJgkrQlLMCIZb3+A9iyrKxpCzDJ2wKMpi3ApG9LWIARy3r9h7AlLMCIZb3+Q9gStgdgTqwAw8alE9eJhgWY9G0J2wcwYd8SYJK0JSzAiGW9/gPYsqysaQswydsCjKYtwKRvS1iAEct6/YewJSzAiGW9/kPYEjZlYNi4dOI60bAAk74tYfWBmc8BJllbwgKMWNbrP4Aty8qatgCTvC3AaNoCTPq2hAUYsazXfwhbwvYBzCHApGpL2KSBKc4V2/2JVXnVnPzaEhZgxLJe/yFsCasPzGLfEmCStCVs4sA8AhhNsaysaQswydsCjKYtwKRvS1iAEct6/YewJSzAiGW9/kPYEjZpYIqdy+5PrMqr5uTXlrAAI5b1+g9hS1h1YMq3kgFMkraEBRixrNd/AFuWlTVtASZ5W4DRtAWY9G0JCzBiWa//ELaE7QGYQ4BJ1pawSQPDxqUP14mGtQdM8f7+7k+syqvm5NeWsOrAlPuWAJOkLWEBRizr9R/AlmVlTVuASd4WYDRtASZ9W8LaBub8293d/Sx7t7u7+9VbgBnflrCmgfn0/avs/LtX2Zt9ehgbtoQ1DcyHr/P/vdm//OlVS2DYuHThOtGw/cxh8l7m04vdYmSWZZ/n2nj0Qg1gBMcjNIIabyU7lD9uIzCXPz4rRmWrXkaALj3MVP9oD+Cq3MOsgNn+iwXAfHrxrLy0nMcIkgAMy8q9uZoG5vzb5XQfYAzYAoymrT4wJS8fHr7PLn9us6wMMC5cJxpWH5iw/xKm+/nPL5cLZYIkADPVNjiAqzowh4Z2+gHGhetEwxoEJnwgpvsTq/KqOfm1JSzAiGW9/kPYElYbmGrfEmCStCUswIhlvf4D2LKsrGkLMMnbAoymLcCkb0tYgBHLev2HsCUswIhlvf5D2BI2bWDCzmX3J1blVXPya0tYgBHLev2HsCWsMjDLt5IBTJK2hAUYsazXfwBblpU1bQEmeVuA0bQFmPRtCasOzCHAJGxLWIARy3r9h7AlbNrAsHHpwnWiYQ0CE0592f2JVXnVnPzaElYZmOW+JcAkaUtYgBHLev0HsGVZWdMWYJK3BRhNW4BJ35awACOW9foPYUtYgBHLev2HsCVs4sCwcenBdaJhASZ9W8LqArN67yXAJGlLWIARy3r9B7BlWVnTFmCStwUYTVuLwOTEdH9iVV41J7+2hNUG5hBgUrYlLMCIZb3+Q9gSNnFg5nOAse860bAAk74tYXWBWe1bAkyStoQFGLGs138AW5aVNW0BJnlbgNG0BZj0bQkLMGJZr/8QtoQFGLGs138IW8ImDgwblx5cJxoWYNK3JawqMLX3XgJMkraEBRixrNd/AFuWlTVtASZ5W4DRtB0IGIEawKi7I6SixnsvD+WPo4fxY0tY5R7mkCFZ0raETR0YpVF3kVfNya8tYQFGLOv1H8KWsKrA1PYtASZJW8ICjFjW6z+ALcvKmrYAk7wtwGjaAkz6toQFGLGs138IW8ICjFjW6z+ELWFTB4aNSweuEw0LMOnbElYTmPp7L60Ao0iM9foPYUtYgBHLev0HsGVZWdMWYJK37QLMk5Xu3n1S1/qwVw6LlPXKNt5KBjBJ2nZwfbJWAAMwqdoCDMCIZb3+Q9gCTOrAKJ7Jz3r9h7AFGIARy3r9h7AFGE1g6vuWAJOkLcAAjFjW6z+ALcvKmrYegOnyCliv/wC2NwIjrCnAXBXAtJIr25tc44G5IoABmFRtAQZgxLJe/yFstYC52xDA2AHmysYlwOi7Cmt6V6i11tGvnfXKAkwrubKNA2YDFp/VtOGwpIFpvFkZYJK0FQKzlZGGnj//w/OlhNi0DasggBHKev0HsN2+rNxs6c9ranQp64FpaD03krTWKwswreTJdgMw60npAMxNQ7aUgDkEmHg5sV1fOeHAa722AXPjVGeEEmjbAkwrObHdOlNpQUocMNexGbgE2rYA00pObDfMVNqj0gKYa9wMVwJtW5vANIfdANNW81LXZypdeWkFTLh3GzTWKwswrWTEtgJi2/UKlep61aqr61WLvnpdCkwsbRuoMVLZtWrsWwKMMVspEDddvzb8eh4PzAaAOgGzARrrLxjAtFJPtrGArFETld/8ZglMbBOvAxPMatc7A/P85gGa9RcMYFpJybYlEJvUmKx81hGYum4CJmIIdzMw16Gx/YIBTEu1tN0GSNM27nlrLoFJ1WFId0XiZeeWAhihxgUmtgdpBUxzuXgwWG5Se2AkC2hdlDgwj+b1nRg/wHQdYsUBc4UUpfXiHrVxDeHaAE0Vm8SBefwoqLrHLjDacxApMDeR0mH+PZQ2dEDL/H94fvXtm22LuaGyHWQUmB9+qCHzZK3Wu3WBbL2WQKwBpmuercB461M2ax1A1/qauzFv3xRUNupFaaj5ZmVLwJTIPLr6vEwBI1dnYNZ3KmM2eV01gVktwil0N1MAJiAzDx/vHw+YtT3IgMBsGH7JgKmWlTtqwOWEq8B8Vq6iLQGKrb4WMOGXHxoG5ocfqhINBIx4TjIIMFtRSR2Yq0/vao8TUf/uwJyUv9w2MBUyYXAWntUVgNa73VyLbUCI5yR9AXP3Bm1uRFNUWRjhq7e+9PJ7TnIF619/PbQOzHIyU+gKMOtLtvn+J2seL9bqCXXu8W5ipIHKQgBzTc0Op/7XtPFqr3+xhcCcLPRrof6A+fRi9+F7FWBqK81XVNWgWh/YBszmKkUoGpgNWKwdeDVaB8Bc002LaXVVPG3561phUV0v+Whe/7VSb8Bc/rifvftaEZjH67mJkrTpDqXOT6ipv+/aDE98Kr7wzQ5q0/Vfa+oNmE9/fJud/+6tHBid5mdVylgguXp9XfWAOf/9++zT96/yS5/n2nZ0rl6fV01j/ynsWWM3zwS1qTUJGrYQmA8PK2CCBD1MP29isfVu5VFsO31aYKVrf015wSS2YmBWPQzAjGsLMGPaioGJnsNQf2rQm6sDYC5/fBa3Skb9qUFvrg6Aid6Hof7UoDdXD8A0pPK72yi1+ttxJazIFmD82RIWYMRKrf52XAkrsgUYd7YsK49pCzDubAFmTFuA8WdLWIARK7X623ElrMgWYPzZEhZgxEqt/nZcCSuyBRh/toQFGLFSq78dV8KKbAHGnS3LymPaAow7W4AZ07YlMAJJPsZsRoTtS67CRqQFGD8ibG8CGKEI25dchQUYqQjbl1yFHRUYhBIWwCAUIYBBKEIAg1CEAAahCCkBszoXU3mpcXImY1plO/92d3c/y97t7u5+9XbLo0bSKmyZ0kVlQ9ZQWsuVzVWdoTKi0eoAs/pOjPJS80sybGmVLZwC9/y7V9mb/ZEjrVetkIuUPiobFM7KbbiyuT6UMMc0Wh1gVueTLS81TzBrS6tsH77OQju8/OnVloeMp1XYMqWPymaLv0aWK5u/8l/+5yJsTKPVAWZ1xvLyUvMU5rbUzJZfynviYmRmUauwZUo3lQ1/qy1XNqjEI6bR6gCz+k6M8lLzSzJsqZEtnDo6jMqs/i1chS1Teqls8dNyZYNKYGIa7cR7mE8vnpW3Gh1tXynkm30vlf2wnD4brWzQaD2M0zlMdv7t8sU0+rJeKeSbfSeVzd5Uf4msVjbofKw5zOo7McpLzS/JsKVVtpKX8Mfw8mebbXAVtkzpo7LlQMxyZYNKPGIareo+TPj9bvZh8rC13YIvbQ5y6pUtU7qobDW0sVzZbAFMZKNlpx+hCAEMQhECGIQiBDAIRQhgEIoQwCAUIYDpURcHs0IParcdz269bOt39sXL/F/tWuPetc5/+8u1Y1FbAUyPujgoUDm7t7e86ePTvbWHb9VGYNY6A4umAKZHlcBkRzvLmzq13o3ArHUGGE0BTI+qAxOGZ7dfn92bzXbKy9nZ/X/Pf1RXvvhzft/e4sCdrHrAwig8LB/YlcAsH/GydlDhvHAsjy6dwrUH1bE72eoXoTYCmB5VAnOaN8+Lg7ytHt/5ZdFyy8v3CpCqK3d+yY4DPzvF6Kq6PRgUo638vgUwNav6QcG5cKyOrpyKOxbHhv+qXzRmYRwLYHpUNenPG/BpaKF58w2td3n5Xv2OcKU25qpuD5f/LxCxuC//V7OqH7QAZq9xdLa8ozz2NPQ/ewzT2gtgelTRw5zdC93McbleFlpq/XLzStWuaw9YOJ3mF2+VwNQeUT+o5GJ1dOVUARM6ooojgGkrgOlRiyFZGJFV46aipdYvN6/UgSlvD/r49NbLVQ9Te0T9oCUw1dEA04cApkeVc5ijvOWe3lqNj+qXm1fqQ7LapkrR1k+rHqb2iPpBS2BqR9fvKI49LSZCANNeANOjlvswYbqdN+OyFdcvN6/U5ubV7cEg/Di7VwFTe0T9oBUw5dGVUznxX076AaaLAKZHVcvKx/ksIywA3CqbdP1y40pt9be6vdBRfvE/iml+dUT5iNpBqzlMefTS6Wi201hWBpgOAhiEIgQwCEUIYBCKEMAgFCGAQShCAINQhAAGoQgBDEIRAhiEIgQwCEXo/wGhJewEmZ7/CQAAAABJRU5ErkJggg==" width="100%" style="display: block; margin: auto;" /></p>
<p>The plotting function contains some customization options, the plot
above mainly using default colors with the PMF of the shifted fit split
up according to genotype. One can e.g. choose to plot one combined
unshifted PMF, and to not plot the unshifted fit, via the code below
(once again using default colors, though these can all be changed):</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>loc_plot <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">MAGE_EMfitplot</span>(<span class="at">ref_counts=</span>PlotData<span class="sc">$</span>ref_count, </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">var_counts=</span>PlotData<span class="sc">$</span>var_count, <span class="at">pr=</span>PlotData_eqtl<span class="sc">$</span>rho_rr, <span class="at">prv=</span>PlotData_eqtl<span class="sc">$</span>rho_rv,</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">pv=</span>PlotData_eqtl<span class="sc">$</span>rho_vv, <span class="at">theta_hom=</span>PlotData_eqtl<span class="sc">$</span>theta_hom, </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_het=</span>PlotData_eqtl<span class="sc">$</span>theta_het, <span class="at">pr_NoShift=</span><span class="fu">mean</span>(PlotData<span class="sc">$</span>prr_H0), </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">prv_NoShift=</span><span class="fu">mean</span>(PlotData<span class="sc">$</span>prv_H0), <span class="at">pv_NoShift=</span><span class="fu">mean</span>(PlotData<span class="sc">$</span>pvv_H0), </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_hom_NoShift=</span>PlotData_eqtl<span class="sc">$</span>theta_hom_NoShift,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_het_NoShift=</span>PlotData_eqtl<span class="sc">$</span>theta_het_NoShift, </span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">probshift=</span><span class="fu">as.numeric</span>(PlotData_eqtl<span class="sc">$</span>probshift), <span class="at">SE=</span>SEmedian, </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">ScaleCount =</span> PlotData_eqtl<span class="sc">$</span>coverage, <span class="at">ScaleHist =</span> <span class="cn">TRUE</span>, <span class="at">plot_NoShift =</span> <span class="cn">FALSE</span>,</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">SplitPeaks =</span> <span class="cn">FALSE</span>, <span class="at">nbins =</span> <span class="dv">30</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>loc_plot</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAzAAAAHgCAMAAABNd3gvAAAAxlBMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6ADo6AGY6kNtNTU1NTW5NTY5NbqtNjshZWVlmAABmADpmtv9uTU1uTW5uTY5ubqtuq+SOTU2OTW6OTY6OyP+QOgCQkDqQtpCQ27aQ2/+gIPCkgjGrbk2rbm6rbo6ryKur5P+2ZgC22/+2///Ijk3I///LYoTbkDrb/7bb///kq27k///r6+v00oH/swD/tmb/yI7/25D/27b/29v/3Yz/5Kv//7b//8j//9v//+T///9xPKDvAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAW0klEQVR4nO2dDVvb1hmGnZWmq9M0GW2XdGRbPtqRFcaarJBsJQT9/z81yZZsCfzxSDo6532P7+faGgPi5vFr3Ug6BjMrCCFyZqkLEOIpCENIjyAMIT2CMIT0CMIQ0iMIQ0iP9BTmP/ujbNM/01B9YSmbEIsw/rCURRg5uc1/QF69CoK5vLzzDh4whMkSizApsQjjD0tZhJGT2/ztUCkrYRHGH5ayCCMnt/nboVJWwiKMPyxlEUZObvO3Q6WshEUYd1iWlVNiEcYdFmFSYhHGH5ayCCMnt/nboVJWwiKMPyxlEUZObvO3Q6WshEUYf1jKIoyc3OZvh0pZCYsw7rAsK6fETibMvccjTHKb/4AgTErsdEeYaYzJbf52qJSVsAjjD0tZhJGT2/ztUCkrYRHGH5ayCCMnt/nboVJWwiKMPyxlEUZObvMfEJaVU2IRxh0WYVJiEcYflrIIIye3+duhUlbCIow/LGURRk5u87dDpayERRh/WMoijJzc5m+HSlkJizDusCwrp8QijDsswqTEIow/LGURRk5u87dDpayERRh/WMoijJzc5m+HSlkJizD+sJRFGDm5zd8OlbISdqAwQi4vgyNJlVevgmB4fEaFI4wbLM/DpMQijD8sZTMV5sk64+9K0zcYyS+WsggjJ7f526FSVsIijD8sZRFGTm7zt0OlrISdTpjXbWPG35WmbzCSXyxlEUZObvMfEJaVU2IRxh0WYVJiEcYflrIIIye3+duhUlbCIow/LGURRk5u87dDpayERRh/WMoijJzc5m+HSlkJizDusCwrp8QijDsswqTEIow/LGURRk5u87dDpayERRh/WMoijJzc5m+HSlkJizD+sJRFGDm5zd8OlbISFmHcYVlWTolFGHdYhEmJRRh/WMoijJzc5m+HSlkJizD+sJRFGDm5zd8OlbISFmH8YSmLMHJym78dKmUlLMK4w7KsnBKLMO6wCJMSizD+sJRFGDm5zd8OlbISdkJh2saMvytN32Akv1jKIoyc3OZvh0pZCYsw/rCURRg5uc3fDpWyEhZh3GFZVk6JRRh3WIRJiUUYf1jKIoyc3OZvh0pZCYsw/rCURRg5uc3fDpWyEhZh/GEpizBycpu/HSplJSzCuMOyrJwSizDusAiTErtfmHfzKieLf785R5j0WMqaFqbKh8fvi7MTjjA2sJS1LszNj6fF7U+nCGMDS1nrwrz7tpTm+eLErMyXZXYej5bpCCNsT4ij7BKmOsAU19+3jzKCrBxh+KY9GdX2Eaa6gllmdR0jfG2EYR+cjGpbmLNnq1sIkx7LsnJKrCDM8kSsOszc/syycnoswqTECsIsLmEWz8c8Wi2UCV8bYdgHJ6OaFmZThK+NMOyDk1ERRk1u87dDpayERRh/WMoijJzc5m+HSlkJizD+sJRFGDm5zX9AWFZOiZ1SmJYx4+9K0zcYyS0WYVJiEcYflrIIIye3+duhUlbCIow/LGURRk5u87dDpayERRh/WMoijJzc5m+HSlkJizDusCwrp8QijDsswqTEIow/LGURRk5u87dDpayERRh/WMoijJzc5m+HSlkJizD+sJRFGDm5zd8OlbISFmHcYVlWTolFGHdYhEmJRRh/WMoijJzc5m+HSlkJizD+sJRFGDm5zd8OlbISFmH8YSmLMHJym78dKmUlLMK4w7KsnBKLMO6wCJMSizD+sJRFGDm5zd8OlbISFmH8YSmLMHJym78dKmUl7KTCrI0Zf1eavsFIfrGURRg5uc3fDpWyEhZh3GFZVk6JRRh3WIRJiUUYf1jKIoyc3OZvh0pZCYsw/rCURRg5uc3fDpWyEhZh/GEpizBycpu/HSplJSzCuMOyrJwSO1AYIR1hgtMPOK9eBcFcXgbBHGo4wvjBUtbfEUb42gjDPjgZFWHU5DZ/O1TKSliE8YelLMLIyW3+dqiUlbAI4w9LWYSRk9v8B4TnYVJiEcYdFmFSYhHGH5ayCCMnt/nboVJWwiKMPyxlEUZObvO3Q6WshEUYf1jKIoyc3OZvh0pZCYsw7rAsK6fEIow7LMKkxCKMPyxlsxVmZcz4u9L0DUbyi6UswsjJbf52qJSVsAjjD0tZhJGT2/ztUCkrYRHGH5ayCCMnt/kPCMvKKbEI4w6LMCmxCOMPS1mEkZPb/O1QKSthEcYflrIIIye3+duhUlbCIow/LGURRk5u87dDpayERRh3WJaVU2IRxh0WYVJiNwnz6emDlwhjF0tZY8IUxdvZ7ItfEcYolrLmhKmOMrPZ1whjEktZg8IslfnDLwhjD0vZgNgn7dy7rtvwhbcIczGbHZWnZttOzIQmCHOo+2AEqjFhPr+YzY6rG1fbDjFCE4Q51H0wAtWWMJ+ebj0VQ5j0WJaVQ2JDnZIhjFkswoTEBjol+7r8/xHC2MRS1powb48W1uwwRmiCMIe6D0ag2hLm09PdV/wIkxZLWYSRY33+MbCUNSZMcVGp8ukpz/TbxFLWmjDF1Wy28ydjECYllrLmhNkbocnrtjHj71jTNxjJLZZl5ZBYhMkeizAhsUGE+fiwOiVrfvTy3Xw+/+a8uHk+f/weYQxgKWtMmDvPwJydVP+9fXNSvPsWYQxgKWtMmHpZuc7tT6fVPzd/OS+ufzhHmPRYyhoT5vOLtjDlqdh8flJc//l9cfNj5c6XZe5e02xIRxhhe0ISpOuL/nmda5jOU5bX359WR5kPjxthqgjqcoQ51G/aEajTHGFKYfZ/4Q3CVL9r2bror3J2sj7CIExiLGWNCbMhZydcw1jBsqwcEhtemOpc7Pbn89s3z1glM4FFmJDYMMJczGbHF83v87+bzx+dFjwPYwZLWWvCvP3i30+Pg/w+DMK4oR5o2TAX/cfVUzEhfrwfYdxQD7QswuSPpawxYYqL6pQsyO/DIIwb6oGWDXPRH+z3YRDGDfVAy070PAzC2MGyrBwSizDZYxEmJDbMRf+9H41BGENYyhoTptbmTzv+qpLQBGEOdR+MQLUoTHG1428qCU0Q5lD3wQhUm8JwSmYTS9lphLm8fD1KmK1/HAZh0mIpa0yY+qJ/11+GFZogzKHugxGotoQRIjRBGJaVJ6MijBrr84+ARZiQ2ICnZLueihGaIMyh7oMRqLaEKa6qyxd++NIqlrLGhAn45y4Qxg31QMsiTP5YyhoTJuDfh2mMGX/Hmr7BSH6xlLUmTPUiGLPZ8T1NEMYElrLmhNkboQnCsKw8GRVh1FiffwQswoTEBjslW78uGcIYw1LWmjABX5cMYbxQD7RsoGXlYC+zhDBeqAdatuMLwmSJpawxYUK+LhnCeKEeaNkgwgR8XTKE8UI90LJhhNkboQnCsKw8GdWWMN2/cYkwxrAIExIb6KIfYQxjKWtMmGLnc5YIkxpLWWPChHzlS4TxQj3Qslz054+lLMLIsT7/GFjKWhJGuOJHmLRYypoTZtdPxSBMYizLyiGxCJM9FmFCYhEmfyxlEUaO9fnHwFIWYeRYn38MLGWnEKbyZZgwe18nFmHSYilrSRgtQhOEOdR9MAIVYdRYn38MLGURRo71+UfAsqwcEosw2WMRJiQ2kjBCOsIEpxMSJHeEkTPZEWZpzPjvBI3gwUh+sZT1d4QRmiDMoe6DEagIo8b6/GNgKYswcqzPPwaWsggjx/r8Y2ApizByrM8/ApZl5ZBYhMkeizAhsQiTP5ayCCPH+vxjYCmLMHKszz8GlrITCLPwBWGyxFIWYeRYn38MLGURRo71+cfAUhZh5FiffwQsy8ohsQiTPRZhQmIRJn8sZRFGjvX5x8BSFmHkWJ9/DCxlEUaO9fnHwFIWYeRYn38MLGURRo71+cfAUhZh5FiffwQsy8ohsQiTPRZhQmIRJn8sZRFGjvX5x8BSNrwwS18QJkssZRFGjvX5x8BSFmHkWJ9/DCxlEUaO9fnHwFI2a2EWxoy/Y03fYCS3WJaVQ2IRJnsswoTEIkz+WMoijBzr84+BpSzCyLE+/xhYyiKMHOvzj4GlLMLIsT7/GFjKIowc6/OPgaUswsixPv8IWJaVQ2IRJnsswoTEIkz+WMoijBzr84+BpWxwYWpfECZLLGURRo71+cfAUhZh5FiffwwsZRFGjvX5x8BSFmHkWJ9/BCzLyiGxCJM9FmFCYhEmfyxlbQtz/d18flIU7+bz+TfnCJMeS1nTwtz8eFpcf39anJ1whLGBpaxpYT58W/7n7OT2p1OEsYGlrGlh6qPMzfP54sysKL4ss3PrZTrCCNsTkiB3hJGzU5jbN88WZ2Xro4ygLkeYQ/2mHYFq+whz8/xZfWt1HSM0QRiWlSejmhbm+rvV5T7CGMAiTEhseGFqXz48fl/c/syysgEsZU0LUz3/Ul3ul/8+Wi2UCU0Q5lD3wQjUwMI0vvBMf5ZYyuYtTGXM+DvW9A1G8oulLMLIsT7/GFjKIowc6/OPgaUswsixPv8IWJaVQ2IRJnsswoTEIkz+WMoijBzr84+BpSzCyLE+/xhYyiKMHOvzj4GlLMLIsT7/GFjKIowc6/OPgaUswsixPv8IWJaVQ2IRJnsswoTEIkz+WMoGFmblC8JkiaUswsixPv8YWMoijBzr84+BpSzCyLE+/xhYyiKMHOvzj4GlLMLIsT7/CFiWlUNiESZ7LMKExCJM/ljKIowc6/OPgaUswsixPv8YWMoijBzr84+BpSzCyLE+/xhYyiKMHOvzj4GlLMLIsT7/CFiWlUNiESZ7LMKExFoU5vW9x2Z4rM8/BpayYYVZ+4IwWWIpizByrM8/BpayCCPH+vxjYCmLMHKszz8GlrIII8f6/GNgKYswcqzPPwKWZeWQWITJHoswIbEIkz+Wsggjx/r8Y2ApizByrM8/BpayCCPH+vxjYCmLMHKszz8GlrIII8f6/GNgKYswcqzPPwKWZeWQWITJHoswIbGRhBHSESY4nZAg6fjyWv88jjB+sJQNeoTpCLP/CyOMOyxlEUaO9fnHwFIWYeRYn38MLGURRo71+cfAUhZh5FiffwQsy8ohsQiTPRZhQmIRJn8sZRFGjvX5x8BSFmHkWJ9/DCxlEUaO9fnHwFIWYeRYn38MLGVzFyagMdbnHwNLWYSRY33+EbAsK4fEIkz22DHCPFnn8vJJO9vL3tmsZ6xPtuMLwmSJHUF9sjUIgzC5YhEGYeRYn38MLMIgjBzr84+BRRiEkWN9/jGwCIMwcqzPPwYWYRBGjvX5R8CyrBwS60GYMY+A9flHwG4URpwpwtwNwgyKK+wman9h7gRhECZXLMIgjBzr84+BDSXMZScIgzCZYkcIcylmK7r3Y2d9sggzKK6w/YTZoUX3IdoahGl/YYRxhxWF2etIJ7/99vtvq4jaDC0bIAgjxvr8I2D3Lyt39/TfWuk8JtuF6WS7N0pb65NFmEHxhN0hzHZTRgiz6ZQtH2G6Y0CYvLDbJyeeeG3PPmFWj2Q7CUYQGoswg+IEu/dKZYAp/YS5r03kEYTGIsygOMHuuFIZrsoAYe55E28EobEIMyhOsNuvVMb6MkiY7pJAnBGExiLMoDjBbr+oDytM30/eYY31ySLMoDjA3jv9Eg8DYkYJs3yMN0pjfbIIMyi2sV1V/vWvZhcfJcyuZeX+tLrOfWtsTxZhBsYutnMGthimZWG60lQPu93JLoMwgxIF2+9+h1wCi5/u1dbUkx0RhBkUS8J09zWHslTZdLCZbrIjYlWY9sAQZvP9vmNKoKv5NFmdoP0uLD2PnuyI2BTmdWdaCHP3fm8yJdTyV5p0hFGerxk12RExKkw9sOVHnmzNdtoYyYT0nr/WZ68wGR1TdmTTesDIR3LjAzZkJzErTDOtu/frUIXZflCJui9HSeeQGWY54BCEWSlz2MLsOP3ShGmWlUcm4nLChrs3VpswwlRfu9PUmDCNMwcqzF5VDkiYVYmB3owXpvminT72hGkdkg9BmMsN2bcTHV42TWm7PdtHr39k/V2r28SiMMN+53XIlHpkfYdGC7zx0b//zQxh7mXH5MbuCp2PrGS5P/qQwtw8nz9+H0SYveMJNKUe6S3M/vbtR2XDY4Mw97Lju+qAnUR8XLoVAgpz++akePdtQGHqqIPxk9Y9/X37/UaYe4m2k2wffUBhbv5yXlz/cB5cmO2bBZ5S4Gyt3c7v4/agA8uQ7y47HqEh36sCCnP95/fFzY+n5a0vy+zbmpC8s1eYD48bYaoIR5hpftrE84/GBMqYvw/Tyr11KB4wBSsLsz7CIExaLMKkxMrC9L6GYf7MYDKqA2Fu3zzrt0rG/JnBZFQHwvR+Hob5M4PJqB6E6STI1x6S3OZvh0pZCYsw/rCURRg5uc3fDpWyEhZh3GFZVk6JRRh3WIRJiUUYf1jKIoyc3OZvh0pZCYsw/rCURRg5uc3fDpWyEhZh/GEpizBycpu/HSplJSzCuMOyrJwSizDusAiTEjtQGCGufo2ZslPFVdkebRHGTyg7WRBGDGWniquyCKOGslPFVdmkwhCScRCGkB5BGEJ6BGEI6RGEIaRHAgmzfi2m+lbnxZmMZd3t+rv5/KQo3s3n82/O93xWoqzL1i1dTLbqWo3W8mTLNK9Q2WOnDSPM+m9i1Le6fyTDVtbdqpfAvf7+tDg7Sd1pa1qDXLb0Mdkq1atyG55smQ+1zH122jDCrF9Ptr7VfYFZW1l3+1BN5+zk9qfTfZ+TLOuydUsfky2W340sT7Z85B/9Y1m2z04bRpj1K5bXt7ovYW4r3W7lrfJIvDgzs5h12bqlm8lW36stT7ZKrUefnTaMMOu/iVHf6v6RDFvpdKteOro6K7P6vXBdtm7pZbKLfy1PtkotTJ+d9sCPMDfPn9XvNXq2fWeQZydeJvthdflsdLJVkh1hnF7DFNffrR5Mow/rnUGenTiZbHHWfCeyOtkq16muYdZ/E6O+1f0jGbay7lb7Un0zvP3Z5j64Llu39DHZ+kTM8mSr1Hr02WmDPg9TfX03z8OUZVvPFjyyeZLTnmzd0sVkm1Mby5MtlsL03Gl5pp+QHkEYQnoEYQjpEYQhpEcQhpAeQRhCegRhJsznF7NFvm6972L24OVQ3sevXpb/a73V+ehW8n//eW9bMjQIM2E+v1io8vHh8epdn54eb918b3YKs5WMLCGDMBOmFqZ4e7R616i9d6cwW8kIEzIIM2HawlSnZ3/45ePD2eyovl18/ONfy3+aN776W/mx4+WGR0XzCUtQ9WnliV0tzOozXrY2WpCXxHrrmlS99XWz7VGx/kJkSBBmwtTCXJW75+cX5b568cWvyz23vv1wIVLzxhe/FheVP0eLs6vm/RVgcbZVfmwpTAvV3qgiL4jN1g1p8YHlttX/my+UcjCOgzATprnoL3fgq2oPLXffau9d3X7Y/kD1Ruucq3l/dft/lRHLj5X/a6HaGy2FOe5sXaw+UG97VR1/jjlNGx6EmTCLI8zHh9Vh5qJeL6v21Pbt7hvNft36hCXpqrz5oBam9RntjWov1ls3pEaY6kDUeIQwQ4MwE2Z5SladkTXnTYs9tX27+0ZbmPr9VT49ffByfYRpfUZ7o5UwzdYIM0UQZsLU1zBvyz336sH6/Kh9u/tG+5Ss9aTKYl+/ao4wrc9ob7QSprV1+wOLba8WF0IIMzwIM2FWz8NUl9vlblzvxe3b3Tda1+bN+ytA9c/Hh40wrc9ob7QWpt66IdUX/quLfoQZE4SZMM2y8kV5lVEtADyod+n27c4brdXf5v2LvC1v/n1xmd9sUX9Ga6P1NUy99Yr0dnbUWVZGmBFBGEJ6BGEI6RGEIaRHEIaQHkEYQnoEYQjpEYQhpEcQhpAeQRhCegRhCOmR/wPh5CSmfm55rgAAAABJRU5ErkJggg==" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="differential-allelic-divergence-detection" class="section level1">
<h1>Differential allelic divergence detection</h1>
<p>In the context of MAGE, we see tumor-specific events such Copy Number
Alterations, aberrant hypermethylation, promotor mutation and silencing
events of tumor suppressor genes as mechanisms that will affect one
allele, or both alleles in a different way, per sample. However, these
mechanisms show no preference towards either allele on a
population-level, acting mainly as an added source of variance on the
observed reference allele fraction captured by <span class="math inline">\(\theta_{het}\)</span> while leaving <span class="math inline">\(\pi_{het}\)</span> unchanged. Calling this
variance <em>Allelic Divergence (AD)</em>, we term its change (increase)
in disease due to, amongst others, the mechanisms listed above
<em>differential Allelic Divergence (dAD)</em>.</p>
<p>Detecting dAD between to populations can be done via likelihood ratio
test, comparing: (1) a model fitting ALL data, both control and case, as
one beta-binomial mixture model, with (2) a model allowing the control-
and tumor data to have their separate <span class="math inline">\(\theta\)</span> parameters for the heterozygous
samples, yet keeping the <span class="math inline">\(\pi\)</span>
parameter constant. These two models’ parameters are distinguished by
respectively <code>H0</code> (null hypothesis) and <code>H1</code>
(alternative hypothesis) in the code below. The pivotal new function is
which fits the model assuming different heterozygote-thetas while
keeping all other parameters shared (i.e. the <code>H1</code>
model):</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>ParCASE <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">&quot;list&quot;</span>, <span class="at">length=</span>NC)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>NC){ <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  ParCASE[[i]] <span class="ot">&lt;-</span> caseList[(spl[i]<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span>(spl[i<span class="sc">+</span><span class="dv">1</span>])]</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>ParTOT <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">&quot;list&quot;</span>, <span class="at">length=</span>NC)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>NC){ <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  ParTOT[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>(ParCTRL[[i]], ParCASE[[i]])</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>dAD_analysis <span class="ot">&lt;-</span> <span class="cf">function</span>(data){</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  controlListP <span class="ot">&lt;-</span> data[[<span class="dv">1</span>]]</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  caseListP <span class="ot">&lt;-</span> data[[<span class="dv">2</span>]]</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>  positions <span class="ot">&lt;-</span> <span class="fu">names</span>(controlListP)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>  dAD_res <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">LocName =</span> <span class="fu">names</span>(controlListP), <span class="at">PiFitH0 =</span> <span class="dv">0</span>, <span class="at">PiFitH1 =</span> <span class="dv">0</span>, <span class="at">ThetaH0 =</span> <span class="dv">0</span>,</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">ThetaCTRL_H1 =</span> <span class="dv">0</span>, <span class="at">ThetaCASE_H1 =</span> <span class="dv">0</span>, <span class="at">RhoH0 =</span> <span class="dv">0</span>, <span class="at">RhoCTRL =</span> <span class="dv">0</span>, <span class="at">RhoCASE =</span> <span class="dv">0</span>, <span class="at">LRTpval =</span> <span class="dv">0</span>, </span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">NumHetCTRL =</span> <span class="dv">0</span>, <span class="at">NumHetCASE =</span> <span class="dv">0</span>, <span class="at">RobFlagCTRL =</span> <span class="st">&quot;&quot;</span>, <span class="at">RobFlagCASE =</span> <span class="st">&quot;&quot;</span>, <span class="at">HWECTRL =</span> <span class="dv">0</span>,</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">HWECASE =</span> <span class="dv">0</span>, <span class="at">CovCTRL_mean =</span> <span class="dv">0</span>, <span class="at">CovCASE_mean =</span> <span class="dv">0</span>, <span class="at">CovCTRL_med =</span> <span class="dv">0</span>, <span class="at">CovCASE_med =</span> <span class="dv">0</span>,</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">NumOutCTRL =</span> <span class="dv">0</span>, <span class="at">NumOutCASE =</span> <span class="dv">0</span>, <span class="at">QualityCTRL =</span> <span class="st">&quot;N&quot;</span>, <span class="at">QualityCASE =</span> <span class="st">&quot;N&quot;</span>, <span class="at">pr =</span> <span class="dv">0</span>, </span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">prv =</span> <span class="dv">0</span>, <span class="at">pv =</span> <span class="dv">0</span>, <span class="at">ThetaHom =</span> <span class="dv">0</span>)</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (LOC <span class="cf">in</span> positions) {</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>    CTRL_DF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">&quot;ref_count&quot;</span> <span class="ot">=</span> controlListP[[LOC]]<span class="sc">$</span>ref_count, </span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&quot;var_count&quot;</span> <span class="ot">=</span> controlListP[[LOC]]<span class="sc">$</span>var_count, <span class="st">&quot;isCase&quot;</span> <span class="ot">=</span> <span class="dv">0</span>)</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>    CASE_DF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">&quot;ref_count&quot;</span> <span class="ot">=</span> caseListP[[LOC]]<span class="sc">$</span>ref_count, </span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&quot;var_count&quot;</span> <span class="ot">=</span> caseListP[[LOC]]<span class="sc">$</span>var_count, <span class="st">&quot;isCase&quot;</span> <span class="ot">=</span> <span class="dv">1</span>)</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Previously, we used EMfit_betabinom_robust() to ROBUSTLY fit our models by removing</span></span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># outliers via Cook&#39;s distance. Now however, we&#39;re working with data from two different</span></span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># sources (control- and case tissue) and have no way of knowing in advance how similar</span></span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># these are. As such, outlier detections should happen ON BOTH SETS SEPARATELY yet the</span></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># two hypotheses we&#39;ll be fitting share all or some parameters between the two. </span></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># For this, we can use the EMfit_betabinom_robust() function with its &quot;fitH0&quot; set to </span></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># FALSE, which will not complete the entire eqtl-detection pipeline, but will cut it </span></span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># short before fitting the unshifted model. By running this function on both our </span></span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># datasets separately in advance of the dAD-relevant fits, we ensure correct outlier</span></span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># detection ánd the use of the same dataset in our upcoming likelihood ratio tests.</span></span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1. Detect and extract outliers using EMfit_betabinom_robust()</span></span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>    OUTfitCTRL <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">EMfit_betabinom_robust</span>(<span class="at">data_counts =</span> CTRL_DF, <span class="at">SE =</span> SEmedian, </span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">inbr =</span> Fmedian, <span class="at">fitH0 =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a>    OUTfitCASE <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">EMfit_betabinom_robust</span>(<span class="at">data_counts =</span> CASE_DF, <span class="at">SE =</span> SEmedian,</span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">inbr =</span> Fmedian, <span class="at">fitH0 =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a>    OUTfitCTRL_DH <span class="ot">&lt;-</span> OUTfitCTRL<span class="sc">$</span>data_hash; OUTfitCASE_DH <span class="ot">&lt;-</span> OUTfitCASE<span class="sc">$</span>data_hash</span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a>    CTRL_DF<span class="sc">$</span>Outlier <span class="ot">&lt;-</span> OUTfitCTRL_DH<span class="sc">$</span>Outlier; CASE_DF<span class="sc">$</span>Outlier <span class="ot">&lt;-</span> OUTfitCASE_DH<span class="sc">$</span>Outlier</span>
<span id="cb16-48"><a href="#cb16-48" aria-hidden="true" tabindex="-1"></a>    CurDF <span class="ot">&lt;-</span> <span class="fu">rbind</span>(CTRL_DF, CASE_DF)</span>
<span id="cb16-49"><a href="#cb16-49" aria-hidden="true" tabindex="-1"></a>    <span class="co"># These results are also handy for an estimation of the number of heterozygotes in </span></span>
<span id="cb16-50"><a href="#cb16-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># controls and tumors AND the number of outliers, both good filter criteria.</span></span>
<span id="cb16-51"><a href="#cb16-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># We can also include a &quot;RobFlag&quot; which gives more information about outlier detection </span></span>
<span id="cb16-52"><a href="#cb16-52" aria-hidden="true" tabindex="-1"></a>    <span class="co"># (e.g. none detected, or so unreasonably many that none were removed)</span></span>
<span id="cb16-53"><a href="#cb16-53" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>NumHetCTRL[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> <span class="fu">sum</span>(OUTfitCTRL_DH<span class="sc">$</span>prv)</span>
<span id="cb16-54"><a href="#cb16-54" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>NumHetCASE[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> <span class="fu">sum</span>(OUTfitCASE_DH<span class="sc">$</span>prv)</span>
<span id="cb16-55"><a href="#cb16-55" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>NumOutCTRL[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> <span class="fu">sum</span>(CTRL_DF<span class="sc">$</span>Outlier)</span>
<span id="cb16-56"><a href="#cb16-56" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>NumOutCASE[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> <span class="fu">sum</span>(CASE_DF<span class="sc">$</span>Outlier)</span>
<span id="cb16-57"><a href="#cb16-57" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>RobFlagCTRL[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> OUTfitCTRL<span class="sc">$</span>RobFlag</span>
<span id="cb16-58"><a href="#cb16-58" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>RobFlagCASE[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> OUTfitCASE<span class="sc">$</span>RobFlag</span>
<span id="cb16-59"><a href="#cb16-59" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-60"><a href="#cb16-60" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-61"><a href="#cb16-61" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 2. Perform one fit on the entire (non-outlying) data, i.e. all parameters shared,</span></span>
<span id="cb16-62"><a href="#cb16-62" aria-hidden="true" tabindex="-1"></a>    <span class="co"># i.e. assuming no dAD; the null hypothesis in dAD detection.</span></span>
<span id="cb16-63"><a href="#cb16-63" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Remark this uses the non-robust fitting function, since outliers were already detected.</span></span>
<span id="cb16-64"><a href="#cb16-64" aria-hidden="true" tabindex="-1"></a>    NOdAD_fit <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">EMfit_betabinom</span>(<span class="at">data_counts =</span> CurDF[CurDF<span class="sc">$</span>Outlier <span class="sc">==</span> <span class="dv">0</span>,], </span>
<span id="cb16-65"><a href="#cb16-65" aria-hidden="true" tabindex="-1"></a>      <span class="at">SE =</span> SEmedian, <span class="at">inbr =</span> Fmedian, <span class="at">fitH0 =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-66"><a href="#cb16-66" aria-hidden="true" tabindex="-1"></a>    NOdAD_fit_DF <span class="ot">&lt;-</span> NOdAD_fit<span class="sc">$</span>data_hash</span>
<span id="cb16-67"><a href="#cb16-67" aria-hidden="true" tabindex="-1"></a>    PiH0 <span class="ot">&lt;-</span> NOdAD_fit<span class="sc">$</span>AB</span>
<span id="cb16-68"><a href="#cb16-68" aria-hidden="true" tabindex="-1"></a>    rho_rr <span class="ot">&lt;-</span> NOdAD_fit<span class="sc">$</span>rho_rr; rho_rv <span class="ot">&lt;-</span> NOdAD_fit<span class="sc">$</span>rho_rv; rho_vv <span class="ot">&lt;-</span> NOdAD_fit<span class="sc">$</span>rho_vv</span>
<span id="cb16-69"><a href="#cb16-69" aria-hidden="true" tabindex="-1"></a>    ThetaHomH0 <span class="ot">&lt;-</span> NOdAD_fit<span class="sc">$</span>theta_hom; ThetaHetH0 <span class="ot">&lt;-</span> NOdAD_fit<span class="sc">$</span>theta_het</span>
<span id="cb16-70"><a href="#cb16-70" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-71"><a href="#cb16-71" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>PiFitH0[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> PiH0</span>
<span id="cb16-72"><a href="#cb16-72" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>ThetaH0[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ThetaHetH0</span>
<span id="cb16-73"><a href="#cb16-73" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-74"><a href="#cb16-74" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-75"><a href="#cb16-75" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 3. Perform a fit on the (non-outlying) data allowing separate theta_het parameters </span></span>
<span id="cb16-76"><a href="#cb16-76" aria-hidden="true" tabindex="-1"></a>    <span class="co"># for control- and case-data</span></span>
<span id="cb16-77"><a href="#cb16-77" aria-hidden="true" tabindex="-1"></a>    FullFit <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">EMfit_betabinom_popcomb</span>(<span class="at">data_counts =</span> CurDF[CurDF<span class="sc">$</span>Outlier <span class="sc">==</span> <span class="dv">0</span>,],</span>
<span id="cb16-78"><a href="#cb16-78" aria-hidden="true" tabindex="-1"></a>      <span class="at">SE =</span> SEmedian, <span class="at">inbr =</span> Fmedian, <span class="at">probshift_init =</span> PiH0)</span>
<span id="cb16-79"><a href="#cb16-79" aria-hidden="true" tabindex="-1"></a>    ParamVec <span class="ot">&lt;-</span> FullFit<span class="sc">$</span>ParamVec</span>
<span id="cb16-80"><a href="#cb16-80" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>PiFitH1[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ParamVec[<span class="st">&quot;probshift&quot;</span>]</span>
<span id="cb16-81"><a href="#cb16-81" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>ThetaCTRL_H1[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ParamVec[<span class="st">&quot;theta_het_control&quot;</span>]</span>
<span id="cb16-82"><a href="#cb16-82" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>ThetaCASE_H1[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ParamVec[<span class="st">&quot;theta_het_case&quot;</span>]</span>
<span id="cb16-83"><a href="#cb16-83" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-84"><a href="#cb16-84" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-85"><a href="#cb16-85" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 4. Perform the Likelihood Ratio Test for dAD detection</span></span>
<span id="cb16-86"><a href="#cb16-86" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Likelihood of fit with all parameters shared:</span></span>
<span id="cb16-87"><a href="#cb16-87" aria-hidden="true" tabindex="-1"></a>    LikTot <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">pmf_betabinomMix</span>(CurDF[CurDF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>ref_count, </span>
<span id="cb16-88"><a href="#cb16-88" aria-hidden="true" tabindex="-1"></a>      CurDF[CurDF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>var_count, <span class="at">probshift =</span> PiH0, SEmedian, rho_rr, rho_vv, rho_rv, </span>
<span id="cb16-89"><a href="#cb16-89" aria-hidden="true" tabindex="-1"></a>      <span class="at">theta_hom =</span> ThetaHomH0, <span class="at">theta_het =</span> ThetaHetH0)</span>
<span id="cb16-90"><a href="#cb16-90" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Likelihood of fit with separate theta_het (calculated in two steps because of the </span></span>
<span id="cb16-91"><a href="#cb16-91" aria-hidden="true" tabindex="-1"></a>    <span class="co"># different theta)</span></span>
<span id="cb16-92"><a href="#cb16-92" aria-hidden="true" tabindex="-1"></a>    LikCTRL <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">pmf_betabinomMix</span>(CTRL_DF[CTRL_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>ref_count, </span>
<span id="cb16-93"><a href="#cb16-93" aria-hidden="true" tabindex="-1"></a>      CTRL_DF[CTRL_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>var_count, <span class="at">probshift =</span> ParamVec[<span class="st">&quot;probshift&quot;</span>], SEmedian,</span>
<span id="cb16-94"><a href="#cb16-94" aria-hidden="true" tabindex="-1"></a>      ParamVec[<span class="st">&quot;pr&quot;</span>], ParamVec[<span class="st">&quot;pv&quot;</span>], ParamVec[<span class="st">&quot;prv&quot;</span>], <span class="at">theta_hom =</span> ParamVec[<span class="st">&quot;theta_hom&quot;</span>],</span>
<span id="cb16-95"><a href="#cb16-95" aria-hidden="true" tabindex="-1"></a>      <span class="at">theta_het =</span> ParamVec[<span class="st">&quot;theta_het_control&quot;</span>])</span>
<span id="cb16-96"><a href="#cb16-96" aria-hidden="true" tabindex="-1"></a>    LikCASE <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">pmf_betabinomMix</span>(CASE_DF[CASE_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>ref_count, </span>
<span id="cb16-97"><a href="#cb16-97" aria-hidden="true" tabindex="-1"></a>      CASE_DF[CASE_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>var_count, <span class="at">probshift =</span> ParamVec[<span class="st">&quot;probshift&quot;</span>], SEmedian,</span>
<span id="cb16-98"><a href="#cb16-98" aria-hidden="true" tabindex="-1"></a>      ParamVec[<span class="st">&quot;pr&quot;</span>], ParamVec[<span class="st">&quot;pv&quot;</span>], ParamVec[<span class="st">&quot;prv&quot;</span>], <span class="at">theta_hom =</span> ParamVec[<span class="st">&quot;theta_hom&quot;</span>], </span>
<span id="cb16-99"><a href="#cb16-99" aria-hidden="true" tabindex="-1"></a>      <span class="at">theta_het =</span> ParamVec[<span class="st">&quot;theta_het_case&quot;</span>])</span>
<span id="cb16-100"><a href="#cb16-100" aria-hidden="true" tabindex="-1"></a>    lrtstat <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">2</span> <span class="sc">*</span> (<span class="fu">sum</span>(<span class="fu">log</span>(LikTot)) <span class="sc">-</span> <span class="fu">sum</span>(<span class="fu">log</span>(<span class="fu">c</span>(LikCTRL, LikCASE))))</span>
<span id="cb16-101"><a href="#cb16-101" aria-hidden="true" tabindex="-1"></a>    LRTpval <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(lrtstat, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> F)</span>
<span id="cb16-102"><a href="#cb16-102" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>LRTpval[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> LRTpval</span>
<span id="cb16-103"><a href="#cb16-103" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-104"><a href="#cb16-104" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-105"><a href="#cb16-105" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 5. Fill out the results dataframe</span></span>
<span id="cb16-106"><a href="#cb16-106" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>QualityCTRL[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> <span class="co"># spot bad quality data</span></span>
<span id="cb16-107"><a href="#cb16-107" aria-hidden="true" tabindex="-1"></a>      OUTfitCTRL<span class="sc">$</span>quality; dAD_res<span class="sc">$</span>QualityCASE[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span>OUTfitCASE<span class="sc">$</span>quality </span>
<span id="cb16-108"><a href="#cb16-108" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Test HWE on both the control and tumor data:</span></span>
<span id="cb16-109"><a href="#cb16-109" aria-hidden="true" tabindex="-1"></a>    HWEtest_CTRL <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">HWE_chisquared</span>(<span class="at">Fmedian =</span> Fmedian, <span class="at">data =</span> OUTfitCTRL_DH)</span>
<span id="cb16-110"><a href="#cb16-110" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>HWECTRL[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> HWEtest_CTRL<span class="sc">$</span>PVAL</span>
<span id="cb16-111"><a href="#cb16-111" aria-hidden="true" tabindex="-1"></a>    HWEtest_CASE <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">HWE_chisquared</span>(<span class="at">Fmedian =</span> Fmedian, <span class="at">data =</span> OUTfitCASE_DH)</span>
<span id="cb16-112"><a href="#cb16-112" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>HWECASE[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> HWEtest_CASE<span class="sc">$</span>PVAL</span>
<span id="cb16-113"><a href="#cb16-113" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Mean and median coverages:</span></span>
<span id="cb16-114"><a href="#cb16-114" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>CovCTRL_mean[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> </span>
<span id="cb16-115"><a href="#cb16-115" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mean</span>(CTRL_DF[CTRL_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>ref_count <span class="sc">+</span> CTRL_DF[CTRL_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>var_count)</span>
<span id="cb16-116"><a href="#cb16-116" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>CovCASE_mean[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> </span>
<span id="cb16-117"><a href="#cb16-117" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mean</span>(CASE_DF[CASE_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>ref_count <span class="sc">+</span> CASE_DF[CASE_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>var_count)</span>
<span id="cb16-118"><a href="#cb16-118" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>CovCTRL_med[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> </span>
<span id="cb16-119"><a href="#cb16-119" aria-hidden="true" tabindex="-1"></a>      <span class="fu">median</span>(CTRL_DF[CTRL_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>ref_count <span class="sc">+</span> CTRL_DF[CTRL_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>var_count)</span>
<span id="cb16-120"><a href="#cb16-120" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>CovCASE_med[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> </span>
<span id="cb16-121"><a href="#cb16-121" aria-hidden="true" tabindex="-1"></a>      <span class="fu">median</span>(CASE_DF[CASE_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>ref_count <span class="sc">+</span> CASE_DF[CASE_DF<span class="sc">$</span>Outlier<span class="sc">==</span><span class="dv">0</span>,]<span class="sc">$</span>var_count)</span>
<span id="cb16-122"><a href="#cb16-122" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>pr[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ParamVec[<span class="st">&quot;pr&quot;</span>]</span>
<span id="cb16-123"><a href="#cb16-123" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>prv[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ParamVec[<span class="st">&quot;prv&quot;</span>]</span>
<span id="cb16-124"><a href="#cb16-124" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>pv[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ParamVec[<span class="st">&quot;pv&quot;</span>]</span>
<span id="cb16-125"><a href="#cb16-125" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-126"><a href="#cb16-126" aria-hidden="true" tabindex="-1"></a>    dAD_res<span class="sc">$</span>ThetaHom[dAD_res<span class="sc">$</span>LocName <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> ParamVec[<span class="st">&quot;theta_hom&quot;</span>]</span>
<span id="cb16-127"><a href="#cb16-127" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-128"><a href="#cb16-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-129"><a href="#cb16-129" aria-hidden="true" tabindex="-1"></a>  dAD_res<span class="sc">$</span>HWECTRL[<span class="fu">is.na</span>(dAD_res<span class="sc">$</span>HWECTRL)] <span class="ot">&lt;-</span> </span>
<span id="cb16-130"><a href="#cb16-130" aria-hidden="true" tabindex="-1"></a>  dAD_res<span class="sc">$</span>HWECTRL[<span class="fu">is.na</span>(dAD_res<span class="sc">$</span>HWECTRL)] <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span></span>
<span id="cb16-131"><a href="#cb16-131" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-132"><a href="#cb16-132" aria-hidden="true" tabindex="-1"></a>  dAD_res<span class="sc">$</span>RhoH0 <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>((<span class="dv">1</span><span class="sc">/</span>dAD_res<span class="sc">$</span>ThetaH0)<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb16-133"><a href="#cb16-133" aria-hidden="true" tabindex="-1"></a>  dAD_res<span class="sc">$</span>RhoCTRL <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>((<span class="dv">1</span><span class="sc">/</span>dAD_res<span class="sc">$</span>ThetaCTRL_H1)<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb16-134"><a href="#cb16-134" aria-hidden="true" tabindex="-1"></a>  dAD_res<span class="sc">$</span>RhoCASE <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>((<span class="dv">1</span><span class="sc">/</span>dAD_res<span class="sc">$</span>ThetaCASE_H1)<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb16-135"><a href="#cb16-135" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-136"><a href="#cb16-136" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(dAD_res)</span>
<span id="cb16-137"><a href="#cb16-137" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-138"><a href="#cb16-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-139"><a href="#cb16-139" aria-hidden="true" tabindex="-1"></a>cl <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">makeCluster</span>(<span class="fu">getOption</span>(<span class="st">&quot;cl.cores&quot;</span>, NC))</span>
<span id="cb16-140"><a href="#cb16-140" aria-hidden="true" tabindex="-1"></a>parallel<span class="sc">::</span><span class="fu">clusterExport</span>(cl, <span class="fu">c</span>(<span class="st">&quot;Fmedian&quot;</span>, <span class="st">&quot;SEmedian&quot;</span>))</span>
<span id="cb16-141"><a href="#cb16-141" aria-hidden="true" tabindex="-1"></a>dADFinData <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">parLapply</span>(cl, <span class="at">X =</span> ParTOT, <span class="at">fun =</span> dAD_analysis)</span>
<span id="cb16-142"><a href="#cb16-142" aria-hidden="true" tabindex="-1"></a>parallel<span class="sc">::</span><span class="fu">stopCluster</span>(cl)</span>
<span id="cb16-143"><a href="#cb16-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-144"><a href="#cb16-144" aria-hidden="true" tabindex="-1"></a>dAD_res <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">&quot;rbind&quot;</span>, dADFinData)</span>
<span id="cb16-145"><a href="#cb16-145" aria-hidden="true" tabindex="-1"></a>  </span></code></pre></div>
<p>In <code>dAD_res</code>, both - and -values are listed, which both
represent the overdispersion parameter; <span class="math inline">\(\theta\)</span> just has a range from 0 to
infinity while <span class="math inline">\(\rho\)</span> has been
rescaled to go from 0 to 1, which is more convenient for some analyses,
such as plotting. When studying differential allelic divergence, one
usually tries to look for “disturbed regions” by plotting the results in
<code>dAD_res</code> among the length of chromosomes (e.g. plotting
tumor rho-values along the chromosomes while also marking the actually
statistically relevant loci in some way), as will be demonstrated at the
end of this section.</p>
<p>That being said, you can also just extract loci showing differential
allelic divergence from <code>dAD_res</code> using the
<code>LRTpval</code> column to retain statistically significant results.
Defining an effect size is a bit trickier; you’d expect to be able to do
it on the overdispersion-parameter-values of cases compared to controls,
but how would one go about it? When both are very small (e.g. rho ~
10^-7 range), one being 10 times as large as the other doesn’t mean that
much, but when both are quite large (e.g. rho ~ 10^-2 range) this does
make a very large difference. Subtracting one from the other has an
equally unclear interpretation regarding effect size. The effect of the
overdispersion parameter on the actual variance also depends on the pi
parameter, which further complicates things.</p>
<p>These kind of difficulties makes us prefer a “looser” approach such
as detecting affected regions on plots rather than strict statistical
procedures. But if you wanted to e.g. only retain loci that show a
strong deviation from a regular binomial regarding its variance, using a
rho-cutoff of, say, 0.1, can do. Such cutoffs are based on experience
and can be played around with. In the code below, we retain such
good-quality loci which are also statistically significant and have a
median coverage of at least 10, at least 15 heterozygous samples, and
samples conform Hardy-Weinberd Equilibrium in both control- and
tumor-data. Also, note that for true induced allelic imbalance,
overdispersion in the tumor samples needs to be larger than in the
control samples.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(dAD_res<span class="sc">$</span>LocName[dAD_res<span class="sc">$</span>RhoCASE <span class="sc">&gt;=</span> <span class="fl">0.1</span> <span class="sc">&amp;</span> <span class="fu">p.adjust</span>(dAD_res<span class="sc">$</span>LRTpval, </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>) <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> dAD_res<span class="sc">$</span>CovCTRL_med <span class="sc">&gt;=</span> <span class="dv">10</span> <span class="sc">&amp;</span> dAD_res<span class="sc">$</span>CovCASE_med <span class="sc">&gt;=</span> <span class="dv">10</span> <span class="sc">&amp;</span> </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  dAD_res<span class="sc">$</span>NumHetCTRL <span class="sc">&gt;=</span> <span class="dv">15</span> <span class="sc">&amp;</span> dAD_res<span class="sc">$</span>NumHetCASE <span class="sc">&gt;=</span> <span class="dv">15</span> <span class="sc">&amp;</span> dAD_res<span class="sc">$</span>HWECTRL <span class="sc">&gt;=</span> <span class="fl">0.001</span> <span class="sc">&amp;</span> </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  dAD_res<span class="sc">$</span>HWECASE <span class="sc">&gt;=</span> <span class="fl">0.001</span> <span class="sc">&amp;</span> dAD_res<span class="sc">$</span>RhoCASE <span class="sc">&gt;</span> dAD_res<span class="sc">$</span>RhoCTRL], <span class="at">collapse =</span> <span class="st">&quot;, &quot;</span>))</span></code></pre></div>
<p>[1] “loc0, loc6, loc7, loc68, loc80, loc84, loc85, loc86, loc88,
loc89, loc91, loc93, loc95, loc96, loc97, loc98, loc100, loc158, loc211,
loc212, loc218, loc258”</p>
<p>The previously used function can also be used to vizualize the dAD
detection results via seperate plots of the control- and case-population
utilizing their separately fitted overdispersion parameters. Here, it’s
important to set the input argument to the same value for both plots, so
the width of the PMFs can actually be compared correctly.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>CTRL_DF <span class="ot">&lt;-</span> controlList[[<span class="st">&quot;loc85&quot;</span>]]</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>CASE_DF <span class="ot">&lt;-</span> caseList[[<span class="st">&quot;loc85&quot;</span>]]</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>PlotData_dAD <span class="ot">&lt;-</span> dAD_res[dAD_res<span class="sc">$</span>LocName<span class="sc">==</span><span class="st">&quot;loc85&quot;</span>,]</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>dAD_plot1 <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">MAGE_EMfitplot</span>(<span class="at">ref_counts=</span>CTRL_DF<span class="sc">$</span>ref_count, <span class="at">var_counts=</span>CTRL_DF<span class="sc">$</span>var_count,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pr=</span>PlotData_dAD<span class="sc">$</span>pr, <span class="at">prv=</span>PlotData_dAD<span class="sc">$</span>prv, <span class="at">pv=</span>PlotData_dAD<span class="sc">$</span>pv, </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_hom =</span> PlotData_dAD<span class="sc">$</span>ThetaHom, <span class="at">theta_het =</span> PlotData_dAD<span class="sc">$</span>ThetaCTRL_H1, </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">probshift =</span> PlotData_dAD<span class="sc">$</span>PiFitH1, <span class="at">SE =</span> SEmedian, <span class="at">ScaleCount =</span> <span class="dv">50</span>, <span class="at">ScaleHist =</span> <span class="cn">TRUE</span>, </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">nbins =</span> <span class="dv">30</span>, <span class="at">plot_NoShift =</span> <span class="cn">FALSE</span>)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>dAD_plot2 <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">MAGE_EMfitplot</span>(CASE_DF<span class="sc">$</span>ref_count, CASE_DF<span class="sc">$</span>var_count, <span class="at">pr=</span>PlotData_dAD<span class="sc">$</span>pr,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">prv=</span>PlotData_dAD<span class="sc">$</span>prv, <span class="at">pv=</span>PlotData_dAD<span class="sc">$</span>pv, <span class="at">theta_hom =</span> PlotData_dAD<span class="sc">$</span>ThetaHom, </span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">theta_het =</span> PlotData_dAD<span class="sc">$</span>ThetaCASE_H1, <span class="at">probshift =</span> PlotData_dAD<span class="sc">$</span>PiFitH1, <span class="at">SE =</span> SEmedian,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">ScaleCount =</span> <span class="dv">50</span>, <span class="at">ScaleHist =</span> <span class="cn">TRUE</span>, <span class="at">nbins =</span> <span class="dv">30</span>, <span class="at">plot_NoShift =</span> <span class="cn">FALSE</span>)</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(dAD_plot1, dAD_plot2, <span class="at">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAzAAAAHgCAMAAABNd3gvAAAA6lBMVEUAAAAAADoAAGYAOpAAZrYAoIcqr7IxeW4zMzM6AAA6ADo6AGY6kNtNTU1NTW5NTY5NbqtNjshNu9VUdmVUhZFYWsFZWVlmAABmADpmtv9uTU1uTW5uTY5ubo5ubqtuq+SByb6M1MmOTU2OTW6OTY6Obk2OyP+QOgCQkDqQtpCQ27aQ2/+ZU0mgIPCk1eGrbk2rbm6rbo6ryKur5OSr5P+v4Oy2ZgC22/+2///Ijk3I///bkDrb/7bb///kq27k///mSzXpo5nr6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T////gnz/GAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAdH0lEQVR4nO2dDXvbxnKF5Rurtm9Nf93Eru3U8leZtomjJFXdK7mR5Vi+pRXh//+dAiRIghIIYLFnwcHue57EoqjhwQCDl7sYkMBehhDqrL1dJ4DQmAQwCDkIYBByEMAg5CCAQchBAIOQgxyB+eOKrj2xRV3j5IajS9Crmj2TEK9TFHbb6gIwbnEAk4hdosC8e/fHH/v7wuUCjCysa20ARhEIMADTLw5gBo0DmETsAEYTBzCJ2AGMJg5gErEDGE0cwCRiBzCaOIBJxA5gNHEAk4hdosDQVg6Zq18YbWX/LAEGYPraAYzJOIBJxA5gNHEAk4gdwGjiACYRux7AnE4mk/sn2cXryYOPADPUggHGhl0PYI6nxb+Xh9Ps9CHADLVggLFh5w7M5S9HxY+LtyfZ7PsTgBlowQBjw84dmHwqNplMs9mLj9nFm4Kd27m8yrkDvXuXZfv7u85iMIXdh2grZ03AzJ4fFaPM+YMlMB4V2dkbuPnzMGdnV18HMP3stNltrUtLl+x4uh5hxgiM+QQTBsa0XX9gOIYJaggwNu16AFPMxS5/Pbk8PKBLFs4QYGza9RlhTieTe0cZ52GCGgKMTbu+UzJRRQBmmwDGph3AiOIAJg27VIGhrSzMlbYywDgvF2BkYQDjnWUAQ+sJJgyMaTuAEcUBTBp2ACOKA5jo7R4XAhhRHMBEbwcwyjiAid4OYJRxBoCZPSu+drH8BIbigxgj38PFdkkDE2Fbufjo+Oz5UflNWMkXYmkrVwUwcQFzXuCx+hS55MPkAFNV0sCYT7DfMUw+ypTfUxr3F2JNqgRmy18Bxi3OBDDFVy7Kb8JKvhA78oMOsR0jjDLOAjAXrw+y7NoIAzCiMIBRxhkAZvasuP7V+I9hrNoBjDJu98AseMnKb8JKvhA78j1cbAcwyjihYXNhtqq4IGlxIobzMIHskgbGclu5JzABykNbuSqAARjvXEPYAYxPlqEMDSeYOjBW7QBGGQcw0dsBjDIOYKK3AxhlHMBEbwcwyjiAid4OYJRxABO9XdLA0FbW5BrCjrayT5aBDAFGk2sIO4DxyTKUoeEEUwfGqh3AKOMAJno7gFHGAUz0dgCjjAOY6O0ARhkHMNHbSYFBOjVfbKGXQu1DSdklPcLQVtbkGsKOtrJPloEMAUaTawg7gPHJMpSh4QRTB8aqHcAo4wAmejuAUcYBTPR2AKOMA5jo7QBGGQcw0dsBjDIOYKK3SxoY2sqaXEPY0Vb2yTKQIcBocg1hBzA+WYYyNJxg6sBYtQMYZRzARG8HMMo4gIneDmCUcQATvR3AKOMAJno7gFHGAUz0dkkDQ1tZk2sIO9rKPlkGMgQYTa4h7ADGJ8tQhoYTTB0Yq3YAo4wDmOjtAEYZBzDR2wGMMg5gorcDGGUcwERvBzDKODEwZ2dX4wBm13bNdYkcGOttZTEwo9K7d1m2v7/rLK6rrMsWAYzbcm0D0zPZ3QwJls/DJDvCWE4wdWCs2gGMMg5gorcDGGUcwERvBzDKOICJ3g5glHEAE70dwCjjACZ6u6SBoa2syTWEHW1lnywDGQKMJtcQdgDjk2UoQ8MJpg6MVTsPYC4Pp1l28Xry4CPA6A0BxqadBzCnk+kcmtOHAKM3BBibdv2Bmf3rv02zi7cn2ez7E4CRGwKMTbvewFz+8l/56DJ78TG7eHOU/347l1c50YaaPxXbS6H2oaTsegNzelBMx84fLIHxqAgjTI0YYWza9QUmH1ouN0aYMQJDW1mTawi72NrKp5NCB+M+hgEYTa4h7GIDJlu0lS8PD+iSBTFMHRirdpyHUcYBTPR2nOlXxgFM9HYAo4wDmOjtAEYZBzDR2wGMMg5gordLGhjayppcQ9hF2FYGmP6BANMmgPHJMpSh4QRTB8aqHcAo4wAmejuAUcYBTPR2AKOMA5jo7QBGGQcw0dsBjDIOYKK3SxoY2sqaXEPY0Vb2yTKQIcBocg1hBzA+WYYyNJxg6sBYtQMYZRzARG8HMMo4gIneDmCUcQATvR3AKOMAJno7gFHGAUz0dkkDQ1tZk2sIO9rKPlkGMgQYTa4h7ADGJ8tQhoYTTB0Yq3YAo4wDmOjtAEYZZwKY+aV7i0v53j+RXGdx5Hu42A5glHEWgDkvQMmOp8Vjyf2uRr6Hi+0ARhlnAJjje7/lI8zlL/M7KkiuFT/yPVxsBzDKOAPALKZk+VRsMuF+VwFU1mWLIgcmzrZyAczs+VExykjud7WbIYG2sk+WgQzjBWau46nkflcAU1XSwFhOUAIMxzByO4BRxpkBppiLXf56Irnf1cj3cLEdwCjjzABTnIe5d6S539XI93CxnRQYpFNzN6aXQu1DSdkxwijjTIwwABPSDmCUcQATvV3SwMTZVo4DGNrKPlkGMgQYTa4h7ADGJ8tQhoYTTB0Yq3YAo4wDmOjtAEYZBzDR2wGMMk4LzNnZTwBjza65LgDjFgcw0dsBjDIOYKK3SxoY423lpIEx3FYGGNlyAUYWBjA+WYYyNJxg6sBYtQMYZRzARG8HMMo4gIneDmCUcQATvR3AKOMAJno7gFHGAUz0dkkDQ1tZk2sIO9rKPlkGMgQYTa4h7ADGJ8tQhoYTTB0Yq3YAo4wDmOjtAEYZBzDR2wGMMg5gorcDGGUcwERvBzDKOICJ3i5pYGgra3INYUdb2SfLQIYAo8k1hB3A+GQZytBwgqkDY9UOYJRxABO9HcAo4wAmejuAUcYBTPR2AKOMA5jo7QBGGQcw0dslDQxtZU2uIeyiayufTyb3TzR3HQWYGgFMXMAUt+o9fZhdHk7972vNlKxGqQNj1c5nSpZDc/H2ZHGba4ARGwKMTTsfYPKhZfbiY3bx5ij/5XYur3KiDZWFUSrUPpSUXX9gZs/uHWXnD5bAeFSEEaZGjDA27XxGmJyU9QgDMFrDAMAggZpH/pa28vGUY5hQhowwNu36jjDlXOzy8GDEXTLayppcQ9jF1lbOTieT/BiG8zD9AgGmTeMH5uuTGz+GqQhTshp1BqZDXcYIjFU7lxHm/d7eN78DzDCGDiNMa10AZldTsq9P9vbuAswQhk5Tspa6AMzujmGK0vzl7wAT3NDxGKaxLgCzK2A+7O3dzKcA2yYAPZMEmBo5AdNSF4DZCTB//rC397R48HnbW1nPJAGmRt2Baa8LwOymS9Y25I8OmCjayh3qMkZgxt9WDlcRgKkR52FGD8yfP9zN/78ZETCWE3SZkrXVZYzAWLVzAOb9zXl1GirTM0mAqVF3YNrrAjC7OYZpPbLsmSTA1MjhGKbTET/AaMIARhkHMNHbOUzJPhQl+fqEM/2DGHafkrXXBWB20yX7vLfX/AmMnkkCTI0cumStdQEY2sqSONrKmlxD2EXQVgYYXSDAtCkCYL7cKYb+po/49UySKVmNugPTXpcxAmPVrjswHTr9PZMEmBp1BqbTOUuAGR6Ysn0JMMMYuraVAWYYO5cRBmAGNHQYYQBmQDuHY5j2U2M9kwSYGnU/hulwyhJgdjIl2+OgfzhDhylZa10AhrayJI62sibXEHYRtJUBRhcIMG2KAZgPe3tPPzR9b7xnkkzJauQATGtdxgiMVTuX78N8879PnvJ9mIEM53Xp9n2Y1roAjBKYprpcOQ9TtPz5eP8whp2B6VAXgAEYm3EAE71dd2CyD8XQz/dhBjLsPiVrrwvA7AQYvg8zpGF3YPg+zJB2LsCEqght5Ro5ABO4PLvZw+22lQFGt1yAkYWNHxg+GjOoocNBPx+NGdDOdYT5+i8d796D/FQWpqu61CXUPpSUnfOU7HPDKeWeSTLC1Mh1StZUF4DZJTBMyQYxdAaGKdkgds7ANN2EpGeSAFMjV2Dabg4DMJow54P+pjuQ9kwSYGrketDf4c6wPZMd+R4utqOtTFvZO9cQduNvKwOMMDAkMLPvT7Ls4vXkwcfVD4DZ4XmYhpZ/zySZktXI+TzMsi7nk/sn2eXhNDt9uPwxRmCs2jmMMJ+LaTIfvhzIsPsIs1mX43u/5SPMxduTYqQpfwDMbkYYbncxoKHLx/s361IQMnvxMbt4c1T+yJ+8nWv7Ox3qrOYTygDjFmcGmPMHc1LKH37lGfmQILZz+T4M94cZ0NDh+zBX6lI7wgCMKMylS/ahOLBsusxizyQBpkYOXbIrdZlxDBPQjrZylG3ly8ODRZfsYKxdsvG3lQFGGBgaGM7D9IxTT8m4LtlQhm5TMq5LNpSdAzBcl2xIQ65LZtOOyywp43bUVuYyS8PZAYwyDmCit+O6ZMq43ZyH4bpkA9pxXTJl3G4O+rku2YB2tJWjayvHAcz428oR3uMyCmAivcfl+IHhLsqDGrp++DI2YKzaOR30t50b65kkwNTI5aC//ZwlwOxihOHKlwMacuVLm3ZJH/RbTjD1g36rdgCjjAOY6O06AtPtyLJnkgBTo27AdD/iBxhNmAswrZ++6JkkbeUaOQDT4VMx4wNm5G1lgJEuGGBaBTA+WYYyNJxg6sBYtQMYZRzARG8HMMo4gInerjMwrdeJBRipYVdgutQFYAYHppt6JgkwNeI8jE273sDMnk0mU81lSQCmRgBj064vMMX1FGfPjySXh6etXKPUgRl5W/mazgtKjqeSSysCTI0AJi5gCnF5+IByvO14F4Xah5Ky8wCmuA6p5PLwHMPUKPURxqpdf2AuXh9kmeTy8ABTI4CxaefRJZsW1Iz6GMZyggBj064vMAteNJeHB5gaAYxNu77AnE4KTTkPE8oQYGzaJX2mn7ayJtcQdjG2lQGmdyDAtAlgfLIMZWg4wdSBsWoHMMo4gIneDmCUcQATvR3AKOMAJno7gFHGAUz0dgCjjAOY6O2SBoa2sibXEHa0lX2yDGRoGpiiLgCjstMB01iXyIGxnGDqwFi1AxhlnG1gkEBlXbYIYNzibAPTM9mRDwliO0YYZRzARG8HMMo4gIneDmCUcQATvV3SwNBW1uQawo62sk+WgQwBRpNrCDuA8ckylKHhBFMHxqodwCjjACZ6O4BRxgFM9HYAo4wDmOjtAEYZBzDR2wGMMg5gordLGhjayppcQ9jRVvbJMpAhwGhyDWEHMD5ZhjI0nGDqwFi1AxhlHMBEbwcwyjiAid4OYJRxABO9HcAo4wAmejuAUcYBTPR2SQNDW1mTawi7JNrKY9O7d1m2v7/rLOrVfHWSXgq1DwWxSwKYnkkyJatR6iOMVTuAUcYBTPR2AKOMA5jo7QBGGQcw0dsBjDIOYKK3AxhlHMBEb5c0MJyH0eQawo62sk+WgQwBRpNrCDuA8ckylKHhBFMHxqodwCjjACZ6O4BRxgFM9HYAo4wDmOjtAEYZBzDR2wGMMg5gordLGhjayppcQ9jRVvbJMpAhwGhyDWEHMD5ZhjI0nGDqwFi1AxhlHMBEbwcwyjiAid4OYJRxABO9HcAo4+wAczqZTO6fZBevJw8+AozQDmCUcXaAOZ4W/14eTrPThwAjtEsamHjbype/HBU/Lt6eZLPvT8YIDG1lnywDGcYLTD4Vm0ym2ezFx+ziTcHO7VxuxdyxrF4zrvl6cZEDYzlBP2Bmz4+KUeb8wRIYj/KMfA4ltkt6hLGcoKBLdjxdjzAAIwoDGGWcNWBGfAxj1Q5glHFmgCnmYpe/nlweHtAl09oBjDLODDDFeZh7RxnnYeR2AKOMswOMsDwj38PFdkkDE29befzAcB7GJ8tAhokD83iuAdfJJQxgfLIMZWg4wdSBsWoHMMo4gJGFWbUDGGUcwMjCrNp5ADM/I6boWwJMjQDGpl1/YM6L71tIPj8OMDUCGJt2vYE5vvdbPsJIPnsBMDWa1wVgzNk116V1Sjbiz49ndj9CXqgsjFJXa28aGKttZU9gGj8//uhRtyQ5D1Oj8CPMo0cA4x4mG2FqKtKZGKZk1zXAlGxBzIDrFIOdJzDNxzAA0z8OYGzaeQLT/PlxgOkfBzA27TyBaT4PAzD94wDGpp0HMF0qIsrSNRBgACaQHcAo4wBGFmbVLlVg5h3VtNvKtoGJsK0MMP0DAaZNAOOTpWtgR2AeMyUzC4xVO4BRLhdgZGFW7YIC898diQGY6wIYm3YAo1wuwMjCrNoBjHK54wLm2zkxA65TDHYAo1wuwMjCrNoFBeZb8aBPWzkdYJJsKwNM/ziAARjflWFKlhAwVu0ARrlcgJGFWbUDGOVyAUYWZtUOYJTLBRhZmFU7gFEuF2BkYVbtwgIjLgnAAMyu7ZIGhrayXWDSbCsDTO84gAEY35VhSiYGZl6eAdcpBjuAUS4XYGRhVu0ARrlcgJGFWbUDGOVyAUYWZtUOYJTLBRhZmFU7gFEuF2BkYVbtUgBmAcfj68+1ti6Xr4wDmKtalEdqqZPVe/c037fHGxjtexjAeNX66iJMjzCJnoexDUzrlAxgpOsUgx3ANAlgpOsUgx3ANAlgpOsUg13awOyX2uYDMNJ1isEuYWD2r+q6D8BI1ykGOykwV/Wtjcbl46rK50pGXs21QmYdvPHKXSQ9wG3HF+VR7WpW93CxXUIjTGWcWeBx67vvXr7c339ZajHKbAxHsY8whoGhrey/MiJgSlpybQJTIgMwwdbJJQxg/FdGAswCl5e35nr18opWyGy8EmA06zRuu8W+kBowc1pybQFmgQzAhFmncdsNA4y0JP7A7C9nYFuBuXVrjszGKwFGs07jtksQmP0VL03A3Fr3mAFGuk7jtksPmPmxfgdgFshUXgkwmnUat11ywJStsa2qknRrQcyV/lqoBOsEMNbsUgOmhpfvvivIqAXmZfWkDMBo1sklzF5bOS1g9uvGlyZgXq7PyQCMaJ1cwgDGf2U8gKnlpWlKVnaYAWbEcyixXULAdMGlBpgrxIRLsE7DAPNth7uQjnYPF9ulA0w3XmqA2ZyWhUuwTgBjzS4ZYLrhUgvMrSoxwRKsFcBYsxsIGGVJem2drrzUAnOrgkyoBOsFMNbsUgCm29F+s1bEhEhwuwDGml0SwDTzstlWbkIGYBTr5BLm1FbuMAUAmC5xLcNLR2BuVT/ArE1wuwAGYHxXxrl23tOxCjH7AKOJy7rs3l3tup9ZBpjWOP/Dl01i5Ak2CmC6mAGMJrCIU/Jya/lpTGWCzUoEmO37+la7x9ukyW6r4gZm8dUXITDdiJHtXIu6AEyNdgpMS11GC0z5VTElMJ2mZYkD02Wy1QpM+VRW+eWPmr8DTI/V2dhQ619W36yUArP4xH/zTiEAplqX8MB0KM8QwFy7qGJXDQxMW13GCcz6m8gtCHRtK5fAVD9b5rkmrXUZMzBb3sZWYb3p6Kb65Y4HGGFJOgFTbLPlV4/VwKyJ8VyTjnWJAJiOe/lqCy+LVl4z7lblDzXl7EhP97VoLkxrXUYITLGVXnYExk0bH2D2XJOOdRknMN3o2LqFy19eVX7pVs51gdro6bqyVw6SIgRmjsvLkMCU1/rzXJN4gGk9CLm6BV/1qIoLMBtPOcPTCEx7XUYGzH7JS1Bgqpdh6r0mjcCs6zICYLYyUlOC4YF5udKr+rGneWXjBmZRq+XWCQhMAzIKYM4GBqbpmKw+15aB5Nrm2niqDzAuBer21Gby603QBEyHuowImGXVBgFmOdZfS0MDzE9WgelESt3mqjxlA5jq/Hr7oBMtMM2lk9ZjtfUrW7cGmMePr++HlaeyzbBqcJUXO8Bs7lM9340CA7NtiXULrHn3q8FmGy8jB6YsovvmdG8rb2795ZYVAnN2ZgaY7cNJ3+HbBRi32jQusRWYjeyuz9VKXIwCU7Oj1QeWweu3PPfN6QvMEhkZMGdXeBkOmI1861DRT3GbpQFmq7YDUzNXm9PS6Y3MOjDVYgavZh0w1ZswbSTYB5h5VT79tEtgaoaT6sqH3sTDqdsR1uZ7xnDAPGphxh2YjbKGOMLvto1XZ/+Xk7PewCzew84+fRoemEcFM7sdTgwrL0oNNv2BuXg9efCxGZhFTZqYyTb2nWZgNrOfr9OOgbnaa3EB5mxDnz6pgelQnq0bNPT2NK2NuvxtrutF7gHM5eE0O33YAszPP//8qF37TqrWdFhgzrbJKf9rr87p+CQHprk81Xw6FKiD/il0DRq1tTJe+tumPIG5eHuSzb4/6VARlTSVDSjt6voB06U8i91i11ttbOoLzOzFx+zizVH+6Hauq3912jU6Ar/rLSVS163iSIhTefTvxrveqsNpywZvBeb8wbIiha6+hQnOewc17Hopn50leC3OERhJecTrJK7NbrLrfdC/fgsDGJfAgYCRlAdg2sM6A9M8SbYOzPgSdARGUp7d7JK27XoDc3l40NCGGd3+6BtnDRhJeaLYw60A09zoH93+6BtnDRhJeaLYw80As6GeS7ezP/rGmQNGUZ4o9nCAMRkHMInYAYwmDmASsUsUmOjbyiMGJtK2MsD0CwSYNgGMf5ZMyRICxrYdwGjiACYRO4DRxAFMInYAo4kDmETsAEYTBzCJ2AGMJg5gErFLFBjayiFz9QujreyfJcAATF87m8Bc1bUvxfpKbZheggHNU7Lb5gYwOzYEGJt2AGPUEGBs2gGMUUOAsWkXCBiE0hLAIOQggEHIQQCDkIMABiEH9QFmfWWf8tHGpX68DGfPJpNplp1OJpP7Jy2v6uJXGskSLPyKDD0TzFYX3xNtwarE5REXR1wbdWVa69IDmPUdFspHm7dc8DEsLno6e36UHU/7u23cAmJhpEuwUHE5Y78Ec4+yqKItWJW4POLiiGujrkx7XXoAs746aflo83KlPobnRXLH08tfjtpe082vNNIlmC32G88Es+N7vy3sRFuwKnF5xMUR10ZcmQ516QHM+vrX5aPNC2L7GM5zfXOUD4Tzwd/brzSSJli843gmmK2GftEW3HDWlkdcHHFt5JVprUsPYNZ3WCgfbd5ywccwW1wsuBj4Pd4q1n6lkTLB+U/PBLNVYURbsCpxecTFEddGXpnWulgbYS5eH5TP9p+MXsnneKpM8Hx1EOg1Wx7lCCMojrg28sqEGGECHsNks2erde2/1lfyOZ4qj2GOl/uMBJgxHcNIiiOujbwyrXXp1SU7WLVhDha9hAPPJtTy9WVJiveKy1977z9rv9JIl2A53HsmmK0KI9qCVYnLIy6OuDbyyrTWpf95mMJaeh4mN6w00+95TFDWCZZGsgSXA7RngovCCLdgVeLyiIsjro26Mq114Uw/Qg4CGIQcBDAIOQhgEHIQwCDkIIBByEHmgPnzh7257lae+7B348e+fl/++mP+X+W3jb9udf7H/1yLRdTGJDDzcny583T11NcnT7eGt6qxKFudgaVO1MYuMNn7m6unvLZQY1G2OgNMnaiNcWCKKcBf/v7lzt7ezfJx9uWf/z3/sfzlr/+R/+3pIvBmtnzBwqh4WT55KIuyesWPlaC588KxjC6dit/uLmNvZusFpS1qYxeYz/km+POHfHt8+Ob3xdYpH9+ZF2v5yze/Zx+KGt2cj+DL5wuD+Yie/21RlIpVNahwnjsuo5dO8z8sYov/lwva5YYxIGpjEpjFgWW+kT4XWyHfRMUWWj2+U/1D8UtlXF8+Xzz+v2KrL/6W/1exqgYtivJ0Izpb/aGM/Vy8xz3d9VTAgqiNSWDuFuNx8Vb2oezJFFuj+njzl+W2q7xg4fQ5f3ijLErlFdWgctuvo5dOy6IUb3bLWgEMtbEKzHzUX47N861Rfbz5S7Uo5fOFvj658eP6XazyimrQqijLaBtFsSpqYxaY7H2+dT7fWI/B1cebv1SH/Urjfr49Py/fxSqvqAatilKJrv5hHvt5PtkGGGpTyCowxeHenz/km6rcUtXHm79Ujv+WzxcGxY8vd5ZFqbyiGrQuShm9dCoPLlcHlgBTiNrYBSaf896dH2TeKDdb9fHGL1/WHcbl83O9zx/+5/xQchlRvqIStJ4nl9Erp/d7NyvGADMXtTEIDEKWBTAIOQhgEHIQwCDkIIBByEEAg5CDAAYhBwEMQg4CGIQcBDAIOej/Af3h/6i6Z/AxAAAAAElFTkSuQmCC" width="100%" style="display: block; margin: auto;" /></p>
<p>The plot above is of a locus showcasing very clear differential
allelic divergence, with the reference allele fraction in heterozygous
case samples being way more variable, probably due to case-specific
effects.</p>
<p>As promised, to conclude, this section showcases a plot of dAD
results across a chromosome using the function. The toy data isn’t very
fit for this, however (no chromosomal location, too small), so included
with the toy data is some thoroughly processed data to illustrate the
plot on. The inputs needed are <span class="math inline">\(\rho\)</span>
values in both controls and cases along with differential expression
data (log2 fold changes) which can also be inferred from raw RNAseq
counts. optionally even accepts additional hypermethylation and
copy-number-alteration data for even more extensive plots. As this is
for illustrative purposes only, we just make the plot using the provided
datasets below, but more details can be learned from ’s help page. The
function returns several plot components (including legends) as separate
objects due to its complexity, so it’s up to the end user to arrange
them into a suitable composition.</p>
<p>As a final remark, many of the measures in this plot (differential
expression, hypermethylation, copy number alterations) are usually
obtained at the gene-level rather than the SNP/locus level. To make the
results of different analyses more compareable, MAGE’s per-locus
p-values can be combined using the function. For this, we recommend
using the geometric mean as a nice balance between more conservative
(arithmetic mean, maximum) and liberal (harmonic mean, minimum) methods
for combining p-values, and the different loci should be weighted as
well, for which we recommend <span class="math inline">\(\small
\sqrt{median \ locus \ coverage \ * \ estimated \ number \ of \
heterozygous \ samples }\)</span>, the latter being an output of MAGE’s
beta-binomial mixture model fits via the heterozygote genotype frequency
times the number of samples. For p-values concerning a comparison
between a control- and case-population, one can e.g. use the minimum of
this weight across populations.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>ChromPlot <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">MAGE_ADChromplot</span>(AD_Data, DE_Data, Meth_Data, CNAgain_Data, CNAloss_Data,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">pvalSIG =</span> <span class="fl">0.05</span>, <span class="at">roll_median =</span> <span class="dv">15</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>ChromPlot[[<span class="st">&quot;ADDE_plot&quot;</span>]] <span class="sc">/</span> ChromPlot[[<span class="st">&quot;MethCNA_plot&quot;</span>]] <span class="sc">/</span> (ChromPlot[[<span class="st">&quot;LEG1&quot;</span>]] <span class="sc">+</span> </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  ChromPlot[[<span class="st">&quot;LEG2&quot;</span>]] <span class="sc">+</span> ChromPlot[[<span class="st">&quot;LEG3&quot;</span>]]) <span class="sc">+</span> </span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="fl">0.5</span>), <span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAzAAAAHgCAIAAAD1yx9KAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nOzdf3wT550v+u8kTku2yd1LMpxdkJVYVIKUOu22AURkYNd0m66NV0DqsC69AU5OVorP9UZ2MEtfB/LKK6+Ee0txYquXe6jdXBacU+JL3AS0YG2au3gXsMBA2m3ipQQpyCALtmsBezbdLdm4nfvHSPJIGv2a3yN/3i//MZqfj54ZzXz9PM88D8NxHAEAAACAfu7QOwEAAAAAMx0CMgAAAACdISADAAAA0BkCMgAAAACdISADAAAA0BkCMgAAAACdISADAAAA0BkCMgAAAACdISADAAAA0BkCMgAAAACdISADAAAA0BkCMgAAAACdISADAAAA0FmV3gmQJZFIqLp/lmXVPoTBIQcImUBEyATkABEhE4gImUBEhs+E/W3LOgcieqeibCghAwAAANAZAjIAAAAAnSEgAwAAANAZAjIAAAAAnSEgAwAAANAZAjIAAAAAnekbkI12u91ut9vtG4znXSc+6HN3j2qYKAAAAKhI117b37ItnDP75pF1XS321J/IChrQsR+y+KBvJ20PBJwUH/S1dlsDHU6xlb7XH6V67VMHAAAAleTctq5Xfkz0zdwlN2JjD22JNC3RPk3TdCwhm4hF611OIiLLoytswyGRUrD44PdO1tTbtE4ZAAAAVJLwHkfXK5eWb/kOK7IweiNWy1o0T1Mm/UrI4rFxm3U9P22x1tDJWJycGdkRH/zeyRV/+Zf0veFY5qaLFy/mJ86fP692MllW7NzNJMgBQiYQETIBOUBEyAQiQiYQkeEzoavFTkTC/vodbeHONqJrr10UWT2auDJ28Tn7KSIi0quoTL+AbCIWJWuB5Xw45rfQYM6idByGoZPUhhwgZAIRIROQA0SETCAiZAIRmSETyhs66drlBNUuf/XtZfOIKHpm27ozFn5aU/pVWVZbC1VF8uFYs94FiAAAAFDZ5j29eSAdgdnut44l8r9pqB79Ssgs1ppoaILIQkTx2DjVuATRV/z0yWg02uruT37e6R7fuNeP+AwAAAAqkY6N+quttuGDg3Hiw69k+/4kS7M/kLR3o43qtwcQjQEAAIDyzm3r2vbazeSH6I3Y4wtnVhsysjT7t8fcre5+ItvGvX4nEVF80Nca2yDaAQYAAACAUq69tv+5cN3ALseSXU9dW7ev5btERPTw8lffduiRHIbjOD2Oqww06lcbcoCQCUSETEAOEBEygYiQCURk+EzY37asvEb9xoChkwAAAAB0hoAMAAAAQGcIyAAAAAB0hoAMAAAAQGcIyAAAAAB0hoAMAAAAQGcIyAAAAAB0hoAMAAAAQGcIyAAAAAB0hoAMAAAAQGcIyAAAAAB0hoAMAAAAQGcIyAAAAAB0hoAMAAAAQGcIyAAAAAB0hoAMAAAAQGcIyAAAAAB0xnAcp3caAAAAAJTR1WLvHIjonYqyVemdAFkSiYSq+2dZVu1DGBxygJAJRIRMQA4QETKBiJAJRIRMUAeqLAEAAAB0hoAMAAAAQGcIyAAAAAB0Jj0gi/jrmBx1fvM1owMAAADQl4SALOhlGIZhNtEBLscB2oTADAAAAIzo2mv7W7aFRRYcP9ri6Gqxd2177abmieKVG5AFvczhtRzHcdyIz5672O4b4TiO43ZccHiDiiQQAAAAQL5z27qe+67Y26HRM9u8tCXcORB5ynVs357jmqeMqPyArKGX621QbDUAAAAA1YX3OLpeubR8y3dYkYXRxJXHFy4hIrpvyWr21DtiRWiqQ6N+AAAAqChdLfaulox6PEdbuHPg7WUWsbWvXU486Lifn543n6VLN66pn8QcsgKyrHb9aDgGAAAAuusciJTRWX88bIRebmW9Zelorx2abtA/VNvuQEwGAAAAZmJxiNVjak1yQBbc3R7yDAkbijX0DnlC7bvRlh8AAGBm2t+2bH/bMr1TUaZ589kr4Rv89LXLCVpw/zwdUoE2ZAAAADCT2dgH3xo5EiWim+eOJZZ/w6FHIiQHZA1be1x9jcK+LYLexj5Xz1a8XAkAAABGN90nmW3Zrl72jce6Wuz7QqufalulS3KqJG9p942Eqc7BMOk5rp6waOdkAAAAYEyiNYyb95zRPiVamPf05gGxaVrVNBBu0iVJKbKqLFPdwCYhGgMAADCRfO29zNcOzPykl5ABAABABcgqD0M0pgsZJWSio4szGDEJAAB0Y8q3/ABklJBF/JvaQ54hjJAEAAAAIJPkErLwhRB51iIaAwAAAJBLckDmWORSMh0AAAAAM5bkgMzu2+HJ7IdMgtFut9vtdrt9g/H8C93u7lFZRwEAAAAwtHIDsqA33Xy/sY+or1F6o/74oG8nbQ8EAoG9K062ZgVd0wsDezeO70RIBgAAqsF7AKC7cgOyhl6usNIb+U/EovUuJxGR5dEVtuFQRsxlafYHOpyUZykAAIBC+GgMMRnoS+F+yCJ+f9jnKykmi8fGbdb1/LTFWkMnY3FyWsRWPH0yWr/BKZizePFifuL8+fPy0lscyxphDHg9IQcImUBEyATkABGZJxP2ty3rHIiotJVZMqF0ot+o8NeUsxRESQnIgl6msY+IiDK6veBne4Z8pe1lIhYla5F14oO+1v4o2TbuFcZj03FYIpEoPdkSsCyr9iEMDjlAyAQiQiYgB4jIbJkgLalFtzJXJpRI9BsV+JpFM6HyskgDZTfqj/jrGvs8QxzHceGescY6f4SIb1rW2EfkGSq5yrLaaiu6jqXZHwgEAhtiraLN/gEAAAAqQbkBWeTYoZCrZ2sDEf+iZejQsaC/jmnsI1dPuJwWZGSx1kRjE/x0PDZONVax+koiInK66tOrAgAAgBGl2+GhQZ4E5QZk4Qshql2QGkTcscgVam9sD7l6wuUPLV5ttQ0fHIxTspWYK6NWcrRb0BfGaGjYZq0uM6WKwlgcAAAABeApKZMCjfpdPeGygzEiIkuzf3vM3eruJ7Jt3Ot3EvGNxmIbAh1OZ8femK/V3U9E/OLmvOVnqis35M8apRUAAGCG2LznDCIzaeQHZK71qyVEYzxnRyDQIZxhafYHBJPNslKmoLKusP1tyxCTAQDAzMQ/MfEoLJfC3V5UJP6SKv3CwoUIAAAAZZE8dBIAAMDMgkbroB4pAdn0eEmO9hCF2h3Shk4CAAAwDT4I62qxK9izP94YgzQdh04CAACAii1vQ7hZFrQhAwAArfHPaQO2tc0XQBgwqVBhyi0hC3pLqpQscTUAAAATUPtVLZQkQbklZA29HHkZpjFrHMuUiL/O0R7i+yab4XWXhv3/DwAAchW+Xata+4Z+IjRx/GiL9yJx9OB3ntr19H2Zy24eWbfvjQ9Sn765ZmCXQ+vkSWnUn2xGtvYwk2sTHeA4jiu/2/5KhX96AACymPfGqGzAZN58kEy/rxw9s81LW8KdA5GnXMf27TmetfhGbOyhLZHOAf5Ph2iM5LQha+jluF4FU1I50r/YGfhjAwCobJ0DkUQioeDtPd3VZaUWkmXllU7fMZq48vjCJURE9y1Zzb7xTrhtlSDqit6I1bKPa56oTGjUDwAAGsG/qTNN7jg32kRjXS12IuociCQ/X7uceNDh4qfnzWfp2I1r5JiXXj2auDJ28Tn7KSIiemhLpGmJBmnMhoAMAAC0kH4wY7hDUNt0KMaLhxOUvx7y2uUE1S5/9e1l84goembbujMWflpT6KkfAAC0IxyMDmEZaMTiYAssnff05oF0BGa73zqWiGuSqkwoIQMAAAB16dw2bt589so7N4juI748bMFCKQVgqfc085L1eiYCMgAAAK3l6xpJWLGrdZoqmY198K2RI8841thunjuWWP4XGWHTuW1dg45UXxjRG7HHF7bl3ZF6LcxkVFlG/HUiHV+gP1gAAMimfe2k2etDzZ5+g7Et29XLvvFYV4t9X2j1U22riIiuvba/ZVuYiJbsesp1bF+LvavF3tXyHPlM1u1FxL+pPSTaOSwAAIAYjUt9zNWLRFZSi/aCISFiM1FuqGBV00C4SThj3tObB5KT9615u3ONDmkSklxCFr4QIs9aRGMAAGA4ikcem/ecqYBoJrdLMFIzSquAHNOS5BIyxyIXXVAyJVKwbKHXJgxyCA0SqSqzp18RyARCJiAHiEhqJvCdQolurlKupndb+v4lrKnBAyLfCtn9bJWmq8W+v21Z1lb5DlE4bYW3yk0Yfj5FSQ7I7L4dHqbRu1bXOstEIqHq/lmWlX8ItROpKkVywOyQCTQzMqFwacFMyIGiZGbC5j1ncjdXKVfTuy19/yWuKcwEDR4Qoiukr1VpCcjaKt9OCuy8wJUgYW/Ak1xlGfQ29hH1NaJRP8x4fFMPtMA1tfTpw3kEUzBabWCBH47RkirDqqaBsHqd+EsuIcNQlgBEmbchc7UghnxwHnWkdqumCmDk/xnynbhKOqE3j6zb98YH6Y/st97dvMamyJ7RUz+AAtINfo18r4Si0o8NnEcFlZ6Zpiun1CudlRTfmMvNI+v2vbFgzUCkM/nXx77x2P4jUUV2Lisgy+qJrM5fXutCAACjQUymEm1iCM3Oml4hkcaXJX4FmY6H3hh7aIuwl7JVTVseT7zxg7ASe5feU3/EX+dorx3iRlKN+oNexlFH4RGfXYmEAQDoIz30NeouTYc/d53P/lC0AbnZwwuN63NL/BWYPVeNQ3qj/t3tIc+Q8BXLht4hT6h9Nxr1p6EOC0AR2r82gThMKbqX6JRYDWqWM26QdPLdZ9CMe8Ctcn2r9uIr2wTlYcePvvIW+61nFOnZH2NZageNVQEk0LGZTm4UiN+vNFrWV3Y2/tfc+ekEmPcMGjD0mYHDbt635u2naN2+lumaQAUb9Ut/y3Jrj8uR0Q9Z0NvY5+oJo/P+whSvBEGcBxVM34dQuu4yDT83zaC+WAgXnnGoN8iS9BIyu28kTHUOhknPcfWgAZnWDPg/E4DicgMjLQ/NT/A9YeIXVxZp2aXj6TYy40RjODsqkVVlafeNcD6lUgIAsgjvksa5d8uRWyqgbKlJev94wKiqMq5GCZS6XHF96u/aa/uf20Xf+snmNTY6t63rlR+nlzy0JaJUV7FoQ2Zi+JVCWm7NmsGfgkWrYHIbb6l9wZeeY8bPXtCdsm3ecb3pKnrGn4zG0v2QpVrxHz/a4tjPB2qylfuWZdCbHB0p6M0eNAlDJ+kAv1IQSvdPSyaJ1/Ml0rCNhXVPjylOK89ESTUyo2Wj0dKjiWjiCrHzbEbrh6yhl+PSExg6yTjwL/sMkVsvmXt/NHsdnGGjMSFdfnFm7B1NTlJlflND/RBkfhEFU6KIzoFIVwuajCsMQyfpQJF7hCkeWqCIdC9cufWS+dqNGbwPvAIJ0/7CLjeX8ItTlujlrVQmm6vAOAufeFxvBrBq4XK6OPjaTZX7IZMTkOVWX6K+UlP4oc5MfL1kiWffjM8hKqdtmWQmzRmQw3T3TNMluFI52sJrrLv2tdj3vfEB0Y+PtNi7kn+ei8TpP3RS0NvY5xniGijif7mPPENcr8Nf5/AGuV70RKauos1uSoSfuuLSXYmq8Sag6OEo/3k37Pg/RYvH8qVWjRooHSvUJBxOs2PJV1ZqC8ffMjNZ34vfUNWm8lXSdymXoy3c2abuISQHZMHDfeQZaiCKHDsUIs+OBiJaUEvth4O9DYjINCD/LoMuyI2vlN4gC9zxjfwwyNfnqilon7EmyhwhyZeukNH+o8hl0rMjmcFPh3nJ7/YifCFErvWK1J9WrALDq1GZF7foE1raz6PA4xC/N+Mo5VwUWEeNQjtlmfofA40zNl0matgTqmA5vZH/nRAyRSKzmDHNxnD8aIv34oPbntr19H3C2dde2//crsTy3s62VXL2Ljkgcyxy0aFLEaLDfeTqWW0nQaEZpJVyTyn93qrs6BlZ+xGmE8GZEVTkTVORAZ6NEI6YJVzQ0kzLEOE1UDiy1CpFxZV43YquNtPOb15XvruvJbxmQND5xbynNw987cy2x7r2yIrJJAdkdt+BnkMOB0NEniGfnYJepnGsJzyCeCxb4V9juZ0HqvfbzuoPXXTaUHeWylbZQ9dJfn1MqTDIXI8WU6RW1ddj8VK5gkrMwwK96iifJlN5aEtk4ajjyDZHZjmZbVnz46deeSfctkpyjaGMKsvMgZPSHZSVY7TbvXOYiGwb9/qbLVkL44O+1v4oERHVbw90OKWn1MjKesBo9ksQjt8n7GzG1LVLJlLZ0Zgi5FSNKRvfaFZcZ4rrQdlE5sYE+GnowhT/EpTg+NEW70Xi6MHvZNc5Fl+awdEWXrPHkV1OJpvkgCzir3Nc2CHnlcr4oG8nbQ8EnBQf9LV2WzNjrtHu1pMr9gb8Fj4y8w2KhGwVwuA3lwJPMlULz3DnrTD63tPTYZOCZS3C91iV2qdJqfprLb0vMe3rkSv+jBd+Z0jjxMgTPbPNS1vCnUvo5pF1+/bMz6xbLLxUhKMt3Onc1tViT41lGT0z+BYt75UTn0kOyMIXQjIOS0Q0EYvWu5xERJZHV9j6Q6MdzumIbDQ0XL8hwEdgluYN9f2hCaKcgCzfXUCp+XzhkHr7N938Epud8eQfl/IUP+ieD+ZKZ7pdf/qjLunJpVk+lPiQLryfzoGI6PysOenDSUu/8DQVTnbuJWeQ6029+Vl5krs+P6dr6L+rmp6Z9lzgGSc90tJPFE1ceXzhEiKi+5asZt/IqlssvDSPJbs6B75xtMXRRRwR0fI+vRr1N2ztcTle9m9t8EkcPSEeG7dZ1/PTFmsNnYzFyZkOuZwdgYzoLL0qEdHixYv5Cb4yjWVZ0SNgviLz0x/zrc8/qHKH0ch6nKSfZwb5XmXNzzet1Pqi8/OdiHL3U9Z8/jwKg4/0ecyKSErcv/DCyI1pJKdTwvysC7Lc/eRbJPxSXS329FGU+t3lHqXwrozwe1F7vuh08tJ1+7oCfoOkE/N1nJ99K7t2OfGgw8VPz5vP0rEb18gxj0paOm1V00A4d06TaFLKx3ASmn4REQW9TGOfyHzPUInVmKPd7oPWVD1kxodM8UFf68kV4ssSiURZiS4Xy7JqH8LgpOVA4f/sSyzoNk6VZeFMyEqnIslW/LuXvkNFzl2+Hep4NuU3fyzltyD/xJVYPaTXryM3EzROSeFSaiLqdPto6lOyfyWx4GGV0oDnAhk+E/a3Lcv+H+bctq7BdDP840db/i/21beXzSttqVakl5DJHVy82morvtJot3vnuFiLfzCyArdLMJfSm+8YnBEi+6LMnskAxmVx5C36LrpU4OaRdfve+CBrJvutdzevKSGiKUZ+x7BSWaw10VTDsHhsnGpcWUFXfNDX2l+zPeCv0PcrQQPGKWYzC+2bRUOWshq94doGKMm8+eyVd24Q3UdE1y4naMHCeSUvpVTXr8Sx33q3cyA79rp5ZF1XywdE35T50qWsgCzir3O0h4jIM8RtvVTnOLQ+PFJ6k7Jqq2344OB6Z7MlfvpktH5DRtyVisYqtbcLqGxKVREqq5RUqZQexA3K2mzUgUoBjMrGPvjWyJFnHGtsN88dSyz/C0cZS48ffS5cNxDOF2zdt+btzjVEdPxoyzaSEZPJGVycaRzrCXM7djONlO4o1rug5J4wLM3+7TF3q7uf74fMScTHYbENgY7q0yejRNGd7uHU2hXcFdkMV2GFWFnvysncj1JKLHQp97VEEKXI2S+6k6xONyrmFwSgDtuyXb2Jlse63uB7GltFxJd7hesGdjlEl05b1TRQyuuTJa6Wl8zBxX12Cqbm2H07PO2NZQ0u7uwIBDqEMyzN/gARETX7A81SkwaGJPr8SD91ZuY/+oZ6lAoTw58s+Qkz1Bc0rwI9EQh/QQXWBACRNyLnPb15IP9SzenXhgxmntznR2UTfQEzi2GfoFkJ6xyI8H06GDbBRqBg5pS4q9x/bNRIjAHl9quHkl1QX6o3/7xkNSOT3w/ZgtScoLexz9UTxmCWUEBudKJIv51GfvYU+HZGTjaYiGjQb/DfhTRZvRwLIRoD9WWVoh0/2uKlLeGmJUREdG5b1yuy9i69hMzuG+EWeBmmnYiokekjcvWU06YfZrbCcVjhRbl9c5vl2WOKRAJQaWNvGMf0f2s/eUPvtABIdIesrRt6OQFEY1AWYV0D5QnCROtiTPT/MT/eixGeZEZIA2hM8o8i30+Mr7YW/pEBLi2DJANAJrQhAz3l9j9eoFNZE7URyW0Ub6LEg3kJL7MCtXul70340SAXcNERLQFMSpl+yHglj5oEkC3rHUzRFSTvXJeWzrkHwjMDNCN8W1bBfRpkwJzcRgsAFUB6QBbx1znaa4e4kVQMFvQyTB2akYFUWQ+PoiGaNLkvZ8nfGyItMI6sq7HcixPxDUDJcscal0V6P2S720OeoRFBiVhD75CHadwd9KGYDBQm+malhM1FG6IhnAIAALmW7OocKL5WAfIa9QMoSjQ2ymqJJXP/uU3shc2TAQAAdCG9H7LeIQ/TWLdouooS/ZCBAoqGX4o0NC5QZoYCsxKZpasRKAvOKUBxqbHGU58fXv7q28vmFdqiFNKrLL2NfUTU7uA7IkuZ/owm/mACuZEZqjJLYa7u36CC4f1l0Fb0zLbHTl15fE3GWOPHj7Y4upb3drbpM5ZlQy/H9co5MoCR5PaFlp7uHIjokyZ1mGsEJ4CicPWCZm4eee4UbXtq4On7Mmavahr4Cbvtsf1HfrJ5jU3yztEPGcA00arMrpYKeXG4QFkCyhgAAIqJXgqNsa5X7xNZZFvgqj0V+tuba54WW1oSOQFZ0Ms00hDX2xD0Mo19RKimhIpReERw8yrQ7y4AABQUTVwhtlm8DOy+eQvoSvgGkQ4BWdDb2OcZ4hoo4n+5jzxDXK/DX+fwBhGSge4U7AnWOJ1hqgc1PgAAupPeqP9wH3mGGogixw6FyLOjgYgW1FL74WBvAyIygJkJhW0AULlWLVxOR0aPNy0RabwfHn2Llvc6cheUTH4/ZOELIXItkpMGADCrooO+o/gNACqGo633oVPerj3Hs+bfPLLuyKna5Y/r85alY5GLDl2KEB3uI1fPajsJCs00w7JsBRzC4EyUA6JJVST9JsoEtQmzonMg0tVi39+2TPgiKj9TdH2zq6TvIplZMkHVdJolE1Q1UzMh+UJl17ZtT+1Ktd8/t23fGwvWDOySWTQlOSCz+w70HHI4GCLyDPnsFPQyjWM94RFNqyvVbtlT8Y2HijJXDogmVX76zZUJKknffIVZkZ4WnSn60bxwGZAZMiEdI6iXTuNnggZmdibYlu0KZ1QHyB40iSfjLUu7b4TzpT819HJcgZUBAAAAIA/0QwaVDA2YFCfaUCyry35kOwBUkONHW95ZWLxGssTV8iq3UX/QyzDeYHJCjDcoNSkAYGSFwyy8XwkAFWpV08A3PmxxdLXY9x+Jiiw/t62rxd4lLxqj8kvI0jWTGDoJYMbJjckqtQddAACBVU0D4Saim0fWdbV8kL1weV/nwC75x0CVJQAogB+XCZWVAGAGx4+2eC8SRw9+Z/plyZSbR9bteyMddX1T+PrkfWve7lyjVprKDcjSoyTlg9GTQDcopNEXojEAMIPomW1e2hLuXEI3j6zbt2d+Z1tG/2E3YmMPbYk0LdE6WeW2IWvo5ZLCPS4izxCXNuQh8gwhGgNNIQgAAIByRBNXHl+4hIjoviWr2VPvhDOX3ojVshYdkiV96KTd7SHPkLDbsYbeIQ/T6F2LEjLQFmIyAAAQ4runFvZZPe3a5cSDDhc/PW8+S8duXCPHvPTiaOLK2MXn7KeIiEjLojK0IQMAAICKIh6K8eLhBOV/G/La5QTVLn/17WXziCh6Ztu6MxZ+WnWSA7KGrT0uR0Z5WNDb2OfqCaN4DAAAAIwjvMdx5BRHRLS8r/NxR6FRn+Y9vXng6dQH2/3WsQ/jRMYOyMjuG+EWeBmGSc/xDHHajpwEAAAAUISjLdzZNv2RvfLODaL7iC8PW7BQk3irqHIb9WeabuLPcRzajgEAAIDB2dgH3xo5EiWim+eOJZZ/I6P+8ty2rm2v3Ux+iN6IJZv/a0ByCVnEX+e4sANBGAAAAJiIbdmu3kTLY11v8P2QrSIiuvba/ufCdQO7HEt2PXVt3b6W7xIR0cPLX31bTuf7ZZEckIUvhJRMBwAAAIAmkj3vT5v39OaB5KSqvb8WILnKsmFrj6vvZX/+1xgAAAAAoCSSA7Lg7vYQhdodGFwcAAAAQB7p3V5gcHEAAAAARch7yxIAAAAAZJMVkEX8del6yoi/jqlDkzIAAACAskkPyIJexnFofZgb8hARkd13oIfaHWhCBgAAAFAm6Y36D/eRZ4fPPj3H7tvhob7D5URko91ut9vtdvsG4/lWiQ/63N2jUlMJAAAAYHw6tiGLD/p20vZAIBDYu+Jkq2jQNdrtbu2Pap4yAAAAAC0p2A8ZP7j41pK77p+IRetdTiIiy6MrbMOhrIhstNvt3jm+cftGm9QkAgAAAJiC/MHF24mIGpk+IldPeERYh1lYPDZus67npy3WGjoZi5PTMr3c2REIdBDFB0/mbLp48WJ+4vz585LTXyKWLTQq/EyAHCBkAhEhE5ADRGSeTFA1nWbJBFUhExQnPSAjktcZ2UQsSlZpm6bjsEQiIfHopWFZVu1DGBxygJAJRIRMQA4QkRkyIR0jqJdO42eCBpAJatCvDVm1FXWRAAAAACTnLUsvI6/fMYu1Jhqb4KfjsXGqsVoKbwAAAABQmWQ16k8PZSmp97Fqq2344GCciOKnT6ba9wMAAMjF/vK63kkAKI/0Kku7b4TjOI7jhjzU18gHZmUVmVma/dtr+lvdbnfryRV7O5xE6HUMAAAAZiJ5jfqJSNCyP+hlGh3eBVxvyT1fJN+knGZp9gcyV8mdAwAAAFBZFAjIgl6msY+f9AyVEY0BAAAAAJHMsSx5jX2unjBfe1mZ0Rgbw2ABAAA6YE/+NTv8tt6pANCCvLEsh/g4rIzuYE2K/adreicBAAAAKtsz+3oAACAASURBVJbkKsuGXo5TMiEAAAAAM5XUErKIv44RktUlmSlc+gWGiQAA0AJ74T32A7xwDzOLlIAs6GUYx6H1yXZjHMdxXHj9IYfE7sgAAAAAZrjyA7Kgt7HPM5TVbszuG+GGPH2NCMkAAADABK69tr9lW1hkwfGjLY6uFnvXttduapiccgOyiP/lPlfPVrG3KRu29rj6Xq74qksAAAAwuXPbup77rtgA6dEz27y0Jdw5EHnKdWzfnuOapajcgCx8IUS1C8RfqrQvqKXQBbFgEwAAAMAQwnscXa9cWr7lO2Jtw6OJK48vXEJEdN+S1eypdzSLaqT3QwYAACDZtX/Gm1Kglq4We1dL3h65HG3hzoG3l1nEll27nHjQcT8/PW8+S5duaNXtVbndXjgWuejQpQg1iHzPyKUxcq13KJIuAACoROzVj6jqM/SZu69RkYCMPf03yYm/fZOfSHztCXUTB5Wic0Bq+6l4OEH6BDLllpDZfTs8ofbdYm33g7vbQ54dFd9HLAAASMReQTNj0EF4j4MvM+sq2ibM4tCr5Lb8jmEbeoc8TCMz1hMWvGgZ8dc52qknPFKRYycBAACAaTnawp1tpa06bz575Z0bRPcR0bXLCVqwcJ6qSZsmpQ1ZQ2+q47E0x4UdM2EEJQAAACIiYv/xvN5JMDr26mUTlona2AffGjkSJaKb544lln9Ds/pLqUMn2X0jnE/RlJhA+EORgkzHQrHXZgEAoHzsL/+JGEbvVMDMc+21/c+F6wZ2Oci2bFdvouWxrjc4evA7T+1apVkSJI9lCQAAJsNeeC+x6BGtDzoxrvERAUoy7+nNA2LTtKppINykeXLQ7QUAKO/EEHo0gOLY8TAbi6b/iq9/7IAGqQL5Tr0zfQdgw2M6psREzF1CxrKq3/RZlqWCtwkN0qCviv+CpUAmkCATTv4dEdGKPyp1/YpRGd9I5rcoafOfvyd957/8JwUSUOaa5UrvuTIuCWkKfPe3XyfmDmJZlq5eLroypJk7IEsk1G2/xbJsIpEofB2pnQZ98TmgdyqkeP88e8edVPsVBRJv3kxQUGYmsFT8yq+0TDP1ZTDd/Jxh6O+D9Lv/aXrZb3+Tseqdd9JvphIPfF58P6VlguRnb9H7LRW78ISbq3S+0pnAlnwINvrh9IdPP0lO3FFFd3+O7rgzMbda+VSqjGVZ+ulZ+uTfE45a0eVEdPhHtPYLqRl/N0R3fXaM/drv32/WH5EGzB2QGUEknH0D+eTfiYi++GVcdlBpxq/g39wZgb0+QcwdxP32ry/8ARH97pyMhfel+jdPP1zZxCT99jc0NaVF2no6E+1dyu7z3bdYhqE/Xqf8TXvyf7JfSJygO+4ssE6yjR1zBxHR1H/Qb6bors8SEX36SWL+Q4onSZp//p/Tv/07q+jGv9FCIvrtFHvp/cSCL7EfXUh8fpF+qasQCMgAtBa9nLy12eaLPACuXGH/7V/ojioiooe+qHNYz46Hieindz56dYJ+8ynLCBqdfhRhP2/Hfx36YC8kS7wSixZPz/zFT+mfY4k/XMOOnSWizHcVK+q9RXb7k4mdryen/4//mvhv/z05HXyd7hR5qLGRMSImYf+idkmsCGM/Y++4kxiG5tQUWo396EJ6+r0Rduo/Cq18dYJ9oBr3DXEIyNTy09MsEX310dIKtCfG6Y47iTjhzMS8B1RJGejnwwssEX1mVvLjlavT/3R+8m/Jic/+jtap4rET4zT1H4maBUTExi7Tb7mim1QM9vJF+uTfE1/4qt4JKY798H369DbdkQyN2V/8NHuFE0fovrmap0sZ7MWf0eySEs92rifrg1RVRUTsq88lnntV5aQJjL3PTsYSDzu1O2J+ed+EyNd1CMfxRXGJ6gdVSxRIhIBMXeNX2JoHM2Iy9vIviIg+czcREccRpUqqQQl/fywZ4sz+PX0TYjJsLMrfwdnoJbqDKVzDkivxMfvQf3z44ayF938uQakXrN7ez67bnPjh99jfm0vuJxNENPITlojqHjPo/8fs5YuFaogujCWvrd/+NjmHE8Ssd9xJ//YvVLhwIEtW6H1nFVV9JjnNMET0Pj1MRF/61cnE55MtcdgPf15hZV2qYiNjVPXZdCO5ZOzCRyqf/Rw/MzFnjvjGSrg+yaZPqSn89f+Y/hfxoSXlbJnZEvHxjQwRvf0/OOYRJ3f+tDKJmwEQkBXHfHUp98HPJW/+iw9YomQ5umvOxdI3ZBOTuTMvfvYL7L0GfZ5BYXzx2Izy1l8lv/Kxg+zs1IPvaox9wIpruAzslY+I+y3dfa/eCTEu9vXdRETs75e7IX9/ln7c6IfEccI4fuKaoK3VXXL2TZTbYYRy1wD7y38iosTvlZ1jeXf40S/OV61Qam8zEwIyML1zJ9glKxNob66XxK/Kznn+ZHG/oSWf+yA5K13sVADDEFHi97UaWC6rPuizd2t2XLNgr08k/9css0i1ArD/eK4CSivZ6xNmfMezUiEgM59//hf2P/2vKGCY9tvf0M/OsKW1PAGFRaPsvYI6n1/eZKP/oF9qysRe/Yh+k/FiIFusEyyoYD8ZZIkyXhX91b8SEf3O54iImr6Nuy6oCwEZaOenp9nfuZeI6JNf06ef0Gc+Szf+ie68k+bZyb5Ays3u52eVLBWbuC6+t4nrRMRWz8XtGGYQ5uEvc+//TO9USMTUreLG3ldwh4XfcJn8VxTPgwIQkJWEefjLIycM9NIZe+MGESXuv1/vhJTqxBA763NUoH3rhxfYhYsQ8WQ7d4Ilok8FLcVdfyw9l9joJboreQ4S1TWyUqYf9nr8/Tu/zE+nr6ipT+hffkmLvpSROWf/fvox2fjFXxARcRxx3Ie/80UiWvivZo02QBc//n/YTz6hz3+BVn/hmt5pUdK1f2b/4bgqe173v5m+SldjCMhM7KNI8nnzmykKZ7/8Pm31Bp0DnRNBJf99fO8Uy6Taq9xZrOHKeyMs/z4r38SFldGRyAfvTX+LzwhaE5UYR37wUzbd24URsFc/mv4g1nVTKT76iL0DrwgDQH7M4ke50LDeqTAHBGRgFBd+zhLRr39Fn5lFDz9imtKywmV7Z46zRDTrHskxT7bjgWRcyP2WvrZW5Ljpl7yE5ZF8/wwsXVImEQAAJWNc9dzwMb1TYQIIyCrfuz/OW0BV2kgnGZvn626wgN/B2/rqGHqDJaLGbyWI6L0Rdt4CvRMEAABSISArVd1KJnTKQM3IKtvPTrPVX6TffJoc0i1xNTl/6pMCG+nm5N+w3G9pZWMi9P+xfEnYf/yaVjRkF18t/0Pm1N9PX0JLncz59yReUX/2XxgiOvB9joiC/y+7sY35m8MmvjiZ2i8VXYd7f/rtzS9/Mfvfgp+Pmfjrg7KYTVv0TgKAFAjIAHSw1Cm9uSsfjaVtbGOI6E/WMu9fMEFQsvbbycSfO19eapkv/YGc/pkrBrP8awWWcoOvJVerX11otZ+dL+lYX/qK8OPf/60JLjDQXdN6tOWXCAFZGVzLy77Oyn3qwEygYDRmCv/lOWa/n1uzISPlSxYzZ0c5Kq14rEQf/JT91xvTH+8y1ag1WmK+khySvMTIjPeHX0uewQ8uKnBbY76QHA+K++lZ+XsDpTzVkTzLuTVCX7Rn/IS590Y1StOMgYBMN8xXltCMvxmtf4ohotMjyoetX2tghn9i9Gh48SPl1Vqu+ZY5ojH+dfe+3RwR/bfdDBFt9omkfKmTKauzKObhL6OQTFnTkVk5D9eHH0qeTe79nzFffDg5/Y8f5N9CcMTFj5aTQOmY2i9xp/FyX0k2/UWhG8vKVeJLmUec3Kl3s2b+yVpz3KOMCQGZupYsLnJ1Ml9dKmG33MUyxsQ0poZ10znzaF1G4yplrfrG9IHGLqkYog3/NcsPrsMPMH3sIPu791NVzmB2y/9Q4g0rNxrb9Kwy9z7mDxYTEfezc4VXW7o0+3BnznDpmdzFiw/VZKzg2Vo8eQoWj80EzPI/VmvPjzilbCWo00xHZsbBPFrPnQvpnQoiImbpctH53D+cp9Q/50TE/cN72qVJJ498hZmcFBmmGQgBmUkxDz1UfCVjeP3/no6BvuFmTg4rHxLla7JQ/5jO/6vl9ouYG40tfqR4Io+8wZVeNvaVL6fqlX5RdlYzS+qIiDsv8gx75CuG+MeXefjLwtb9Cuww9aTkTryj4G7lYB7+cu7MEgMLpvlppZNTOaT99yvkf1HdQnfmDx7hzp9W9RDJA6VKRguQ0ESnqHTZKohCQAbqevJ/z/gFrqgX/0GKlhuJNlX5k3XSf9K1C7S4HRz+EUd5KumkKT0a+9Ki6TUf/gJDgrDsS4sY7uxI+n/xXOlFzGIX99OSKrCWLZs+nIn+SRDFrPyG3kkohFni4sZKqhM0nXScZPz2G74XUuXBx48yf9TI/V1Q5g75wmmNlRKNlbqr5V/nRv42/bFAjVDht0yAh4AMjMuk/06t/TbzVr8OzdeE0VgaH5aVi/mqlAosbTBf+gO9k6APptZwdYLKYr66lDv5k9yvyR3qk77PJS5u9CQRMY/+kZy0Ze92VRMRMX/UwH+88JFiv/esNnZF+7iXX+wnE1P3NSLiQuqMvjTD6BuQjXa7dw4TkW3jXn+zpbylAMb1+EZThpLm9eXa6Qw/fLDUpyPzyDJ1kgMSMSseE5m53iP8yG3flHfzjc+JzHSukJ+wwhZ9Xq3fO+OqV2nPymJcq/ROgiTXXtv/XLhuYJcjc/bNI+v2vZEukP7mmpwVVKJjQBYf9O2k7YGAk+KDvtZua6DDWfJSACgbs7RO7yRoYe0Gkacj94sL/ATzxVptkwMKY3Ye4Dx/kjHn/3xdr8SAiZ3b1vXKj4m+mbvkRmzsoS2RprztO9SiY0A2EYvWu5xERJZHV9j6Q6MdTmepSwEAysB8YZHeSQDFMH1/o3cSwNzCexxHTtUu3/Kdi6+EcxZGb8Rq2cd1SJV+AVk8Nm6zruenLdYaOhmLk9NS0tLFi5NtEs+fL6NXQ2lYluU4o3dnBQAAAGldLXYi6hyIiC51tIU724iuvSbWg1Q0cWXs4nP2U0REpGVRmX4B2UQsSlZpS9NxWCKRPVygsliWVfsQBoccIGQCESETkANEhEwgImQCEZkhE/KFYsVdu5yg2uWvvr1sHhFFz2xbd8bCT6tOv4Cs2mqTvBQAAACgNOE9jiP8YFDL+zrbCr6CMO/pzQPp/vxs91vHPowTVXhAZrHWREMTRBYiisfGqcZlKXkpAAAAQGmSdZTGdod+h6622oYPDsaJKH76ZKoFf4lLAQAAAJR2blvXttduJj9Eb8QeX1j5bcjI0uzfHnO3uvv5nsacRETxQV9rbEOgwym6FAAAAEBx6T7Jlux66tq6fS3fJSKih5e/+rY2nZAREWPqVwjRqF9tyAFCJhARMgE5QETIBCJCJhCR4TNhf9sy6Y369aNjlSUAAAAAECEgAwAAANAdAjIAAAAAnZm7DZnaFi9erMFgAGBwuAyAcBlACq4EUAlKyAAAAAB0hoAMAAAAQGeosgQAAADQGUrIAAAAAHSGgAwAAABAZwjIAAAAAHSGgAwAAABAZwjIAAAAAHSGgAwAAABAZwjIAAAAAHSGgAwAAABAZwjIAAAAAHSGgAwAAABAZwjIAAAAAHSGgAwAAABAZwjIAAAAAHSGgAwAAABAZ1V6J0CWRCKh6v5ZllX7EAaHHCBkAhEhE5ADRGSGTPh0c6Pw4137hxQ/hPEzQQMGz4RPNzfOPXpW71SUDSVkAAAAADrTNiCLD/rc3aMiC0a73W632+32DcaLzAQAAACQZKJ/sunF2xkfl15vWnq9aektsehESxoGZKPd7tb+qMiC+KBvJ20PBAKBvStOtiYDNtGZAAAAAJKMvnj9mT1T059P3Hrm3Vk/ODv36Nm5P2ibeunJjyf0S5tmAdlot9u9c3zj9o02kYUTsWi9y0lEZHl0hW04NJp3JgAAAEDZbr+y9PpLkXueb5tuOz86fLv+z++tJiKi6o331n84pWt9nEaN+p0dgUAHUXzwZO6yeGzcZl3PT1usNXQyFicnic20JLdYvHgxP3H+/Hm1U86yrNqHMDjkACETiAiZgBwgIsNnwvXMjyql1uCZoA0jZ8J1outNS4lI2LR/1pazc7cQTfRP11c6X5jrTH848evhhVV/pmUysxngLcuJWJSsJc1MScdheMtSbcgBQiYQETIBOUBEJswENVJrukxQg/Ezocy3LK9+/Gzn1KbB2dVqpacUBnjLstoqUo8pOhMAAABASSduNTXfXjE454kH9E2HAQIyi7UmGks2pIvHxqnGaskzEwAAAEApE/2TTZ30/FndozEyREBG1Vbb8MHBOBHFT59MNeUXnQkAAACghIn+yWf2VD1/drYxIgz92pDFB32tsQ2BDidZmv3bY+5Wdz+RbeNeP58xojMBAAAAFDB1+t0poqmXlqbfBZlVanB29eNnla/l1DYgszT7A2LTqbcwM4nOBAAAAJCkeuOco8nJqiden/uEronJZIQqSwAAALmyBrIEMBcEZAAAUDnu2j+kxpjiAGozQD9kAAAAAKbxwL3fP3uv0jtFCRkAAACAzuSUkAW9TCMNcb0NQS/T2EdE5BniehsUShkAAACAYZy41dR5u9AKq2cffWGW1L1LD8iC3sY+zxDXQBH/y33kGeJ6Hf46hzeoZUimwVhaRh6uSxvIAUImEBEyATlARMbOBL7zAj6FwmnFGTkTNGPkTLhefBXpSu4ao2ySA7Lg4T7yDDUQRY4dCpFnRwMRLail9sPB3gbNIjKMZak25AAhE4gImYAcICKTZIIwhRjLUiXIBDXIb0MWvhAi1yKHAmkBAAAAmJEkl5A5Frno0KUI0eE+cvWstpOg0AwAAAAASic5ILP7DvQccjgYIvIM+ewU9DKNYz3hEcRjAAAAAOWR8Zal3TfC+dKfGno5ToH0AAAAAMw46BgWAAAAoKiVs4+eVW/vchr1B70M4w0mJxiGSX4CAAAAqERTbz55vWlp+m/yzatK7dnc/ZABAAAAaGPqzScnD9hnH3091fvriVtNzZM0OOeJB+TvXLN+yEa73TuHici2ca+/2SJYEB/0tfZHhavWbw90OLPm8/MAAAAAyjfRP/nMnikiGf3pn/j4wIeznn9dsO3K2c+vvv7SX91+QnoH/Wny25CFL4TItb5wP2TxQd9O2h4I8GFWt1UYXFma/YHm1IfRbvdB63onEdFELIooDAAAAGQ7ceuZPVXPn53jpKk3n5x8tn/O9zcarRG9Nv2QTcSi9S4nEZHl0RW2/tBoh1Ms0Brt3jm+cW+HhYgoHhu3WddLTR0AAEClmzNnjt5JKM/k5KQ+B54Yn5rfxg95VPXEn8868MNfT2y8t7rcvay8d9PCyZdevD1dwHbi1kvHqjYNyi8eI436IRMGVxZrDZ2MxclpyVlr8OBw/YZAcv5ELBodbnX3E1FOheXixYv5ifPnz0tNf6mMPFyXNpADhEwgImQCcoCIjJ0Jhh3LkmEYNZJhRtpcP9eJrjctJaK5Sr8UWfXE63PoycmmpdNzNinTgIw06odsIhYla7HdjR7qp417U2FXPDY+3dwsPujzDVZPNz1Lx2EYy1JtyAFCJhARMgE5QEQmyQSVxrIUlkWVWMyjTfGVLmVOkq8Eza6f7FCsuqbqcufHoxtnO2nqzR/eJrpH6p6rnnh97hOy0ye6Z1X2mqXaaiu6zmho2LZib7rYLKNlmcVaEw1NEOUUqgEAAAjxYdC11UtItZBIwm51q6qDpJWzf9A2+czS60RVm7rumf9DvdMjQlZAFvHXOdpD6Y+unvCIzy6ynjCgisfGqcaVE1plxWMAAFB5VC004oMw9fBBVVlfAXGYgVRvnHN0IxERnbh1wH53WQ3IJvonn9lDfO3k6IvXXzqWXjLr+bOzFXr7UHpAFvHXOdprh7h0q7Ggl3HUkWhMVm21DR8cXO9stsRPn4zWb8hJfDw2bluxXhCPjXa7D1pTHWTEY+P1rg7JKQUAAN1o3PCcj4E+3dxI6sRDHMcZv94Wcpy41fTDqh+8fm81Tb35w9v1fz67jG2vfvy9ZDSW6ofsrKAfsqWTCjUjk94P2e72kGdI2Ia/oXfIwzTuDvpye4a1NPu3x9yt7n6+HzInEd//WGxDsrF+TiszZ8femC/Vpj+9DQAAmEC+IEzVEiM+CAMQs3L288PXn1n6KyKa3zbn+yvL2XZ86jJVffsB4/dDViJnRyCQUchlafYHhEuzI66MVmQAAKAhBYu1UG0HhuB8Ye7RF/RORCGSx7Js2Nrj6msUDl4Z9Db2uXq2YuAkAACTmpMif1eTKfJ3BaCrlXfX0+0f9U/Ryns3Lbz90ou3pxfx/ZD9Z337ISO7byRMdQ5B9yp5G/UDAIBhlBhv5cZSpuj2AkBxs7acnf3K0smmPURE9OGtpmMZiw8YoMoysycyAAAwBMlFXCjQAhA1a8vZuVvUPYTRhnICAACJSo/DEHgBGE25AVnQyzT2FVrBM8TlvmUJAAAqyg3FEHIBKG7qzScnD3yYNVOp0ZMYLv+IRwAAYGS5gyTO5Fu6cPhClYYyBFO43rRU4VM/0T/5zJ6pPLFXKkpbPfuorJZkkkrIDNN6H2NZqg05QMgEIkImGCkHCnfxpWoijZMJBag0lmWaKTJBbTMsE07ceiZ673RnsNlSo1ueuNX0IsmIycrt9qKhl+O4A7SJSarzR6QeGgAAyiNaNYneJQDUtLK0oq8SV8tLUj9kdt8Il4TYDABAY5MCeqfFoO7aP0Toux9MRe5bloKeLyL+OoYJoVE/AIBiNB4IEgBynLjV1Hm70ApyW4/xFOz2Ar2SAQBIhMALwKhWzk69IpAaXDwVfk30Tz6zp+p5BaIxkjF0UrJELIdwMCUAACii6FBFkzk0SxsACPCDiwvCr+qNc55fnTmYknSSS8gi/k3tqJ8EAJAsKw5DpAUwk0kuIQtfCJFnLaIxAAAphNEYyr0AzMCgg4s7FrnoghIpAACYIdCfPoCpVT3x+twHXuS7HebNev7sHKdCO5e6od23w8M0eteWWmc52u3eOUxEto17/c2WzGXxQV9rfzT1qX57oMNZbBMAAOMr0DgMoRiAtq5+/Gzzry4T0cJ7fvD6vdVSd+N8Ye7RFxRMVprkgCzobewjokYma2RL0WZl8UHfTtoeCDgpPuhr7bYmI66UiVg0HYWVuAkAgKGU8pokgjAAvdx+pXnq22fnOolGX7z+vf67v7+x5ADoxK2m4buV6NiiMMltyBp6OVGiBWYTsWi9y0lEZHl0hW04NJqxNB4bt1mzY9XCmwAAGEWB1yTxdiSAMZz49fDqu/mCHecLc8uIxrSjSZLisXGbdT0/bbHW0MlYnJzTdZATsWh0uNXdT0TpCsuCmyxevJifOH/+vNppZ1lW7UMYHHKAkAlEhEzIkwPC4b1nwsDeRr4MrhORIIVZHxVk5EzQjJEz4Xru6PIT41PzbZ++svTWMMmtslSNlMHFaYjrbQh6mcY+kRXEqiwnYlGy5t1lPDY+3UwsPujzDVb7mwtuko7DMLi42pADhEwgImSCWA5kvSZJ6t+RdGeKyyArhYon2BSZoDbjZ8J0KJZ2ec/Ut8/O3UI00T9ZXpWlRspNT0Nv8l/Ahl6O6y1xo2qrrcBSS7M/0Jz+YK2JhiaKbQIAoKvcaAwADG1+2718lWV1TdXlH/56YqPRCsk0CRBTUZaF+PKwGlfRdyYlbAIAoKii7fQRigGYQ3VNFY3rnYgitBk6qdpqGz44GCei+OmTqcb6KaPdbt9gPPkhHhvnFxfcBABANXNSCq+GaAzAPFbe/eCej/nXAyfGp+Z//W6DFY+RzKGTqCfM+chf57iwI9mqjIZE37K0NPu3x9yt7n6+UzEnEd/5WGxDoMPp7Ngb86Xa9KcXi20CAKAa0QgsHXUZv9EMABQwa0vXr5uWXiciWj376AtGa0BGMgKy8IUQ1e6wE9GCWjp0KUINDb1DHuZl/9YGn11kfWdHINAhnGFp9gcEk80lbAIAoKgC3VVonBIAUNnK2bkt/UvdcKXCaRGjQIzoWOQKXQgT2YmIpqcAAAwNAxkBQDlO3GrqvC0yf/VsJbqNlRyQNaz1UOPhYG9Dg31BLbUfDvY2OC6NkWu9Q3aaAABUgiAMAKTKLWM7cavph1U/UKYTf+mN+ht6wz1jjXX+CDX0Dnn6GhnG0U49B0TrKwEA9JSvnT6iMQCQYeXs5+2/+l7/lBL7klNlafeNcD4iKq9TMgAA7SAIAwB1XY5OKdECzIDvGQAAyIU4DAA0MHU1QvO/rkgsJXkn+YZOIiLx8ZPUoMFYWkYerksbyAFCJhCRsTNBOKBkLqWGmDRyDmjGyJmAsSy1ZORMuK7WjkUb9S+85wfKjMIkvVH/1h5XX3utIPIKepnGsZ7wiI/8dQ7Gq0VMhrEs1YYcIGQCERk7E4p2XaFIyo2cA5oxRSZgLEsNzNRMEG3U3zkVJ1Kim1nJjfqDu9tDnoxuYBt6hzyh9t1Bsvt2eKjvsFiX/QAAihG205/MoW/aAGAGWDn7+dW3f6R/o34AAB0UHdQIAEA7ejfqb9ja43I0etcKqywb+1w94QaiCDokAwBFoUt9ADCg26eO0fw2fduQkd03wi3wCtvSeoa4kQaK+Osc7dQTRodkACARBvYGAOMxaKN+IhLvf2y6dzIAmPGUql5EBAYAepM8GmZJ0IZMeXo1cKmwJ5YB2wkpm8OKfEEFk2SoDK+wixl09OnmxtyZd+0fylonaw6A9hCQlcRQz6p8TJFIUzNgDhswSbkQXYEu7to/JBqNkVgEhpgMSjT15pOTBz5Mf5z1/NnZTmX2rFlANtrt3jlMRLaNe/3NlqyF8UFfa3+UiIjqtwc6nFnzhLNNQK/HjymezWUxyIM83eOOsjks/9spfsYLJGmmdjsE5pYvxvp0c2M6AksHbYjJtLi8MQAAIABJREFUKtzoi9dfOkZERKtnH5U4HPjUm09OHrDPPvr6rNEXr79Es4/+50+fXTp5dXDOEw/IT6A2AVl80LeTtgcCTooP+lq7rZnB1Wh368kVewN+Cx+F+Qb3+pstNBGLGicKM0hkUJgaicRjWMhol4HR0gNgOnw0VqAsDSrERP/kSzT76NlZfFD1yom5W1aWv5cTHx/4cNbzrwuCuQfu/fbqX730V7efkBjhCUnuGDbir2O8pfb9OhGL1rucRESWR1fYhkOjwoWjoeH6DclCM0vzhvpobIKI4rFxm7VY17f8fznqzb/etFTV/WM+5mM+5pto/vWmpYZKjyLzKU87MzwXKmN+UvXGOalSsapHv141PJzzsqT+JJeQhS+ESl43Hhu3Wdfz0xZrDZ2MxcmZrrZ0dgSmi8FGQ8P8qhOxaHS41d1PRDkVlosXL+Yn/vr37yCxEbXyDWGG+RLmsyxrqPToMp+fNk56dJlfdIjAyp6fu8iY6ZzJ84UjGBZen5/OfX7n7s0I38uY86m09bOCeG3SeT113LliL0VOnX53qv7PJRVorby7nm79qH/Kme7n4urHPzpG9V3yi8eIiJE87G7EX+c4tD48UkJ/Y6Pd7oPWVNOxjA+Z4oO+1pMr9vqbLcJJovig73v0lyKbYCxLtSEHqFIy4dNU1Yy0zSsjE+RADpDhM6H0izwdiqVXLn1bg2eCNkrMhNyQV5tWep9ubhQNxWiif/KZPVO08J4fvH6vxNEnr378bPPtFYNzHvgrvkVa1SZlGpCRjBKy4O72EFHIwbRnzveIDCpebbUV3+Fot3vneLrFv6XZH2hOLbJYa6KhCSKRGA4AyvApmi0DpOC3oI2slyf0VL1xztGNRCduNT35scSY7IF7v3/2XiKiF+YefUHZ1EluQ9bQy4nKicaID6hiE/x0PDZONdas0Co+6HPvpO0B0XIzAAAABd21fygrGjNQ0ACqW3l3/YdTcYV2Nvri9aYXFWmRps1bltVW2/DBwfXOZkv89Mlo/YaMVyfjg77W/prt2W9eCio247HxeleHJikFAACACjP64vUf2eZ8n2/7deLXwwur/kzKbm6/svTWsMj8W03HSHafZOUGZEEv09hXaAWxKkuyNPu3x9yt7n6+H7J0P2OxDYGO6tMno0TRne70d6zfHuhwduyN+VJt+tPbAAAAgPnoXEHsfGHO1Scnm/YQEUlvQ3bi18M5UVeyQ7IXZhFNTVwlkt6erNyATGz0SiJKRWqunq0iVZZERM6OQCCjkMvS7A8QEZGwtVjWcrH5AKaQ22q48uS2g0YbNTC7ArWWuNRNreqJ1+c+ofIhqmW17pfchkwo6GWYxj5XT5gr5Z1LgIonvKfzveMU6SPHhNJfJ2uiwr4mQBou9Rlv5eyjOZWSzhfmSu33P4vsNmSpkrGS+r8AmElymwlX3n/VfBfnlfe9ZjiZnaSYVOHvi/ALiE7cauos2H5f+qBMJDMgSzYoE202BgAAZoY4GyDTytnZ3ZuduNXUSQqNLy5r6KTGPvIMifd0UUkqr7IJAAAAZFs5+/nVt3/UP6XEviQFZBF/HeNoD82EYAwAAADKNoMKMi5HFQnIyq6yjPjrHO0hg1RT5htdS9lDFB7Jq+IZ54sXGJtMbeVmQsY1c/SscDQ3vfIz34CMpcs3kmPWRHK6aakuZ6osokNl50p/EVXPnaFyrMDVYpwbQi5V79Xp4S8Lj5g5cxT+7ukfly5X9fXiqyjm9qljNL9NkT5dy+6HbHd7iIior5ER745M00hNy7EsZ+bgZQYcte1601KN27VIzgTRrTTLz6x/T/nW95ITkC8TRH8g/L1Y+zNFkkYzLCr9n4B6545PjC45VljWVzbgDSGXSilM/4KEHw14yrRR4pVw1/4h418w5RBt1L/wnh9s1CUgy9sPGYCqCjw+8y0y5o2ylP7JSokVsrpEKv24iku/aCl6UF3OQunHLbpa+nup9OjNyjcNcmxmvkGpCD7TTBGVgjpyG/UrSZuhkwBkST9Ccp/6RQM1wz54RB+9pYdNpTy50yvMoMYcKUpFNsK+SwpfUTIzOX0gVWOyrJ608h0o/VvDi5YAmkFABuZQOHYRfWYYJwpJP96yirWET7vcGsbC+0xvXkrcOUMGThaWPiob2Qj/GZAZSYvuPOtA6UOoPd5Dvoun4i8VgHKcuNU0fHfxDsZKXC0vBGRgdPnqwtLTJv0PPusZL5wvYVfKpKki8LkhWp4qB9+GrPA+FTwRWWGfsiVVWdGeaNineAYCmNbK2UfpVtPSW0RVmwbnPJEzPtLoi9dfOiazV1hCQFY6yfcmPCzl39az8rCsYgOD17nIDMUk5K200yH5raXcw2l5OlQ6+6pGKsLqQlUPKjqSRO5qwpm5l0G52Vv0W5S+Q4PEi4WvMYM3nMinaN7KvxLMJtl6bOrNJ683fZi9sL5r7tEX5B9DkbEsK5zM68wgdw3zyvqvXXS6MM269i33QNK+ToGdaCz30Ok5Ml9ZkCBf5GfGH2BW7ql6iu/aP8T/STtiWdmr4NsnRjitRdtoihaBG5zku2VuZF+RIVrVE6/PPXo2+2/LSkV2znAcp8iOdKFltxeSqfEevuno+MvUK1e1eVdOgwPxlHqzTL07deHmUKXPzwfv1pFYJpTV9rFonpd7t9TlxpKVCQVSYq6WFWWdysKZoHtA9unmRuP06lc6lJCprsR/0xGNqXd0zRJQSvmQSkec4Qr8fIRvLwr/tEtcRcvXnCBXpea58ALTOy0SZSVewj3TvKXRhqJZG7LRbvfOYSKybdzrb7aUtLTwJmZSevNYoz1fK6ZUQE7GlpUJpgjF5syZIzp/cnJS7fWzOi3TpsEQ2qerqsC7w+mZWqdJJ7qXDElgrmI8eVLdus5vm/N9ZbpyVZY2SYoP+nbS9kDASfFBX2u3NdDhLLa08CamZPDW5VA6HQMaRdY3mtz0X1u9pNydzDt2ziz5X8Hr5+u2gzfv2Dl+e2XTozvRt1aNf7efSaEYEV39+NlOev7sXCdNvfnk5Cs1SjX8UpA2AdlELFrvchIRWR5dYesPjXY4nUWWFt5Ea4rcsK6tXlL0P8WsDQ14w8X6JjJnzhzRr1Du95qcnCyrmLDw/oUNTfhfRCnhl/CBke98gRHcJTZyQzIUK02+60H0/mmcm8Nd+4fmzJmTTrzaL8kqKP3jkpM5/Bcv/DXz3ZE0MT51efXdTiKiqke/XnVg+PaWlXK6qFCDJgFZPDZus67npy3WGjoZi5PTUnApFdpk8eLF/MT58+fVTnvh8VPLGll23rFz5f7fX+7ItVhfpfWzNiz3VRijrS+N8uMoZw67LlS4Qa7R8nNGrc+ybPH1BWeWP5WlHiD/JZGr8EWi182EDz2Ft3oJd37NlNjyXfTLFs6xfHvWZjj264IhaJOzJsan5tuSEU91TRW9++kEzarWIC1l0CQgm4hFyVre0oKbpOMwzd6yzBfU50uA5H8CsjYrd/+Kr59VLqJ7enRZv2Ia0smhUibkqygxYIbjMqByMiF9ZsvNtNLrzpS9OZe+Pp8JouuLF0iL7tcYsr5F6ZnDZ0KBJ12+/NHsR5QdEcajU2TT5tCSafKWZbW1UDaILi28CQAAAECJLDYDtuLPoklAZrHWRGMT/HQ8Nk41VkuxpYU3AQAAAChRdU3V5egUPz0xPkX2uwxWX0la9UNWbbUNHxyME1H89MlUY/3CSwtvAgAAAFCimqr5xz5+8yoRTZ1+d6q+3mgt+knDnvpzOhWLD/paYxuSvVkYtB+yxYsXa/DeABgcLgMgXAaQgivBtIzeD5m5h05SG354QLgMgIhwGUAKrgRQCYZOAgAAANAZSsgAAAAAdIYSMgAAAACdISADAAAA0BkCMgAAAACdISADAAAA0JkBe+JQUOGezGT2c2bQvtMgS3zQ19ofTX2q357s+i5N4csg83DihwTdZHR/KKTC3WD6UsAlYDx5roRit4uicCWADFzFmnjz2T999UzmVKlLpe1c5j5BFWdeLXAq1LgMMo/97JsTZe4U1HLm1T/90z/9U5HzpMZlMH3uJ958FpeBseS9EgrfLorClQCyVHCV5UQsNeKS5dEVtuHQaElLR7vdvO7M9Ue7M+aIbl74iKCLeGzcZs07Zpkal4Fg5Z3jG/8SBaWGMNrtdu8c37h9o01koQqXQTw2blvxqIWIyNK8oT49Mi/oreCVkO92gSsBtFC5AZnwl2Wx1tB4LF586Wj3zvGNewOBwN6N4zu78wdUopsXPiLoYyIWjfa3it9NVbkM0gsHDw7Xb0A4ZhDOjkAg4G8WDc3VuAws1proydNxIv5KKPBPAWir0JWQ53aBKwG0UbltyCZiUbKWuXQ0NGxbsddC/L8y7tBoh9OZahJARMPuVNsAq9jmhY8IuojHxjPGT/UNVguaCKlxGaR2cqifNu5FcxEzUOUycHYEqNvtdhNalJqH+O0CVwJopHIDsmqrWIl00aXR/lZ3f3LaZo2T09kRCHQQjXa7Q650e8y42OaFjwi6sDT7A83pD9aaaGiCKH1HVOMy4E3fw8Hw1LgM4oO+1pMr9gYCFr6+qxutuU0g7+0CVwJooXIDMuGvKR4bpxqXpZSlpb0FI755wSOCAalyGRAhHjMXNS6D0VjUtuIv+f04XfV0MBYnJy4Is8KVAFqo3DZkVG21DR8cjBNR/PTJVFPLwkudrvrkTBrtdvsGBS2CnB0ZP0jRzQsfEfSQcR7jsfGss6LGZUAZ7XjBBFS4DKqttlTLIRoNDVONFZeD8YnfLnAlgEYqe3DxnC5hMvqeKdiRWPHafvRDZg6CjoWEfYWpfRkctOIKMCDhqVf7MphuZoTbgfHkuRJEbhdEuBJAG5UdkAEAAACYQAVXWQIAAACYAwIyAAAAAJ0hIAMAAADQGQIyAAAAAJ0hIAMAAADQGQIyAAAAAJ0hIAOYQSL+OmZanT8iXCD4WNnEv6z6WRAJBiMaHQoATAcBGcAMEfHXMY5D68NcUriH2h2MN6h3umaIiL/O8fIlftruG+FGfHZ9EwQAxoKADGBGCHod7dQTFoQBdt9IuMfV9zJKagAA9IeADGAmCB7uI8+O7EKZ7JKaY950dWay5Czir2O8/mRFJz9PUO2ZLl7j6+CC00vq/BHRFcU3p+zK1CLrl364vHW0ZcmT5tK+SzJtjvYQhdod6aSm01LSF0RBJkDFQ0AGMANELo2Ra5Gj8Eqh9kOLwsnaTFdfYzoG6Gu/sIPjOI7rbeAji9qh5FpjjYJIIdT+Mh1IbR5qdzCbpj+m9pZn84i/LrMyNbVE5uGC3vTWydV2lx/Y5EuDcP6QZzq/xA7Kl0aSqyecVVVZ6hcUnA4AqFAcAFS8cI+LPENFVhCskYoesucPeYSfpj+KbC62t0KbiyVP5uHC4bBwbxlfytWTsYzLN7dgGkRWL+mgJWWI2DcCgMqFEjKAGWLsUpH6unxFaJnzhZ8ci1zTu83aPM/eRDe3+3Z4+hpFK+fkHM5utxNRMFUP62gPiX6/okTTEL4QotoFuQ3zyz1oqV8QACocAjKAGcC+oJZCF8K5C4JeY/S/0NDL/4s4lArMFKmhC3oZhmkcE5RLaUCXgwKA6SEgA5gJGraKvVAZ8b/cJ1rMU0Do0LH0XvKVEknfnA/MhjzUdzgo+3DBw33CVlvhCxJLyETTkFGaJeOgMvMTACoFAjKAGcHuO9BD7Y7MvmAd7SHPUG9D6Xtp2CpoGR/xv9xHnrUKbB70Zr5YmXoFQebhSBDuRPx1jX1lbVokzfbV64Xz6wTfQPSgooWU8r8gAFSKKr0TAADasPtGuAVexsG0p+a4esJcmd2T2n0jYapzMAwREXmGuHLCubybN/RyQ14mOZdPGF/CJO9wDb1DHqYx+YVdPeGhWkfjhTBR/rZZoXZB9qTSkScNeb+L+EHtDWs91NjI9HmGuK3FMgQAZh6G4zi90wAAAAAwo6HKEgAAAEBnCMgAAAAAdIaADAAAAEBnCMgAAAAAdIaADAAAAEBnCMgAAAAAdIaADAAAAEBnCMgAAAAAdIaADAAAAEBnCMgAAAAAdIaADAAAAEBnCMgAAAAAdIaADAAAAEBnCMgUFfHXMQzDeIMZc4NehmHq/BENE6Hc0SL+uuzvI0XQm5UtfFZlpjOZfRkyNgp6NcxHs2IYBf4AAEBbVXonoCKNXYpQgz31KXJpTM/EyBLxb2qvHeIalN/xsUMhj8fT17476OsV7t7VEx7xTeedv87B1KVnNfQOHWY2+VcL1oBsC+7XOwUAAFA2lJApz+OpPXRsuhgncuxQrcejY3pkCO5up56tyodjFNzdHvKs3brIRX2HCxW/2X0jQ55Q++70Og1be0jwEXKx/4sCfwAAoC0EZCpYu1YQkUWOHaJFi4SLhRVzggq5zPq6ZM0cX/0YnF4kr/YwkmdHQW/6sMHpKsqI/+U+1/rV9vSmdX6/NzN9hQl2eylzweE+ci1y2FevLxaRETVs7XEJ1rH7dnj6Xka9ZX6/e48CfwAAoC0EZGpwLKILYX4ycuwQrV+9IL0o4q9ztNcOcRzHcVy4Z6wxGdkEvdOzuXCPS1AqFGp/mQ6k5vc1Sg7JhIcO94w1puKuoJdpHOsJcxzHcTsuNLaHUusfOxRKx2PJlLTTEL85tTuKJCRzt+19wkWH+8i1frWdSorI7AtqaezSdATmWOQKCYsgIdO99yjwBwAA2kJApgb76vVjySgjfIGEMU1wd3vIM5RqNGX37UjWx0UcWzku3ZbKvnq9SxCDeHYkm0xlzS9PxqHtvh2eZCQUPNw3fYSGrT2u1AbhCyGqXZDRWCu1vd13oMdVuJwq724zFpUUkRFRKBXgEh+gCT9Dpns+p8Cf7oLeVDFszvsgeiRF1wQAwIyAgEwV9gW16WhHGNP8/+2dP4jrWn7Hf3rjeS83+3jFNkmVhSDNLsYLCezdLFJ3U0nTuEgGQlimWqmUmiEsTGmS4jZSKaWabBHwpnAzUnVvZxGYQAJrnH0jwW4IBNJe0rzdN1aKcySdo3+Wx5blP78Pl4slnfPTz7LG/up3fuf84ucFgKflI5OaR6YAiKIIzBCflIWpAEAeSjtyi7UkDYm2i58X7H7x+kau7Fvov04VNZgNZh7oY5U5sF6RcdcgdR2p5Dtf7uDfAaG6zLPKWTqAIMhZgIKsG9SxvniOIZh5mfRIkekwXsbcFKkWy4b4IrteFRVhc8+OYU2IYOaxolSyQmhWZMHMK0bqkAb+8Ds7+IcgCILsFxRkHaGOR9NHp6jHxKsRVOY/BTNPtqMkXc4hWoalJnWI5pzXdvWwp04HJIuhrqZTM+3i5wWUtCbrVY3Z2Jl4oPusII3sxhgZF1DjXEcq6VqQZXFcbupJeZYIe7D8wBAY+RyRbBKJkjd85mzlxvm5L0xPbu5L4cmkrQ+VxstDluUrkDfLj+EoJ4IgG4GCrCukIViWVxxuVO9sObRu029x5ncrV0uxo2ge7Bxy6vf5BEoqc9SxDlk6WGDkp1bHenFkMJ1SEDu3Vtikx6hZLZs3QM3Gj9OwKOSaRi1jR9E8mV95ozAaihToVJAFhqB5qaD2R5YkGEH5E8xmbcSOIlmQxX3BkhgFE1rTYZQkZDwwMIR8Wos/sqyqPwFuSkyS+LqXTU2B0NLSQ74eMnNO2vrQZHzdFUjxtAkxF9mypx1DwBpBkIMBBVlXiNc3MvCTFAEARHMe2WBJNIVsQdc8VV1fD+leaXoT+fp2eeupLfY5XjTnkb3Q6EDhyE/zYlSX/K4IgiBMhna+ZJo61vlwnmzboBW716G65DeNNVulx1JFRjUh77k0vYkKcb/y5E+E482XO/hXTexMPNmO0k9edX0i5guKLNNjwXsrlO2HdE5KcVG5/GOkgVPOcAXRkrt9VDdhsrtyx1Q3yuectPah0fi6K5D5QE8kXt/IOPUEQZBNwJX6d4pozhOz4rXqMovdswdyVDdJXGY7SQAAgG9b3bXBiZaHmJMHhpWFn9Sxrk0eYzMXRNduYroVBqrh3pNpAgCoVefP3Vr/9uLHaajfz1GP1fKmswyw+HEawuieufbSUAZvGYFp3uuWNgtcVaV6zL4WIXhPX3DtLdoMmHHnaBmCfs+oH3Wsg1eqcKGOddA0ASqfBthR7DQ3wDSjWVsfGo2vvQIg8tbEq1GdEQRBkCowQnb2sMOmJFLB/EYd3Lr4ndUOOB06jJCVyGWHOmbXUcnjToWIp1WZodi6upjqksHAijythlHsVj40Gq8HhReCIDsCI2Rnj+pGtiJJgkU2dS5AIJoPtiIZQeJW/trFjlL988ZXpNwZgaEt7MjF8FgT7eXU9sTPCwCiSNSxDtoscGHmge5nH36rO2ETVZOHUwND0DRlGM3XRVXb341Vxps7MlfgYPn48WN3xt+9e9edcWTn4M1wyKAgQ5oHQkVzXjV42qJnB3BDv0g13Q1Zitc3sjV9jkFlhxrlGyLWiSIzwAPdV7M9+XBeE/wwIrSLmamur3vaMgKQAOgIJT1TNmoKYmsf6oyzHZuvwAHz9u3bLsw+PT11YRbpFLwZDhYcskSQ06LDIUtSWkLK14PQmHIMRH957CIl6h0/2TAwyktSUMMPbMvAqA68csPrWUlUupXPXub8au1Do/F2V6CGfDWN0gIfpR0FR3H9DAQ5H1CQIchp8Qdf7uBfHezcWbKQMTPArY51AG4WLT+pmKwXUT0IyLXUFrZdNc2SnRDMTlIGAADdvplK+WnyQmQtfWg03vIKVBEY0vQmSivIGgHEzi3d4es07e59tmLHiORsxo5CF9BIfHh9/dqTJg4cBy8MAEAcGE6AS6ycAEJCZ/MhCHIS/Oo/dmDkh3+2AyN7I11C7BALHMWOcgsPBWmXJ1/KdjQ3Id3MhGRgCMxihFumZH78+LG7Uap+0obiwHj/fOd2kKd6rASGMRu76/8CTvBmOCEwhwxBTopfXexAS/1wexNIHYEhaOAniQqBIUwA0lzM2FEkTfBA9xMXitNrjgr6Drdwn5suVNKjgXE7G8/zyT28fM3hLyE/BamjeUe9obp3z4riSCf1ps4OHLJEkJPif3+zg3/IzhCvRtnyyoFBc8TIsG4wo/UrHEUwAlIEzSflMaRhurItPXhOEOVEi/5Gthzy5RACQ/NGpTIhpRrB3Lq+gUFXmc4KMYws6Shq/7ZHNO9HTIUK5Ag5rwjZH/8FfHcAXw7gzQDeDOCLAXw+gC8GcDmAzwdwOYDLAQwu4HIAl5cwuIDBAAYDuBjQF+T1xQAuLmBwmb4ewMUFXAzons8u6M7P0gafDej+wSX84E8Bfv2fkCSQrGC1AkhglQAksFpBsqL76f/kRcLtpF1WAFDVBZiOK0iIWd7CX/0Mpu6aNpl9zrGksUvjZgKcnb//J/i7v01NZW+f6ZhdmRXrZ2mTtFxVbvJdViv6mvwf/gZ+/CfwQjZf6IuXFbys4NsVvLzAtyv4/Qv8bgXfvMA3q35v2k35nzOUU/ue7bsRqhs9K3RZGd1PTBECXdMED0DXSUEO05z7hiAIAEAKDogA5txfCqSX7p/X3GJSz8OnoR7RnEegSNkcXFrUYZMrEjsKn1YIZLWfhWTdOtcnFFJSxzpoE+dOPZ231BvbR3lfw3kJMgQ5ef77t317gBQo6kWmgoXrlnZVtDpumAHFwjghd8QeWRaZSFGU1+LVCGBBFhshau1+Iz32OA1l+6HYRTTv/StVPSnpoo518KZccRXkdahuH+n1KMgQ5KT49W/79gBBMtJa7K5KNiRhmW9wRzQPQK62MfMA9CsRaPRss3XfyuWuUlS16/CHIOx52pw0lFGRHTEoyBDkpPjFv/TtAYJQKmqxe2RMDcrl5L3KxHyy1ptM66VFy7BOttVxJEv37pDwFQshdwadAe2DRj9e2Y7m14/Z/Ap2KJmddlEfTGW7kIHF3HjdTJiaWOxay/njQpuzbM8ZCTLhR/BHF3078fXXPTvwywMYBPn5T3t24Cff69kBBOkD92d/vhM7xj/+e6t2DbXYoV05efJrmSWUkQIObBn5jNDKyr8Rep+luvfwWMaCrSVRx1dffbWTk3369GltG0+b2FGSiERySYKl+0miEgWmGePEpa8tsKPEFMmGpAAVTfm8ZNqFq2nGGa8qd9YQi222XPsW2hRVexU4yxJBEATZC2nV0lbl5AND0Dx5/dK7AFAxyzLtJA1lEjNCekO2H+gUjesbGbJ4J9nMFkcOs2YgmnNfD8kiyRDMvKwH6cJ9nrzx0kcdl2Ox2bE1luvfQlc31BlFyBAEQc6ZtpGt7khrsa8tJ58tfcEFIsSrEUDYKgCU9ynVH03hIieHRmAoz3drozCVyw5XhxCLtIls7YqCQxX+5dVnU/L6tkwJ43R8kb17cmuVd1XUEItdY7na57X37hZghAxBEATpAPH6RobFM7PYV7QMSYVQaZiGRihszIyosaoSV9JwswSy1Is02MIQO5M8N23nbDte2SqEWEO5BOtxEOaFywRBYBbxzUu9auBH9ka3QPOF3MZyF6AgQxAEQbqgvha7aN7r4Gn5kfz3N3ZurZCZCsAZfM14kWjOfd3T2HVg0wBcOkTWbQH4elvUVOwohmHkTWLn1gpDS1KcOHYUxTCU7IBS9IGDzGC47iC7aQ9UrO1LksturVD3i2v9tqQpnrWd5S5AQYYgCIJ0Q30tdv4I+L5OxoXix2lYjJbkCoTJOtrQDa50PEkfzwJwnRaArzJObfkj65YqKw/G+elE88GWs/HacDF8SJLEVSs7cqeaeUeqx9SxXie0o2VIS1tkm5vQEIvd0nIHoCBDEARBdojqsgEH1c0CHoUhSOZI3lw056UaSExXElrjFZnqlkzXeFW2BwAQPy9SGSOa88RVQTTnD3ArCIKgeWTQVRrKRD0a4BJvo2WYysa0URVl49EypDJAHeupCqHjjNJQLlnKJFZlR4ZglgYohU0MAAAOCklEQVQgjw/1zpY9rRB8VJy4oKgCQ/MA6q92mYZY7JaWOwAFGYIgCLJfYkcRigUqWyZ0qa6vLybdlqEMDEFa3pOQFdlDdGJkE1lGPU+HuwoCr58FL2JnsshnEx4dojmPbMiCmJqX5hCK5gO96oIgTIZR4usQVgYIa6iLxW5veeegIEMQBEH2S5rXlaJBRQp/Hao7v1/e7q6MdqcF4MvGpaFM4zLBzNskA7+pY+y8X97PD0+Oiea8FP+s2+SDo5VB07kpklAnuVn4WGxxE7gDJbOtLbc9y/bgshcIgiDI3tmqWKfqzqXAcYJrcxfVKDstAF82Dg+2IglCvl0OyYhXo1CTFIge2J1mdcc4cB7h2nWPcqyya2ipAGYJfs2T7ejglCsAAPS1lHAPkJX6vzuALwfwZgBvBvDFAD4fwBcDuBzA5wO4HMDlAAYXcDmAy0sYXMBgAIMBXAzoC/L6YgAXFzC4TF8P4OICLgZ0z2cXdOdnaYPPBnT/4BJ+8PuvIVlBkkCygtUKIIFVApDAapXvp/+TFwm3k3ZZAUBVF2A6riAhZpOSWVjTJrPPOZY0dmncTKBoJ9vM3z7TMbsyK9bP0iZpuarc5LusVvT1Km2zSuCFbL7QFy8reFnBtyt4eYFvV/D7F/jdCr55gW9W/d63CLIlHz9+fPv2bReWn56e3r1714Xl46W/BfpbcY43A1sd6QBKODSAETIEQRAEQU6UrWKxewUFGYKcFOmwylYc8iM+8gqenp76duFcOPy/HbwZDhYUZAiCICdOd6NUXZhFOgVvhoMFZ1kiCIIgCIL0DAoyBEEQBEGQnjkjQZb8W98eAMD3v9+zA39t9OwAAPzDL3p24F//q2cHEATZGXHgODtbk+zoiAPDCfpcXR7ZHWckyBAEORoCIy1fGBjrajfvwZVeHUDqiQPDeJRM81DXMegeUXVNeG/gHXoKoCBDEOSQ6XBd7CNx4CTZicwNjNvZ2O2ieONxqXDVvRtOlB5L/iC7AQUZgiAIcnwEhuaNxiiVAQBE835kSccjIffIMWlrFGQIgmxCYGQFCOkzebFQNDvgSA7y7WkLxXGIKdqVaak4z5yt3DhrjuupOAHbnwsWtPWh0nj5C718BfJm+bFj+Q04VmJn4tGCkwgAqGMdvI5rriMdc0aCTPhR3x4AwNdf9+zALw9gweKf/7RnB37yvZ4dOF4CQ9A83SfleP2RJQlGAOL1jQy06jFpNfNAvrkWSR05sKMkSZIkssGSGAUTWtNhlJb7DQxBska5YcurOHtalo7i656W6Z7Q0tJDvh4ywYK2PjQZX3cFUjxtQsxFtuxpOIbUmmqNWzriMEL/cRoW9FidLm98YGDsH7mGVsc6ZGXMkeMkQRDkhOjyayGyZZCptEmSJEl8HUC2oySyZYBUpWR78xdse9KMP8T3p4fp8bwP+7LkF3cixtHWPtQZ5w7UXoG176hHPnz48KkbPnz48FqnCh9rdq3WbKTXuPSRp/ugwmzxw+BuUPajzS3W3g0HTNU1KbHvm4Fee/rhAfN9AeyHlbdlmjHk/UufJGu86iPjP0rGEHcK9txs+7r93XBGETIEQbYifpyGMLpiUqiloQzhMgLRvNezGFkWH8tecO3zUFpuKloWoh3qmPkC5vZWx604awDi1YgGC9r70Gh87RUoWhOvRg12euL//uYvd/Jvp07FzsST7SidNqG6fjr0RgYlfe5ISrQMq83lplQ3smVv4sTFEC57g8r2HTUvXt+wH+Wx0vodfFXDrtqzsJHj0JIEaXmfqh1Py/MO6iLZgSFokOvmQux5g7B0XXi7LjreMmq+O1CQIQjyWnLZoY5TRcZLoNCSmLQsyar8GY2fF+3Op7rkW7dijEkeSnW9WvnQaLyeroRXYLQZ8Iwd5djHRRs0boNMJ3cM141SqcurHxhAdZM5naIZGPU3xnGxeD68G0K2H8h1Fq9vZIBMBTNKOXhvhVkzEM25r4fW+wAA1ulm3niDIK2X/oU7TXUTkkhRt78zsJYlgiCvJX5eABBFoo510GaBCzMPdD9biUC2o/naZQk2UTWiOU9MACCPu5oyjObmmi6tfKg13tyRuQK7pLVE3Ygv//lDB1aLBIagVWUAgu63+jVL74ZXXdo6XZ7enqoazDyQbfLAEDtKKsN0P7InkrXh6VgCQ3m+m5ti7Ci38NB849S22cTIdnz69KnT9iwF5VwhpNlPhSANZbBmgauqqpukt016b7G3RduwNJH+98Vg+TKid4cmQPEGrdvfGRghQxCkHeL1jcw/gEfLMPsJJDEyY5ZPfVPHessRFH4YEdoJEtX1GftcOjMTBGntQ5PxlOYrUEc+ebM4nzTPOjcMI28SO7dWGFqS4sSxoyiGoWQH+I4HierWJMi0/FFLP/umn1dybJNgUBrCDWYe6PemCPQypwNYW//i7kRDZ0ZEc/5KNVYVNjwOaiPZ+R+QBj6f1LUVTIC/Jjr+qqj5FqAgQxCkJaJ5z05gDAwt+3ED+pPnsUsRqHd8Vkdg1GgJ0XxgWwZG9fARMzkOgMquTAyF1m3WnfGrtQ+NxttdgUoCQ5re0ByXhWYEEDu3NC/FH2U+ezCmWSqzgFyOLK4XLoYPCZkHWtHx2GnQuE0yXRpW/yzX6HKoemAoDEhVpqXlgjifwFmcn1nQ0AAAy/dK/gMeGAJ3/zG/6rxWLwhxkkfXrN1LrHs6OGAq5iO46q51MwunoUVznqWKAfONUbe/E1CQIQjSGtUlma3kcXVhR+w3JEnxYbN+RHMe2ZA9+Gqe7ifVD/5cS21h25VJ/S7JxBXy8+fWdPtmKuWnyfxq60Oj8ZZXoIL4eZFqAtGcJ64K0TKk10gd62kQjv6OSkO5FPbJJEVlx6OnXuOSzC8tP8LI9Np8oRpdDhUPDJzgCwzNg8qgGxHEkQ3WBMiL6WMcOwpNJU980IygUkMTeQ3qnQ00HSp2Jl7+BxI7t1SrVwlx0mCNdmeJlmFhAsvx0BDJbqOb29A6vF0dHa/fv0NQkCEIsgnMiFRRsFQlveYPmNwDbkU9Iqbl3DTd1DzfkhsP489/lfd/pQ/VxpuaMR4UmmHBpdbUa1z+CPi+ng3JlRa/o9Tpcig/MNC4LLE+GUaJrzN6LiPTOKzYiZZhOsKmeSUZR1um8lq8vpG9WQAQP04hS08HEM35A9zW2KBnaa/dyxOKj4n6SHZL3byeeulfFx1vFTXfJSjIEARBOkO8GmXDaGTsShrK9PclmHmbfL+/uuMhUquzCyqfOcLrW27mJEO9Li8/MHDPACI5XvUYUAmzJtW6dC/RvNe9WRA/ToFVTIGRLgDhVwWEN4NJjjtKaiPZLXVzG+qkf110vGXUfHecgSDLV2LuLRE2rq3p0oMHfS1H3f9FyP0oXAKuYs4hp0sjR4jqRlnMBvy5KYrmg73Q8u2qTuLVKORKCgCQX6Z1HU+J4gr7gaEx6x+QIaRFT8WCpKFM6xRVfKFUoI51b3I7HRUVEwmABbPKaakbSPDYmSzyJdsOCNGcs9K2cbMmkt1eN1fK6HbSvy703hCS74BTX/YiMIgOpmFJSYGOFW6ZdG25uQo9+RAYwmQYJYlIN/Y5jTd3QfNk+kHEjiIJy327kDoiWSGwj6Nk5jtN3IgdRbJ6cw55JflqFQdJyb3iDtGczwuvVDdJSJmz9FBzx1NENOf+UtCETKzo9Gs0Q3XnYChGMN/7nyvxTRIs4pYKAFejUJMUiB4qO6h39kRa8qU31bGuaYIHoOs0OUnljYjmg61IgkBOYooA1eozdt4v7/d/EZBdIyQ7qrVykMSOIk1vcvlDF/zd631LJWHqQyrP9udD8SLs34PSRSj5tD8vyHc7k1tSuhx9XJ9dIgjC9kZO+mvh7Pj48ePbt2+7sPz09PTu3bsuLG9CHDiPcG2qBx0y7GxhsQ3f/anfDMfNSQ9ZltaALk2j3gOqu4dAZwPx4zTkMj0LMeN9UMo2zSvb7BBSWrjOZhajK2RrlC6HeDU6zNWuEQSpQlTNw1ZjgSEIklUartwNB//ukfactCADKKzK0nt9udi5tUI2C6J7omUIoyvxADLpiux6/rBozpP7pVTzHpm8g2aCmXfMiyvW1QXfjL7fBIKcECQP6Vhj7sj+OOkcsmgZAtwU9y6eY+jhgSIdL5PtaJ/xsvh5ARBqAvg0JaWPBK6sQkX6xrspDANZ5k3sKIIQtq7VwhAYmgf7lcwI0jlPT099u4AcCngzHCwnLcgOilwqSMK068mzBZiCqiCa97pFirrt7fzknFllwNi57biML819jh1FUJ7bX2simvVzmMCGnBGY2YNk4M1wyJz0kGV1dY1+R6NE817fffrUGvi33FMmna+nCylKy/vE17tcYY+uYyEt79vn72VJZjiygCAIguydkxZkAIVEpc5GyjZBGlaX/OiGmrS5HlaV5BZ4jJ8XnShjmip3Cw/JJlkbsaMQNYaxMQRBEKQXTlqQlWpXFYpi7YPi2oZ7d4KU1C060HPSevw4DXeuCWNHEWh5uY10FVmIrLbIIoIgCILsgZ3MyTpcfD2vIc++7smFJIlsed9O8KfswQFaGCSrNNLTB8GcnSl6Etlyn94gCIIgSJIkyakn9atuZCsSWU8Z+hmSUt3EN+iSzlCx2HTniOY8uTKy9UL7uAqqm/iQL7l9QGOD8eM0BAAr+3wIB+QhgiAIcg6c9kr9CIIgCIIgR8BJ55AhCIIgCIIcAyjIEARBEARBegYFGYIgCIIgSM+gIEMQBEEQBOkZFGQIgiAIgiA9g4IMQRAEQRCkZ1CQIQiCIAiC9Mz/A4xUu2gIlG7uAAAAAElFTkSuQmCC" width="100%" style="display: block; margin: auto;" /></p>
<p>The histograms in the top panes show (rolling medians) of the
overdisperstion (<span class="math inline">\(\rho\)</span>) parameters
in controls (black) and cases (colored). The case-<span class="math inline">\(\rho\)</span> is colored according to the (already
5% FDR corrected) p-value testing for significant dAD in cases versus
controls, while the relative heigts of these histograms provide an
illustration of the effect size. The remaining plot elements are
explained in the legend.</p>
</div>
<div id="detecting-imprinted-and-differentially-imprinted-loci" class="section level1">
<h1>Detecting imprinted and differentially imprinted loci</h1>
<p>Imprinting analyses in MAGE are quite different from the analyses up
until now in that they don’t rely on beta-binomial modelling and make
different assumptions about the data. They are, generally speaking, a
bit more robust because imprinting is usually an extreme phenomenon,
i.e. heterozygous samples of imprinted loci will almost always “seem
like” homozygous samples because one of the alleles has been silenced
almost entirely. Partial imprinting can exist and MAGE can detect it,
but this phenomenon is not modeled as thoroughly as the previous ones in
this vignette.</p>
<p>First off we assume samples show no AB and are conform
Hardy-Weinberg-Equilibrium while taking imprinting into account, which
is tested by . It does this by assessing whether the amount of samples
with a reference allele fraction &gt; 0.5 (and so also the amount with
the fraction &lt;= 0.5) is equal to that expected under HWE assuming no
AB using a chi square test. This may seem kind of heuristic at first and
prone to fail if there is even a little AB, but remember we want to
mainly retain imprinted loci, meaning heterozygous samples should behave
as one of both homozygotes (and so either clearly &gt; 0.5 or &lt;= 0.5
in terms of reference allele fraction) so this filter is very fitting
for retaining HWE-conform samples that are at the same time likely to
showcase imprinting. A good filter setting is requiring the p-value to
be at least 0.05.</p>
<p>After making these assumptions, detects imprinted loci in the
remaining control samples by first assuming an unshifted binomial
mixture model, then splitting up the heterozygous samples in two
separate groups according to a degree of imprinting <span class="math inline">\(i\)</span> and varying this <span class="math inline">\(i\)</span> from 0 to 1 in a for-loop, calculating
likelihoods along the way to perform a likelihood-ratio test to check
for differential imprinting (i.e. compare the most likely model with the
non-imprinted one; <span class="math inline">\(i=0\)</span>). For
completeness, the imprinted PMF is:</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(&amp;counts_{ref}|counts_{total}) =  P_{rr} * {\tt
pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=1-SE) \ + \\
&amp; 0.5P_{rv} * {\tt pbinom} \left( x=counts_{ref} \ |\
n=counts_{total},
p=\frac{0.5-\frac{i}{2}}{1-\frac{i}{2}}(1-SE)+\frac{0.5}{1-\frac{i}{2}}SE\right)
\ + \\
&amp; 0.5P_{rv} * {\tt pbinom} \left( x=counts_{ref} \ |\
n=counts_{total},
p=\frac{0.5-\frac{i}{2}}{1-\frac{i}{2}}SE+\frac{0.5}{1-\frac{i}{2}}(1-SE)\right)
\ + \\
&amp; P_{vv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=SE)
\end{aligned}
\]</span></p>
<p>After retaining only significantly and sufficiently imprinted loci
(5% FDR level, degree of imprinting at least 0.6 and median degree of
imprinting at least 0.8 for robustness; see for the latter’s
calculation) using the function, we use to check whether their
corresponding case samples show a significantly higher number of
apparently heterozygous samples via a logistic regression test (more
details can be found on the function’s help page). If this is true, it
indicates that a significant fraction of heterozygous samples in tumor
tissue “lost their imprinting” and actually started expressing both
alleles again. We call these loci “differentially imprinted”. Comment in
the code below provide additional insight into the imprinting analysis
pipeline:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform filtering using symmetry_gof()</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Notice we use allelefreq_0 as input in this function, which is the allele frequency as </span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># estimated by an UNSHIFTED binomial mixture model using AllelicMeta_est() earlier in this </span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># vignette, which is what symmetry_GOF assumes as well so it&#39;s only fitting.</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We also enforce that this allelefreq_0 can not be more extreme than 0.15 or 0.85, </span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co"># because detecting imprinting would be very hard otherwise.</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>ImprData <span class="ot">&lt;-</span> controlList</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(LOC <span class="cf">in</span> <span class="fu">names</span>(controlList)){</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (ImprData[[LOC]]<span class="sc">$</span>allelefreq_0[<span class="dv">1</span>] <span class="sc">&lt;=</span> <span class="fl">0.15</span> <span class="sc">||</span> </span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>      ImprData[[LOC]]<span class="sc">$</span>allelefreq_0[<span class="dv">1</span>] <span class="sc">&gt;=</span> (<span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.15</span>)) {</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    ImprData[[LOC]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    ImprData[[LOC]]<span class="sc">$</span>sym <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">symmetry_gof</span>(ImprData[[LOC]]<span class="sc">$</span>ref_count, </span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>      ImprData[[LOC]]<span class="sc">$</span>var_count, ImprData[[LOC]]<span class="sc">$</span>allelefreq_0[<span class="dv">1</span>])</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (ImprData[[LOC]]<span class="sc">$</span>sym[<span class="dv">1</span>] <span class="sc">&lt;=</span> <span class="fl">0.05</span>) {</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>      ImprData[[LOC]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Detect imprinted control loci</span></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>impr_res <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(LOC <span class="cf">in</span> <span class="fu">names</span>(ImprData)){</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>  i_results <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">imprinting_est</span>(ImprData[[LOC]]<span class="sc">$</span>ref_count, ImprData[[LOC]]<span class="sc">$</span>var_count, </span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>                <span class="at">allelefreq =</span> ImprData[[LOC]]<span class="sc">$</span>allelefreq_0[<span class="dv">1</span>], <span class="at">SE =</span> SEmedian, <span class="at">inbr =</span> Fmedian)</span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># An additional robustified &quot;median imprinting&quot; across samples to be used as possible </span></span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># additional filter criterion:</span></span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>  med_imp <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">median_imprinting</span>(ImprData[[LOC]]<span class="sc">$</span>ref_count, ImprData[[LOC]]<span class="sc">$</span>var_count, </span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a>              <span class="at">allelefreq =</span> ImprData[[LOC]]<span class="sc">$</span>allelefreq_0[<span class="dv">1</span>], <span class="at">inbr =</span> Fmedian)</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>  results_z <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">&quot;position&quot;</span> <span class="ot">=</span> ImprData[[LOC]]<span class="sc">$</span>locus_id[<span class="dv">1</span>], <span class="st">&quot;LRT&quot;</span> <span class="ot">=</span> i_results<span class="sc">$</span>LRT, </span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;p&quot;</span> <span class="ot">=</span> i_results<span class="sc">$</span>p_value, <span class="st">&quot;estimated.i&quot;</span> <span class="ot">=</span> i_results<span class="sc">$</span>est_i, <span class="st">&quot;allele.frequency&quot;</span> <span class="ot">=</span> </span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>    ImprData[[LOC]]<span class="sc">$</span>allelefreq_0[<span class="dv">1</span>], <span class="st">&quot;reference&quot;</span> <span class="ot">=</span> ImprData[[LOC]]<span class="sc">$</span>ref[<span class="dv">1</span>], <span class="st">&quot;variant&quot;</span> <span class="ot">=</span> </span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a>    ImprData[[LOC]]<span class="sc">$</span>var[<span class="dv">1</span>], <span class="st">&quot;med_cov&quot;</span> <span class="ot">=</span> ImprData[[LOC]]<span class="sc">$</span>coverage[<span class="dv">1</span>], <span class="st">&quot;nr_samples&quot;</span> <span class="ot">=</span> </span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">nrow</span>(ImprData[[LOC]]), <span class="st">&quot;GOF&quot;</span> <span class="ot">=</span> i_results<span class="sc">$</span>GOF_likelihood, <span class="st">&quot;symmetry&quot;</span> <span class="ot">=</span> </span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a>    ImprData[[LOC]]<span class="sc">$</span>sym[<span class="dv">1</span>], <span class="st">&quot;med_impr&quot;</span> <span class="ot">=</span> med_imp, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-37"><a href="#cb20-37" aria-hidden="true" tabindex="-1"></a>  impr_res <span class="ot">&lt;-</span> <span class="fu">rbind</span>(impr_res, results_z)</span>
<span id="cb20-38"><a href="#cb20-38" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-39"><a href="#cb20-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Retain significantly imprinted loci (5% FDR) utilizing some additional filters, amongst</span></span>
<span id="cb20-40"><a href="#cb20-40" aria-hidden="true" tabindex="-1"></a><span class="co"># which a custom  Goodness-Of-Fit which more or less corresponds to a locus&#39; likelihood of</span></span>
<span id="cb20-41"><a href="#cb20-41" aria-hidden="true" tabindex="-1"></a><span class="co"># the imprinted model*coverage; 0.8 is a good cutoff. Other filter criteria are imprinting </span></span>
<span id="cb20-42"><a href="#cb20-42" aria-hidden="true" tabindex="-1"></a><span class="co"># (0.6) and median imprinting (0.8)</span></span>
<span id="cb20-43"><a href="#cb20-43" aria-hidden="true" tabindex="-1"></a>impr_res_FIN <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">final_filter</span>(<span class="at">data_hash=</span><span class="cn">NULL</span>, impr_res, <span class="at">results_wd=</span><span class="cn">NULL</span>, <span class="at">gof_filt =</span> <span class="fl">0.8</span>, </span>
<span id="cb20-44"><a href="#cb20-44" aria-hidden="true" tabindex="-1"></a>  <span class="at">med_impr_filt =</span> <span class="fl">0.8</span>, <span class="at">i_filt =</span> <span class="fl">0.6</span>, <span class="at">file_all =</span> <span class="cn">FALSE</span>, <span class="at">file_impr =</span> <span class="cn">FALSE</span>, </span>
<span id="cb20-45"><a href="#cb20-45" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_all_counts =</span> <span class="cn">FALSE</span>, <span class="at">file_impr_counts =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-46"><a href="#cb20-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-47"><a href="#cb20-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-48"><a href="#cb20-48" aria-hidden="true" tabindex="-1"></a><span class="co"># From among actually imprinted loci, detect differential expression in case data</span></span>
<span id="cb20-49"><a href="#cb20-49" aria-hidden="true" tabindex="-1"></a>pos_impr <span class="ot">&lt;-</span> <span class="fu">as.character</span>(impr_res_FIN<span class="sc">$</span>position)</span>
<span id="cb20-50"><a href="#cb20-50" aria-hidden="true" tabindex="-1"></a>p_DI_df <span class="ot">&lt;-</span> impr_res_FIN</span>
<span id="cb20-51"><a href="#cb20-51" aria-hidden="true" tabindex="-1"></a>p_DI_df<span class="sc">$</span>DI_pval <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb20-52"><a href="#cb20-52" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(LOC <span class="cf">in</span> pos_impr){</span>
<span id="cb20-53"><a href="#cb20-53" aria-hidden="true" tabindex="-1"></a>  CData <span class="ot">&lt;-</span> controlList[[LOC]]</span>
<span id="cb20-54"><a href="#cb20-54" aria-hidden="true" tabindex="-1"></a>  TData <span class="ot">&lt;-</span> caseList[[LOC]]</span>
<span id="cb20-55"><a href="#cb20-55" aria-hidden="true" tabindex="-1"></a>  p_DI <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">LOItest_logreg</span>(CData<span class="sc">$</span>ref_count, CData<span class="sc">$</span>var_count, </span>
<span id="cb20-56"><a href="#cb20-56" aria-hidden="true" tabindex="-1"></a>                                  TData<span class="sc">$</span>ref_count, TData<span class="sc">$</span>var_count)<span class="sc">$</span>p.value</span>
<span id="cb20-57"><a href="#cb20-57" aria-hidden="true" tabindex="-1"></a>  p_DI_df<span class="sc">$</span>DI_pval[p_DI_df<span class="sc">$</span>position <span class="sc">==</span> LOC] <span class="ot">&lt;-</span> p_DI</span>
<span id="cb20-58"><a href="#cb20-58" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>We can take a look at the loci that are significantly and
sufficiently imprinted (i.e. the output of from above):</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(impr_res_FIN<span class="sc">$</span>position, <span class="at">collapse =</span> <span class="st">&quot;, &quot;</span>))</span></code></pre></div>
<p>[1] “loc42, loc81, loc138, loc166, loc167, loc168, loc169, loc202,
loc215”</p>
<p>From among this set, a part is actually differentially imprinted in
case samples, which is indicated by the p-value in the
<code>DI_pval</code> column in the <code>p_DI_df</code> dataframe. We
can control loss-of-imprinting detection, for a change, at the 5% FWER
level (actually feasible here due to the small number of actually
imprinted loci remaining, though we’re looking at a small toy dataset
here; this isn’t necessarily the case in larger experiments):</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(p_DI_df<span class="sc">$</span>position[<span class="fu">p.adjust</span>(p_DI_df<span class="sc">$</span>DI_pval, <span class="at">method =</span> <span class="st">&quot;holm&quot;</span>) <span class="sc">&lt;</span> <span class="fl">0.05</span>], </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">collapse =</span> <span class="st">&quot;, &quot;</span>))</span></code></pre></div>
<p>[1] “loc138, loc166, loc167, loc202, loc215”</p>
<p>We can plot histograms of both the control- and tumor allele
fractions; locus 202 is obviously imprintend in controls and loses a lot
of this imprinting in cases (controls are plotted using to visualize
imprinting results, but since an imprinting-estimation never happens on
the case data - we just detect a change in heterozygosity via logistic
regression - we simple make a histogram for cases instead).</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>HistCTRL <span class="ot">&lt;-</span> MAGE<span class="sc">::</span><span class="fu">MAGE_imprintplot</span>(controlList[[<span class="st">&quot;loc202&quot;</span>]]<span class="sc">$</span>ref_count, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  controlList[[<span class="st">&quot;loc202&quot;</span>]]<span class="sc">$</span>var_count, </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">allelefreq =</span> impr_res_FIN<span class="sc">$</span>allele.frequency[impr_res_FIN<span class="sc">$</span>position<span class="sc">==</span><span class="st">&quot;loc202&quot;</span>],</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">impr =</span> impr_res_FIN<span class="sc">$</span>estimated.i[impr_res_FIN<span class="sc">$</span>position<span class="sc">==</span><span class="st">&quot;loc202&quot;</span>],</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">SE =</span> SEmedian, <span class="at">inbr =</span> Fmedian, <span class="at">plot_NoImpr =</span> <span class="cn">TRUE</span>, <span class="at">SplitPeaks =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggtitle</span>(<span class="st">&quot;Control&quot;</span>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>RatioCASE <span class="ot">&lt;-</span> caseList[[<span class="st">&quot;loc202&quot;</span>]]<span class="sc">$</span>ref_count <span class="sc">/</span> </span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  (caseList[[<span class="st">&quot;loc202&quot;</span>]]<span class="sc">$</span>ref_count <span class="sc">+</span> caseList[[<span class="st">&quot;loc202&quot;</span>]]<span class="sc">$</span>var_count)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>CASEDat <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">&quot;Ratio&quot;</span> <span class="ot">=</span> RatioCASE)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>HistCASE <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>() <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">geom_histogram</span>(<span class="at">data =</span> CASEDat, ggplot2<span class="sc">::</span><span class="fu">aes</span>(Ratio),</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;Reference allele fraction&quot;</span>, <span class="at">y=</span><span class="st">&quot;Frequency&quot;</span>) <span class="sc">+</span> </span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggtitle</span>(<span class="st">&quot;Case&quot;</span>)</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(HistCTRL, HistCASE, <span class="at">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAzAAAAHgCAMAAABNd3gvAAABTVBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYzMzM6AAA6ADo6AGY6OgA6OmY6OpA6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshZWVlmAABmADpmAGZmOgBmOjpmOmZmOpBmZjpmkJBmkNtmtttmtv9uTU1uTW5uTY5ubqtuq8huq+SOTU2OTW6OTY6ObquOyP+QOgCQOjqQZgCQZjqQZmaQkDqQkGaQkLaQtpCQttuQ27aQ2/+gIPCkgjGrbk2rbm6rbo6rjk2ryKur5P+2ZgC2kDq2kGa2tma2tpC2ttu225C229u22/+2/9u2///Ijk3I///bkDrbkGbbtmbbtpDb25Db27bb29vb2//b/7bb/9vb///kq27k///r6+vy8vL00oH/swD/tmb/yI7/25D/27b/29v/3Yz/5Kv//7b//8j//9v//+T///+e3i1kAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAciElEQVR4nO2d638bx3lGV3RYmHaSxiVk2XTctGkoW5Xskk4aNylp2TWZNG1VXtq0aUopFi+uedv//2N39gIsSGIxszs7eAY8zy+OcHlxOJwXh7MXcJmkhBDrJPMeACExBWEIcQjCEOIQhCHEIQhDiEMQhhCHIAwhDkEYQhziLszVb7+fJG/8bNrTZz/Zu/2SzeUj569D3DKjL8RPnIU5W0nyvDPl+cMlhJlHZvWF+ImrMBdryTv/k6b/9WGyfncBwswlM/tC/MRVmMNkNf/3bMU48O2HSfK9F7kRv/t+8uBvshvZD7nVq83BYbL0ov40wvSbyb78Z7Z1lnUjm/lfZDfyzTTTix+8mOcQFyOOwlxtVgvIH7L/TvLNgAdbhSdZ1ith3lhJlo9qTyNMv5nsy2HZjaov69UWW9YL0i2OwlysDcZ3snb8zHRn+Si7NTjKbg2KTbLs7urk0wjTb270JVvdM0UG2X/vHOU3rjbNOvP7lcF0BLFKF2HOivnffbBV/IDLnyyEMXfrTyNMv5noS5Y//O7vVxIjzBs//Tdz/2wl32K7aweTOKWbMPkO5kllRLbrUglj7t58mvSXSWHKI2bZQ7vm3x+9SE+KjTS2yTqn9T7Mf/zoBcLIZKIvF2vJn//Dv/wxd+j32b5+srSHML7S9ijZxdrS3s1trlvCsEkWLJN9MT+pzqodlv/+ebJe/vAindP+PMzqrb36QpjRUTF2+gOm3pdMlSNzGHn56CT5i6P06jfmR9aDX6fpt5vsw3RN6zP95sfXjePGhTD5YeVcEA4rB0ytL9VB/vzoZXGj2olZnfcwo4/7Z8m+/cXK6DNLZxNnJnNhLj5MBv9bCnLGictwqfUlP2H8612zcWxOXL5jJt/0gk+adQ+fVibEIQhDiEMQhhCHIAwhDkEYQhyCMIQ4BGEIcQjCEOIQhCHEIQhDiEMchfnT7dz12F2xrFsIXIev2q6NfsfQoUwb56EvCNMHDmE0cQgjikMYTRzCiOIQRhOHMKI4hNHEIYwoDmE0cQgjikMYTRzCiOIQRhOHMKI4hNHEIYwoDmE0ceGFIZrx++boUKaNY4URxbHCaOIQRhSHMJo4hBHFIYwmDmFEcQijiUMYURzCaOIQRhSHMJo4hBHFIYwmTlKYdh3vHP9T1HfZ+3luvdTXd78Q73C9viCMUx3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI1DmMaRWU+RfR3CRI2TFuZibd38c7K0Nw327Yv07K2tO5+a9rjTyKynyL4OYaLGxS2MkcJdDIeRWU+RfR3CRI1DmMaRWU+RfR3CRI2LRJjdQfbv4eDsrc+TZPkoTa82k2SQnq0kyWr+WLJePGbUMo9m93OZipvtR2Y9RfZ1CBM1LhJhzCJztbl+trK0d7U5SKv/Cikygw6Lx9PD5aN8xTlbWc/+rW62H5n1FNnXIUzUOHFhkjxLe0ads7f38jd/pkG+kZb9XyFM7bGsLivLX2yEeXvq0QLLkVlPkX0dwkSNExdmtA9zODD/y5eM7NETs1lWLiHFMvLW1mEh12q6a7bWiv2b8mb7kVlPkX0dwkSNi0WYs7f/fXO9WRjzWPW6pFh9ypvtR2Y9RfZ1CBM1LhZhrjZ/mm1fFZtfb++dPNgab5KVwuSP1V5ZHT0rIe1GZj1F9nUIEzUuFmHSQ7NxdXOnP9+1KYW52syWmMyavLpYfaqb7UdmPUX2dQgTNS4aYcp9+8+LvZLisHKa7aQMKmHyx/KVJ8n/NQ+VN9uPzHqK7OsQJmqctDD15Ee8/J2ktByZ9RTZ1yFM1LhohDlcTRHmZnwKQ/yl6Mu0Z0MIk5+dRJibYYUJjotmhQmQLlNkX4cwUeMQpnFk1lNkX4cwUeMQpnFk1lNkX4cwUeMQpnFk1lNkX4cwUeMQpnFk1lNkX4cwUeOkhfnm7rR7Q7QamfUU2dchTNS4foU5fzwcbqTp5bPho1cOjalKEWZGECY4rldhLj/ZSc8/2rl+vpG+fM+hMVUpwswIwgTH9SrMqbFkf+Py04P0/OMD+8ZUr0eYGUGY4Lje92GyVeb8yat8sUnTN7M49DCwMPGl+SMYbmn/HtJ+h0cmzPXzp+npo0oYy8ZUL2aFmRFWmOC4noW5fPY02/V/gjD9lCFMcFzfR8k2sv9nH6avMoQJjutVmMKXfLOsh6NkU68JM37CXGjG/soxXabIvg5hosb1KszLoclGT+dhLFRwu85Slymyr0OYqHG9HyVr05iq9JtvkluZEObqs8+TZPXEXFzp4se/zH8x+QcfLv1rtq788OdJsm4uGTO6k19B5nt/2fALNV2myL4OYaLGRS6MufrlysDcvFhbPjLXkVkpr/i3mp4sHxWbZOWddHc1PWn6Df8uU2RfhzBR48SFmbFJdvXZVmr+u/hgy1wwI7tpNsLKPZfSltGdix/v5cUuI7OeIvs6hIkatzjCfJCpsLveIMzZD48QplNfOo11QXALJQwrTPHSVmk/VO13uF5fRIRZG5R7LdOEYR+ma186jXVBcIsjzAd/nR8lmxDmajM/SlYKc7GW/BkrTIe+dBrrguAiFmYy+T7MzDSemOkyRfZ1CBM17h4JU11K1mVk1lNkX+e3McfHCBMUtzDCeEiXKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocQjTODLrKbKvQ5iocRELY359/8bnj7tdeanLFNnXIUzUuKiFWdqzFcaqpMsU2dchTNQ4cWFq14rJM3nVmH8epNVvhi3tFQ91ufJSlymyr0OYqHFRC7O3uz763ePlo7TrlZe6TJF9HcJEjRMXpnGTLFPir47Kq1sUvz/W7cpLXabIvg5hosbFLUx6uFq++4u3fbcrL3WZIvs6hIkaF7kwV5/9qmGFcbzyUpcpsq9DmKhxkQtTHCqb2Idpf+WlLlNkX4cwUeNiFyY9XLp5lKz1lZe6TJF9nYIw54+H7x64/3XrTmNdEFzEwsyM65WXukyRfZ2AMJfPNtKXj145/3XrTmNdENz9Fmbyyktdpsi+TkCY8m8oOv9luE5jXRDcIgvjmi5TZF+nI0zrv25NZqQQZtqzCONUJyBMvkn2cMf5r1t3GuuC4FhhGkdmPUX2dQLCmJ3+v/1yx/mvW3ca64LgwgvjkMDCxJdCmFYvzXZg2Ifpq2xeK0zgdJki+zqBFcYsKy/fc//r1p3GuiA4hGkcmfUU2dcJCJOeDvMTMJyH6akMYWwSkTCW3/1CvMMRprd0mSL7OoSJGocwjSOzniL7OoSJGocwjSOzniL7OoSJGocwjSOzniL7OoSJGicpTKfBLggOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRxCCOKQxhNHMKI4hBGE4cwoji7xhhfECYkDmFEcQijiUMYURzCaOIQRhSHMJo4hBHFIYwmDmFEcQijiUMYURzCaOIQRhSHMJo4hBHFIYwmDmFEcQijiUMYURzCaOIQRhRnKcz2NsIExSGMKA5hNHEII4pDGE0cwojiEEYThzCiOITRxCGMKI6jZJo4hBHFIYwmDmFEcQijiUMYURzCaOL6Fib/s9Yvh8Phu/x5a/9lCBMc17Mwp7ko+xs0ppcyhAmO61eY/YdfZyvM9Zc7NKaXMoQJjguxSXb5LNskyxeZN7O0aye5K4Uwfljt30Pa7/AIhTn/aKe2yrQfq/ZMzmeF4cRlYFyQnX6T0X5M+7FqzyTCeCrTxiGMKA5hNHEhhDl99Cq9/orDyv7LfApD/KUQZtqzVudhHo4OlLV/D2m/w2MXpv1QtScyshWGxvRZhjDBcQgjikMYTRzCiOIQRhOHMKI4S2E40x8YhzCiOITRxCGMKA5hNHEII4pDGE0cwojiEEYThzCiOITRxCGMKA5hNHEII4pDGE0cwoji7BpjzlsiTEgcwojiEEYThzCiOITRxCGMKA5hNHEII4pDGE0cwojiEEYThzCiOKuyY4QJjUMYURzCaOIQRhRnJ8zxFwgTFocwojiE0cQhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRxCCOKQxhNnGdhLtYebNGYYGXWJy7pS9AylxVmN0mWj2hMmDKHM/30JWCZ2ybZxVqSrNKYEGVOH42hL8HKXPdhTGuW9mhM72WOnyWjL4HK3IQ5TJJBtgkwbQOg/Vi1Z1JeGPoSrMxBmKvNJFk3N06m/ShrP1btmdQWhr6ELHM5SjZ1yacx/sscjpLRl4BlnIcJj+tjH4a+BCpz2iRbzf4b0JggZfYnLulLyDIHYXYHeXeaOkO8pRDGpnJ2X9q/h7Tf4drCXKw171nSmPmsMPQlaBnChMchjCbO9ybZoWnJxRpnlIOU2e/D0JeQZS5HyU6SpPETGDTGY5nDp5XpS8AyDiuHx/kWhr4ELEOY8DiE0cT5FuZsxSz90z/iR2N8lm1v2wpDX0KW2QtjcQam/Vi1Z1JamBt9uX4+fLiTppfPho9e0Rf/Zc6HlREmTJm1MDf6sr+Rnj56df18I335Hn3xX+aywiBMwDKHFabel8tPD6p/zj8+oC/eyxz2YZpOjdEY32X2+zATfTl/8k9mk+z8yav08pNs0yx9M8uMthGHFMJMe3Zykyxh5zJcmcMmWb0v5483jC3ZVlkpDH3hsHLsOM8rzETKpWW8wtAXhIkd16cwl3+Xm8I+TF9lLsIcJsn6YdMFfdqPVXsm59EYhxOXE33ZzzfJrp8/5ShZL2Uuvw+z/Me1dX7vIlCZvTCTfbl8Nnz3gPMwvZU5nYcxh/z5GHmYMpeP99OXcGUIEx6HMJo435tkh2bp5/cuApU5/D4MfQlYxu/DhMf53umnLwHLOKwcHudbGPoSsAxhwuMQRhPnfaefj8YELGv50Rj60nOZ6wpz8UHDX+9pP1btmZQWhr4ELXPeJDtpONXffqzaMxmBMPQlUJm7MCz9IcqOnYWhL0HKnIWZ+kdIaMxchaEvYcqcd/qb/gJp+7Fqz6S0MPQlaBmHlcPjellh6EuYMoQJj0MYTVw/m2RNh/zbj1V7JuchzPEXjudh6EuQMpeLYJjNZD7kF6bMXhj6ErTM+bpkHL4MUuaywtCXgGUIEx6HMJo435tk/B2SgGUOm2T0JWSZ40Uwyr8IT2P6LnMQhr6ELOOwcnicb2HoS8CyDsLk173i6iT9lCFMeFwfm2S165Kdmsv5cJX4nsq2t502ybheXKiy1tcl23/4dbbCcIXFnsochOF6cSHLOlxmyYjCVeJ7SimMRSWXWQpa1lEYrhLfU5n9CoMwQcs6XJdscoWhMfPaJOO6ZCHLOlyX7Jx9mP7KXHb6uS5ZwLKOh5W5SnxPZS7CzEz7oWpPpLYwt/7GJedheiyzF4a/PRq0zPnDlzQmSJnzhy/pS5gyp53+hnNjNMZvmcuHL+lLwDL337jk8GWIMvffuKQvQcr48GV4nOcVhr6ELEOY8DiE0cT5FMZiz5LGzEEY+hK6zEWYpk9f0Bi/ZS7C0JeAZQgTHocwmjiEEcXZlFmeuKQvocsQJjwOYTRxCCOKQxhNnF9hZl6PlMbMRRj6EraM8zDhcR6FoS+hyxAmPA5hNHEII4oLLgzxl0KYac8iTB84VhhNHCuMKM6mjM+ShcchjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojibMk5chschjCjOouwYYcLjEEYUhzCaOIQRxSGMJg5hRHEIo4lDGFEcwmjiEEYUhzCaOIQRxSGMJg5hRHE2whx/gTChcQgjikMYTRzCiOIQRhOHMKI4hNHEhReG+EoljB9a+/eQ9js8emHaj1V7JllhPJVp4xBGFIcwmjiEEcUhjCYOYURxFmXmvCXCBMYhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiLMk5czgGHMKI4hNHEIYwoDmE0cQgjikMYTRzCiOIQRhOHMKI4hNHEIYwoDmE0cQgjikMYTRzCiOIsyjhxOQccwojiEEYThzCiOITRxCGMKA5hNHEII4rrVZjT4fDdgzS9fDZ89Iq++C9DmPC4PoU5//ggffleev18w/xDX7yXIUx4XN+bZJk0l58e5O7QF99lCBMe17cw2dJy/uRVevnJTnbnzSxuXSRNKYSZ9izC9IGzKOtw4vL88cOd9PRRJQx9YYWJHdevMKkxZbzC0BeEiR3XtzDp/gb7MH2VIUx4XJ/ClNti18+fcpSslzKECY/rdYV5ORxm+zCch+mrDGHC43rfJKMv/ZUhTHgcwmjiEEYUN7vs+Pg1wgTHIYwobnbZ9jbChMchjCgOYTRxQYR5OSw+HUtjfJYhzDxwQYTZ36Ax/ssQZh64EMJcf7lTu9d+rNoziTCeyrRxIYS5fJZtkuWLDJ+K9Zjt7e8KYfzg2r+HtN/hEQpz/tFObZVpP1btmWSF8VSmjQt2lGy0H9N+rNoziTCeyrRxCCOKm13Gict54EIIYz4ce/0Vh5U9lyHMPHBBVpjyw7E0xmsZwswDF2yTjMb4LkOYeeAQRhSHMJo4hBHFIYwmDmFEcQijiUMYURzCaOIQRhQ3u4wTl/PAIYwoDmE0cQgjikMYTRzCiOIQRhOHMKI4hNHEIYwoDmE0cfMX5v276a0He28agzDzwCGMKA5hNHEKwhwfI0yLMk5czgOnIExujLfB3pvGIMw8cAgjikMYTRzCiOIQRhOHMKI4hNHEIYwoDmE0cQgjikMYTRzCiOIQRhOHMKK4mWXb2wgzBxzCiOIQRhMnIcwXmTHeBntfGoMwc8EhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRxCCOKm13m+cQl8ZdCmGnPIkwfuODCtB+q9kSywvgq08YhjCZOQ5jMGG+DvTeNQZh54BBGFDezLPMFYcLjEEYUhzCaOIQRxSGMJi68MDdTCtMNcg9zfPzdd4Uwfnjt30Pa7/DohbkDzgrjXmbOW7LChMchjCgOYTRxCCOKmy3Ma4SZAw5hRHEIo4kTEcbOGOmZRBhPZdo4hBHFzV2Y9+/ueo/fUhQ4hBHFIYwmDmFEcQijiUMYUdyMMnOeH2HmgEMYURzCaOIQRhSHMJo4hBHFIYwmDmFEcQijiVMRxsoY6ZlEGE9l2jiEEcU1lxW+IEx4HMKI4hBGE4cworjmsm2EkexLGYQJj5shzDbCKPalDMKExyGMJg5hRHEIo4lDGFFcY9k2wkj2pQrChMfNEOY1wgj2pUoIYV5bGCM9kwjjqUwbhzCiOITRxCGMKA5hNHEII4prLCtP9CNMeBzCiOKayipfECY8DmFmZbuMry+LMFHjEGZWECbEt+RUV4wsvf1Qly+LMB3qJgTR+0mGMAijI8ytFQVhAnxLTnUIY4SxMCaYME1lUzfPEMZPGcLMbMz2tpAws8qm7tMEa0x1FgZhbj7U5cvGJcx2PMKYIIznb8mt7t4L80X+6dt5CtN0FExuHwZhXISx+zasv1sNYXJj5ibMjOPGM3CjFyKMnzKEsRDGGDMnYWaeZZkpTPlyhPFThjA2wuTGeBlsC2G64aoFKpAw22NfFkGYiS8zYcL7t/OnGWUTz991Z+q3piXM5bPho1czhHkddIXJLZHFNZQdb3sUZlZfEMbym7j7m7r1Ulthrp9vpC/fmyVMtg9zPMOZVo25eRh4fL/HLbzal2uJu7Msm6CaLl2FmdkXhLH8Ju7+pm691FaYy08P0vOPD2YJUxhzXL3jjo97vH/cW/oe/uvX3oSZ2ReEsfwm7v6mbr3UVpjzJ6/Sy092sltvZmks7eEdNu/4/HZsVbCLQ1+Ix8wU5vRR1RiTttba1y0ErsNXtW0cfQn7Va2FGf8kozFKwtCXsF/VWpjmbeVug+hQpo0LIAx9CftVrYW5fv604WhMt0F0KNPGBRCGvoT9qtbCNB/v7zaIDmXauADC0JewX9VemIn4HUSHMm1cCGHoS9CvijB94hBGE4cwojiE0cQhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRxCCOKQxhNHMKI4hBGE4cwojiE0cQhjCgOYTRx4YW5I55/2w+cJhtcEYRZIFyfbHBFEGaBcH2ywRVBmAXC9ckGV6S7MITcoyAMIQ5BGEIcgjCEOARhCHFIS2HGlywpb01cw6QD7vzxcLiRpi+Hw+G7BzNeZYErOZ5GZ2hmeN1Gl46uJ+Zl7iZCX/ruSzthxpeOL29NXku+Pc5cy/H8o510f6MtawJXcnyNzsRcorXT6DJE2VUvczcR+tIlVn1pJ8z4sovlrcnrMLbHnZoB7m9cf7kz6zVWuJLja3Rp8c7pNrp0/+HXBc3L3E2EvnSIXV/aCTO+sG95a/JKv+1x+Xg/2ckWw3wLoCuu5HgcnfmR02106Wjp9zJ3k2D60ndf2gkzvnR8eWvyWvLtcWlxDVSz+rf/eTHGlRx/o8v/7Ta6dNQYL3M3Zaj0pUVs+iK3wlw+e1o+2nqL9MZg9jf8je50tBfYZXs5xhWGvhQR24dJzx+PvuHW3/qNwexv+NtW3q/eNT4aE9E+DH2p0vYo2dPR0ZinxRGFp52Od1SvLvtifmBcf9V2Jse4kuNrdOWC32106agxXuZuIvSl9750Og9jvoDH4/0ZrnZE/WH7jZTx6EqOp9FVK3S30RWN8TZ304ZKX5xj0xfO9BPiEIQhxCEIQ4hDEIYQhyAMIQ5BGEIcIifM1WaSZ7X22GHyYKst7+ytrex/tXsTz04lf/viVu39Dn0pIihM3pKzlfXRQxdr61PLZ6axMVPJyHIz9KWIqjDp7mD0UKdZamzMVPL8G6MW+lJEWhizGbC0d7aSJIPydnr29i+zf6o7b/0qe269KByk1QsKkHlZtgFRNmb0iq1aUU4uiGV1STL3VqvaQTr+Qvc39KWIqjAn2TRcbWZzcrh8VMxQeXslb1h1Z/koPTR9GuSrePW4AeSrevZc0Zgaql5kyDmxqq5I+RNFrfmv+kLznJg5h74UERSm2LnMJurEzEQ2TWaWRrdX6k+YO7W1vXrc3P4/M/PFc9n/aqh6UdGY9YnqdPREWXtifs6tK2wOzDP0pYigMKtmTTY/zg7L4zJmRuq3J+9U81d7QUE6yW4+KBtTe0W9qJz/cXVFqhpjfuBV/brvwtAXE01h8pW/Wp/zGanfnrxTb0z5uMnF2oOt8U+y2ivqRaPGVNU6jVELfSkiKky6m83QyYPxOly/PXmnvvTXDt7nc3pS/SSrvaJeNGpMrbr+RF57km9wIwx9MVEVxuzyXW1m01XOVv325J3aPmD1uAGYf85WqsbUXlEvGjemrK5I5Q7maOcSYehLEVVhsu3e1XxH80E5dfXbE3fOxkcZq8fz7GY3/zHfnawqylfUisbbymX1iLSbDGpghKEvZeSEIUQ5CEOIQxCGEIcgDCEOQRhCHIIwhDgEYQhxCMIQ4hCEIcQhCEOIQ/4feLLOAuLf2K0AAAAASUVORK5CYII=" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="session-info" class="section level1">
<h1>Session info</h1>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running under: Windows 10 x64 (build 18363)</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] LC_COLLATE=Dutch_Belgium.1252 </span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2] LC_CTYPE=Dutch_Belgium.1252   </span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3] LC_MONETARY=Dutch_Belgium.1252</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [4] LC_NUMERIC=C                  </span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5] LC_TIME=Dutch_Belgium.1252    </span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] stats    </span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2] graphics </span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3] grDevices</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [4] utils    </span></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5] datasets </span></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [6] methods  </span></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [7] base     </span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] gtools_3.8.2       </span></span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [2] zoo_1.8-8          </span></span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [3] tidyselect_1.1.0   </span></span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [4] xfun_0.32          </span></span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [5] bslib_0.2.4        </span></span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [6] purrr_0.3.4        </span></span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [7] lattice_0.20-41    </span></span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [8] colorspace_1.4-1   </span></span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [9] vctrs_0.3.8        </span></span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [10] generics_0.1.0     </span></span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [11] htmltools_0.5.1.1  </span></span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [12] hash_2.2.6.1       </span></span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [13] yaml_2.2.1         </span></span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [14] gmp_0.6-2          </span></span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [15] utf8_1.1.4         </span></span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [16] rlang_0.4.10       </span></span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [17] jquerylib_0.1.3    </span></span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [18] pillar_1.7.0       </span></span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [19] glue_1.4.2         </span></span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [20] DBI_1.1.1          </span></span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [21] MAGE_1.0.0         </span></span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [22] lifecycle_1.0.0    </span></span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [23] stringr_1.4.0      </span></span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [24] munsell_0.5.0      </span></span>
<span id="cb24-49"><a href="#cb24-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [25] gtable_0.3.0       </span></span>
<span id="cb24-50"><a href="#cb24-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [26] evaluate_0.14      </span></span>
<span id="cb24-51"><a href="#cb24-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [27] labeling_0.4.2     </span></span>
<span id="cb24-52"><a href="#cb24-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [28] knitr_1.31         </span></span>
<span id="cb24-53"><a href="#cb24-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [29] parallel_4.0.2     </span></span>
<span id="cb24-54"><a href="#cb24-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [30] fansi_0.4.2        </span></span>
<span id="cb24-55"><a href="#cb24-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [31] highr_0.8          </span></span>
<span id="cb24-56"><a href="#cb24-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [32] Rcpp_1.0.8         </span></span>
<span id="cb24-57"><a href="#cb24-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [33] scales_1.1.1       </span></span>
<span id="cb24-58"><a href="#cb24-58" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [34] jsonlite_1.7.2     </span></span>
<span id="cb24-59"><a href="#cb24-59" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [35] alabama_2015.3-1   </span></span>
<span id="cb24-60"><a href="#cb24-60" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [36] farver_2.0.3       </span></span>
<span id="cb24-61"><a href="#cb24-61" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [37] gridExtra_2.3      </span></span>
<span id="cb24-62"><a href="#cb24-62" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [38] ggplot2_3.3.3      </span></span>
<span id="cb24-63"><a href="#cb24-63" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [39] digest_0.6.25      </span></span>
<span id="cb24-64"><a href="#cb24-64" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [40] stringi_1.5.3      </span></span>
<span id="cb24-65"><a href="#cb24-65" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [41] dplyr_1.0.4        </span></span>
<span id="cb24-66"><a href="#cb24-66" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [42] numDeriv_2016.8-1.1</span></span>
<span id="cb24-67"><a href="#cb24-67" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [43] rbibutils_2.0      </span></span>
<span id="cb24-68"><a href="#cb24-68" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [44] grid_4.0.2         </span></span>
<span id="cb24-69"><a href="#cb24-69" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [45] Rdpack_2.1.1       </span></span>
<span id="cb24-70"><a href="#cb24-70" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [46] tools_4.0.2        </span></span>
<span id="cb24-71"><a href="#cb24-71" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [47] magrittr_2.0.1     </span></span>
<span id="cb24-72"><a href="#cb24-72" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [48] sass_0.4.2         </span></span>
<span id="cb24-73"><a href="#cb24-73" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [49] patchwork_1.1.1    </span></span>
<span id="cb24-74"><a href="#cb24-74" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [50] tibble_3.0.3       </span></span>
<span id="cb24-75"><a href="#cb24-75" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [51] crayon_1.4.1       </span></span>
<span id="cb24-76"><a href="#cb24-76" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [52] pkgconfig_2.0.3    </span></span>
<span id="cb24-77"><a href="#cb24-77" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [53] ellipsis_0.3.2     </span></span>
<span id="cb24-78"><a href="#cb24-78" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [54] MASS_7.3-53        </span></span>
<span id="cb24-79"><a href="#cb24-79" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [55] data.table_1.13.0  </span></span>
<span id="cb24-80"><a href="#cb24-80" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [56] ggnewscale_0.4.7   </span></span>
<span id="cb24-81"><a href="#cb24-81" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [57] assertthat_0.2.1   </span></span>
<span id="cb24-82"><a href="#cb24-82" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [58] rmarkdown_2.15     </span></span>
<span id="cb24-83"><a href="#cb24-83" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [59] R6_2.5.0           </span></span>
<span id="cb24-84"><a href="#cb24-84" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [60] compiler_4.0.2</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
