<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>maelstRom imprinting tutorial • maelstRom</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="maelstRom imprinting tutorial">
<meta property="og:description" content="maelstRom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>maelstRom imprinting tutorial</h1>
                        <h4 data-toc-skip class="author">Cedric
Stroobandt</h4>
            
            <h4 data-toc-skip class="date">2025-04-08</h4>
      
      
      <div class="hidden name"><code>maelstRom_imprinting_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="maelstrom-for-imprinting-analyses">maelstRom for imprinting analyses<a class="anchor" aria-label="anchor" href="#maelstrom-for-imprinting-analyses"></a>
</h2>
<p>maelstRom includes functionalities for the detection of
(differential) imprinting, which were previously published by our lab
(<span class="citation">Goovaerts et al. (2018)</span>). This extreme
form of allele-specific expression (ASE) requires a seperate and robust
analysis pipeline, as maelstRom’s beta-binomial mixture models (as
introduced in the “maelstRom Allelic Dispersion tutorial”) rely on the
occurrence of observeable heterozygosity in RNAseq data to study ASE.
The latter is absent in fully imprinted loci, and this will results in
e.g. numerical failure during maelstRom’s beta-binomial mixture model
fit, or otherwise categorization as a low-quality locus. This vignette
covers an analysis pipeline retaining these loci and quantifying their
degree of (differential) imprinting.</p>
</div>
<div class="section level2">
<h2 id="a-toy-dataset">A toy dataset<a class="anchor" aria-label="anchor" href="#a-toy-dataset"></a>
</h2>
<p>This vignette utilizes the same toy dataset as the “maelstRom Allelic
Dispersion tutorial”. As such, the code chunk below re-runs the code
found there up to and including “Enabling parallellization”; please
consult this tutorial for an in-depth explanation of input data
collection and pre-processing.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"maelstRom"</span>, package <span class="op">=</span> <span class="st">"maelstRom"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># maelstRom expects lists:</span></span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">caseList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LOC</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>                 <span class="co"># For every locus...</span></span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">ControlCountsToy</span><span class="op">[</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">==</span><span class="va">LOC</span>,<span class="op">]</span> <span class="co"># extract nucleotide counts,</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span>                             <span class="co"># and put it into the list</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LOC</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">CaseCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>                    <span class="co"># Do the same for case data</span></span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">CaseCountsToy</span><span class="op">[</span><span class="va">CaseCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">==</span><span class="va">LOC</span>,<span class="op">]</span></span>
<span>  <span class="va">caseList</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/standard_alleles.html">standard_alleles</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span> </span>
<span><span class="co">#&gt; Warning:</span></span>
<span><span class="co">#&gt; replacing</span></span>
<span><span class="co">#&gt; previous import</span></span>
<span><span class="co">#&gt; 'data.table::yearmon'</span></span>
<span><span class="co">#&gt; by</span></span>
<span><span class="co">#&gt; 'zoo::yearmon'</span></span>
<span><span class="co">#&gt; when loading</span></span>
<span><span class="co">#&gt; 'maelstRom'</span></span>
<span><span class="co">#&gt; Warning:</span></span>
<span><span class="co">#&gt; replacing</span></span>
<span><span class="co">#&gt; previous import</span></span>
<span><span class="co">#&gt; 'data.table::yearqtr'</span></span>
<span><span class="co">#&gt; by</span></span>
<span><span class="co">#&gt; 'zoo::yearqtr'</span></span>
<span><span class="co">#&gt; when loading</span></span>
<span><span class="co">#&gt; 'maelstRom'</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">caseList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref_alleles</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_alleles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">var</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref_count</span> <span class="op">&lt;-</span> <span class="va">interDF</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">==</span><span class="va">interDF</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">var_count</span> <span class="op">&lt;-</span> <span class="va">interDF</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">==</span><span class="va">interDF</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/prior_filter.html">prior_filter</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span>, min_median_cov <span class="op">=</span> <span class="fl">0</span>, </span>
<span>    min_nr_samples <span class="op">=</span> <span class="fl">20</span>, checkref_filter <span class="op">=</span> <span class="cn">FALSE</span>, prior_allelefreq_filter <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    min_PriorAlleleFreq <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co"># Filter out case data if corresponding control data was removed:</span></span>
<span>    <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> </span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="va">NC</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># Number of Cores, CHANGE THIS for a speed-up when working on a linux machine</span></span>
<span></span>
<span><span class="va">NS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span> </span>
<span><span class="va">spl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span>,<span class="va">NC</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">NS</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">NC</span><span class="op">-</span><span class="va">NS</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">ParCTRL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the split input data into a list for parallellisation</span></span>
<span>  <span class="va">ParCTRL</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ParCASE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the split input data into a list for parallellisation</span></span>
<span>  <span class="va">ParCASE</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="metaparameter-estimation">Metaparameter estimation<a class="anchor" aria-label="anchor" href="#metaparameter-estimation"></a>
</h2>
<p>Like in the “maelstRom Allelic Dispersion tutorial”, some
metaparameters need to be calculated (Sequencing Error, and inbreeding
coefficient); the code and explanations here is identical to that
tutorial, with the exception that we also update the control data
(<code>ParCTRL</code> and <code>controlList</code>) using
<code>AllelicMeta_est_par</code>’s output, to include a preliminary
per-locus reference allele fraction estimate among the population’s
genotypes. This can be used as a prior filter during imprinting
detection.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">GenoFinData</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParCTRL</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va"><a href="../reference/AllelicMeta_est_par.html">AllelicMeta_est_par</a></span>,</span>
<span>  MinAllele_filt <span class="op">=</span> <span class="fl">0.15</span>, SE_filt <span class="op">=</span> <span class="fl">0.035</span>, NumSamp_filt <span class="op">=</span> <span class="fl">20</span>, MedianCov_filt <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Update control list with a prior allele frequency estimate:</span></span>
<span><span class="va">ParCTRL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SE_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">F_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">SEmedian</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">SE_vec</span><span class="op">)</span></span>
<span><span class="va">Fmedian</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">F_vec</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">SEmedian</span>, <span class="va">Fmedian</span><span class="op">)</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>0.00248190279214063, 0.00872807570598289,</p>
</div>
<div class="section level2">
<h2 id="detecting-imprinted-and-differentially-imprinted-loci">Detecting imprinted and differentially imprinted loci<a class="anchor" aria-label="anchor" href="#detecting-imprinted-and-differentially-imprinted-loci"></a>
</h2>
<p>Imprinting analyses in maelstRom are quite different from the
analyses up until now in that they don’t rely on beta-binomial modelling
and make different assumptions about the data. They are, generally
speaking, a bit more robust because imprinting is usually an extreme
phenomenon, i.e. heterozygous samples of imprinted loci will almost
always “seem like” homozygous samples because one of the alleles has
been silenced almost entirely. Partial imprinting can exist and
maelstRom can detect it, but this phenomenon is not modelled as
thoroughly as the previous ones in this vignette.</p>
<p>First off we assume samples show no AB and are conform
Hardy-Weinberg-Equilibrium while taking imprinting into account, which
is tested by <code>symmetry_gof</code>. It does this by assessing
whether the number of samples with a reference allele fraction &gt; 0.5
(and so also the number with a fraction &lt;= 0.5) is equal to that
expected under HWE assuming no AB using a chi square test. This may seem
kind of heuristic at first and prone to fail if there is even a little
AB, but remember we want to mainly retain imprinted loci, meaning
heterozygous samples should behave as one of both homozygotes (and so
clearly feature reference allele fractions &gt; 0.5 or &lt;=0.5) so this
filter is very fitting for retaining HWE-conform samples that are at the
same time likely to showcase imprinting. A good filter setting is
requiring the p-value to be at least 0.05.</p>
<p>After applying these filters, <code>imprinting_est</code> detects
imprinted loci in the remaining control samples by first assuming an
unshifted binomial mixture model, then splitting up the heterozygous
samples in two separate groups according to a degree of imprinting <span class="math inline">\(i\)</span> and varying this <span class="math inline">\(i\)</span> from 0 to 1 in a for-loop, calculating
likelihoods along the way to perform a likelihood-ratio test to check
for differential imprinting (i.e. compare the most likely model with the
non-imprinted one; <span class="math inline">\(i=0\)</span>). For
completeness, the imprinted PMF is (see <span class="citation">Goovaerts
et al. (2018)</span> for a full elaboration):</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(&amp;counts_{ref}|counts_{total}) =  P_{rr} * {\tt
pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=1-SE) \ + \\
&amp; 0.5P_{rv} * {\tt pbinom} \left( x=counts_{ref} \ |\
n=counts_{total},
p=\frac{0.5-\frac{i}{2}}{1-\frac{i}{2}}(1-SE)+\frac{0.5}{1-\frac{i}{2}}SE\right)
\ + \\
&amp; 0.5P_{rv} * {\tt pbinom} \left( x=counts_{ref} \ |\
n=counts_{total},
p=\frac{0.5-\frac{i}{2}}{1-\frac{i}{2}}SE+\frac{0.5}{1-\frac{i}{2}}(1-SE)\right)
\ + \\
&amp; P_{vv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=SE)
\end{aligned}
\]</span></p>
<p>After retaining only significantly and sufficiently imprinted loci
(5% FDR level, degree of imprinting at least 0.6 and median degree of
imprinting at least 0.8 for robustness; see
<code>median_imprinting</code> for the latter’s calculation) using the
<code>final_filter</code> function, we use <code><a href="../reference/LOItest_logreg.html">LOItest_logreg()</a></code>
to test for differential imprinting. More specifically, logistic
regression is used to evaluate whether case samples feature a
significantly higher number of apparently heterozygous samples than
control samples (by considering the least expressed allele as a success;
more details can be found on the function’s help page. Note that the
logistic regression approach takes into account variable coverage
between samples, contrasting our previously published strategy by <span class="citation">Goovaerts et al. (2018)</span>). If this is true, it
indicates that a significant fraction of heterozygous samples in tumor
tissue “lost their imprinting” and actually started expressing both
alleles again. We call these loci “differentially imprinted”, as we
reserve the term “loss-of-imprinting” for the phenomenon where
differential imprinting co-occurs with increased total expression,
indicative of re-expression of the silenced allele. Comments in the code
below provide additional insight into the imprinting analysis
pipeline:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform filtering using symmetry_gof()</span></span>
<span><span class="co"># Notice we use allelefreq_prel as input in this function, which is the allele frequency as </span></span>
<span><span class="co"># estimated by an UNSHIFTED binomial mixture model using AllelicMeta_est() earlier in this </span></span>
<span><span class="co"># vignette, which is what symmetry_GOF assumes as well so it's only fitting.</span></span>
<span><span class="co"># We also enforce that this allelefreq_prel can not be more extreme than 0.15 or 0.85, </span></span>
<span><span class="co"># because detecting imprinting would be very hard otherwise.</span></span>
<span><span class="va">ImprData</span> <span class="op">&lt;-</span> <span class="va">controlList</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LOC</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">allelefreq_prel</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;=</span> <span class="fl">0.15</span> <span class="op">||</span> </span>
<span>      <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">allelefreq_prel</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fl">0.15</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sym</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/symmetry_gof.html">symmetry_gof</a></span><span class="op">(</span><span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span>, </span>
<span>      <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span>, <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">allelefreq_prel</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sym</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co"># Detect imprinted control loci</span></span>
<span><span class="va">impr_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LOC</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ImprData</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">i_results</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/imprinting_est.html">imprinting_est</a></span><span class="op">(</span><span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span>, <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span>, </span>
<span>                allelefreq <span class="op">=</span> <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">allelefreq_prel</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, SE <span class="op">=</span> <span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span><span class="op">)</span></span>
<span>  <span class="co"># An additional robustified "median imprinting" across samples to be used as possible </span></span>
<span>  <span class="co"># additional filter criterion:</span></span>
<span>  <span class="va">med_imp</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/median_imprinting.html">median_imprinting</a></span><span class="op">(</span><span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span>, <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span>, </span>
<span>              allelefreq <span class="op">=</span> <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">allelefreq_prel</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span><span class="op">)</span></span>
<span>  <span class="va">results_z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"position"</span> <span class="op">=</span> <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Locus</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"LRT"</span> <span class="op">=</span> <span class="va">i_results</span><span class="op">$</span><span class="va">LRT</span>, </span>
<span>    <span class="st">"p"</span> <span class="op">=</span> <span class="va">i_results</span><span class="op">$</span><span class="va">p_value</span>, <span class="st">"estimated.i"</span> <span class="op">=</span> <span class="va">i_results</span><span class="op">$</span><span class="va">est_i</span>, <span class="st">"allele.frequency"</span> <span class="op">=</span> </span>
<span>    <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">allelefreq_prel</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"reference"</span> <span class="op">=</span> <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"variant"</span> <span class="op">=</span> </span>
<span>    <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"med_cov"</span> <span class="op">=</span> <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">coverage</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"nr_samples"</span> <span class="op">=</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">"GOF"</span> <span class="op">=</span> <span class="va">i_results</span><span class="op">$</span><span class="va">GOF_likelihood</span>, <span class="st">"symmetry"</span> <span class="op">=</span> </span>
<span>    <span class="va">ImprData</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sym</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"med_impr"</span> <span class="op">=</span> <span class="va">med_imp</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">impr_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">impr_res</span>, <span class="va">results_z</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Retain significantly imprinted loci (5% FDR) utilizing some additional filters, amongst</span></span>
<span><span class="co"># which a custom  Goodness-Of-Fit which more or less corresponds to a locus' likelihood of</span></span>
<span><span class="co"># the imprinted model*coverage; 0.8 is a good cutoff. Other filter criteria are imprinting </span></span>
<span><span class="co"># (0.6) and median imprinting (0.8)</span></span>
<span><span class="va">impr_res_FIN</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/final_filter.html">final_filter</a></span><span class="op">(</span>data_hash<span class="op">=</span><span class="cn">NULL</span>, <span class="va">impr_res</span>, results_wd<span class="op">=</span><span class="cn">NULL</span>, gof_filt <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  med_impr_filt <span class="op">=</span> <span class="fl">0.8</span>, i_filt <span class="op">=</span> <span class="fl">0.6</span>, adj_p_filt <span class="op">=</span> <span class="fl">0.05</span>, file_all <span class="op">=</span> <span class="cn">FALSE</span>, file_impr <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  file_all_counts <span class="op">=</span> <span class="cn">FALSE</span>, file_impr_counts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Amongst actually imprinted loci, detect differential expression in case data</span></span>
<span><span class="va">pos_impr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">impr_res_FIN</span><span class="op">$</span><span class="va">position</span><span class="op">)</span></span>
<span><span class="va">p_DI_df</span> <span class="op">&lt;-</span> <span class="va">impr_res_FIN</span></span>
<span><span class="va">p_DI_df</span><span class="op">$</span><span class="va">DI_pval</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LOC</span> <span class="kw">in</span> <span class="va">pos_impr</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">CData</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">TData</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">p_DI</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/LOItest_logreg.html">LOItest_logreg</a></span><span class="op">(</span><span class="va">CData</span><span class="op">$</span><span class="va">ref_count</span>, <span class="va">CData</span><span class="op">$</span><span class="va">var_count</span>, </span>
<span>                                  <span class="va">TData</span><span class="op">$</span><span class="va">ref_count</span>, <span class="va">TData</span><span class="op">$</span><span class="va">var_count</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>  <span class="va">p_DI_df</span><span class="op">$</span><span class="va">DI_pval</span><span class="op">[</span><span class="va">p_DI_df</span><span class="op">$</span><span class="va">position</span> <span class="op">==</span> <span class="va">LOC</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p_DI</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We can take a look at the loci that are significantly and
sufficiently imprinted (i.e. the output of <code>final_filter</code>
from above):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">impr_res_FIN</span><span class="op">$</span><span class="va">position</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>Locus76, Locus77, Locus172, Locus173, Locus184, Locus185,
Locus186,</p>
<p>From this set, a part is actually differentially imprinted in case
samples, which is indicated by the p-value in the <code>DI_pval</code>
column in the <code>p_DI_df</code> dataframe. We can control
loss-of-imprinting detection, for a change, at the 5% FWER level
(actually feasible here due to the small number of actually imprinted
loci remaining, though we’re looking at a small toy dataset here; this
isn’t necessarily the case in larger experiments):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">p_DI_df</span><span class="op">$</span><span class="va">position</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">p_DI_df</span><span class="op">$</span><span class="va">DI_pval</span>, method <span class="op">=</span> <span class="st">"holm"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>Locus172, Locus173,</p>
<p>We can plot histograms of both the control and case allele fractions;
locus 202 is obviously imprinted in controls and loses a lot of this
imprinting in cases (controls are plotted using
<code>maelstRom_imprintplot</code> to visualize imprinting results, but
since an imprinting-estimation never happens on the case data - we just
detect a change in heterozygosity via logistic regression - we simply
make a histogram for cases instead).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HistCTRL</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_imprintplot.html">maelstRom_imprintplot</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="st">"Locus172"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span>, </span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="st">"Locus172"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span>, </span>
<span>  allelefreq <span class="op">=</span> <span class="va">impr_res_FIN</span><span class="op">$</span><span class="va">allele.frequency</span><span class="op">[</span><span class="va">impr_res_FIN</span><span class="op">$</span><span class="va">position</span><span class="op">==</span><span class="st">"Locus172"</span><span class="op">]</span>,</span>
<span>  impr <span class="op">=</span> <span class="va">impr_res_FIN</span><span class="op">$</span><span class="va">estimated.i</span><span class="op">[</span><span class="va">impr_res_FIN</span><span class="op">$</span><span class="va">position</span><span class="op">==</span><span class="st">"Locus172"</span><span class="op">]</span>,</span>
<span>  SE <span class="op">=</span> <span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span>, plot_NoImpr <span class="op">=</span> <span class="cn">TRUE</span>, SplitPeaks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Control"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">RatioCASE</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="st">"Locus172"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span> <span class="op">/</span> </span>
<span>  <span class="op">(</span><span class="va">caseList</span><span class="op">[[</span><span class="st">"Locus172"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span> <span class="op">+</span> <span class="va">caseList</span><span class="op">[[</span><span class="st">"Locus172"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span><span class="op">)</span></span>
<span><span class="va">CASEDat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"Ratio"</span> <span class="op">=</span> <span class="va">RatioCASE</span><span class="op">)</span></span>
<span><span class="va">HistCASE</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">CASEDat</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">Ratio</span><span class="op">)</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Reference allele fraction"</span>, y<span class="op">=</span><span class="st">"Frequency"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Case"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">HistCTRL</span>, <span class="va">HistCASE</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="maelstRom_imprinting_tutorial_files/figure-html/unnamed-chunk-6-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 10 x64 (build 19045)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Dutch_Belgium.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Dutch_Belgium.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Dutch_Belgium.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                  </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Dutch_Belgium.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Brussels</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats    </span></span>
<span><span class="co">#&gt; [2] graphics </span></span>
<span><span class="co">#&gt; [3] grDevices</span></span>
<span><span class="co">#&gt; [4] utils    </span></span>
<span><span class="co">#&gt; [5] datasets </span></span>
<span><span class="co">#&gt; [6] methods  </span></span>
<span><span class="co">#&gt; [7] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.5       </span></span>
<span><span class="co">#&gt;  [2] xfun_0.45          </span></span>
<span><span class="co">#&gt;  [3] bslib_0.7.0        </span></span>
<span><span class="co">#&gt;  [4] ggplot2_3.5.1      </span></span>
<span><span class="co">#&gt;  [5] htmlwidgets_1.6.4  </span></span>
<span><span class="co">#&gt;  [6] lattice_0.22-5     </span></span>
<span><span class="co">#&gt;  [7] numDeriv_2016.8-1.1</span></span>
<span><span class="co">#&gt;  [8] vctrs_0.6.5        </span></span>
<span><span class="co">#&gt;  [9] tools_4.3.3        </span></span>
<span><span class="co">#&gt; [10] Rdpack_2.6.1       </span></span>
<span><span class="co">#&gt; [11] generics_0.1.3     </span></span>
<span><span class="co">#&gt; [12] parallel_4.3.3     </span></span>
<span><span class="co">#&gt; [13] rgl_1.3.1          </span></span>
<span><span class="co">#&gt; [14] tibble_3.2.1       </span></span>
<span><span class="co">#&gt; [15] fansi_1.0.6        </span></span>
<span><span class="co">#&gt; [16] highr_0.11         </span></span>
<span><span class="co">#&gt; [17] pkgconfig_2.0.3    </span></span>
<span><span class="co">#&gt; [18] data.table_1.16.0  </span></span>
<span><span class="co">#&gt; [19] ggnewscale_0.5.0   </span></span>
<span><span class="co">#&gt; [20] desc_1.4.3         </span></span>
<span><span class="co">#&gt; [21] alabama_2023.1.0   </span></span>
<span><span class="co">#&gt; [22] lifecycle_1.0.4    </span></span>
<span><span class="co">#&gt; [23] compiler_4.3.3     </span></span>
<span><span class="co">#&gt; [24] farver_2.1.2       </span></span>
<span><span class="co">#&gt; [25] stringr_1.5.1      </span></span>
<span><span class="co">#&gt; [26] textshaping_0.4.0  </span></span>
<span><span class="co">#&gt; [27] munsell_0.5.1      </span></span>
<span><span class="co">#&gt; [28] carData_3.0-5      </span></span>
<span><span class="co">#&gt; [29] matlib_0.9.6       </span></span>
<span><span class="co">#&gt; [30] htmltools_0.5.8    </span></span>
<span><span class="co">#&gt; [31] sass_0.4.9         </span></span>
<span><span class="co">#&gt; [32] hash_2.2.6.3       </span></span>
<span><span class="co">#&gt; [33] yaml_2.3.9         </span></span>
<span><span class="co">#&gt; [34] gmp_0.7-5          </span></span>
<span><span class="co">#&gt; [35] pillar_1.9.0       </span></span>
<span><span class="co">#&gt; [36] pkgdown_2.0.9      </span></span>
<span><span class="co">#&gt; [37] car_3.1-2          </span></span>
<span><span class="co">#&gt; [38] jquerylib_0.1.4    </span></span>
<span><span class="co">#&gt; [39] MASS_7.3-60.0.1    </span></span>
<span><span class="co">#&gt; [40] cachem_1.1.0       </span></span>
<span><span class="co">#&gt; [41] abind_1.4-5        </span></span>
<span><span class="co">#&gt; [42] gtools_3.9.5       </span></span>
<span><span class="co">#&gt; [43] tidyselect_1.2.1   </span></span>
<span><span class="co">#&gt; [44] digest_0.6.35      </span></span>
<span><span class="co">#&gt; [45] stringi_1.8.3      </span></span>
<span><span class="co">#&gt; [46] dplyr_1.1.4        </span></span>
<span><span class="co">#&gt; [47] purrr_1.0.2        </span></span>
<span><span class="co">#&gt; [48] labeling_0.4.3     </span></span>
<span><span class="co">#&gt; [49] fastmap_1.2.0      </span></span>
<span><span class="co">#&gt; [50] grid_4.3.3         </span></span>
<span><span class="co">#&gt; [51] colorspace_2.1-0   </span></span>
<span><span class="co">#&gt; [52] cli_3.6.2          </span></span>
<span><span class="co">#&gt; [53] magrittr_2.0.3     </span></span>
<span><span class="co">#&gt; [54] patchwork_1.2.0    </span></span>
<span><span class="co">#&gt; [55] base64enc_0.1-3    </span></span>
<span><span class="co">#&gt; [56] utf8_1.2.4         </span></span>
<span><span class="co">#&gt; [57] withr_3.0.0        </span></span>
<span><span class="co">#&gt; [58] scales_1.3.0       </span></span>
<span><span class="co">#&gt; [59] rmarkdown_2.27     </span></span>
<span><span class="co">#&gt; [60] gridExtra_2.3      </span></span>
<span><span class="co">#&gt; [61] ragg_1.3.2         </span></span>
<span><span class="co">#&gt; [62] zoo_1.8-12         </span></span>
<span><span class="co">#&gt; [63] memoise_2.0.1      </span></span>
<span><span class="co">#&gt; [64] evaluate_0.24.0    </span></span>
<span><span class="co">#&gt; [65] knitr_1.48         </span></span>
<span><span class="co">#&gt; [66] rbibutils_2.2.16   </span></span>
<span><span class="co">#&gt; [67] rlang_1.1.3        </span></span>
<span><span class="co">#&gt; [68] Rcpp_1.0.12        </span></span>
<span><span class="co">#&gt; [69] xtable_1.8-4       </span></span>
<span><span class="co">#&gt; [70] glue_1.7.0         </span></span>
<span><span class="co">#&gt; [71] rstudioapi_0.16.0  </span></span>
<span><span class="co">#&gt; [72] maelstRom_1.1.11   </span></span>
<span><span class="co">#&gt; [73] jsonlite_1.8.8     </span></span>
<span><span class="co">#&gt; [74] R6_2.5.1           </span></span>
<span><span class="co">#&gt; [75] systemfonts_1.1.0  </span></span>
<span><span class="co">#&gt; [76] fs_1.6.4</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-goovaerts2018" class="csl-entry">
Goovaerts, Tine, Sandra Steyaert, Chari A Vandenbussche, Jeroen Galle,
Olivier Thas, Wim Van Criekinge, and Tim De Meyer. 2018. <span>“A
Comprehensive Overview of Genomic Imprinting in Breast and Its
Deregulation in Cancer.”</span> <em>Nature Communications</em> 9 (1):
1–14.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
