<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>maelstRom Allelic Dispersion tutorial • maelstRom</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="maelstRom Allelic Dispersion tutorial">
<meta property="og:description" content="maelstRom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>maelstRom Allelic Dispersion tutorial</h1>
                        <h4 data-toc-skip class="author">Cedric
Stroobandt</h4>
            
            <h4 data-toc-skip class="date">2024-10-29</h4>
      
      
      <div class="hidden name"><code>maelstRom_Allelic_Dispersion_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="maelstrom-modeller-of-allele-specific-transcriptomics">maelstRom: Modeller of Allele-Specific Transcriptomics<a class="anchor" aria-label="anchor" href="#maelstrom-modeller-of-allele-specific-transcriptomics"></a>
</h2>
<p>Welcome to the tutorial vignette of the Modeller of Allele-Specific
Transcriptomics, maelstRom! This R package offers extensive
likelihood-based modelling of Allele-Specific Expression (ASE)
phenomena, which includes statistical tests for their detection
(e.g. genome-wide cis-eqtl mapping) or differential occurrence
(e.g. copy-number alterations and/or hypermethylated loci in diseases
such as cancer). All of these functionalities rely solely on regular
RNAseq data, without the requirement of additional experiments such as
genotyping, nor specialized data-types such as paired RNAseq.</p>
<p>This vignette is a step-by-step guide for a basic ASE-analysis using
maelstRom starting from raw allelic counts, from input pre-processing up
until vizualization of final results. For instructions on installing
maelstRom, please consult maelstRom’s github home page at <a href="https://biobix.github.io/maelstRom/" class="external-link uri">https://biobix.github.io/maelstRom/</a>.</p>
</div>
<div class="section level2">
<h2 id="a-toy-dataset">A toy dataset<a class="anchor" aria-label="anchor" href="#a-toy-dataset"></a>
</h2>
<p>maelstRom relies on per-locus per-sample nucleotide counts (A/T/C/G),
relying on single-nucleotide polymorphisms (SNPs) to differentiate
between alleles. Such counts are easily obtained from RNAseq BAM or SAM
files, either by custom scripting or extisting functionalities such as
SAMtool’s mpileup, or GATK’s CollectAllelicCounts (<a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037594071-CollectAllelicCounts" class="external-link uri">https://gatk.broadinstitute.org/hc/en-us/articles/360037594071-CollectAllelicCounts</a>);
the latter already filters these nucleotide counts down to just one
reference- and variant-allele.</p>
<p>A toy data example of such input data is included in maelstRom, and
is used throughout the remainder of this vignette. This toy data
consists of both a hypothetical control (healthy) and case (diseased)
dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">maelstRom</span><span class="op">)</span></span>
<span><span class="co">#&gt; Package 'maelstRom' version 1.1.11</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"maelstRom"</span>, package <span class="op">=</span> <span class="st">"maelstRom"</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ControlCountsToy</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">Gene</th>
<th align="left">Locus</th>
<th align="left">Sample</th>
<th align="left">ref_alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">5295</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample1</td>
<td align="left">G/A/C</td>
<td align="right">63</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">74</td>
</tr>
<tr class="even">
<td align="left">5296</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample2</td>
<td align="left">G/A/C</td>
<td align="right">214</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">5297</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample3</td>
<td align="left">G/A/C</td>
<td align="right">249</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">5298</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample4</td>
<td align="left">G/A/C</td>
<td align="right">54</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">62</td>
</tr>
<tr class="odd">
<td align="left">5299</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample5</td>
<td align="left">G/A/C</td>
<td align="right">80</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">87</td>
</tr>
<tr class="even">
<td align="left">5300</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample6</td>
<td align="left">G/A/C</td>
<td align="right">57</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">41</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CaseCountsToy</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">Gene</th>
<th align="left">Locus</th>
<th align="left">Sample</th>
<th align="left">ref_alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">11070</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample129</td>
<td align="left">G/A/C</td>
<td align="right">194</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">321</td>
</tr>
<tr class="even">
<td align="left">11071</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample130</td>
<td align="left">G/A/C</td>
<td align="right">74</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">62</td>
</tr>
<tr class="odd">
<td align="left">11072</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample131</td>
<td align="left">G/A/C</td>
<td align="right">62</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">49</td>
</tr>
<tr class="even">
<td align="left">11073</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample132</td>
<td align="left">G/A/C</td>
<td align="right">333</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">11074</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample133</td>
<td align="left">G/A/C</td>
<td align="right">226</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">137</td>
</tr>
<tr class="even">
<td align="left">11075</td>
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">Sample134</td>
<td align="left">G/A/C</td>
<td align="right">171</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>This toy data contains nucleotide counts spanning 200 SNP loci across
71 genes (a gene can contain multiple SNP loci), covered by 128 control-
and 268 case-samples:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] 71</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] 200</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] 128</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">CaseCountsToy</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] 268</p>
</div>
<div class="section level2">
<h2 id="input-pre-processing">Input pre-processing<a class="anchor" aria-label="anchor" href="#input-pre-processing"></a>
</h2>
<p>The nucleotide counts above are formated as one giant dataframe. In
order to use them as a maelstRom input, these need to be split into a
list of seperate dataframes per locus, as maelstRoms models ASE on a
per-locus level. In general, every maelstRom function has its own help
page detailing the expected format of its input.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># maelstRom expects lists:</span></span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">caseList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LOC</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>                 <span class="co"># For every locus...</span></span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">ControlCountsToy</span><span class="op">[</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">==</span><span class="va">LOC</span>,<span class="op">]</span> <span class="co"># extract nucleotide counts,</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span>                             <span class="co"># and put it into the list</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LOC</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">CaseCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>                    <span class="co"># Do the same for case data</span></span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">CaseCountsToy</span><span class="op">[</span><span class="va">CaseCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">==</span><span class="va">LOC</span>,<span class="op">]</span></span>
<span>  <span class="va">caseList</span><span class="op">[[</span><span class="va">LOC</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="determine-one-reference-and-one-variant-allele">Determine one reference and one variant allele<a class="anchor" aria-label="anchor" href="#determine-one-reference-and-one-variant-allele"></a>
</h2>
<p>maelstRom’s models are built on the beta-binomial distrubution. As
the name implies, this distribution models repeated observations of a
binary outcome and is thus limited to two possible alleles per locus.
While this may seem overly restrictive, it’s rather commonplace for SNPs
in human populations to exhibit, at most, two commonly occurring SNP
variants. But in the case that three or even four SNP variants should be
considered, pairwise maelstRom fits of said variants are an option;
beta-multinomial models, which could model all possible variants at
once, are not a feature of the current version of maelstRom due to their
limited added value in practical research comparative to the complexity
of their implementation.</p>
<p><code><a href="../reference/standard_alleles.html">standard_alleles()</a></code> takes our list of per-locus
dataframes as an input, and picks the most- and second-most occurring
nucleotide (at population-level) as reference- and variant nucleotide
respectively, adding the resulting choice and the corresponding
reference- and variant-count to each individual dataframe. If a
“ref_allele” column is provided in these dataframes (as is the case for
our toy data), <code><a href="../reference/standard_alleles.html">standard_alleles()</a></code> will only consider the
nucleotides listed therein as possible reference- or variant candidates.
This can be used to e.g. integrate expert knowledge about the current
population (it is known only the listed nucleotides should occur at a
certain location, and anything else is some sort of artefact), or limit
all further analyses to certain alleles of particular interest.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/standard_alleles.html">standard_alleles</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="st">"Locus10"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="6%">
<col width="7%">
<col width="9%">
<col width="9%">
<col width="14%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="12%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">Gene</th>
<th align="left">Locus</th>
<th align="left">Sample</th>
<th align="left">ref_alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">ref</th>
<th align="left">var</th>
<th align="right">ref_count</th>
<th align="right">var_count</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">6254</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample1</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">113</td>
<td align="right">176</td>
<td align="right">1</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">176</td>
<td align="right">113</td>
</tr>
<tr class="even">
<td align="left">6255</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample2</td>
<td align="left">C/T</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">630</td>
<td align="right">1</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">630</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">6256</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample3</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">202</td>
<td align="right">273</td>
<td align="right">3</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">273</td>
<td align="right">202</td>
</tr>
<tr class="even">
<td align="left">6257</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample4</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">381</td>
<td align="right">1</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">381</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">6258</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample5</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">683</td>
<td align="right">0</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">683</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">6259</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample6</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">418</td>
<td align="right">2</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">418</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>While <code><a href="../reference/standard_alleles.html">standard_alleles()</a></code> could be run seperately on the
case data, it makes sense (from both a data-analytical and biological
viewpoint) to pick the same reference- and variant nucleotide as the
control data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">caseList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref_alleles</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_alleles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">var</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref_count</span> <span class="op">&lt;-</span> <span class="va">interDF</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">==</span><span class="va">interDF</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">var_count</span> <span class="op">&lt;-</span> <span class="va">interDF</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">==</span><span class="va">interDF</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">caseList</span><span class="op">[[</span><span class="st">"Locus10"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="7%">
<col width="7%">
<col width="9%">
<col width="11%">
<col width="14%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="11%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">Gene</th>
<th align="left">Locus</th>
<th align="left">Sample</th>
<th align="left">ref_alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">ref</th>
<th align="left">var</th>
<th align="right">ref_count</th>
<th align="right">var_count</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">13114</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample129</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">457</td>
<td align="right">0</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">457</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">13115</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample130</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">382</td>
<td align="right">0</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">382</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">13116</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample131</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">111</td>
<td align="right">177</td>
<td align="right">0</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">177</td>
<td align="right">111</td>
</tr>
<tr class="even">
<td align="left">13117</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample132</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">547</td>
<td align="right">0</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">547</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">13118</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample133</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">332</td>
<td align="right">1</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">332</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">13119</td>
<td align="left">Gene4</td>
<td align="left">Locus10</td>
<td align="left">Sample134</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">33</td>
<td align="right">88</td>
<td align="right">0</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">88</td>
<td align="right">33</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="prior-filtering">Prior filtering<a class="anchor" aria-label="anchor" href="#prior-filtering"></a>
</h2>
<p>Some rudimentary prior filtering is always worth considering, as it
avoids wasting time on trying to fit models on inherently low-quality
data. maelstRom’s <code><a href="../reference/prior_filter.html">prior_filter()</a></code> provides various filtering
options (see its help page; minimal median coverage across a locus,
minimal number of samples, minimal minor allele count frequency i.e. the
% abundance of the variant allele over all RNAseq counts, etc.). For
this tutorial, we use a mild prior filter, simply requiring at least 20
control samples for a locus to be retained. We perform no prior
filtering on the case data itself, but do remove loci that were removed
in the control-data, as the main use of the case-data is differential
ASE when compared to controls.</p>
<p>This 20-control-sample requirement removes 5 loci, corresponding to
two genes, from the toy dataset:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/prior_filter.html">prior_filter</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span>, min_median_cov <span class="op">=</span> <span class="fl">0</span>, </span>
<span>    min_nr_samples <span class="op">=</span> <span class="fl">20</span>, checkref_filter <span class="op">=</span> <span class="cn">FALSE</span>, prior_allelefreq_filter <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    min_PriorAlleleFreq <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co"># Filter out case data if corresponding control data was removed:</span></span>
<span>    <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> </span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># These loci were filtered out:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCountsToy</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>Locus30, Locus31, Locus187, Locus188, Locus189,</p>
</div>
<div class="section level2">
<h2 id="enabling-parallellization">Enabling parallellization<a class="anchor" aria-label="anchor" href="#enabling-parallellization"></a>
</h2>
<p>maelstRom can run for quite some time on larger datasets (especially
when applied genome-wide), so this tutorial illustrates (optional)
parallelization using R’s <code>parallel</code> package. This package
allows the use of multiple cores for maelstRom’s computations, at least
on Linux (on e.g. a local windows installation, it only allows for one
core). While the number of cores <code>NC</code> is kept to 1 in the
code below, picking a higher number should allow for a proportionate
speed-up.</p>
<p>The following code chunk splits maelstRom’s input into a list of
<code>NC</code> elements, which is required for parallelization using
the <code>parallel</code> package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">NC</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># Number of Cores, CHANGE THIS for a speed-up when working on a linux machine</span></span>
<span></span>
<span><span class="va">NS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span> </span>
<span><span class="va">spl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span>,<span class="va">NC</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">NS</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">NC</span><span class="op">-</span><span class="va">NS</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">ParCTRL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the split input data into a list for parallellisation</span></span>
<span>  <span class="va">ParCTRL</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ParCASE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the split input data into a list for parallellisation</span></span>
<span>  <span class="va">ParCASE</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="metaparameter-estimation">Metaparameter estimation<a class="anchor" aria-label="anchor" href="#metaparameter-estimation"></a>
</h2>
<p>maelstRom’s ASE models rely on two population-wide (constant)
metaparameters: the inbreeding coefficient <code>F</code> of the
population under study, and the sequencing error rate <code>SE</code>
controlling the occurrence “faulty” nuceotide counts in homozygous
individuals (i.e. counts of the allele for which they are not
homozygous). These can very well be set at values based on expert
knowledge (e.g. most natural populations should be approcimately
panmictic, i.e. exhibit an inbreeding coefficient of zero) and/or the
used sequencing technogy’s and alignment algorithm’s propensity to
report faulty nucleotides.</p>
<p>Nevertheless, maelstRom has a functionality to quickly and robustly
estimate these parameters from the supplied data, which can be used in
the abscence of prior knowledge or, even should it be available, is a
safe option anyway (as both a sanity check, or to incorporate error
sources and data biases which were possibly overlooked in a would-be
prior estimate). This happens by fitting the following extremely basic
binomial mixture model on the supplied control data:</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(counts_{ref}|counts_{total}) = &amp;\ \phi_{rr} * {\tt
pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=1-SE) \ + \\
&amp;\ \phi_{rv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total},
p=0.5) \ + \\
&amp;\ \phi_{vv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total},
p=SE) \\
\end{aligned}
\]</span></p>
<p>With <span class="math inline">\((\phi_{rr}, \phi_{rv},
\phi_{vv})\)</span> to-be-fitted genotype frequencies from which the
inbreeding coefficient can be estimated, and SE the sequencing error
rate estimate. This model does not incorporate any ASE at all (perfectly
50-50 allelic expression in the heterozygous peak, no variance beyond
technical binomial sampling variance allowed) but should be sufficient
to produce robuts, median-based metaparameter estimates so long as it
fits a large portion of the supplied data somewhat reasonably.</p>
<p>The following code chunk uses <code><a href="../reference/AllelicMeta_est_par.html">AllelicMeta_est_par()</a></code> to
produce the wanted estimates, and to check for the aforementioned
“reasonable fit” only includes control loci with at least 20 available
samples, having a median coverage of at least 4 across these samples,
not returning an estimated SE over 0.035 (as this is unreasonably high
and most likely indicates low-quality data or a failed model fit) and
having an estimated minor allele frequency of at least 0.15 based on the
above mixture model fit (anything lowes implies a rather low occurrence
of at least one of the possible genotypes, and so less reliable
estimates of the population’s metaparameters):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">GenoFinData</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParCTRL</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va"><a href="../reference/AllelicMeta_est_par.html">AllelicMeta_est_par</a></span>,</span>
<span>  <span class="co"># Filter criteria mentioned in the text above:</span></span>
<span>  MinAllele_filt <span class="op">=</span> <span class="fl">0.15</span>, SE_filt <span class="op">=</span> <span class="fl">0.035</span>, NumSamp_filt <span class="op">=</span> <span class="fl">20</span>, MedianCov_filt <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract metaparameter estimates from the parallelly-produced output:</span></span>
<span><span class="va">SE_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">F_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SEmedian</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">SE_vec</span><span class="op">)</span></span>
<span><span class="va">Fmedian</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">F_vec</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#estimates of resp. sequencing error rate and inbreeding coefficient:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">SEmedian</span>, <span class="va">Fmedian</span><span class="op">)</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span>  </span></code></pre></div>
<p>0.00248190279214063, 0.00872807570598289,</p>
<p>These metaparameter estimates seem reasonable. Please note that an
extremely low sequencing error metaparameter would make fitting of
homozygous peaks later down the pipeline way too strict, failing to
accomodate for even a single faulty read. As such, <em>setting SE below
0.002 is not recommended</em>. An inbreeding coefficient close to zero
is a sign of a panmictic population and commonplace when working on
human data, but it could be more extreme in specialized experimental
setups.</p>
</div>
<div class="section level2">
<h2 id="maelstroms-ase-fit-allelic-bias-detection">maelstRom’s ASE fit &amp; allelic bias detection<a class="anchor" aria-label="anchor" href="#maelstroms-ase-fit-allelic-bias-detection"></a>
</h2>
<p>maelstRom’s main functionality models allele-specific counts as a
beta-binomial mixture of three possible genotypes. This allows for both
Allelic Bias (AB), i.e. unequal allelic expression in heterozygotes, via
its <span class="math inline">\(\pi\)</span>-parameter (beta-binomial
mean); and Allelic Dispersion (AD), i.e. the population-level
variability in allelic expression, via its <span class="math inline">\(\theta_{het}\)</span>-parameter (beta-binomial
overdispersion):</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(counts_{ref}|counts_{total}) = &amp;\ \phi_{rr} * {\tt
pBetaBinom}(x=counts_{ref} \ |\ n=counts_{total}, \pi=1-SE,
\theta=\theta_{hom}) \ + \\
&amp;\ \phi_{rv} * {\tt pBetaBinom}(x=counts_{ref} \ |\
n=counts_{total}, \pi=\pi_{het}, \theta=\theta_{het}) \ + \\
&amp;\ \phi_{vv} * {\tt pBetaBinom}(x=counts_{ref} \ |\
n=counts_{total}, \pi=SE, \theta=\theta_{hom}) \\
\end{aligned}
\]</span></p>
<p>With the remaining parameters being: <span class="math inline">\((\phi_{rr}, \phi_{rv}, \phi_{vv})\)</span> the
fitted genotype frequencies, and <span class="math inline">\(\theta_{hom}\)</span> the beta-binomial
overdispersion parameter in homozygotes. By fitting the former through
expectation-maximization, which iteratively assigns each sample to the
three possible genotypes, then estimates all genotype-specific
beta-binomial parameters, maelstRom does not require additional
genotyping-data.</p>
<p>The code chunk below uses the <code>EMfit_betabinom_robust</code>
function to fit this mixture model, and writes the results to
<code>ASEfit_res</code>. Additionally, it uses the previously calculated
inbreeding coefficient <code>Fmedian</code> to test every locus for
significant deviation from Hardy-Weinberg Equilibrium (HWE), which could
indicate a failed model fit and can thus be used as a goodness-of-fit
filter criterion later:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ASEfitter</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">positions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span>  <span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">z</span> <span class="kw">in</span> <span class="va">positions</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">maelstRomres</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/EMfit_betabinom_robust.html">EMfit_betabinom_robust</a></span><span class="op">(</span>data_counts <span class="op">=</span> <span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                            SE <span class="op">=</span> <span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">maelstRomres</span><span class="op">$</span><span class="va">data_hash</span></span>
<span>    </span>
<span>    <span class="va">res_loc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"Gene"</span> <span class="op">=</span> <span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> ,<span class="st">"Locus"</span> <span class="op">=</span> <span class="va">z</span>, </span>
<span>      <span class="st">"reference"</span> <span class="op">=</span> <span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"variant"</span> <span class="op">=</span> <span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>      <span class="st">"phi_rr"</span> <span class="op">=</span> <span class="va">maelstRomres</span><span class="op">$</span><span class="va">rho_rr</span>, <span class="st">"phi_rv"</span> <span class="op">=</span> <span class="va">maelstRomres</span><span class="op">$</span><span class="va">rho_rv</span>, </span>
<span>      <span class="st">"phi_vv"</span> <span class="op">=</span> <span class="va">maelstRomres</span><span class="op">$</span><span class="va">rho_vv</span>, <span class="st">"pi"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">maelstRomres</span><span class="op">$</span><span class="va">AB</span><span class="op">)</span>, </span>
<span>      <span class="st">"AB_pval"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">maelstRomres</span><span class="op">$</span><span class="va">AB_p</span><span class="op">)</span>, <span class="st">"theta_hom"</span> <span class="op">=</span> <span class="va">maelstRomres</span><span class="op">$</span><span class="va">theta_hom</span>, </span>
<span>      <span class="st">"theta_het"</span> <span class="op">=</span> <span class="va">maelstRomres</span><span class="op">$</span><span class="va">theta_het</span>, <span class="st">"allele.frequency"</span> <span class="op">=</span> <span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">allelefreq</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>      <span class="st">"coverage"</span> <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span><span class="op">+</span><span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span><span class="op">)</span>,</span>
<span>      <span class="st">"nr_samples"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="va">z</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>      <span class="st">"quality"</span> <span class="op">=</span> <span class="va">maelstRomres</span><span class="op">$</span><span class="va">quality</span>,</span>
<span>      stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">results</span>, <span class="va">res_loc</span><span class="op">)</span> <span class="co"># results; one position per line</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/HWE_chisquared.html">HWE_chisquared</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span>, <span class="va">Fmedian</span>, results <span class="op">=</span> <span class="va">results</span><span class="op">)</span></span>
<span>  <span class="va">results</span><span class="op">$</span><span class="va">HWEpval</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">HWEpval</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fl">1</span></span>
<span>  <span class="va">results</span><span class="op">$</span><span class="va">HWE_TestStat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">HWE_TestStat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fl">1</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">results</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">clusterExport</a></span><span class="op">(</span><span class="va">cl</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Fmedian"</span>, <span class="st">"SEmedian"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ASEfit</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParCTRL</span>, fun <span class="op">=</span> <span class="va">ASEfitter</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ParCTRL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ASEfit</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ASEfit</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ASEfit_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ASEfit</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><code>ASEfit_res</code> contains all per-locus parameters of the
beta-binomial mixture model discussed above. Also, besides the just
mentioned results of the chi square test for deviation from HWE, it
contains additional quality metrics which can be used in filtering of
results, those being each locus’ reference allele frequency among
genotypes in the population (<code>allele.frequency</code>), median
count coverage (<code>coverage</code>), the number of samples involved
in the fit (<code>nr_samples</code>) and lastly, a quality flag which
equals “!” in the case maelstRom’s fit fails for this particular locus,
which is usually due to no apparent heterozygotes being present
(<code>quality</code>). Finally, <code>AB_pval</code> contains the
result (raw p-value) of a likelihood ratio test for whether <span class="math inline">\(\pi_{het}\)</span> is significantly different from
0.5, thus detecting whether there’s an Allelic Bias (AB) in expression
in a given locus.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ASEfit_res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="3%">
<col width="4%">
<col width="5%">
<col width="4%">
<col width="5%">
<col width="5%">
<col width="5%">
<col width="5%">
<col width="5%">
<col width="5%">
<col width="5%">
<col width="10%">
<col width="5%">
<col width="6%">
<col width="4%">
<col width="5%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left">Gene</th>
<th align="left">Locus</th>
<th align="left">reference</th>
<th align="left">variant</th>
<th align="right">phi_rr</th>
<th align="right">phi_rv</th>
<th align="right">phi_vv</th>
<th align="right">pi</th>
<th align="right">AB_pval</th>
<th align="right">theta_hom</th>
<th align="right">theta_het</th>
<th align="right">allele.frequency</th>
<th align="right">coverage</th>
<th align="right">nr_samples</th>
<th align="left">quality</th>
<th align="right">HWEpval</th>
<th align="right">HWE_TestStat</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Gene1</td>
<td align="left">Locus1</td>
<td align="left">A</td>
<td align="left">G</td>
<td align="right">0.3904620</td>
<td align="right">0.4845264</td>
<td align="right">0.1250116</td>
<td align="right">0.4898086</td>
<td align="right">0.2898711</td>
<td align="right">0.0602616</td>
<td align="right">0.0173647</td>
<td align="right">0.6327252</td>
<td align="right">190.5</td>
<td align="right">128</td>
<td align="left"></td>
<td align="right">0.9189807</td>
<td align="right">0.1689803</td>
</tr>
<tr class="even">
<td align="left">Gene1</td>
<td align="left">Locus2</td>
<td align="left">A</td>
<td align="left">G</td>
<td align="right">0.7031276</td>
<td align="right">0.2812474</td>
<td align="right">0.0156250</td>
<td align="right">0.5647487</td>
<td align="right">0.0000103</td>
<td align="right">0.0366680</td>
<td align="right">0.0130104</td>
<td align="right">0.8437513</td>
<td align="right">96.0</td>
<td align="right">128</td>
<td align="left"></td>
<td align="right">0.8149959</td>
<td align="right">0.4091444</td>
</tr>
<tr class="odd">
<td align="left">Gene1</td>
<td align="left">Locus3</td>
<td align="left">C</td>
<td align="left">A</td>
<td align="right">0.7031522</td>
<td align="right">0.2812201</td>
<td align="right">0.0156277</td>
<td align="right">0.5093620</td>
<td align="right">0.3614079</td>
<td align="right">0.0837429</td>
<td align="right">0.0066971</td>
<td align="right">0.8437623</td>
<td align="right">116.5</td>
<td align="right">128</td>
<td align="left"></td>
<td align="right">0.8151967</td>
<td align="right">0.4086518</td>
</tr>
<tr class="even">
<td align="left">Gene2</td>
<td align="left">Locus4</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0.4163886</td>
<td align="right">0.4900306</td>
<td align="right">0.0935808</td>
<td align="right">0.6226814</td>
<td align="right">0.0000000</td>
<td align="right">0.0481506</td>
<td align="right">0.0202838</td>
<td align="right">0.6614039</td>
<td align="right">27.0</td>
<td align="right">128</td>
<td align="left"></td>
<td align="right">0.7073233</td>
<td align="right">0.6925348</td>
</tr>
<tr class="odd">
<td align="left">Gene2</td>
<td align="left">Locus5</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0.5469213</td>
<td align="right">0.3828949</td>
<td align="right">0.0701838</td>
<td align="right">0.5143016</td>
<td align="right">0.3560640</td>
<td align="right">0.0449810</td>
<td align="right">0.0180381</td>
<td align="right">0.7383688</td>
<td align="right">34.0</td>
<td align="right">128</td>
<td align="left"></td>
<td align="right">0.9999981</td>
<td align="right">0.0000037</td>
</tr>
<tr class="even">
<td align="left">Gene2</td>
<td align="left">Locus6</td>
<td align="left">G</td>
<td align="left">T</td>
<td align="right">0.1370091</td>
<td align="right">0.5524136</td>
<td align="right">0.3105773</td>
<td align="right">0.7119506</td>
<td align="right">0.0000000</td>
<td align="right">0.0207748</td>
<td align="right">0.0159835</td>
<td align="right">0.4132159</td>
<td align="right">15.0</td>
<td align="right">128</td>
<td align="left"></td>
<td align="right">0.4919931</td>
<td align="right">1.4185814</td>
</tr>
</tbody>
</table>
<p>In addition to this dataframe of per-locus fitting results, the
per-locus allelic count data in <code>controlList</code> has been
updated to include, for every sample, its
Expectation-Maximization-assigned likelihood to be originating from a
reference homozygous, heterozygous, or variant homozygous individual
(prr, prv, pvv), listing the most likely of these options in the
“genotypeN” column.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="st">"Locus27"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="st">"ref"</span>, <span class="st">"var"</span>, <span class="st">"ref_count"</span>, <span class="st">"var_count"</span>, </span>
<span>                                           <span class="st">"prr"</span>, <span class="st">"prv"</span>, <span class="st">"pvv"</span>, <span class="st">"genotypeN"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="7%">
<col width="9%">
<col width="4%">
<col width="4%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">Sample</th>
<th align="left">ref</th>
<th align="left">var</th>
<th align="right">ref_count</th>
<th align="right">var_count</th>
<th align="right">prr</th>
<th align="right">prv</th>
<th align="right">pvv</th>
<th align="left">genotypeN</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">16915</td>
<td align="left">Sample1</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">7</td>
<td align="right">0</td>
<td align="right">0.9954662</td>
<td align="right">0.0045338</td>
<td align="right">0.0000000</td>
<td align="left">rr</td>
</tr>
<tr class="even">
<td align="left">16916</td>
<td align="left">Sample2</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">2</td>
<td align="right">7</td>
<td align="right">0.0000025</td>
<td align="right">0.9987087</td>
<td align="right">0.0012888</td>
<td align="left">rv</td>
</tr>
<tr class="odd">
<td align="left">16917</td>
<td align="left">Sample3</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">6</td>
<td align="right">6</td>
<td align="right">0.0000987</td>
<td align="right">0.9998962</td>
<td align="right">0.0000050</td>
<td align="left">rv</td>
</tr>
<tr class="even">
<td align="left">16918</td>
<td align="left">Sample4</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">9</td>
<td align="right">0</td>
<td align="right">0.9985476</td>
<td align="right">0.0014524</td>
<td align="right">0.0000000</td>
<td align="left">rr</td>
</tr>
<tr class="odd">
<td align="left">16919</td>
<td align="left">Sample5</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">2</td>
<td align="right">8</td>
<td align="right">0.0000010</td>
<td align="right">0.9980893</td>
<td align="right">0.0019098</td>
<td align="left">rv</td>
</tr>
<tr class="even">
<td align="left">16920</td>
<td align="left">Sample6</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">3</td>
<td align="right">3</td>
<td align="right">0.0008466</td>
<td align="right">0.9991102</td>
<td align="right">0.0000432</td>
<td align="left">rv</td>
</tr>
</tbody>
</table>
<p>We can now list loci showcasing significant AB (i.e. deviation from
<span class="math inline">\(\pi_{het} = 0.5\)</span>), both
statistically (5% FDR level) and in terms of effect size (a deviation
from 0.5 of at least 0.1). The code chunk below also includes some basic
quality filter criteria (a median counts coverage of at least 10 across
at least 20 expected heterozygotes to actually fit AB on; with the
latter calculated as the number of samples times the fraction of
heterozygotes <span class="math inline">\(\phi_{rv}\)</span>).
Furthermore, to ensure a good model fit, we require no problems
regarding the quality flag (i.e. not equal to “!”), a detected AB that’s
not more extreme than 0.9 or 0.1, and a HWE-deviation-checking p-value
of at least 0.001 (a filter criterion often used to check for HWE
conformity: <span class="citation">Sha and Zhang (2011)</span>, <span class="citation">Teo et al. (2007)</span>, <span class="citation">Rohlfs
and Weir (2008)</span>); any of these last three criteria not being met
indicates a possibly faulty model fit.</p>
<p>A final, optional, filter, is throwing out loci for which either
<span class="math inline">\(\pi_{het}/\theta_{het} &lt;= 1\)</span> OR
<span class="math inline">\((1-\pi_{het})/\theta_{het} &lt;= 1\)</span>,
as any of these inequalities being met leads to stange behavior of the
beta-binomial distribution (either getting its mode stuck at 0 or 1 or,
if both inequalities are true, becoming bimodal). This being the case
usually suggests a failed model fit as well (for an illustration, try to
plot Locus172’s fit - which is normally removed by this final filter -
using the plotting function in the next code chunk after the code chunk
below).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Statistical evidence for significant AB at the 5% FDR level,</span></span>
<span><span class="co"># only retaining reliable (high-quality) loci with a large enough effect size:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">[</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">AB_pval</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span></span>
<span>  <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">pi</span> <span class="op">-</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">ASEfit_res</span><span class="op">$</span><span class="va">coverage</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">ASEfit_res</span><span class="op">$</span><span class="va">nr_samples</span> <span class="op">*</span> <span class="va">ASEfit_res</span><span class="op">$</span><span class="va">phi_rv</span> <span class="op">&gt;=</span> <span class="fl">20</span>  <span class="op">&amp;</span>   </span>
<span>  <span class="va">ASEfit_res</span><span class="op">$</span><span class="va">quality</span> <span class="op">!=</span> <span class="st">"!"</span> <span class="op">&amp;</span></span>
<span>  <span class="op">(</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">pi</span> <span class="op">&lt;</span> <span class="fl">0.9</span> <span class="op">&amp;</span>  <span class="va">ASEfit_res</span><span class="op">$</span><span class="va">pi</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">ASEfit_res</span><span class="op">$</span><span class="va">HWEpval</span> <span class="op">&gt;=</span> <span class="fl">0.001</span> <span class="op">&amp;</span></span>
<span>  <span class="va">ASEfit_res</span><span class="op">$</span><span class="va">pi</span><span class="op">/</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">theta_het</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">&amp;</span></span>
<span>  <span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">pi</span><span class="op">)</span><span class="op">/</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">theta_het</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span>, </span>
<span>  </span>
<span>  collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>Locus4, Locus6, Locus10, Locus45, Locus73, Locus84, Locus104,
Locus114,</p>
<p>maelstRom’s <code><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot()</a></code> allows us to plot the
previously obtained beta-binomial mixture model fits, together with an
allele fraction histogram depicting the actual data that’s being fit. In
addition to to-be-plotted data and parameters, the function requires two
additional inputs.</p>
<ul>
<li>“nbins”: the number of bins for the histogram of observed per-sample
(reference) allele fractions</li>
<li>“ScaleCount”: the beta-binomial distribution is a discrete
probability distribution that contains a parameter for the total number
of observations, <span class="math inline">\(n\)</span>. The
distributions being plotted as lineplots by
<code><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot()</a></code> are beta-binomials, but with their
x-axis range rescaled from 0-to-<span class="math inline">\(n\)</span>
to 0-to-1. However, even when rescaled, the shape (variance) of this
distribution still depends on the underlying <span class="math inline">\(n\)</span>-parameter being used, which is thus
asked as an input argument, ScaleCount. In order to most closely fit the
underlying histogram of observed data, the code chunk below uses the
locus’ median coverage as ScaleCount.</li>
</ul>
<p><code><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot()</a></code> accepts a lot of other optional
arguments for further plotting functionalities and customization.
Additionally, it returns the plot as a ggplot object, which can be
freely customized even further.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PlotData</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="st">"Locus104"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">PlotData_eqtl</span> <span class="op">&lt;-</span> <span class="va">ASEfit_res</span><span class="op">[</span><span class="va">ASEfit_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">==</span><span class="st">"Locus104"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">loc_plot</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>ref_counts<span class="op">=</span><span class="va">PlotData</span><span class="op">$</span><span class="va">ref_count</span>, </span>
<span>  var_counts<span class="op">=</span><span class="va">PlotData</span><span class="op">$</span><span class="va">var_count</span>, pr<span class="op">=</span><span class="va">PlotData_eqtl</span><span class="op">$</span><span class="va">phi_rr</span>, prv<span class="op">=</span><span class="va">PlotData_eqtl</span><span class="op">$</span><span class="va">phi_rv</span>,</span>
<span>  pv<span class="op">=</span><span class="va">PlotData_eqtl</span><span class="op">$</span><span class="va">phi_vv</span>, theta_hom<span class="op">=</span><span class="va">PlotData_eqtl</span><span class="op">$</span><span class="va">theta_hom</span>, </span>
<span>  theta_het<span class="op">=</span><span class="va">PlotData_eqtl</span><span class="op">$</span><span class="va">theta_het</span>, probshift<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">PlotData_eqtl</span><span class="op">$</span><span class="va">pi</span><span class="op">)</span>, </span>
<span>  SE<span class="op">=</span><span class="va">SEmedian</span>, ScaleCount <span class="op">=</span> <span class="va">PlotData_eqtl</span><span class="op">$</span><span class="va">coverage</span>, nbins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">loc_plot</span></span></code></pre></div>
<p><img src="maelstRom_Allelic_Dispersion_tutorial_files/figure-html/unnamed-chunk-18-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>Besides the previously established statistical evidence, this plot
makes it visually clear that Locus104’s heterozygotes are biased towards
allelic expression of the reference allele.</p>
</div>
<div class="section level2">
<h2 id="differential-allelic-dispersion">Differential allelic dispersion<a class="anchor" aria-label="anchor" href="#differential-allelic-dispersion"></a>
</h2>
<p>maelstRom hypothesizes that early (so potentially causal)
case-specific dysregulation events such as copy number alterations,
aberrant hypermethylation, and gene silencing (incomplete silencing;
either inherent to the mechanism, and/or seemingly observed as
incomplete due to e.g. tumor impurity of the allelic count data) can
cause an allelic expression bias within one individual but will not, at
a population-level, consistently favor the same allele. Thus, rather
than causing an allelic bias (AB), this dysregulation will result in an
increased variance of allelic expression; we call this variance Allelic
Dispersion (AD), and its increase in e.g. disease and cancer
differential Allelic Dispersion (dAD).</p>
<p>Detecting dAD between two populations can be done via a likelihood
ratio test, comparing: (1) a model fitting ALL data, both control and
case, as one beta-binomial mixture model, with (2) a model allowing the
control- and case data to have their separate <span class="math inline">\(\theta_{het}\)</span> parameters. This is done by
the <code><a href="../reference/dAD_analysis.html">dAD_analysis()</a></code> function below, which internally
performs a number of fits:</p>
<ol style="list-style-type: decimal">
<li>A fit on control (<code>OnlyC</code>) and case (<code>OnlyT</code>)
data separately for outlier detection.</li>
<li>A joint fit on control and case data sharing all parameters across
both populations (<code>H0</code>, the null hypothesis of the
dAD-test)</li>
<li>A fit using a separate <span class="math inline">\(\theta_{het}\)</span> parameter for controls and
cases</li>
</ol>
<p>Besides this final and most interesting fit, the <code>dAD_res</code>
object below contains the parameter estimations for the other three fits
(<code>OnlyC</code>, <code>OnlyT</code>, <code>H0</code>), alongside a
selection of quality metrics and a p-value testing for deviation from
HWE in controls and cases (“HWEC”, “HWET”) which can be used as filter
critaria.</p>
<p>The code chunk below only prints the most interesting results: the
actual dAD-detecting pval along with and effect size (<span class="math inline">\(\theta_{het}\)</span> in both cases and controls)
and the most relevant filter critaria in controls (number of estimated
heterozygotes, median coverage, and the HWE-deviation-testing
p-value):</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ParTOT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span>  <span class="va">ParTOT</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ParCTRL</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">ParCASE</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dADFinData</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParTOT</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va"><a href="../reference/dAD_analysis.html">dAD_analysis</a></span>, </span>
<span>                                  SE <span class="op">=</span> <span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dAD_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">dADFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">dADFinData</span> , <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">caseList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">dADFinData</span>, <span class="va">`[[`</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Locus"</span>, <span class="st">"Gene"</span>, <span class="st">"dAD_pval"</span>, <span class="st">"ThetaHetC"</span>, <span class="st">"ThetaHetT"</span>, </span>
<span>                             <span class="st">"RhoC"</span>, <span class="st">"RhoT"</span>, <span class="st">"NumHetC"</span>, <span class="st">"CovC_Med"</span>, <span class="st">"HWEC"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="7%">
<col width="6%">
<col width="10%">
<col width="10%">
<col width="10%">
<col width="10%">
<col width="10%">
<col width="9%">
<col width="9%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="left">Locus</th>
<th align="left">Gene</th>
<th align="right">dAD_pval</th>
<th align="right">ThetaHetC</th>
<th align="right">ThetaHetT</th>
<th align="right">RhoC</th>
<th align="right">RhoT</th>
<th align="right">NumHetC</th>
<th align="right">CovC_Med</th>
<th align="right">HWEC</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Locus1</td>
<td align="left">Gene1</td>
<td align="right">0.0145737</td>
<td align="right">0.0243768</td>
<td align="right">0.0515579</td>
<td align="right">0.0237967</td>
<td align="right">0.0490300</td>
<td align="right">62.01938</td>
<td align="right">190.5</td>
<td align="right">0.9189807</td>
</tr>
<tr class="even">
<td align="left">Locus2</td>
<td align="left">Gene1</td>
<td align="right">0.0001449</td>
<td align="right">0.0141543</td>
<td align="right">0.0751320</td>
<td align="right">0.0139567</td>
<td align="right">0.0698817</td>
<td align="right">35.99966</td>
<td align="right">96.0</td>
<td align="right">0.8149959</td>
</tr>
<tr class="odd">
<td align="left">Locus3</td>
<td align="left">Gene1</td>
<td align="right">0.0000012</td>
<td align="right">0.0073516</td>
<td align="right">0.0785041</td>
<td align="right">0.0072980</td>
<td align="right">0.0727898</td>
<td align="right">35.99617</td>
<td align="right">116.5</td>
<td align="right">0.8151967</td>
</tr>
<tr class="even">
<td align="left">Locus4</td>
<td align="left">Gene2</td>
<td align="right">0.3304979</td>
<td align="right">0.0208067</td>
<td align="right">0.0386687</td>
<td align="right">0.0203826</td>
<td align="right">0.0372291</td>
<td align="right">62.72392</td>
<td align="right">27.0</td>
<td align="right">0.7073233</td>
</tr>
<tr class="odd">
<td align="left">Locus5</td>
<td align="left">Gene2</td>
<td align="right">0.6137441</td>
<td align="right">0.0188209</td>
<td align="right">0.0262031</td>
<td align="right">0.0184732</td>
<td align="right">0.0255340</td>
<td align="right">49.01055</td>
<td align="right">34.0</td>
<td align="right">0.9999981</td>
</tr>
<tr class="even">
<td align="left">Locus6</td>
<td align="left">Gene2</td>
<td align="right">0.3453054</td>
<td align="right">0.0213316</td>
<td align="right">0.0021231</td>
<td align="right">0.0208861</td>
<td align="right">0.0021186</td>
<td align="right">70.70894</td>
<td align="right">15.0</td>
<td align="right">0.4919931</td>
</tr>
</tbody>
</table>
<p><code>RhoC</code> and <code>RhoT</code> are simple rescalings of
<code>ThetaHetC</code> and <code>ThetaHetT</code> respectively:</p>
<p><span class="math display">\[
\rho = \frac{1}{(1/\theta)+1}
\]</span> which ranges from 0 to 1 instead of <span class="math inline">\(\theta\)</span>’s 0 to infinity, and is thus more
suitable for visualization purposes. The code above updated the inputs
in <code>controlList</code> and <code>caseList</code> to include a
column marking outliers as well, which can come in handy for future
analyses.</p>
<p>Significant dAD loci with increased <span class="math inline">\(\theta_{het}\)</span> in cases, and passing some
basic filter criteria, can now be selected:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">dAD_pval</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.001</span> <span class="op">&amp;</span> <span class="co">#Significant at 0.1% FDR</span></span>
<span>    <span class="va">dAD_res</span><span class="op">$</span><span class="va">RhoT</span> <span class="op">&gt;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">RhoC</span> <span class="op">&amp;</span> <span class="co">#Increased AD in cases</span></span>
<span>    <span class="co">#Median coverage of at least 10 in both cases and controls:</span></span>
<span>    <span class="va">dAD_res</span><span class="op">$</span><span class="va">CovC_Med</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">CovT_Med</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> </span>
<span>    <span class="co">#Number of heterozygotes and HWE-conformity are only checked in controls,</span></span>
<span>    <span class="co">#as case-data can be rather erratic, e.g. due to tumorigenesis</span></span>
<span>    <span class="va">dAD_res</span><span class="op">$</span><span class="va">NumHetC</span> <span class="op">&gt;=</span> <span class="fl">15</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">HWEC</span> <span class="op">&gt;=</span> <span class="fl">0.001</span></span>
<span>  <span class="op">]</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>Locus3, Locus18, Locus19, Locus20, Locus69, Locus72, Locus74,
Locus75, Locus116, Locus117, Locus118, Locus138, Locus141, Locus142,
Locus143, Locus147, Locus148, Locus149, Locus154, Locus159, Locus160,
Locus161, Locus164, Locus165, Locus170,</p>
<p>As before, <code>maelstRom_EMfitplot</code> can be used to visualize
the control-case AD difference in these loci:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CTRL_DF</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="st">"Locus19"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">CASE_DF</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="st">"Locus19"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">PlotData_dAD</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">==</span><span class="st">"Locus19"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">dAD_plot1</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>ref_counts<span class="op">=</span><span class="va">CTRL_DF</span><span class="op">$</span><span class="va">ref_count</span>, var_counts<span class="op">=</span><span class="va">CTRL_DF</span><span class="op">$</span><span class="va">var_count</span>,</span>
<span>  pr<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rr</span>, prv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rv</span>, pv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_vv</span>, </span>
<span>  theta_hom <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHom</span>, theta_het <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHetC</span>, </span>
<span>  probshift <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">Pi</span>, SE<span class="op">=</span><span class="va">SEmedian</span>, ScaleCount <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">CovC_Med</span>, nbins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dAD_plot2</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>ref_counts<span class="op">=</span><span class="va">CASE_DF</span><span class="op">$</span><span class="va">ref_count</span>, var_counts<span class="op">=</span><span class="va">CASE_DF</span><span class="op">$</span><span class="va">var_count</span>,</span>
<span>  pr<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rr</span>, prv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rv</span>, pv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_vv</span>, </span>
<span>  theta_hom <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHom</span>, theta_het <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHetT</span>, </span>
<span>  probshift <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">Pi</span>, SE<span class="op">=</span><span class="va">SEmedian</span>, ScaleCount <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">CovT_Med</span>, nbins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">dAD_plot1</span>, <span class="va">dAD_plot2</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="maelstRom_Allelic_Dispersion_tutorial_files/figure-html/unnamed-chunk-21-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="combining-locus-level-results-to-gene-level-results-">Combining locus-level results to gene-level results.<a class="anchor" aria-label="anchor" href="#combining-locus-level-results-to-gene-level-results-"></a>
</h2>
<p>While locus-level results can be informative by themselves, it is -
for biological interpretation - easier to have results (<span class="math inline">\(\theta\)</span> or <span class="math inline">\(\rho\)</span> values, p-values) at the gene-level.
Post-hoc p-value combination (so without directly modelling counts at
the transcript- or gene-level) is especially tricky: the
well-established Fisher combination method is only valid when combining
independent tests (and independency is hard to argue for SNPs that are
close enough to have their count data originate from the same RNA
molecule or RNAseq read) while other combination methods like Cauchy
combination (<span class="citation">Liu and Xie (2020)</span>) behave
like a generalized mean - in the sense that the resulting p-value will
never be lower than the minimal supplied p-value and, as such, is unable
to combine statistical evidence from two tests into an even more
significant p-value.</p>
<p>To this end, maelstRom implements a correlation-corrected Lancaster
p-value combination (the Lancaster combination being, itself, a weighted
version of the Fischer combination), as published by <span class="citation">Dai, Leeder, and Cui (2014)</span>, and further
optimized by implementation of the Score test (instead of the likelihood
ratio test) during the sample-label-permutation-based correlation
calculation (which avoids iterative re-fitting of maelstRom’s
beta-binomial mixture model, as the score test relies solely on the
<code>H0</code> model fit, which is constant across sample-label
permutations).</p>
<p>The previously mentioned (combined) p-values (Fisher, Cauchy,
Lancaster, correlation-corrected Lancaster) are calculated in the code
chunk below, by maelstRom’s <code>AdvancedPvalcomb</code> function
which, besides the sequencing error SE, also takes the number of
permutations to use during correlation-correction <code>nperm</code> as
an input; the higher this number, the longer the code runs. <span class="citation">Dai, Leeder, and Cui (2014)</span> uses 1000
permutations in their publication so we do the same here:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Split results across cores for parallellization in a gene-wise manner</span></span>
<span><span class="va">goi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span><span class="op">)</span></span>
<span><span class="va">NS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span></span>
<span><span class="va">spl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span>,<span class="va">NC</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">NS</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">NC</span><span class="op">-</span><span class="va">NS</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># Helps in splitting input data</span></span>
<span><span class="va">inputlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span>  <span class="va">GenesNow</span> <span class="op">&lt;-</span> <span class="va">goi</span><span class="op">[</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">j</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">ResDF_OIs</span> <span class="op">&lt;-</span> <span class="va">ControlOIs</span> <span class="op">&lt;-</span> <span class="va">CaseOIs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">GenesNow</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">GenesNow</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">ResDF_OIs</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">GenesNow</span><span class="op">[</span><span class="va">k</span><span class="op">]</span>,<span class="op">]</span></span>
<span>    <span class="va">LocsOI</span> <span class="op">&lt;-</span> <span class="va">ResDF_OIs</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Locus</span></span>
<span>    <span class="va">ControlOIs</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">[</span><span class="va">LocsOI</span><span class="op">]</span></span>
<span>    <span class="va">CaseOIs</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">caseList</span><span class="op">)</span><span class="op">[</span><span class="va">LocsOI</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">inputlist</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GenesNow</span>, <span class="va">ResDF_OIs</span>, <span class="va">ControlOIs</span>, <span class="va">CaseOIs</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Call the AdvancedPvalcomb function</span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">AdvPVALs</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">inputlist</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va">AdvancedPvalcomb</span>, SE <span class="op">=</span> <span class="va">SEmedian</span>, nperm <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract gene-level p-values</span></span>
<span><span class="va">FisherPVAL_vec</span> <span class="op">&lt;-</span> <span class="va">LancasterPVAL_vec</span> <span class="op">&lt;-</span>  <span class="va">CorLancasterPVAL_vec</span> <span class="op">&lt;-</span>  <span class="va">CauchyPVAL_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">ii</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">FisherPVAL_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">FisherPVAL_vec</span>, <span class="va">AdvPVALs</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">LancasterPVAL_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">LancasterPVAL_vec</span>, <span class="va">AdvPVALs</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> </span>
<span>  <span class="va">CorLancasterPVAL_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">CorLancasterPVAL_vec</span>, <span class="va">AdvPVALs</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">CauchyPVAL_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">CauchyPVAL_vec</span>, <span class="va">AdvPVALs</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>There are a number of genes with highly correlated SNP-level allelic
counts (suggesting said SNP are located in close proximity to eachother
on the genome or on (processed) transcripts from which RNAseq data is
generated), for which the correlation-corrected Lancaster p-value was
greatly reduced when compared to the (non-corrected) Lancaster
p-value:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">LancasterPVAL_vec</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">23</span>, <span class="fl">40</span>, <span class="fl">51</span><span class="op">)</span><span class="op">]</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>3.67073174475383e-12, 3.22181819373512e-41, 1.42550478480222e-28,</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CorLancasterPVAL_vec</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">23</span>, <span class="fl">40</span>, <span class="fl">51</span><span class="op">)</span><span class="op">]</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>8.58344380865356e-07, 4.93112771598964e-20, 2.06118418072043e-13,</p>
<p>Most genes, however, consist of more independent SNP-level data
(notice the relatively small change):</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">LancasterPVAL_vec</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>2.63624315913471e-09, 0.517344906897836, 4.12379217454491e-06,
0.000739519269403496, 0.8312901339832,</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CorLancasterPVAL_vec</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>2.4461248849389e-08, 0.517344906897836, 5.73902817881098e-06,
0.00095834457979229, 0.815307002744072,</p>
<p>Next to the per-gene correlation-corrected Lancaster p-value, we can
add averages of its per-SNP <span class="math inline">\(\rho_{control}\)</span>s and <span class="math inline">\(\rho_{case}\)</span>s (weighted averages,
according to each SNP’s <span class="math inline">\(\sqrt{median\
coverage \times estimated\ number\ of\ heterozygotes}\)</span> in
controls or cases respectively), as well as some (weighted averages of)
quality metrics:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">dAD_res_Gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Gene<span class="op">=</span><span class="va">goi</span>, Nr_SNPs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  Mean_RhoC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span><span class="op">)</span>, Mean_RhoT<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span><span class="op">)</span>, CorLancasterPVAL <span class="op">=</span> <span class="va">CorLancasterPVAL_vec</span>, </span>
<span>  Mean_NumHetC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span><span class="op">)</span>, Mean_NumHetT<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span><span class="op">)</span>, Mean_CovC_Med<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span><span class="op">)</span>, Mean_CovT_Med<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">goi</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>     </span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">Gene</span> <span class="kw">in</span> <span class="va">goi</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Nr_SNPs</span><span class="op">[</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">CovC_Meds</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">CovC_Med</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span></span>
<span>  <span class="va">CovT_Meds</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">CovT_Med</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span></span>
<span>  <span class="va">NumHetCs</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">NumHetC</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span></span>
<span>  <span class="va">NumHetTs</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">NumHetT</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span></span>
<span>  <span class="va">Cweights</span> <span class="op">&lt;-</span> <span class="va">Tweights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">CovC_Meds</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">Cweights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Cweights</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">CovC_Meds</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">*</span><span class="va">NumHetCs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">Tweights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Tweights</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">CovT_Meds</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">*</span><span class="va">NumHetTs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_NumHetC</span><span class="op">[</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/combine_p_gene.html">combine_p_gene</a></span><span class="op">(</span><span class="va">NumHetCs</span>, weights <span class="op">=</span> </span>
<span>                                                            <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">CovC_Meds</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"arithmetic"</span><span class="op">)</span></span>
<span>  <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_NumHetT</span><span class="op">[</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/combine_p_gene.html">combine_p_gene</a></span><span class="op">(</span><span class="va">NumHetTs</span>, weights <span class="op">=</span> </span>
<span>                                                            <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">CovT_Meds</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"arithmetic"</span><span class="op">)</span></span>
<span>  <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_CovC_Med</span><span class="op">[</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/combine_p_gene.html">combine_p_gene</a></span><span class="op">(</span><span class="va">CovC_Meds</span>, weights <span class="op">=</span> </span>
<span>                                                            <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">NumHetCs</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"arithmetic"</span><span class="op">)</span></span>
<span>  <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_CovT_Med</span><span class="op">[</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/combine_p_gene.html">combine_p_gene</a></span><span class="op">(</span><span class="va">CovT_Meds</span>, weights <span class="op">=</span> </span>
<span>                                                            <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">NumHetTs</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"arithmetic"</span><span class="op">)</span></span>
<span>  <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_RhoC</span><span class="op">[</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/combine_p_gene.html">combine_p_gene</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">RhoC</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> </span>
<span>                                                        <span class="va">Gene</span><span class="op">]</span>, weights <span class="op">=</span> <span class="va">Cweights</span>, method <span class="op">=</span> <span class="st">"arithmetic"</span><span class="op">)</span></span>
<span>  <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_RhoT</span><span class="op">[</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">Gene</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/combine_p_gene.html">combine_p_gene</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">RhoT</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span> <span class="op">==</span> </span>
<span>                                                        <span class="va">Gene</span><span class="op">]</span>, weights <span class="op">=</span> <span class="va">Tweights</span>, method <span class="op">=</span> <span class="st">"arithmetic"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dAD_res_Gene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="5%">
<col width="7%">
<col width="9%">
<col width="9%">
<col width="16%">
<col width="12%">
<col width="12%">
<col width="13%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="left">Gene</th>
<th align="right">Nr_SNPs</th>
<th align="right">Mean_RhoC</th>
<th align="right">Mean_RhoT</th>
<th align="right">CorLancasterPVAL</th>
<th align="right">Mean_NumHetC</th>
<th align="right">Mean_NumHetT</th>
<th align="right">Mean_CovC_Med</th>
<th align="right">Mean_CovT_Med</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Gene1</td>
<td align="right">3</td>
<td align="right">0.0167054</td>
<td align="right">0.0609712</td>
<td align="right">0.0000000</td>
<td align="right">46.44028</td>
<td align="right">87.74274</td>
<td align="right">139.63312</td>
<td align="right">173.500947</td>
</tr>
<tr class="even">
<td align="left">Gene2</td>
<td align="right">3</td>
<td align="right">0.0198453</td>
<td align="right">0.0243574</td>
<td align="right">0.5173449</td>
<td align="right">59.43292</td>
<td align="right">114.12708</td>
<td align="right">24.77530</td>
<td align="right">18.029318</td>
</tr>
<tr class="odd">
<td align="left">Gene3</td>
<td align="right">3</td>
<td align="right">0.0118674</td>
<td align="right">0.0307354</td>
<td align="right">0.0000057</td>
<td align="right">62.15882</td>
<td align="right">116.28909</td>
<td align="right">169.44331</td>
<td align="right">124.286181</td>
</tr>
<tr class="even">
<td align="left">Gene4</td>
<td align="right">2</td>
<td align="right">0.0218546</td>
<td align="right">0.0301046</td>
<td align="right">0.0009583</td>
<td align="right">37.13462</td>
<td align="right">99.43332</td>
<td align="right">240.21210</td>
<td align="right">187.994362</td>
</tr>
<tr class="odd">
<td align="left">Gene5</td>
<td align="right">3</td>
<td align="right">0.0142337</td>
<td align="right">0.0136280</td>
<td align="right">0.8153070</td>
<td align="right">42.79365</td>
<td align="right">93.49767</td>
<td align="right">10.53696</td>
<td align="right">8.449295</td>
</tr>
<tr class="even">
<td align="left">Gene6</td>
<td align="right">3</td>
<td align="right">0.0296196</td>
<td align="right">0.0379265</td>
<td align="right">0.2324265</td>
<td align="right">36.47197</td>
<td align="right">67.56134</td>
<td align="right">16.07402</td>
<td align="right">15.540588</td>
</tr>
</tbody>
</table>
<p>Using similar criteria as for the SNP-level results, we can now
filter out those genes showing the most significant dAD:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">CorLancasterPVAL</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.001</span> <span class="op">&amp;</span> <span class="co">#Significant at 0.1% FDR</span></span>
<span>    <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_RhoT</span> <span class="op">&gt;</span> <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_RhoC</span> <span class="op">&amp;</span> <span class="co">#Increased AD in cases</span></span>
<span>    <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_CovC_Med</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_CovT_Med</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> </span>
<span>    <span class="va">dAD_res_Gene</span><span class="op">$</span><span class="va">Mean_NumHetC</span> <span class="op">&gt;=</span> <span class="fl">15</span></span>
<span>  <span class="op">]</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>Gene1, Gene3, Gene7, Gene24, Gene25, Gene26, Gene34, Gene41, Gene49,
Gene50, Gene52, Gene56, Gene58, Gene60,</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GeneToPlot</span> <span class="op">&lt;-</span> <span class="st">"Gene7"</span></span>
<span><span class="va">LocsToPlot</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Gene</span><span class="op">==</span><span class="va">GeneToPlot</span><span class="op">]</span></span>
<span></span>
<span><span class="va">PlotList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LTP</span> <span class="kw">in</span> <span class="va">LocsToPlot</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">t</span> <span class="op">&lt;-</span> <span class="va">t</span><span class="op">+</span><span class="fl">1</span></span>
<span>  <span class="va">CTRL_DF</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">LTP</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">CASE_DF</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="va">LTP</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">PlotData_dAD</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">==</span><span class="va">LTP</span>,<span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">PlotList</span><span class="op">[[</span><span class="va">t</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>ref_counts<span class="op">=</span><span class="va">CTRL_DF</span><span class="op">$</span><span class="va">ref_count</span>, var_counts<span class="op">=</span><span class="va">CTRL_DF</span><span class="op">$</span><span class="va">var_count</span>,</span>
<span>  pr<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rr</span>, prv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rv</span>, pv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_vv</span>, </span>
<span>  theta_hom <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHom</span>, theta_het <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHetC</span>, </span>
<span>  probshift <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">Pi</span>, SE<span class="op">=</span><span class="va">SEmedian</span>, ScaleCount <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">CovC_Med</span>, nbins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">t</span> <span class="op">&lt;-</span> <span class="va">t</span><span class="op">+</span><span class="fl">1</span></span>
<span>  <span class="va">PlotList</span><span class="op">[[</span><span class="va">t</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>ref_counts<span class="op">=</span><span class="va">CASE_DF</span><span class="op">$</span><span class="va">ref_count</span>, var_counts<span class="op">=</span><span class="va">CASE_DF</span><span class="op">$</span><span class="va">var_count</span>,</span>
<span>  pr<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rr</span>, prv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_rv</span>, pv<span class="op">=</span><span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">phi_vv</span>, </span>
<span>  theta_hom <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHom</span>, theta_het <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">ThetaHetT</span>, </span>
<span>  probshift <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">Pi</span>, SE<span class="op">=</span><span class="va">SEmedian</span>, ScaleCount <span class="op">=</span> <span class="va">PlotData_dAD</span><span class="op">$</span><span class="va">CovT_Med</span>, nbins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span>grobs <span class="op">=</span> <span class="va">PlotList</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="maelstRom_Allelic_Dispersion_tutorial_files/figure-html/unnamed-chunk-29-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 10 x64 (build 19045)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Dutch_Belgium.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Dutch_Belgium.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Dutch_Belgium.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                  </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Dutch_Belgium.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Brussels</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats    </span></span>
<span><span class="co">#&gt; [2] graphics </span></span>
<span><span class="co">#&gt; [3] grDevices</span></span>
<span><span class="co">#&gt; [4] utils    </span></span>
<span><span class="co">#&gt; [5] datasets </span></span>
<span><span class="co">#&gt; [6] methods  </span></span>
<span><span class="co">#&gt; [7] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] maelstRom_1.1.11</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gmp_0.7-5          </span></span>
<span><span class="co">#&gt;  [2] sass_0.4.9         </span></span>
<span><span class="co">#&gt;  [3] utf8_1.2.4         </span></span>
<span><span class="co">#&gt;  [4] generics_0.1.3     </span></span>
<span><span class="co">#&gt;  [5] gtools_3.9.5       </span></span>
<span><span class="co">#&gt;  [6] stringi_1.8.3      </span></span>
<span><span class="co">#&gt;  [7] lattice_0.22-5     </span></span>
<span><span class="co">#&gt;  [8] digest_0.6.35      </span></span>
<span><span class="co">#&gt;  [9] magrittr_2.0.3     </span></span>
<span><span class="co">#&gt; [10] evaluate_0.24.0    </span></span>
<span><span class="co">#&gt; [11] grid_4.3.3         </span></span>
<span><span class="co">#&gt; [12] fastmap_1.2.0      </span></span>
<span><span class="co">#&gt; [13] jsonlite_1.8.8     </span></span>
<span><span class="co">#&gt; [14] ggnewscale_0.5.0   </span></span>
<span><span class="co">#&gt; [15] gridExtra_2.3      </span></span>
<span><span class="co">#&gt; [16] purrr_1.0.2        </span></span>
<span><span class="co">#&gt; [17] fansi_1.0.6        </span></span>
<span><span class="co">#&gt; [18] scales_1.3.0       </span></span>
<span><span class="co">#&gt; [19] numDeriv_2016.8-1.1</span></span>
<span><span class="co">#&gt; [20] textshaping_0.4.0  </span></span>
<span><span class="co">#&gt; [21] jquerylib_0.1.4    </span></span>
<span><span class="co">#&gt; [22] Rdpack_2.6.1       </span></span>
<span><span class="co">#&gt; [23] cli_3.6.2          </span></span>
<span><span class="co">#&gt; [24] rlang_1.1.3        </span></span>
<span><span class="co">#&gt; [25] rbibutils_2.2.16   </span></span>
<span><span class="co">#&gt; [26] munsell_0.5.1      </span></span>
<span><span class="co">#&gt; [27] withr_3.0.0        </span></span>
<span><span class="co">#&gt; [28] MAGE_1.0.0.18      </span></span>
<span><span class="co">#&gt; [29] cachem_1.1.0       </span></span>
<span><span class="co">#&gt; [30] yaml_2.3.9         </span></span>
<span><span class="co">#&gt; [31] tools_4.3.3        </span></span>
<span><span class="co">#&gt; [32] parallel_4.3.3     </span></span>
<span><span class="co">#&gt; [33] memoise_2.0.1      </span></span>
<span><span class="co">#&gt; [34] dplyr_1.1.4        </span></span>
<span><span class="co">#&gt; [35] colorspace_2.1-0   </span></span>
<span><span class="co">#&gt; [36] ggplot2_3.5.1      </span></span>
<span><span class="co">#&gt; [37] hash_2.2.6.3       </span></span>
<span><span class="co">#&gt; [38] vctrs_0.6.5        </span></span>
<span><span class="co">#&gt; [39] R6_2.5.1           </span></span>
<span><span class="co">#&gt; [40] zoo_1.8-12         </span></span>
<span><span class="co">#&gt; [41] lifecycle_1.0.4    </span></span>
<span><span class="co">#&gt; [42] stringr_1.5.1      </span></span>
<span><span class="co">#&gt; [43] fs_1.6.4           </span></span>
<span><span class="co">#&gt; [44] htmlwidgets_1.6.4  </span></span>
<span><span class="co">#&gt; [45] MASS_7.3-60.0.1    </span></span>
<span><span class="co">#&gt; [46] ragg_1.3.2         </span></span>
<span><span class="co">#&gt; [47] pkgconfig_2.0.3    </span></span>
<span><span class="co">#&gt; [48] desc_1.4.3         </span></span>
<span><span class="co">#&gt; [49] pkgdown_2.0.9      </span></span>
<span><span class="co">#&gt; [50] pillar_1.9.0       </span></span>
<span><span class="co">#&gt; [51] bslib_0.7.0        </span></span>
<span><span class="co">#&gt; [52] gtable_0.3.5       </span></span>
<span><span class="co">#&gt; [53] data.table_1.16.0  </span></span>
<span><span class="co">#&gt; [54] glue_1.7.0         </span></span>
<span><span class="co">#&gt; [55] Rcpp_1.0.12        </span></span>
<span><span class="co">#&gt; [56] systemfonts_1.1.0  </span></span>
<span><span class="co">#&gt; [57] highr_0.11         </span></span>
<span><span class="co">#&gt; [58] xfun_0.45          </span></span>
<span><span class="co">#&gt; [59] tibble_3.2.1       </span></span>
<span><span class="co">#&gt; [60] tidyselect_1.2.1   </span></span>
<span><span class="co">#&gt; [61] rstudioapi_0.16.0  </span></span>
<span><span class="co">#&gt; [62] knitr_1.48         </span></span>
<span><span class="co">#&gt; [63] farver_2.1.2       </span></span>
<span><span class="co">#&gt; [64] htmltools_0.5.8    </span></span>
<span><span class="co">#&gt; [65] patchwork_1.2.0    </span></span>
<span><span class="co">#&gt; [66] labeling_0.4.3     </span></span>
<span><span class="co">#&gt; [67] rmarkdown_2.27     </span></span>
<span><span class="co">#&gt; [68] compiler_4.3.3     </span></span>
<span><span class="co">#&gt; [69] alabama_2023.1.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-dai2014modified" class="csl-entry">
Dai, Hongying, J Steven Leeder, and Yuehua Cui. 2014. <span>“A Modified
Generalized Fisher Method for Combining Probabilities from Dependent
Tests.”</span> <em>Frontiers in Genetics</em> 5: 32.
</div>
<div id="ref-liu2020cauchy" class="csl-entry">
Liu, Yaowu, and Jun Xie. 2020. <span>“Cauchy Combination Test: A
Powerful Test with Analytic p-Value Calculation Under Arbitrary
Dependency Structures.”</span> <em>Journal of the American Statistical
Association</em> 115 (529): 393–402.
</div>
<div id="ref-rohlfs" class="csl-entry">
Rohlfs, Rori V, and Bruce S Weir. 2008. <span>“Distributions of
Hardy–Weinberg Equilibrium Test Statistics.”</span> <em>Genetics</em>
180 (3): 1609–16.
</div>
<div id="ref-sha" class="csl-entry">
Sha, Qiuying, and Shuanglin Zhang. 2011. <span>“A Test of Hardy-Weinberg
Equilibrium in Structured Populations.”</span> <em>Genetic
Epidemiology</em> 35 (7): 671–78.
</div>
<div id="ref-teo" class="csl-entry">
Teo, Yik Y, Andrew E Fry, Taane G Clark, ES Tai, and Mark Seielstad.
2007. <span>“On the Usage of HWE for Identifying Genotyping
Errors.”</span> <em>Annals of Human Genetics</em> 71 (5): 701–3.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
