<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>maelstRom results exploration • maelstRom</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="maelstRom results exploration">
<meta property="og:description" content="maelstRom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>maelstRom results exploration</h1>
                        <h4 data-toc-skip class="author">Cedric
Stroobandt</h4>
            
            <h4 data-toc-skip class="date">2024-11-27</h4>
      
      
      <div class="hidden name"><code>maelstRom_results_exploration.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="maelstrom-explorations">maelstRom explorations<a class="anchor" aria-label="anchor" href="#maelstrom-explorations"></a>
</h2>
<p>The results from maelstRom’s Allelic Dispersion tutorial can be
explored even further with auxiliary data analyses, as documented in
this tutorial. It starts from the previously obtained Allelic Dispersion
tutorial results, which are loaded into the workspace below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">maelstRom</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning:</span></span>
<span><span class="co">#&gt; replacing</span></span>
<span><span class="co">#&gt; previous import</span></span>
<span><span class="co">#&gt; 'data.table::yearmon'</span></span>
<span><span class="co">#&gt; by</span></span>
<span><span class="co">#&gt; 'zoo::yearmon'</span></span>
<span><span class="co">#&gt; when loading</span></span>
<span><span class="co">#&gt; 'maelstRom'</span></span>
<span><span class="co">#&gt; Warning:</span></span>
<span><span class="co">#&gt; replacing</span></span>
<span><span class="co">#&gt; previous import</span></span>
<span><span class="co">#&gt; 'data.table::yearqtr'</span></span>
<span><span class="co">#&gt; by</span></span>
<span><span class="co">#&gt; 'zoo::yearqtr'</span></span>
<span><span class="co">#&gt; when loading</span></span>
<span><span class="co">#&gt; 'maelstRom'</span></span>
<span><span class="co">#&gt; Package 'maelstRom' version 1.1.11</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"VignetteImage"</span>, package <span class="op">=</span> <span class="st">"maelstRom"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th align="left">Locus</th>
<th align="left">Gene</th>
<th align="right">dAD_pval</th>
<th align="right">phi_rr</th>
<th align="right">phi_rv</th>
<th align="right">phi_vv</th>
<th align="right">Pi</th>
<th align="right">ThetaHetC</th>
<th align="right">ThetaHetT</th>
<th align="right">RhoC</th>
<th align="right">RhoT</th>
<th align="right">ThetaHom</th>
<th align="right">phi_rr_H0</th>
<th align="right">phi_rv_H0</th>
<th align="right">phi_vv_H0</th>
<th align="right">PiH0</th>
<th align="right">ThetaHetH0</th>
<th align="right">RhoH0</th>
<th align="right">ThetaHomH0</th>
<th align="right">phi_rr_OnlyC</th>
<th align="right">phi_rv_OnlyC</th>
<th align="right">phi_vv_OnlyC</th>
<th align="right">PiOnlyC</th>
<th align="right">ThetaHetOnlyC</th>
<th align="right">ThetaHomOnlyC</th>
<th align="right">phi_rr_OnlyT</th>
<th align="right">phi_rv_OnlyT</th>
<th align="right">phi_vv_OnlyT</th>
<th align="right">PiOnlyT</th>
<th align="right">ThetaHetOnlyT</th>
<th align="right">ThetaHomOnlyT</th>
<th align="right">NumHetC</th>
<th align="right">NumHetT</th>
<th align="left">RobFlagC</th>
<th align="left">RobFlagT</th>
<th align="right">HWEC</th>
<th align="right">HWET</th>
<th align="right">CovC</th>
<th align="right">CovT</th>
<th align="right">CovC_Med</th>
<th align="right">CovT_Med</th>
<th align="right">NumOutC</th>
<th align="right">NumOutT</th>
<th align="right">nrep_H0</th>
<th align="right">nrep_H1</th>
<th align="left">QualityC</th>
<th align="left">QualityT</th>
<th align="right">AD_EX_Corr</th>
<th align="right">AD_EX_Corr_pval</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Locus1</td>
<td align="left">Gene1</td>
<td align="right">0.0145737</td>
<td align="right">0.4266438</td>
<td align="right">0.4312392</td>
<td align="right">0.1421169</td>
<td align="right">0.4952374</td>
<td align="right">0.0243768</td>
<td align="right">0.0515579</td>
<td align="right">0.0237967</td>
<td align="right">0.0490300</td>
<td align="right">0.0216788</td>
<td align="right">0.4251959</td>
<td align="right">0.4319180</td>
<td align="right">0.1428861</td>
<td align="right">0.4973311</td>
<td align="right">0.0424388</td>
<td align="right">0.0407111</td>
<td align="right">0.0208894</td>
<td align="right">0.3904620</td>
<td align="right">0.4845264</td>
<td align="right">0.1250116</td>
<td align="right">0.4898086</td>
<td align="right">0.0173647</td>
<td align="right">0.0602616</td>
<td align="right">0.4472249</td>
<td align="right">0.4033651</td>
<td align="right">0.1494100</td>
<td align="right">0.4947613</td>
<td align="right">0.0539605</td>
<td align="right">0.0168325</td>
<td align="right">62.01938</td>
<td align="right">108.10186</td>
<td align="left">NoOutliers</td>
<td align="left">NoOutliers</td>
<td align="right">0.9189807</td>
<td align="right">0.4769813</td>
<td align="right">199.52344</td>
<td align="right">270.52985</td>
<td align="right">190.5</td>
<td align="right">255</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">18</td>
<td align="right">18</td>
<td align="left"></td>
<td align="left"></td>
<td align="right">0.3159567</td>
<td align="right">0.0008651</td>
</tr>
<tr class="even">
<td align="left">Locus2</td>
<td align="left">Gene1</td>
<td align="right">0.0001449</td>
<td align="right">0.7040331</td>
<td align="right">0.2767914</td>
<td align="right">0.0191755</td>
<td align="right">0.5494551</td>
<td align="right">0.0141543</td>
<td align="right">0.0751320</td>
<td align="right">0.0139567</td>
<td align="right">0.0698817</td>
<td align="right">0.0335140</td>
<td align="right">0.7042999</td>
<td align="right">0.2753414</td>
<td align="right">0.0203586</td>
<td align="right">0.5390438</td>
<td align="right">0.0504500</td>
<td align="right">0.0480271</td>
<td align="right">0.0351454</td>
<td align="right">0.7031276</td>
<td align="right">0.2812474</td>
<td align="right">0.0156250</td>
<td align="right">0.5647487</td>
<td align="right">0.0130104</td>
<td align="right">0.0366680</td>
<td align="right">0.7046997</td>
<td align="right">0.2748898</td>
<td align="right">0.0204105</td>
<td align="right">0.5235947</td>
<td align="right">0.0742229</td>
<td align="right">0.0322330</td>
<td align="right">35.99966</td>
<td align="right">73.67046</td>
<td align="left">NoOutliers</td>
<td align="left">NoOutliers</td>
<td align="right">0.8149959</td>
<td align="right">0.8801815</td>
<td align="right">105.92188</td>
<td align="right">125.97388</td>
<td align="right">96.0</td>
<td align="right">114</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">11</td>
<td align="right">14</td>
<td align="left"></td>
<td align="left"></td>
<td align="right">0.2560135</td>
<td align="right">0.0288009</td>
</tr>
<tr class="odd">
<td align="left">Locus3</td>
<td align="left">Gene1</td>
<td align="right">0.0000012</td>
<td align="right">0.7066314</td>
<td align="right">0.2747330</td>
<td align="right">0.0186356</td>
<td align="right">0.4951042</td>
<td align="right">0.0073516</td>
<td align="right">0.0785041</td>
<td align="right">0.0072980</td>
<td align="right">0.0727898</td>
<td align="right">0.0485708</td>
<td align="right">0.7073771</td>
<td align="right">0.2716372</td>
<td align="right">0.0209857</td>
<td align="right">0.4772722</td>
<td align="right">0.0454193</td>
<td align="right">0.0434460</td>
<td align="right">0.0537090</td>
<td align="right">0.7031522</td>
<td align="right">0.2812201</td>
<td align="right">0.0156277</td>
<td align="right">0.5093620</td>
<td align="right">0.0066971</td>
<td align="right">0.0837429</td>
<td align="right">0.7084733</td>
<td align="right">0.2720498</td>
<td align="right">0.0194768</td>
<td align="right">0.4576387</td>
<td align="right">0.0733042</td>
<td align="right">0.0398699</td>
<td align="right">35.99617</td>
<td align="right">72.63730</td>
<td align="left">NoOutliers</td>
<td align="left">NoOutliers</td>
<td align="right">0.8151967</td>
<td align="right">0.8697809</td>
<td align="right">124.57031</td>
<td align="right">150.34082</td>
<td align="right">116.5</td>
<td align="right">134</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">19</td>
<td align="right">14</td>
<td align="left"></td>
<td align="left"></td>
<td align="right">0.1352185</td>
<td align="right">0.2540302</td>
</tr>
<tr class="even">
<td align="left">Locus4</td>
<td align="left">Gene2</td>
<td align="right">0.3304979</td>
<td align="right">0.3821658</td>
<td align="right">0.4984763</td>
<td align="right">0.1193579</td>
<td align="right">0.6275309</td>
<td align="right">0.0208067</td>
<td align="right">0.0386687</td>
<td align="right">0.0203826</td>
<td align="right">0.0372291</td>
<td align="right">0.0272748</td>
<td align="right">0.3823811</td>
<td align="right">0.4982913</td>
<td align="right">0.1193276</td>
<td align="right">0.6279068</td>
<td align="right">0.0317866</td>
<td align="right">0.0308073</td>
<td align="right">0.0261031</td>
<td align="right">0.4163886</td>
<td align="right">0.4900306</td>
<td align="right">0.0935808</td>
<td align="right">0.6226814</td>
<td align="right">0.0202838</td>
<td align="right">0.0481506</td>
<td align="right">0.3651954</td>
<td align="right">0.5030342</td>
<td align="right">0.1317704</td>
<td align="right">0.6311273</td>
<td align="right">0.0387137</td>
<td align="right">0.0117994</td>
<td align="right">62.72392</td>
<td align="right">134.31014</td>
<td align="left">NoOutliers</td>
<td align="left">NoOutliers</td>
<td align="right">0.7073233</td>
<td align="right">0.7003476</td>
<td align="right">28.30469</td>
<td align="right">20.26592</td>
<td align="right">27.0</td>
<td align="right">20</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">19</td>
<td align="right">13</td>
<td align="left"></td>
<td align="left"></td>
<td align="right">0.3249952</td>
<td align="right">0.0106015</td>
</tr>
<tr class="odd">
<td align="left">Locus5</td>
<td align="left">Gene2</td>
<td align="right">0.6137441</td>
<td align="right">0.5690665</td>
<td align="right">0.3787043</td>
<td align="right">0.0522292</td>
<td align="right">0.5263937</td>
<td align="right">0.0188209</td>
<td align="right">0.0262031</td>
<td align="right">0.0184732</td>
<td align="right">0.0255340</td>
<td align="right">0.0440111</td>
<td align="right">0.5694019</td>
<td align="right">0.3783571</td>
<td align="right">0.0522410</td>
<td align="right">0.5268711</td>
<td align="right">0.0222538</td>
<td align="right">0.0217694</td>
<td align="right">0.0443944</td>
<td align="right">0.5469213</td>
<td align="right">0.3828949</td>
<td align="right">0.0701838</td>
<td align="right">0.5143016</td>
<td align="right">0.0180381</td>
<td align="right">0.0449810</td>
<td align="right">0.5790841</td>
<td align="right">0.3774147</td>
<td align="right">0.0435011</td>
<td align="right">0.5350999</td>
<td align="right">0.0263470</td>
<td align="right">0.0422924</td>
<td align="right">49.01055</td>
<td align="right">100.76974</td>
<td align="left">NoOutliers</td>
<td align="left">NoOutliers</td>
<td align="right">0.9999981</td>
<td align="right">0.7300886</td>
<td align="right">35.84375</td>
<td align="right">25.41573</td>
<td align="right">34.0</td>
<td align="right">24</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="right">27</td>
<td align="left"></td>
<td align="left"></td>
<td align="right">-0.2740332</td>
<td align="right">0.0357078</td>
</tr>
<tr class="even">
<td align="left">Locus6</td>
<td align="left">Gene2</td>
<td align="right">0.3453054</td>
<td align="right">0.1686252</td>
<td align="right">0.4541586</td>
<td align="right">0.3772162</td>
<td align="right">0.6812257</td>
<td align="right">0.0213316</td>
<td align="right">0.0021231</td>
<td align="right">0.0208861</td>
<td align="right">0.0021186</td>
<td align="right">0.0267148</td>
<td align="right">0.1671000</td>
<td align="right">0.4554184</td>
<td align="right">0.3774816</td>
<td align="right">0.6853380</td>
<td align="right">0.0132906</td>
<td align="right">0.0131163</td>
<td align="right">0.0299717</td>
<td align="right">0.1370091</td>
<td align="right">0.5524136</td>
<td align="right">0.3105773</td>
<td align="right">0.7119506</td>
<td align="right">0.0159835</td>
<td align="right">0.0207748</td>
<td align="right">0.1869743</td>
<td align="right">0.4025645</td>
<td align="right">0.4104612</td>
<td align="right">0.6594220</td>
<td align="right">0.0000000</td>
<td align="right">0.0320631</td>
<td align="right">70.70894</td>
<td align="right">106.27703</td>
<td align="left">NoOutliers</td>
<td align="left">NoOutliers</td>
<td align="right">0.4919931</td>
<td align="right">0.2581178</td>
<td align="right">16.69531</td>
<td align="right">11.32197</td>
<td align="right">15.0</td>
<td align="right">10</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">16</td>
<td align="right">37</td>
<td align="left"></td>
<td align="left"></td>
<td align="right">-0.3421650</td>
<td align="right">0.2763032</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="exploration-sample-level-ad-correlations">Exploration: Sample-level AD correlations<a class="anchor" aria-label="anchor" href="#exploration-sample-level-ad-correlations"></a>
</h2>
<p>If AD is reflective of underlying (epi)genetic dysregulation, it
stands to reason that the most extreme (allelically imbalanced) samples
that contribute most to increased AD (i.e. in the plots above: those
heterozygotes deviating most from <span class="math inline">\(\pi_{het}\)</span> - marked in a green vertical
dotted line - and thus contributing most to broadening of the
heterozygous beta-binomial fit) are also the most (epi)genetically
dysregulated ones. We can try to detect such a link by correlating each
heterozygous sample’s extremity to any known (epi)genetic dysregulation
of these samples, such as (promotor) methylation or the occurrence of
copy number alterations.</p>
<p>As this tutorial relies on SNP counts alone, we correlate this sample
extremity, or “sample-level AD” to the amount of expression itself.
After all, the previously mentioned mechanisms (promotor
hypermethylation, copy number gains or losses) are known to impact a
gene’s expression; so, accordingly, those very dysregulated samples
showing extreme sample-level AD could show the most deviant
(differential) expression as well. To take count biases into account,
the actual measure for a sample’s extremity (sample-level AD) is not
e.g. simply deviance from <span class="math inline">\(\pi_{het}\)</span>, but the -log of the minimum of
that sample’s heterozygous cumulative distribution function or its
complement, which will be higher the more extreme of an observation the
sample is.</p>
<p>Though correlating AD with expression is the default setting of the
<code>AD_correlator</code> function below, it accepts any per-sample
measure (if e.g. hypermethylation percentages or any other dysregulation
data are available) as an optional input.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">AD_EX_Corrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">AD_EX_Corr_pvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">LocOI</span> <span class="kw">in</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">CountDF</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="va">LocOI</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">dAD_fitres</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Locus</span> <span class="op">==</span> <span class="va">LocOI</span>,<span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">CorResult</span> <span class="op">&lt;-</span> <span class="fu">AD_correlator</span><span class="op">(</span><span class="va">CountDF</span>, <span class="va">dAD_fitres</span>, MinCount <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">AD_EX_Corrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">AD_EX_Corrs</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">CorResult</span><span class="op">[</span><span class="st">"correlation"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">AD_EX_Corr_pvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">AD_EX_Corr_pvals</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">CorResult</span><span class="op">[</span><span class="st">"p-value"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">dAD_res</span><span class="op">$</span><span class="va">AD_EX_Corr</span> <span class="op">&lt;-</span> <span class="va">AD_EX_Corrs</span></span>
<span><span class="va">dAD_res</span><span class="op">$</span><span class="va">AD_EX_Corr_pval</span> <span class="op">&lt;-</span> <span class="va">AD_EX_Corr_pvals</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">Locus</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">AD_EX_Corr_pval</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">AD_EX_Corr_pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">HWEC</span> <span class="op">&gt;</span> <span class="fl">0.001</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">HWET</span> <span class="op">&gt;</span> <span class="fl">0.001</span><span class="op">]</span><span class="op">)</span>, <span class="st">","</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>Locus1, Locus2, Locus4, Locus5, Locus40, Locus74, Locus85, Locus93,
Locus112, Locus119, Locus141, Locus174, Locus190, Locus191,</p>
<p>A minimum count filter of 20 is used here to avoid low-count-based
biases in the correlations’ computation. Eventually, the returned number
of actually significant correlations is relatively small. But they are
meant as more of a data exploration tool rather than a hard result or
filter criterion anyway, as their analysis isn’t as robust as the
population-level AD estimates to detect biologically relevant genes, and
both AD in expression can be regulated in many different ways and
influenced by many biases (e.g. tumor purity), making overly strong
correlations between them and/or other (epi)genetic data not evident.
Nevertheless, in cases where e.g. hypermethylation is strongly suspected
to cause an allelic imbalance in sequencing data of afflicted samples,
computing these correlations can provide soft evidence to that
effect.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 10 x64 (build 19045)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Dutch_Belgium.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Dutch_Belgium.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Dutch_Belgium.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                  </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Dutch_Belgium.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Brussels</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats    </span></span>
<span><span class="co">#&gt; [2] graphics </span></span>
<span><span class="co">#&gt; [3] grDevices</span></span>
<span><span class="co">#&gt; [4] utils    </span></span>
<span><span class="co">#&gt; [5] datasets </span></span>
<span><span class="co">#&gt; [6] methods  </span></span>
<span><span class="co">#&gt; [7] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] maelstRom_1.1.11</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gmp_0.7-5          </span></span>
<span><span class="co">#&gt;  [2] sass_0.4.9         </span></span>
<span><span class="co">#&gt;  [3] utf8_1.2.4         </span></span>
<span><span class="co">#&gt;  [4] generics_0.1.3     </span></span>
<span><span class="co">#&gt;  [5] gtools_3.9.5       </span></span>
<span><span class="co">#&gt;  [6] stringi_1.8.3      </span></span>
<span><span class="co">#&gt;  [7] lattice_0.22-5     </span></span>
<span><span class="co">#&gt;  [8] digest_0.6.35      </span></span>
<span><span class="co">#&gt;  [9] magrittr_2.0.3     </span></span>
<span><span class="co">#&gt; [10] evaluate_0.24.0    </span></span>
<span><span class="co">#&gt; [11] grid_4.3.3         </span></span>
<span><span class="co">#&gt; [12] fastmap_1.2.0      </span></span>
<span><span class="co">#&gt; [13] jsonlite_1.8.8     </span></span>
<span><span class="co">#&gt; [14] ggnewscale_0.5.0   </span></span>
<span><span class="co">#&gt; [15] purrr_1.0.2        </span></span>
<span><span class="co">#&gt; [16] fansi_1.0.6        </span></span>
<span><span class="co">#&gt; [17] scales_1.3.0       </span></span>
<span><span class="co">#&gt; [18] numDeriv_2016.8-1.1</span></span>
<span><span class="co">#&gt; [19] textshaping_0.4.0  </span></span>
<span><span class="co">#&gt; [20] jquerylib_0.1.4    </span></span>
<span><span class="co">#&gt; [21] Rdpack_2.6.1       </span></span>
<span><span class="co">#&gt; [22] cli_3.6.2          </span></span>
<span><span class="co">#&gt; [23] rlang_1.1.3        </span></span>
<span><span class="co">#&gt; [24] rbibutils_2.2.16   </span></span>
<span><span class="co">#&gt; [25] munsell_0.5.1      </span></span>
<span><span class="co">#&gt; [26] MAGE_1.0.0.18      </span></span>
<span><span class="co">#&gt; [27] cachem_1.1.0       </span></span>
<span><span class="co">#&gt; [28] yaml_2.3.9         </span></span>
<span><span class="co">#&gt; [29] tools_4.3.3        </span></span>
<span><span class="co">#&gt; [30] parallel_4.3.3     </span></span>
<span><span class="co">#&gt; [31] memoise_2.0.1      </span></span>
<span><span class="co">#&gt; [32] dplyr_1.1.4        </span></span>
<span><span class="co">#&gt; [33] colorspace_2.1-0   </span></span>
<span><span class="co">#&gt; [34] ggplot2_3.5.1      </span></span>
<span><span class="co">#&gt; [35] hash_2.2.6.3       </span></span>
<span><span class="co">#&gt; [36] vctrs_0.6.5        </span></span>
<span><span class="co">#&gt; [37] R6_2.5.1           </span></span>
<span><span class="co">#&gt; [38] zoo_1.8-12         </span></span>
<span><span class="co">#&gt; [39] lifecycle_1.0.4    </span></span>
<span><span class="co">#&gt; [40] stringr_1.5.1      </span></span>
<span><span class="co">#&gt; [41] fs_1.6.4           </span></span>
<span><span class="co">#&gt; [42] htmlwidgets_1.6.4  </span></span>
<span><span class="co">#&gt; [43] MASS_7.3-60.0.1    </span></span>
<span><span class="co">#&gt; [44] ragg_1.3.2         </span></span>
<span><span class="co">#&gt; [45] pkgconfig_2.0.3    </span></span>
<span><span class="co">#&gt; [46] desc_1.4.3         </span></span>
<span><span class="co">#&gt; [47] pkgdown_2.0.9      </span></span>
<span><span class="co">#&gt; [48] pillar_1.9.0       </span></span>
<span><span class="co">#&gt; [49] bslib_0.7.0        </span></span>
<span><span class="co">#&gt; [50] gtable_0.3.5       </span></span>
<span><span class="co">#&gt; [51] data.table_1.16.0  </span></span>
<span><span class="co">#&gt; [52] glue_1.7.0         </span></span>
<span><span class="co">#&gt; [53] Rcpp_1.0.12        </span></span>
<span><span class="co">#&gt; [54] systemfonts_1.1.0  </span></span>
<span><span class="co">#&gt; [55] xfun_0.45          </span></span>
<span><span class="co">#&gt; [56] tibble_3.2.1       </span></span>
<span><span class="co">#&gt; [57] tidyselect_1.2.1   </span></span>
<span><span class="co">#&gt; [58] rstudioapi_0.16.0  </span></span>
<span><span class="co">#&gt; [59] knitr_1.48         </span></span>
<span><span class="co">#&gt; [60] htmltools_0.5.8    </span></span>
<span><span class="co">#&gt; [61] patchwork_1.2.0    </span></span>
<span><span class="co">#&gt; [62] rmarkdown_2.27     </span></span>
<span><span class="co">#&gt; [63] compiler_4.3.3     </span></span>
<span><span class="co">#&gt; [64] alabama_2023.1.0</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
