<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>maelstRom tutorial • maelstRom</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="maelstRom tutorial">
<meta property="og:description" content="maelstRom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/maelstRom_expanded_tutorial.html">maelstRom expanded tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_tutorial.html">maelstRom tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>maelstRom tutorial</h1>
                        <h4 data-toc-skip class="author">Cedric
Stroobandt</h4>
            
            <h4 data-toc-skip class="date">2022-10-13</h4>
      
      
      <div class="hidden name"><code>maelstRom_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="mage-modeller-of-allelic-gene-expression">maelstRom: Modeller of Allelic Gene Expression<a class="anchor" aria-label="anchor" href="#mage-modeller-of-allelic-gene-expression"></a>
</h2>
<p>The maelstRom package provides extensive functions for RNAseq-based
allelic analyses, ranging from basic tasks such as genotyping - albeit
allele-specific-expression (ASE) aware and relying on RNAseq data only -
to the analysis of more complex ASE-related phenomena such as allelic
bias (AB) detection, differential allelic divergence (AD) detection in
case (e.g. tumor) populations compared to controls, and detection of
imprinted loci in controls with subsequent detection of loss of
imprinting (LOI) in cases. More information on these phenomena is
provided throughout this vignette. As maelstRom relies on population-level
models, large numbers of samples are extremely beneficial (100 or more
is ideal; 20 as a bare minimum).</p>
<p>This vignette makes use of <em>wrapper functions that both enable
parallellization and streamline commandos</em>, those being
<code><a href="../reference/AllelicMeta_est_par.html">AllelicMeta_est_par()</a></code>, <code><a href="../reference/BetaBinomGenotyping.html">BetaBinomGenotyping()</a></code>,
<code><a href="../reference/dAD_analysis.html">dAD_analysis()</a></code> and <code><a href="../reference/impr_LOI_analysis.html">impr_LOI_analysis()</a></code>.
Under-the-hood, these call several maelstRom functions and perform
intermediary processing, which makes for a more readable vignette and an
easy-to-use plug-in pipeline, but also make the entire process more of a
black box. As such, maelstRom also includes an expanded version of this
vignette foregoing these wrapper functions, which contains more code but
could be interesting for experienced users wanting more insight in, or
control over, every step of the pipeline.</p>
</div>
<div class="section level2">
<h2 id="a-toy-dataset">A toy dataset<a class="anchor" aria-label="anchor" href="#a-toy-dataset"></a>
</h2>
<p>maelstRom’s main input consists of per-locus per-sample nucleotide counts
(A/T/C/G). These are relatively straightforward to extract from BAM/SAM
files via custom scripting or established pipelines such as SAMtools’
mpileup or GATK’s CollectAllelicCounts functionality (<a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037594071-CollectAllelicCounts" class="external-link uri">https://gatk.broadinstitute.org/hc/en-us/articles/360037594071-CollectAllelicCounts</a>),
though the latter already filters the data down to one reference- and
one variant-allele count.</p>
<p>maelstRom includes a small example of such data, including both a control
(e.g. healthy tissue) and case (e.g. tumor tissue) dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"maelstRom"</span>, package <span class="op">=</span> <span class="st">"maelstRom"</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ControlCounts</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">locus_id</th>
<th align="left">reference alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">sample_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">397</td>
<td align="right">0</td>
<td align="left">sample1</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">398</td>
<td align="right">0</td>
<td align="left">sample2</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">184</td>
<td align="right">0</td>
<td align="right">217</td>
<td align="right">0</td>
<td align="left">sample3</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">111</td>
<td align="right">0</td>
<td align="right">230</td>
<td align="right">0</td>
<td align="left">sample4</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">437</td>
<td align="right">0</td>
<td align="right">577</td>
<td align="right">0</td>
<td align="left">sample5</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">280</td>
<td align="right">0</td>
<td align="left">sample6</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CaseCounts</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">locus_id</th>
<th align="left">reference alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">sample_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">582</td>
<td align="right">0</td>
<td align="left">sample1</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">229</td>
<td align="right">5</td>
<td align="right">309</td>
<td align="right">0</td>
<td align="left">sample2</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">847</td>
<td align="right">0</td>
<td align="left">sample3</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">134</td>
<td align="right">0</td>
<td align="left">sample4</td>
</tr>
<tr class="odd">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">8</td>
<td align="right">4</td>
<td align="right">4751</td>
<td align="right">5</td>
<td align="left">sample5</td>
</tr>
<tr class="even">
<td align="left">loc0</td>
<td align="left">C/A/G</td>
<td align="right">722</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">4</td>
<td align="left">sample6</td>
</tr>
</tbody>
</table>
<p>These toy datasets contain nucleotide counts for 269 loci, every
locus being covered by a varying number of samples (e.g. 127 samples
covering loc1 in the control dataset):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCounts</span><span class="op">$</span><span class="va">locus_id</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] 269</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">ControlCounts</span><span class="op">$</span><span class="va">locus_id</span> <span class="op">==</span> <span class="st">"loc1"</span><span class="op">)</span></span></code></pre></div>
<p>[1] 127</p>
<p>Only a control dataset is required for RNAseq-based genotyping,
imprinting detection, and allelic bias detection, but differential
analyses studying e.g. the effect of disease on allelic expression
(differential allelic divergence, loss of imprinting) require a
comparison of cases against controls.</p>
</div>
<div class="section level2">
<h2 id="input-pre-processing">Input pre-processing<a class="anchor" aria-label="anchor" href="#input-pre-processing"></a>
</h2>
<p>maelstRom expects its input data to be formatted as <em>lists of
dataframes, every dataframe containing the per-sample nucleotide counts
of one particular locus</em>, and has some requirements about column
naming as well. The first function we’ll be using is
<code><a href="../reference/standard_alleles.html">standard_alleles()</a></code>, which expects the columns
<code>ref_alleles</code>, <code>A</code>, <code>T</code>, <code>C</code>
and <code>G</code>. Therefore, here, the current
<code>reference alleles()</code> column in the toy dataset should be
renamed to <code>ref_alleles</code> to comply with input requirements;
each of these requirements are listed on the respective function’s help
pages, though besides this initial setup all remaining input
requirements will be met if following this entire vignette from start to
finish:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># maelstRom expects lists:</span></span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">caseList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ControlCounts</span><span class="op">$</span><span class="va">locus_id</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>              <span class="co"># For every locus...</span></span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">ControlCounts</span><span class="op">[</span><span class="va">ControlCounts</span><span class="op">$</span><span class="va">locus_id</span><span class="op">==</span><span class="va">n</span>,<span class="op">]</span> <span class="co"># extract per-sample nucleotide counts</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ref_alleles"</span>                <span class="co"># re-name the reference_allele column </span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span>                          <span class="co"># put it into the list</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">CaseCounts</span><span class="op">$</span><span class="va">locus_id</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">CaseCounts</span><span class="op">[</span><span class="va">CaseCounts</span><span class="op">$</span><span class="va">locus_id</span><span class="op">==</span><span class="va">n</span>,<span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ref_alleles"</span></span>
<span>  <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="determine-one-reference-and-one-variant-allele">Determine one reference and one variant allele<a class="anchor" aria-label="anchor" href="#determine-one-reference-and-one-variant-allele"></a>
</h2>
<p>maelstRom employs beta-binomial modelling of allele counts, which implies
only two alleles are allowed (which is compatible with the large
majority of all loci in most diploid species, including humans).
Accounting for more alleles would require a different approach,
possibilities being the use of multiple beta-binomial models (for all
alleles against one set reference, or even all pairwise combinations) or
of (beta-)multinomial models”. The latter is, however, not supported by
maelstRom at the moment.</p>
<p>So as it stands, we need to select one reference and one variant
allele and retain only their count data. This can be done by hand by a
researcher based on biological knowledge or through some automated
procedure, e.g. retaining only the most common alleles. maelstRom’s
<code><a href="../reference/standard_alleles.html">standard_alleles()</a></code> function combines both approaches by
taking a look at the standard alleles for every locus according to dbSNP
which are supplied by the <code>ref_alleles</code> column, then
selecting the two most common alleles from among these suggested
standard alleles only. Should one wish to solely rely on RNAseq count
files without external annotation, it’s possible to supply
<code><a href="../reference/standard_alleles.html">standard_alleles()</a></code> with <code>A/T/C/G</code> strings in all
of its <code>ref_alleles</code> columns, which will result in the most
abundant alleles being selected.</p>
<p><code><a href="../reference/standard_alleles.html">standard_alleles()</a></code> updates each dataframe in our lists,
retaining only two alleles in the <code>ref_alleles</code> column and
including additional columns specifying which allele is chosen as
reference and which as variant allele, alongside their respective counts
(the reference will be the most prevalent of the two in terms of total
count in all RNAseq data combined for that particular locus):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/standard_alleles.html">standard_alleles</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="st">"loc27"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="6%">
<col width="11%">
<col width="15%">
<col width="3%">
<col width="5%">
<col width="3%">
<col width="3%">
<col width="12%">
<col width="5%">
<col width="5%">
<col width="12%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">locus_id</th>
<th align="left">ref_alleles</th>
<th align="right">A</th>
<th align="right">T</th>
<th align="right">C</th>
<th align="right">G</th>
<th align="left">sample_id</th>
<th align="left">ref</th>
<th align="left">var</th>
<th align="right">ref_count</th>
<th align="right">var_count</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1043</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">67</td>
<td align="right">43</td>
<td align="right">0</td>
<td align="left">sample1</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">67</td>
<td align="right">43</td>
</tr>
<tr class="even">
<td align="left">1067</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">2</td>
<td align="right">101</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="left">sample2</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">101</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">1091</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">80</td>
<td align="right">0</td>
<td align="left">sample3</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0</td>
<td align="right">80</td>
</tr>
<tr class="even">
<td align="left">1115</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">90</td>
<td align="right">0</td>
<td align="left">sample4</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">0</td>
<td align="right">90</td>
</tr>
<tr class="odd">
<td align="left">1139</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">40</td>
<td align="right">63</td>
<td align="right">0</td>
<td align="left">sample5</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">40</td>
<td align="right">63</td>
</tr>
<tr class="even">
<td align="left">1163</td>
<td align="left">loc27</td>
<td align="left">C/T</td>
<td align="right">0</td>
<td align="right">94</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">sample6</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">94</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>We could run <code><a href="../reference/standard_alleles.html">standard_alleles()</a></code> on the case data as
well, but to enable comparison between both datasets, we enforce
selected alleles to be the same for corresponding loci in the control
and case datasets (this makes sense from a biological perspective as
well):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">caseList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">interDF</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref_alleles</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_alleles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">var</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">ref_count</span> <span class="op">&lt;-</span> <span class="va">interDF</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">==</span><span class="va">interDF</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">interDF</span><span class="op">$</span><span class="va">var_count</span> <span class="op">&lt;-</span> <span class="va">interDF</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">interDF</span><span class="op">)</span><span class="op">==</span><span class="va">interDF</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">interDF</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prior-filtering">Prior filtering<a class="anchor" aria-label="anchor" href="#prior-filtering"></a>
</h2>
<p>Before doing any analyses, it might be beneficial (in terms of
runtime) to filter out low-quality loci. maelstRom’s
<code><a href="../reference/prior_filter.html">prior_filter()</a></code> offers this functionality, leaving out loci
based on a minimal median coverage across all samples, a minimal number
of samples, or a crudely estimated minimal minor allele frequency (just
the % abundance of the variant allele over all RNAseq counts). After
all, loci with a too low minor allele frequency will feature
insufficient numbers of heterozygous individuals, which are required to
observe ASE effects. Also, if the <code>checkref_filter</code> argument
is set to TRUE, <code><a href="../reference/prior_filter.html">prior_filter()</a></code> checks whether the total
nucleotide counts truly suggest there are only two occurring alleles as
assumed in the previous step, or whether there’s heuristic evidence to
the contrary, in which case the locus is removed (details are found on
this function’s help page).</p>
<p>For now, we’ll restrict ourselves to just a “minimum number of
samples” filter of 20 in controls (only counting samples with at least
one reference or variant read). The first analysis is genotyping after
all, and researchers might be interested in getting the most likely
genotype for every sample regardless of median coverage or minor allele
frequency across the entire locus (as maelstRom is a population-level
modeller, a minimal number of samples of at least 20 is always advised).
We apply no actual filtering on the case dataset however, but rather
retain a locus’ data if it’s retained in the control set as well, as
comparison against controls is the case dataset’s main purpose. We do
run <code>prior_filter</code> on cases with no actual filter settings
applied though, which reduces the function to simply removing samples
having a zero-count for both reference and variant alleles (not removing
these would give NA-results during several calculations in the
pipeline). This reduces the toy datasets from 269 to 264 loci:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/prior_filter.html">prior_filter</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span>, min_median_cov <span class="op">=</span> <span class="fl">0</span>, </span>
<span>    min_nr_samples <span class="op">=</span> <span class="fl">20</span>, checkref_filter <span class="op">=</span> <span class="cn">TRUE</span>, prior_allelefreq_filter <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    min_PriorAlleleFreq <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/prior_filter.html">prior_filter</a></span><span class="op">(</span><span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span>, min_median_cov <span class="op">=</span> <span class="fl">0</span>, </span>
<span>    min_nr_samples <span class="op">=</span> <span class="fl">0</span>, checkref_filter <span class="op">=</span> <span class="cn">FALSE</span>, prior_allelefreq_filter <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    min_PriorAlleleFreq <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">caseList</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="enabling-parallellization">Enabling parallellization<a class="anchor" aria-label="anchor" href="#enabling-parallellization"></a>
</h2>
<p>Before any further analyses, this vignette first splits input data
lists into <code>NC</code> chunks below as a nested list,
<code>NC</code> being a number of cores as the coding in this vignette
allows maelstRom to be run in parallel on a server using R’s
<code>parallel</code> package. A local (Windows) installation is only
able to use 1 core at a time though, so <code>NC</code> is set to one in
the code below. Feel free to change it when running maelstRom on e.g. a
linux-based server:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">NC</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># Number of Cores</span></span>
<span><span class="va">NS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">)</span> </span>
<span><span class="va">spl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span>,<span class="va">NC</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">NS</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">NC</span><span class="op">-</span><span class="va">NS</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">NS</span><span class="op">/</span><span class="va">NC</span><span class="op">)</span><span class="op">*</span><span class="va">NC</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># Helps in splitting input data</span></span>
<span><span class="va">ParCTRL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span>  <span class="va">ParCTRL</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ParCASE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span>  <span class="va">ParCASE</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">spl</span><span class="op">[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="metaparameter-estimation">Metaparameter estimation<a class="anchor" aria-label="anchor" href="#metaparameter-estimation"></a>
</h2>
<p>Before model fitting, we need to estimate the population-level
metaparameters shared across all loci, those being the inbreeding
coefficient of the population and the sequencing error rate. These
values can be set based on prior knowledge about the population and the
sequencing protocol of choice, but maelstRom’s <code><a href="../reference/AllelicMeta_est.html">AllelicMeta_est()</a></code>
can estimate them based on observed <em>control</em> data, which is
recommended anyway for a good model fit. This function fits a regular
binomial mixture model assuming no allelic bias (i.e. perfectly balanced
expression of the reference- and variant allele in heterozygotes) to
control read count data via expectation maximization:</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(counts_{ref}|counts_{total}) = &amp;\ P_{rr} * {\tt
pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=1-SE) \ + \\
&amp;\ P_{rv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total},
p=0.5) \ + \\
&amp;\ P_{vv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=SE)
\\
\end{aligned}
\]</span></p>
<p>With <span class="math inline">\((P_{rr}, P_{rv}, P_{vv})\)</span>
to-be-fitted genotype frequencies from which the inbreeding coefficient
can be estimated, and SE a to-be-fitted sequencing error rate estimate.
The function also returns a number of outputs that can be used as
reliability measures, including a preliminary reference allele frequency
estimate across genotypes (from which the minor allele frequency can be
calculated), median coverage and number of samples. The SE-estimate
itself shouldn’t be absurdly high either. Default cutoff values for
these quality metrics are suggested in the code below
(<code>pA_filt</code> dictating minimal estimated minor allele
frequency). We <em>don’t</em> filter these loci out however; a
suboptimal fit by <code><a href="../reference/AllelicMeta_est.html">AllelicMeta_est()</a></code> may be remedied by the
later beta-binomial fit allowing unequal heterozygous allele expression
for actual genotyping and allelic bias detection.</p>
<p>Note that the function being called is
<code><a href="../reference/AllelicMeta_est_par.html">AllelicMeta_est_par()</a></code>, a wrapper function of
<code><a href="../reference/AllelicMeta_est.html">AllelicMeta_est()</a></code> handling all steps described above and
providing, besides updated control-count-dataframes, vectors containing
reliable (according to the filters mentioned above) metaparameter
estimates, of which one still has to take e.g. the mean or median to get
final metaparameters estimates. In this vignette, we use the more robust
median. As mentioned, additional insight into this wrapper function can
be learned from its source code, or in maelstRom’s extended vignette.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">GenoFinData</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParCTRL</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va"><a href="../reference/AllelicMeta_est_par.html">AllelicMeta_est_par</a></span>,</span>
<span>  pA_filt <span class="op">=</span> <span class="fl">0.15</span>, SE_filt <span class="op">=</span> <span class="fl">0.035</span>, NumSamp_filt <span class="op">=</span> <span class="fl">20</span>, MedianCov_filt <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ParCTRL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">SE_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">F_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SEmedian</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">SE_vec</span><span class="op">)</span></span>
<span><span class="va">Fmedian</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">F_vec</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#estimates of resp. sequencing error rate and inbreeding coefficient:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">SEmedian</span>, <span class="va">Fmedian</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>[1] 0.0033650786 [2] -0.0004878278</p>
<p>These metaparameter estimates seem reasonable. Please note that an
extremely low sequencing error metaparameter would make fitting of
homozygous peaks later down the pipeline way too strict, failing to
accomodate for even a single faulty read. As such, <em>setting SE lower
than 0.002 is not recommended</em>. An inbreeding coefficient close to
zero is a sign of a panmictic population and commonplace when working on
human data, but it could be more extreme in specialized experimental
setups.</p>
</div>
<div class="section level2">
<h2 id="genotyping-rnaseq-data-allelic-bias-detection">Genotyping RNAseq data &amp; allelic bias detection<a class="anchor" aria-label="anchor" href="#genotyping-rnaseq-data-allelic-bias-detection"></a>
</h2>
<p>For both genotyping and allelic bias detection, we can use maelstRom’s
<code><a href="../reference/EMfit_betabinom_robust.html">EMfit_betabinom_robust()</a></code> function, which fits a
beta-binomial mixture model using expectation maximization while
allowing for unequal allelic expression in heterozygotes:</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(counts_{ref}|counts_{total}) = &amp;\ P_{rr} * {\tt
pBetaBinom}(x=counts_{ref} \ |\ n=counts_{total}, \pi=1-SE,
\theta=\theta_{hom}) \ + \\
&amp;\ P_{rv} * {\tt pBetaBinom}(x=counts_{ref} \ |\ n=counts_{total},
\pi=\pi_{het}, \theta=\theta_{het}) \ + \\
&amp;\ P_{vv} * {\tt pBetaBinom}(x=counts_{ref} \ |\ n=counts_{total},
\pi=SE, \theta=\theta_{hom}) \\
\end{aligned}
\]</span></p>
<p>With <span class="math inline">\((P_{rr}, P_{rv}, P_{vv})\)</span>
fitted genotype frequencies, <span class="math inline">\((\theta_{hom},
\theta_{het})\)</span> fitted overdispersion parameters in homozygotes
and heterozygotes respectively, and <span class="math inline">\(\pi_{het}\)</span> the heterozygous <span class="math inline">\(\pi\)</span> parameter indicating the expected
reference allele fraction, thus allowing for unequal allelic expression
in heterozygotes. <code><a href="../reference/dpqrBetaBinom.html">pBetaBinom()</a></code> is a function included in
the maelstRom package for the beta-binomial PMF, using a <span class="math inline">\(\pi, \theta\)</span> parameterisation (probability
of success, here defined as observing a reference read, and
overdispersion, respectively) instead of the classical <span class="math inline">\(\alpha, \beta\)</span> parameterisation as given
on e.g. <a href="https://en.wikipedia.org/wiki/Beta-binomial_distribution" class="external-link uri">https://en.wikipedia.org/wiki/Beta-binomial_distribution</a>.
The relationship between them is <span class="math inline">\(\pi =
\alpha/(\alpha+\beta)\)</span> and <span class="math inline">\(\theta =
1/(\alpha+\beta)\)</span>, resulting in a <span class="math inline">\(\pi\)</span> between 0 and 1 and a <span class="math inline">\(\theta\)</span> between 0 and infinity.</p>
<p>We term unequal heterozygous allele expression captured by <span class="math inline">\(\pi_{het}\)</span> <em>Allelic Bias (AB)</em>. AB
can be caused by both biological (e.g. cis-eqtl loci) and technical
(e.g. alignment bias) mechanisms. The fit is robust in the sense that
samples with a disproportionately high impact on the heterozygous
distributional parameters (determined using parameter-MLEs leaving out
that particular sample) are not used for the final fit, though they are
retained in the output data and genotyped using the final model (while
being marked as outlier). The reason for focussing on heterozygous
distributional parameters is that these are still being fitted (the
homozygous <span class="math inline">\(\pi\)</span> parameters are
already fixed by the metaparameter SE) and are actually affected by
allele-specific expression effects (<span class="math inline">\(\pi_{het}\)</span> can indicate AB and <span class="math inline">\(\theta_{het}\)</span> can indicate allelic
divergence, see later; meanwhile <span class="math inline">\(\theta_hom\)</span> is a nuisance parameter at
best). Do note, however, that the robust fit is significantly slower
than the non-robust one provided by the <code><a href="../reference/EMfit_betabinom.html">EMfit_betabinom()</a></code>
function (proportional to the number of samples), and its merit for
genotyping and AB-detection is limited, so one might opt for the latter
instead under certain circumstances (refer to maelstRom’s extended vignette
and modify the pipeline accordingly if you wish to de this). A robust
fit is, however, very valuable for statistical inference on the
overdispersion parameter <span class="math inline">\(\theta\)</span>
(see the next section).</p>
<p>maelstRom’s functions have many optional input parameters to customize the
fitting procedure, but throughout this tutorial and for most
applications the defaults should suffice. Feel free to read up on all
options on each function’s help page. After performing the fit and
writing results of interest into dataframes, we also perform a
chi-square test assessing Hardy-Weinberg Equilibrium using our final
genotypes and the previously determined inbreeding coefficient
hyperparameter, to be used for filtering later.</p>
<p>All steps described above are handled by
<code><a href="../reference/BetaBinomGenotyping.html">BetaBinomGenotyping()</a></code>, a wrapper function of
<code><a href="../reference/EMfit_betabinom_robust.html">EMfit_betabinom_robust()</a></code>, <code><a href="../reference/median_AB.html">median_AB()</a></code>, and
<code><a href="../reference/HWE_chisquared.html">HWE_chisquared()</a></code>, which returns updated per-locus
dataframes with information of the separate samples (e.g. genotype
probabilities) as well as a dataframe containing AB-analysis results
(see <code>Geno_AB_res</code> below). Please refer to
<code><a href="../reference/BetaBinomGenotyping.html">BetaBinomGenotyping()</a></code>’s source code or maelstRom’s extended
vignette for additional insight into all component steps.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">GenoFinData</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParCTRL</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va"><a href="../reference/BetaBinomGenotyping.html">BetaBinomGenotyping</a></span>, </span>
<span>                                   SE <span class="op">=</span> <span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ParCTRL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">controlList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">c</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Geno_AB_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">GenoFinData</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>As a sidenote, while this vignette represses warnings, calling upon
maelstRom’s beta-binomial density function (<code><a href="../reference/dpqrBetaBinom.html">pBetaBinom()</a></code>, which
<code><a href="../reference/EMfit_betabinom_robust.html">EMfit_betabinom_robust()</a></code> does) can result in a memory limit
warning which can usually be ignored; most of the time it signifies that
a certain locus’ beta-binomial fit was so close to a regular binomial
fit that the overdispersion parameter is extremely close to zero,
leading to precision errors in beta-binomial calculations which will
trigger maelstRom to use regular binomial functions instead.</p>
<p>The genotyping results can now be retrieved from the
<code>controlList</code> object, being included as the “genotypeN”
column in each locus’ dataframe:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="st">"loc27"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"locus_id"</span>, <span class="st">"sample_id"</span>, <span class="st">"ref"</span>, </span>
<span>                                            <span class="st">"var"</span>, <span class="st">"genotypeN"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">locus_id</th>
<th align="left">sample_id</th>
<th align="left">ref</th>
<th align="left">var</th>
<th align="left">genotypeN</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1043</td>
<td align="left">loc27</td>
<td align="left">sample1</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rv</td>
</tr>
<tr class="even">
<td align="left">1067</td>
<td align="left">loc27</td>
<td align="left">sample2</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rr</td>
</tr>
<tr class="odd">
<td align="left">1091</td>
<td align="left">loc27</td>
<td align="left">sample3</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">vv</td>
</tr>
<tr class="even">
<td align="left">1115</td>
<td align="left">loc27</td>
<td align="left">sample4</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">vv</td>
</tr>
<tr class="odd">
<td align="left">1139</td>
<td align="left">loc27</td>
<td align="left">sample5</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rv</td>
</tr>
<tr class="even">
<td align="left">1163</td>
<td align="left">loc27</td>
<td align="left">sample6</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="left">rr</td>
</tr>
</tbody>
</table>
<p>The <code>Geno_AB_res</code> dataframe contains the results of
AB-detection, i.e. testing whether the reference allele expression ratio
differs from a “perfectly balanced” ratio of 0.5, i.e. is there a
preference to detect one of both alleles? The “probshift” column
contains the actual fitted allelic ratio while the “p” column contains
the p-value of a likelihood-ratio-test assessing significant deviance
from 0.5; both can be filtered on to retain only significantly
allelically biased loci with a big enough effect size.</p>
<p>Other columns contain additional filter criteria, like</p>
<ul>
<li>“median_AB”: a median (robust) reference allele ratio estimate in
case you want to play it extra safe in retaining significantly biased
loci, see <code><a href="../reference/median_AB.html">median_AB()</a></code>’s help page for more information</li>
<li>“quality”: will contain “!” if a locus contained no heterozygous
samples to detect AB on</li>
<li>“allele.frequency”: the reference allele frequency across genotypes;
you may not want to trust loci with extreme allele frequencies as these
probably contain very little heterozygous individuals</li>
<li>“coverage”: contains the median coverage of all samples covering the
locus)</li>
<li>“nr_samples”: number of samples covering the locus</li>
</ul>
<p>The code below first retains loci showcasing statistically
significant AB (controlling the false discovery rate at 0.05) which are
of good quality. Then, we also construct a more high-fidelity set of
loci, retaining only those of good quality with a fitted allelic ratio
greater than 0.6 or smaller than 0.4, yet not more extreme than 0.9 or
0.1 because those could be the product of poorly fitted loci. The median
ratio should also be larger than 0.6 or smaller than 0.4, allele
frequency across genotypes should not be more extreme then 0.9 or 0.1,
median coverage should be over 10 and number of samples should be over
80. These are reasonable filter criteria, though their specific values
depend on each end user’s desires and the specific dataset used
(e.g. requiring over 80 samples is perhaps too stringent for smaller
datasets and too liberal for larger ones).</p>
<p>In BOTH CASES, we <em>only retain loci that are in Hardy-Weinberg
Equilibrium (HWE)</em> according to the chi square test performed
earlier. Since we retain only those loci for which we can’t find
sufficient evidence against the null hypothesis of HWE, this is a
statistically weak conclusion, yet in scientific research a cut-off of
the p-value of this test of 0.001 is often used to filter out loci with
which something seems wrong (see e.g. <span class="citation">Sha and
Zhang (2011)</span>, <span class="citation">Teo et al. (2007)</span>,
<span class="citation">Rohlfs and Weir (2008)</span>); such extreme
deviations from HWE should not occur in a normal population.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Statistical evidence for significant AB at the 5% FDR level:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">position</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">p</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>      <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">quality</span> <span class="op">!=</span> <span class="st">"!"</span> <span class="op">&amp;</span> <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">Chi2PVAL</span> <span class="op">&gt;=</span> <span class="fl">0.001</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] “loc0, loc5, loc8, loc9, loc11, loc12, loc14, loc16, loc18,
loc20, loc24, loc25, loc26, loc27, loc28, loc29, loc30, loc31, loc34,
loc35, loc39, loc41, loc44, loc45, loc46, loc49, loc54, loc55, loc56,
loc62, loc63, loc67, loc71, loc72, loc73, loc74, loc75, loc78, loc80,
loc86, loc90, loc92, loc94, loc95, loc97, loc99, loc101, loc103, loc105,
loc111, loc114, loc119, loc120, loc121, loc123, loc124, loc126, loc131,
loc132, loc145, loc147, loc149, loc151, loc153, loc154, loc155, loc158,
loc161, loc164, loc171, loc174, loc176, loc179, loc182, loc184, loc185,
loc186, loc187, loc189, loc194, loc198, loc199, loc204, loc206, loc218,
loc220, loc221, loc225, loc232, loc236, loc238, loc240, loc241, loc242,
loc244, loc245, loc247, loc248, loc254, loc259, loc260, loc261, loc262,
loc263, loc264, loc265, loc266, loc267”</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Statistical evidence for significant AB at the 5% FDR level,</span></span>
<span><span class="co"># only retaining reliable (high-quality) loci with a large enough effect size:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">position</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">p</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">quality</span> <span class="op">!=</span> <span class="st">"!"</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">probshift</span> <span class="op">&gt;</span> <span class="fl">0.6</span> <span class="op">|</span> <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">probshift</span> <span class="op">&lt;</span> <span class="fl">0.4</span><span class="op">)</span> </span>
<span>  <span class="op">&amp;</span> <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">probshift</span> <span class="op">&lt;</span> <span class="fl">0.9</span> <span class="op">&amp;</span>  <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">probshift</span> <span class="op">&gt;</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">median_AB</span> <span class="op">&gt;</span> <span class="fl">0.6</span> <span class="op">|</span> </span>
<span>  <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">median_AB</span> <span class="op">&lt;</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">allele.frequency</span> <span class="op">&lt;</span> <span class="fl">0.9</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">allele.frequency</span> <span class="op">&gt;</span> <span class="fl">0.2</span> <span class="op">&amp;</span> <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">coverage</span> <span class="op">&gt;</span> <span class="fl">10</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">nr_samples</span> <span class="op">&gt;</span> <span class="fl">80</span> <span class="op">&amp;</span> <span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">Chi2PVAL</span> <span class="op">&gt;=</span> <span class="fl">0.001</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] “loc0, loc9, loc11, loc16, loc24, loc41, loc62, loc67, loc86,
loc92, loc94, loc111, loc119, loc124, loc131, loc149, loc155, loc179,
loc186, loc245, loc254, loc260, loc263, loc264”</p>
<p>maelstRom contains a function to plot histograms of the reference allele
fraction across samples for a specific locus, while also including PMF
plots of both the unshifted (assuming no AB) and shifted models. The
theta-parameters for the unshifted model are included in the previous
<code><a href="../reference/EMfit_betabinom_robust.html">EMfit_betabinom_robust()</a></code> output as they are calculated
anyway to test for a significant shift using a likelihood ratio test.
This plot can illustrate just how much better the shifted model fits to
the data, as illustrated below.</p>
<p>Keep in mind, however, that for a beta-binomial PMF, its shape
(variance) depends on the coverage, even when normalized from the
reference allele count to a 0-to-1 reference allele fraction, such as on
the x-axis in the plot below. As such, we have to choose a coverage to
use as basis for the PMF plots (<code>ScaleCount</code> argument). The
median coverage across samples is a good choice for this, though for
data points with outlying coverages, the resulting visualization may
still be poor (even if those data points are well fitted by the model).
This is because the fitted model allows for a flexible per-sample
coverage parameter, whereas for our PMF plot, we have to choose just
one.</p>
<p>To accomodate for this, one might opt to set the
<code>ScaleHist</code> argument to TRUE, which will re-calculate
observed per-sample reference allele fractions that are to be plotted in
the histogram using the maelstRom’s beta-binomial quantile function given the
observed per-sample data probability as input, but while using
<code>ScaleCount</code> total counts (i.e. we transform observed
reference allele fractions to the fractions we “would have observed” -
i.e. with the same cumulative distribution function value - given
<code>ScaleCount</code> observed reads, assuming the fitted
beta-binomial model is correct). It’s up to the end user to use this
argument, though for a comfortable interpretation of the plots it’s
recommended if the per-sample coverage is very variable.</p>
<p><code><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot()</a></code> directly accepts the output of the
<code><a href="../reference/BetaBinomGenotyping.html">BetaBinomGenotyping()</a></code> wrapper function to fetch
distributional parameters from, which is handy when following this
vignette. All of these parameters can also be inputted manually,
however, which is illustrated in maelstRom’s extended vignette.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PlotDF</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="st">"loc11"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">PlotAB</span> <span class="op">&lt;-</span> <span class="va">Geno_AB_res</span><span class="op">[</span><span class="va">Geno_AB_res</span><span class="op">$</span><span class="va">position</span><span class="op">==</span><span class="st">"loc11"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">loc_plot</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>DataList_out <span class="op">=</span> <span class="va">PlotDF</span>, Geno_AB_res <span class="op">=</span> <span class="va">PlotAB</span>, SE<span class="op">=</span><span class="va">SEmedian</span>, </span>
<span>  ScaleCount <span class="op">=</span> <span class="va">PlotAB</span><span class="op">$</span><span class="va">coverage</span>, ScaleHist <span class="op">=</span> <span class="cn">TRUE</span>, plot_NoShift <span class="op">=</span> <span class="cn">TRUE</span>, nbins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">loc_plot</span></span></code></pre></div>
<p><img src="maelstRom_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>The plotting function contains some customization options, the plot
above mainly using default colors with the PMF of the shifted fit split
up according to genotype. One can e.g. choose to plot one combined
unshifted PMF, and to not plot the unshifted fit, via the code below
(once again using default colors, though these can all be changed):</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loc_plot</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>DataList_out <span class="op">=</span> <span class="va">PlotDF</span>, Geno_AB_res <span class="op">=</span> <span class="va">PlotAB</span>, SE<span class="op">=</span><span class="va">SEmedian</span>, </span>
<span>  ScaleCount <span class="op">=</span> <span class="va">PlotAB</span><span class="op">$</span><span class="va">coverage</span>, ScaleHist <span class="op">=</span> <span class="cn">TRUE</span>, plot_NoShift <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  SplitPeaks <span class="op">=</span> <span class="cn">FALSE</span>, nbins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">loc_plot</span></span></code></pre></div>
<p><img src="maelstRom_tutorial_files/figure-html/unnamed-chunk-16-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="differential-allelic-divergence-detection">Differential allelic divergence detection<a class="anchor" aria-label="anchor" href="#differential-allelic-divergence-detection"></a>
</h2>
<p>In the context of maelstRom, we see case-specific events such copy number
alterations, aberrant hypermethylation, promotor mutation and silencing
events of e.g. tumor suppressor genes as mechanisms that will affect one
allele, or both alleles in a different way, per sample. However, these
mechanisms show no preference towards either allele on a
population-level, acting mainly as an added source of variance on the
observed reference allele fraction captured by <span class="math inline">\(\theta_{het}\)</span> while leaving <span class="math inline">\(\pi_{het}\)</span> unchanged. Calling this
variance <em>Allelic Divergence (AD)</em>, we term its change (increase)
in disease due to, amongst others, the mechanisms listed above
<em>differential Allelic Divergence (dAD)</em>.</p>
<p>Detecting dAD between two populations can be done via a likelihood
ratio test, comparing: (1) a model fitting ALL data, both control and
case, as one beta-binomial mixture model, with (2) a model allowing the
control- and case data to have their separate <span class="math inline">\(\theta\)</span> parameters for the heterozygous
samples, yet keeping the <span class="math inline">\(\pi\)</span>
parameter constant. The pivotal new function is
<code><a href="../reference/EMfit_betabinom_popcomb.html">EMfit_betabinom_popcomb()</a></code> which fits the model assuming
different heterozygote-thetas while keeping all other parameters shared.
It is itself included in the <code><a href="../reference/dAD_analysis.html">dAD_analysis()</a></code> wrapper
function, which also included <code><a href="../reference/EMfit_betabinom_robust.html">EMfit_betabinom_robust()</a></code>,
<code><a href="../reference/EMfit_betabinom.html">EMfit_betabinom()</a></code>, <code><a href="../reference/pmf_betabinomMix.html">pmf_betabinomMix()</a></code>, and
<code><a href="../reference/HWE_chisquared.html">HWE_chisquared()</a></code>. The dAD-detecting pipeline is relatively
complex in the way the various fits and outlier detection are handled,
so checking out <code><a href="../reference/dAD_analysis.html">dAD_analysis()</a></code>’s source code or the
extended vignette is certainly recommended for additional insight. Here,
however, <code><a href="../reference/dAD_analysis.html">dAD_analysis()</a></code> conveniently returns a dataframe
containing per-locus (d)AD results: <code>dAD_res</code>.</p>
<p><code><a href="../reference/dAD_analysis.html">dAD_analysis()</a></code> requires control- and case-input lists to
be combined into one (<code>ParTOT</code> below), which is only done now
so as to contain the updated <code>ParCTRL</code> list as outputted by
<code><a href="../reference/BetaBinomGenotyping.html">BetaBinomGenotyping()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only make ParTOT now so it contains the updated ParCTRL</span></span>
<span><span class="va">ParTOT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length<span class="op">=</span><span class="va">NC</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span><span class="op">)</span><span class="op">{</span> <span class="co"># Put the splitted input data into a list for parallellisation</span></span>
<span>  <span class="va">ParTOT</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ParCTRL</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">ParCASE</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dADFinData</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParTOT</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va"><a href="../reference/dAD_analysis.html">dAD_analysis</a></span>, SE<span class="op">=</span><span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dAD_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="va">dADFinData</span><span class="op">)</span></span></code></pre></div>
<p>In <code>dAD_res</code>, both <code>theta</code>- and
<code>rho</code>-values are listed, which both represent the
overdispersion parameter; <span class="math inline">\(\theta\)</span>
just has a range from 0 to infinity while <span class="math inline">\(\rho\)</span> has been rescaled to go from 0 to 1,
which is more convenient for some applications, such as plotting. When
studying differential allelic divergence, one usually tries to look for
“disturbed regions” by plotting the results in <code>dAD_res</code>
among the length of chromosomes (e.g. plotting case rho-values along the
chromosomes while also marking the actually statistically relevant loci
in some way), as will be demonstrated at the end of this section.</p>
<p>Note that both parameters are perfectly equivalent, and that the
exact parametrization does not impact the statistical testing procedure
(as both will yield exactly the same likelihoods).</p>
<p>That being said, you can also just extract loci showing differential
allelic divergence from <code>dAD_res</code> using the
<code>LRTpval</code> column to retain statistically significant results.
Defining an effect size is a bit trickier; you’d expect to be able to do
it on the overdispersion parameter values of cases compared to controls,
but how would one go about it? When both are very small (e.g. rho ~
10^-7 range), one being 10 times as large as the other doesn’t mean that
much, but when both are quite large (e.g. rho ~ 10^-2 range) this does
make a very large difference. Subtracting one from the other has an
equally unclear interpretation regarding effect size. The effect of the
overdispersion parameter on the actual variance also depends on the pi
parameter, which further complicates things. Moreover, in many cases,
e.g. cancer, case samples will be mixtures of normal and diseased cells,
with varying (gene-dependent) levels of expression in each cell type,
further complicating the creation of appropriate effect sizes.</p>
<p>This type of difficulties made us opt for detecting affected regions
on plots rather than relying on the effect sizes as such. But if you
wanted to e.g. only retain loci that show a strong deviation from a
regular binomial regarding its variance, using a rho-cutoff of, say,
0.1, can do. Such cutoffs are based on experience and can be played
around with. In the code below, we retain such good-quality loci which
are also statistically significant regarding dAD and have a median
coverage of at least 10 in both controls and cases, as well as at least
15 heterozygous samples and data conform Hardy-Weinberg Equilibrium in
controls only (reliable case data genotyping may be difficult when faced
with extreme differential allelic divergence). Also, note that for
genuine differential allelic divergence, overdispersion will be
typically be larger in cases than in controls.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">LocName</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">RhoHetCASE</span> <span class="op">&gt;=</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">LRTpval</span>, </span>
<span>  method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">CovCTRL_med</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">CovCASE_med</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">dAD_res</span><span class="op">$</span><span class="va">NumHetCTRL</span> <span class="op">&gt;=</span> <span class="fl">15</span> <span class="op">&amp;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">HWECTRL</span> <span class="op">&gt;=</span> <span class="fl">0.001</span> <span class="op">&amp;</span> </span>
<span>  <span class="va">dAD_res</span><span class="op">$</span><span class="va">RhoHetCASE</span> <span class="op">&gt;</span> <span class="va">dAD_res</span><span class="op">$</span><span class="va">RhoHetCTRL</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] “loc0, loc6, loc7, loc68, loc80, loc84, loc85, loc86, loc88,
loc89, loc91, loc92, loc93, loc95, loc96, loc97, loc98, loc100, loc158,
loc211, loc212, loc218, loc223, loc258”</p>
<p>The previously used <code><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot()</a></code> function can also
be used to vizualize the dAD detection results via separate plots of the
control- and case-population utilizing their separately fitted
overdispersion parameters. Here, it’s important to set the input
argument <code>ScaleCount</code> to the same value for both plots, so
the width of the PMFs can actually be compared correctly. When following
this vignette, it once again accepts the (d)AD-results dataframe of
<code><a href="../reference/dAD_analysis.html">dAD_analysis()</a></code> as an input to fetch distributional
parameters from (manual entry of these parameters is demonstrated in the
extended vignette). Be aware setting the <code>plot_NoShift</code>
argument to TRUE is not an option here, as the dAD-analysis results do
not include AB-related results such as distributional parameters of the
unshifted fit, and the extra <code>PlotWhich</code> argument (“control”
or “case”) is required for the function to know which parameters to
fetch from <code>dAD_res</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CTRL_DF</span> <span class="op">&lt;-</span> <span class="va">controlList</span><span class="op">[[</span><span class="st">"loc85"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">CASE_DF</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="st">"loc85"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">PlotData_dAD</span> <span class="op">&lt;-</span> <span class="va">dAD_res</span><span class="op">[</span><span class="va">dAD_res</span><span class="op">$</span><span class="va">LocName</span><span class="op">==</span><span class="st">"loc85"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">dAD_plot1</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>DataList_out <span class="op">=</span> <span class="va">CTRL_DF</span>, dAD_res <span class="op">=</span> <span class="va">PlotData_dAD</span>, </span>
<span>  SE <span class="op">=</span> <span class="va">SEmedian</span>, ScaleCount <span class="op">=</span> <span class="fl">50</span>, ScaleHist <span class="op">=</span> <span class="cn">TRUE</span>, nbins <span class="op">=</span> <span class="fl">30</span>, plot_NoShift <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  PlotWhich <span class="op">=</span> <span class="st">"control"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dAD_plot2</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_EMfitplot.html">maelstRom_EMfitplot</a></span><span class="op">(</span>DataList_out <span class="op">=</span> <span class="va">CTRL_DF</span>, dAD_res <span class="op">=</span> <span class="va">PlotData_dAD</span>, </span>
<span>  SE <span class="op">=</span> <span class="va">SEmedian</span>, ScaleCount <span class="op">=</span> <span class="fl">50</span>, ScaleHist <span class="op">=</span> <span class="cn">TRUE</span>, nbins <span class="op">=</span> <span class="fl">30</span>, plot_NoShift <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  PlotWhich <span class="op">=</span> <span class="st">"case"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">dAD_plot1</span>, <span class="va">dAD_plot2</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="maelstRom_tutorial_files/figure-html/unnamed-chunk-19-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>The plot above is of a locus showcasing very clear differential
allelic divergence, with the reference allele fraction in heterozygous
case samples being way more variable, probably due to case-specific
effects.</p>
<p>As promised, to conclude, this section showcases a plot of dAD
results across a chromosome using the <code><a href="../reference/maelstRom_ADChromplot.html">maelstRom_ADChromplot()</a></code>
function. The toy data isn’t fit for this, however (no chromosomal
location, too small), so included in the maelstRom package is some thoroughly
processed data to illustrate the plot on. The inputs needed are <span class="math inline">\(\rho\)</span> values in both controls and cases
along with differential expression data (log2 fold changes) which can be
easily generated by tools for differential expression analysis on the
data at hand. <code><a href="../reference/maelstRom_ADChromplot.html">maelstRom_ADChromplot()</a></code> optionally even accepts
additional hypermethylation and copy-number-alteration data for even
more extensive plots. As this is for illustrative purposes only, we just
make the plot using the provided datasets below, but more details can be
learned from <code><a href="../reference/maelstRom_ADChromplot.html">maelstRom_ADChromplot()</a></code>’s help page. The function
returns several plot components (including legends) as separate objects
due to its complexity, so it’s up to the end user to arrange them into a
satisfactory composition.</p>
<p>As a final remark, many of the measures in this plot (differential
expression, hypermethylation, copy number alterations) are usually
obtained at the gene-level rather than the SNP/locus level. To make the
results of different analyses more compareable, maelstRom’s per-locus
p-values can be combined into per-gene values using the
<code><a href="../reference/combine_p_gene.html">combine_p_gene()</a></code> function. For this, we recommend using the
geometric mean as a nice balance between more conservative (arithmetic
mean, maximum) and liberal (harmonic mean, minimum) methods for
combining p-values, and the different loci should be weighted as well,
for which we recommend <span class="math inline">\(\small \sqrt{median \
locus \ coverage \ * \ estimated \ number \ of \ heterozygous \ samples
}\)</span>, the latter being an output of maelstRom’s beta-binomial mixture
model fits via the heterozygote genotype frequency times the number of
samples. For p-values concerning a comparison between a control- and
case-population, one can e.g. use the minimum of this weight across
populations.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ChromPlot</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_ADChromplot.html">maelstRom_ADChromplot</a></span><span class="op">(</span><span class="va">AD_Data</span>, <span class="va">DE_Data</span>, <span class="va">Meth_Data</span>, <span class="va">CNAgain_Data</span>, <span class="va">CNAloss_Data</span>,</span>
<span>                                    pvalSIG <span class="op">=</span> <span class="fl">0.05</span>, roll_median <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ChromPlot</span><span class="op">[[</span><span class="st">"ADDE_plot"</span><span class="op">]</span><span class="op">]</span> <span class="op">/</span> <span class="va">ChromPlot</span><span class="op">[[</span><span class="st">"MethCNA_plot"</span><span class="op">]</span><span class="op">]</span> <span class="op">/</span> <span class="op">(</span><span class="va">ChromPlot</span><span class="op">[[</span><span class="st">"LEG1"</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> </span>
<span>  <span class="va">ChromPlot</span><span class="op">[[</span><span class="st">"LEG2"</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">ChromPlot</span><span class="op">[[</span><span class="st">"LEG3"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">0.5</span><span class="op">)</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="maelstRom_tutorial_files/figure-html/unnamed-chunk-20-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>The histograms in the top panes show (rolling medians) of the
overdisperstion (<span class="math inline">\(\rho\)</span>) parameters
in controls (black) and cases (colored). The case-<span class="math inline">\(\rho\)</span> is colored according to the (already
5% FDR corrected) p-value testing for significant dAD in cases versus
controls, while the difference in height of these histograms provide an
illustration of the effect size (actual <span class="math inline">\(\rho\)</span> values). The remaining plot elements
are explained in the legend.</p>
</div>
<div class="section level2">
<h2 id="detecting-imprinted-and-differentially-imprinted-loci">Detecting imprinted and differentially imprinted loci<a class="anchor" aria-label="anchor" href="#detecting-imprinted-and-differentially-imprinted-loci"></a>
</h2>
<p>Imprinting analyses in maelstRom are quite different from the analyses up
until now in that they don’t rely on beta-binomial modelling and make
different assumptions about the data. They are, generally speaking, a
bit more robust because imprinting is usually an extreme phenomenon,
i.e. heterozygous samples of imprinted loci will almost always “seem
like” homozygous samples because one of the alleles has been silenced
almost entirely. Partial imprinting can exist and maelstRom can detect it,
but this phenomenon is not modelled as thoroughly as the previous ones
in this vignette.</p>
<p>First off (besides setting a minimal required minor allele frequence
of 0.15 as detecting imprinting can otherwise be unreliable) we assume
samples show no AB (as we’ll be using unshifted models) and are conform
Hardy-Weinberg-Equilibrium while taking imprinting into account, the
latter of which is tested by <code><a href="../reference/symmetry_gof.html">symmetry_gof()</a></code>. It does this
by assessing whether the number of samples with a reference allele
fraction &gt; 0.5 (and so also the number with the fraction &lt;= 0.5)
is equal to that expected under HWE assuming no AB using a chi square
test. This may seem kind of heuristic at first and prone to fail if
there is even a little AB, but remember we want to mainly retain
imprinted loci, meaning heterozygous samples should behave as one of
both homozygotes (and so clearly feature reference allele fractions &gt;
0.5 or &lt;=0.5) so this filter is very fitting for retaining
HWE-conform samples that are at the same time likely to showcase
imprinting. A good filter setting is requiring the p-value to be at
least 0.05.</p>
<p>After applying these filters, <code><a href="../reference/imprinting_est.html">imprinting_est()</a></code> detects
imprinted loci in the remaining control samples by first assuming an
unshifted binomial mixture model, then splitting up the heterozygous
samples in two separate groups according to a degree of imprinting <span class="math inline">\(i\)</span> and varying this <span class="math inline">\(i\)</span> from 0 to 1 in a for-loop, calculating
likelihoods along the way to perform a likelihood-ratio test to check
for differential imprinting (i.e. compare the most likely model with the
non-imprinted one; <span class="math inline">\(i=0\)</span>). For
completeness, the imprinted PMF is (see <span class="citation">Goovaerts
et al. (2018)</span> for a full elaboration):</p>
<p><span class="math display">\[
\small
\begin{aligned}
\mathrm{PMF}(&amp;counts_{ref}|counts_{total}) =  P_{rr} * {\tt
pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=1-SE) \ + \\
&amp; 0.5P_{rv} * {\tt pbinom} \left( x=counts_{ref} \ |\
n=counts_{total},
p=\frac{0.5-\frac{i}{2}}{1-\frac{i}{2}}(1-SE)+\frac{0.5}{1-\frac{i}{2}}SE\right)
\ + \\
&amp; 0.5P_{rv} * {\tt pbinom} \left( x=counts_{ref} \ |\
n=counts_{total},
p=\frac{0.5-\frac{i}{2}}{1-\frac{i}{2}}SE+\frac{0.5}{1-\frac{i}{2}}(1-SE)\right)
\ + \\
&amp; P_{vv} * {\tt pbinom}(x=counts_{ref} \ |\ n=counts_{total}, p=SE)
\end{aligned}
\]</span> Note that this imprinting model consisting of regular binomial
functions also lead us to using preliminary minor allele frequency
estimates as produced by <code><a href="../reference/AllelicMeta_est_par.html">AllelicMeta_est_par()</a></code> towards the
start of the vignette for the aforementioned 0.15 minor allele frequency
filtering, as it assumes a similar (non-beta-binomial) mixture model.
This becomes clear in maelstRom’s extended vignette, but is here hidden under
the <code><a href="../reference/impr_LOI_analysis.html">impr_LOI_analysis()</a></code> wrapper.</p>
<p>After retaining only significantly and sufficiently imprinted loci
(5% FDR level, degree of imprinting at least 0.6 and median degree of
imprinting at least 0.8 for robustness; see
<code><a href="../reference/median_imprinting.html">median_imprinting()</a></code> for the latter’s calculation) using the
<code><a href="../reference/final_filter.html">final_filter()</a></code> function, we use
<code><a href="../reference/LOItest_logreg.html">LOItest_logreg()</a></code> to test for differential imprinting. More
specifically, logistic regression is used to evaluate whether case
samples feature a significantly higher number of apparently heterozygous
samples than control samples (by considering the least expressed allele
as a success; more details can be found on the function’s help page.
Note that the logistic regression approach takes into account variable
coverage between samples, contrasting our previously published strategy
by <span class="citation">Goovaerts et al. (2018)</span>). If this is
true, it indicates that a significant fraction of heterozygous samples
in tumor tissue “lost their imprinting” and actually started expressing
both alleles again. We call these loci “differentially imprinted”, as we
reserve the term “loss-of-imprinting” for the phenomenon where
differential imprinting co-occurs with increased total expression,
indicative of re-expression of the silenced allele.</p>
<p>All of these steps are combined into the
<code><a href="../reference/impr_LOI_analysis.html">impr_LOI_analysis()</a></code> wrapper function, calling
<code><a href="../reference/symmetry_gof.html">symmetry_gof()</a></code>, <code><a href="../reference/imprinting_est.html">imprinting_est()</a></code>,
<code><a href="../reference/median_imprinting.html">median_imprinting()</a></code>, <code><a href="../reference/final_filter.html">final_filter()</a></code>, and
<code><a href="../reference/LOItest_logreg.html">LOItest_logreg()</a></code>. Extra insight can be learned from
<code><a href="../reference/impr_LOI_analysis.html">impr_LOI_analysis()</a></code>’s source code or maelstRom’s extended
vignette, which includes a lot of clarifying comments. In the end,
<code><a href="../reference/impr_LOI_analysis.html">impr_LOI_analysis()</a></code> returns imprinting-detection results
per locus (<code>impr_res</code>) and, for those loci that are
significantly and sufficiently imprinted, LOI-detection results as well
(<code>LOI_res</code>)</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"cl.cores"</span>, <span class="va">NC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ImprFinData</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">parLapply</a></span><span class="op">(</span><span class="va">cl</span>, X <span class="op">=</span> <span class="va">ParTOT</span>, fun <span class="op">=</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="va"><a href="../reference/impr_LOI_analysis.html">impr_LOI_analysis</a></span>, SE <span class="op">=</span> <span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span>,</span>
<span>  MinMinorAllelefreq<span class="op">=</span><span class="fl">0.15</span>, sym_filt<span class="op">=</span><span class="fl">0.05</span>, adj_p_filt<span class="op">=</span><span class="fl">0.05</span>, gof_filt<span class="op">=</span><span class="fl">0.8</span>, med_i_filt<span class="op">=</span><span class="fl">0.8</span>, i_filt<span class="op">=</span><span class="fl">0.6</span><span class="op">)</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">impr_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ImprFinData</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">LOI_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ImprFinData</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can take a look at the loci that are significantly and
sufficiently imprinted (i.e. those retained in
<code>LOI_res</code>):</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LOI_res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="2%">
<col width="6%">
<col width="7%">
<col width="6%">
<col width="8%">
<col width="11%">
<col width="6%">
<col width="5%">
<col width="5%">
<col width="7%">
<col width="6%">
<col width="6%">
<col width="6%">
<col width="6%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">position</th>
<th align="right">LRT</th>
<th align="right">p</th>
<th align="right">estimated.i</th>
<th align="right">allele.frequency</th>
<th align="left">reference</th>
<th align="left">variant</th>
<th align="right">med_cov</th>
<th align="right">nr_samples</th>
<th align="right">GOF</th>
<th align="right">symmetry</th>
<th align="right">med_impr</th>
<th align="right">adj_p</th>
<th align="right">DI_pval</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">28</td>
<td align="left">loc42</td>
<td align="right">31.91760</td>
<td align="right">0.00e+00</td>
<td align="right">0.81</td>
<td align="right">0.7653805</td>
<td align="left">A</td>
<td align="left">G</td>
<td align="right">8.0</td>
<td align="right">128</td>
<td align="right">1.056476</td>
<td align="right">0.9050920</td>
<td align="right">1.0000000</td>
<td align="right">0.00e+00</td>
<td align="right">0.5631353</td>
</tr>
<tr class="even">
<td align="left">62</td>
<td align="left">loc81</td>
<td align="right">42.12370</td>
<td align="right">0.00e+00</td>
<td align="right">0.75</td>
<td align="right">0.8091673</td>
<td align="left">T</td>
<td align="left">A</td>
<td align="right">10.0</td>
<td align="right">123</td>
<td align="right">1.205116</td>
<td align="right">1.0000000</td>
<td align="right">1.0000000</td>
<td align="right">0.00e+00</td>
<td align="right">0.1139792</td>
</tr>
<tr class="odd">
<td align="left">106</td>
<td align="left">loc138</td>
<td align="right">375.46626</td>
<td align="right">0.00e+00</td>
<td align="right">0.94</td>
<td align="right">0.6854238</td>
<td align="left">G</td>
<td align="left">A</td>
<td align="right">62.5</td>
<td align="right">128</td>
<td align="right">2.021628</td>
<td align="right">0.8636647</td>
<td align="right">0.9639640</td>
<td align="right">0.00e+00</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">132</td>
<td align="left">loc166</td>
<td align="right">16.99503</td>
<td align="right">1.87e-05</td>
<td align="right">1.00</td>
<td align="right">0.7826087</td>
<td align="left">T</td>
<td align="left">A</td>
<td align="right">4.0</td>
<td align="right">23</td>
<td align="right">1.047838</td>
<td align="right">1.0000000</td>
<td align="right">1.0000000</td>
<td align="right">2.63e-05</td>
<td align="right">0.0012089</td>
</tr>
<tr class="odd">
<td align="left">133</td>
<td align="left">loc167</td>
<td align="right">1789.87048</td>
<td align="right">0.00e+00</td>
<td align="right">0.94</td>
<td align="right">0.5156250</td>
<td align="left">T</td>
<td align="left">C</td>
<td align="right">1361.5</td>
<td align="right">256</td>
<td align="right">2.921619</td>
<td align="right">1.0000000</td>
<td align="right">0.9883721</td>
<td align="right">0.00e+00</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">134</td>
<td align="left">loc168</td>
<td align="right">88.14848</td>
<td align="right">0.00e+00</td>
<td align="right">0.97</td>
<td align="right">0.7283403</td>
<td align="left">C</td>
<td align="left">G</td>
<td align="right">14.0</td>
<td align="right">127</td>
<td align="right">1.637326</td>
<td align="right">1.0000000</td>
<td align="right">1.0000000</td>
<td align="right">0.00e+00</td>
<td align="right">0.3780830</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">LOI_res</span><span class="op">$</span><span class="va">position</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] “loc42, loc81, loc138, loc166, loc167, loc168, loc169, loc202,
loc215”</p>
<p>From this set, a part is actually differentially imprinted in case
samples, which is indicated by the p-value in the <code>DI_pval</code>
column in the <code>LOI_res</code> dataframe (the <code>adj_p</code>
p-value detects imprinting in controls, FDR-corrected; these should all
be significant in <code>LOI_res</code>, though not necessarily in
<code>impr_res</code>). We can control differential imprinting
detection, for a change, at the 5% FWER level (actually feasible here
due to the small number of actually imprinted loci remaining, though
we’re looking at a small toy dataset here; this isn’t necessarily the
case in larger experiments):</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">LOI_res</span><span class="op">$</span><span class="va">position</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">LOI_res</span><span class="op">$</span><span class="va">DI_pval</span>, method <span class="op">=</span> <span class="st">"holm"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span>, </span>
<span>            collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>[1] “loc138, loc166, loc167, loc202, loc215”</p>
<p>We can plot histograms of both the control and case allele fractions;
locus 202 is obviously imprinted in controls and loses a lot of this
imprinting in cases (controls are plotted using
<code><a href="../reference/maelstRom_imprintplot.html">maelstRom_imprintplot()</a></code> to visualize imprinting results, but
since an imprinting-estimation never happens on the case data - we just
detect a change in heterozygosity via logistic regression during
LOI-detection - we simply make a histogram for cases instead).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HistCTRL</span> <span class="op">&lt;-</span> <span class="fu">maelstRom</span><span class="fu">::</span><span class="fu"><a href="../reference/maelstRom_imprintplot.html">maelstRom_imprintplot</a></span><span class="op">(</span><span class="va">controlList</span><span class="op">[[</span><span class="st">"loc202"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span>, </span>
<span>  <span class="va">controlList</span><span class="op">[[</span><span class="st">"loc202"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span>, </span>
<span>  allelefreq <span class="op">=</span> <span class="va">impr_res</span><span class="op">$</span><span class="va">allele.frequency</span><span class="op">[</span><span class="va">impr_res</span><span class="op">$</span><span class="va">position</span><span class="op">==</span><span class="st">"loc202"</span><span class="op">]</span>,</span>
<span>  impr <span class="op">=</span> <span class="va">impr_res</span><span class="op">$</span><span class="va">estimated.i</span><span class="op">[</span><span class="va">impr_res</span><span class="op">$</span><span class="va">position</span><span class="op">==</span><span class="st">"loc202"</span><span class="op">]</span>,</span>
<span>  SE <span class="op">=</span> <span class="va">SEmedian</span>, inbr <span class="op">=</span> <span class="va">Fmedian</span>, plot_NoImpr <span class="op">=</span> <span class="cn">TRUE</span>, SplitPeaks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Control"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">RatioCASE</span> <span class="op">&lt;-</span> <span class="va">caseList</span><span class="op">[[</span><span class="st">"loc202"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span> <span class="op">/</span> </span>
<span>  <span class="op">(</span><span class="va">caseList</span><span class="op">[[</span><span class="st">"loc202"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ref_count</span> <span class="op">+</span> <span class="va">caseList</span><span class="op">[[</span><span class="st">"loc202"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">var_count</span><span class="op">)</span></span>
<span><span class="va">CASEDat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"Ratio"</span> <span class="op">=</span> <span class="va">RatioCASE</span><span class="op">)</span></span>
<span><span class="va">HistCASE</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">CASEDat</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">Ratio</span><span class="op">)</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Reference allele fraction"</span>, y<span class="op">=</span><span class="st">"Frequency"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Case"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">HistCTRL</span>, <span class="va">HistCASE</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="maelstRom_tutorial_files/figure-html/unnamed-chunk-25-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 10 x64 (build 18363)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Dutch_Belgium.1252 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Dutch_Belgium.1252   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Dutch_Belgium.1252</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                  </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Dutch_Belgium.1252    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats    </span></span>
<span><span class="co">#&gt; [2] graphics </span></span>
<span><span class="co">#&gt; [3] grDevices</span></span>
<span><span class="co">#&gt; [4] utils    </span></span>
<span><span class="co">#&gt; [5] datasets </span></span>
<span><span class="co">#&gt; [6] methods  </span></span>
<span><span class="co">#&gt; [7] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Rcpp_1.0.8         </span></span>
<span><span class="co">#&gt;  [2] ggnewscale_0.4.7   </span></span>
<span><span class="co">#&gt;  [3] lattice_0.20-41    </span></span>
<span><span class="co">#&gt;  [4] zoo_1.8-8          </span></span>
<span><span class="co">#&gt;  [5] gtools_3.8.2       </span></span>
<span><span class="co">#&gt;  [6] assertthat_0.2.1   </span></span>
<span><span class="co">#&gt;  [7] rprojroot_2.0.2    </span></span>
<span><span class="co">#&gt;  [8] digest_0.6.25      </span></span>
<span><span class="co">#&gt;  [9] utf8_1.1.4         </span></span>
<span><span class="co">#&gt; [10] gmp_0.6-2          </span></span>
<span><span class="co">#&gt; [11] R6_2.5.0           </span></span>
<span><span class="co">#&gt; [12] alabama_2015.3-1   </span></span>
<span><span class="co">#&gt; [13] evaluate_0.14      </span></span>
<span><span class="co">#&gt; [14] ggplot2_3.3.3      </span></span>
<span><span class="co">#&gt; [15] highr_0.8          </span></span>
<span><span class="co">#&gt; [16] pillar_1.7.0       </span></span>
<span><span class="co">#&gt; [17] Rdpack_2.1.1       </span></span>
<span><span class="co">#&gt; [18] rlang_1.0.4        </span></span>
<span><span class="co">#&gt; [19] rstudioapi_0.13    </span></span>
<span><span class="co">#&gt; [20] data.table_1.13.0  </span></span>
<span><span class="co">#&gt; [21] jquerylib_0.1.3    </span></span>
<span><span class="co">#&gt; [22] hash_2.2.6.1       </span></span>
<span><span class="co">#&gt; [23] rmarkdown_2.15     </span></span>
<span><span class="co">#&gt; [24] pkgdown_2.0.6      </span></span>
<span><span class="co">#&gt; [25] textshaping_0.3.6  </span></span>
<span><span class="co">#&gt; [26] desc_1.2.0         </span></span>
<span><span class="co">#&gt; [27] labeling_0.4.2     </span></span>
<span><span class="co">#&gt; [28] stringr_1.4.0      </span></span>
<span><span class="co">#&gt; [29] munsell_0.5.0      </span></span>
<span><span class="co">#&gt; [30] compiler_4.0.2     </span></span>
<span><span class="co">#&gt; [31] numDeriv_2016.8-1.1</span></span>
<span><span class="co">#&gt; [32] xfun_0.32          </span></span>
<span><span class="co">#&gt; [33] pkgconfig_2.0.3    </span></span>
<span><span class="co">#&gt; [34] systemfonts_1.0.4  </span></span>
<span><span class="co">#&gt; [35] htmltools_0.5.3    </span></span>
<span><span class="co">#&gt; [36] tidyselect_1.1.0   </span></span>
<span><span class="co">#&gt; [37] gridExtra_2.3      </span></span>
<span><span class="co">#&gt; [38] tibble_3.0.3       </span></span>
<span><span class="co">#&gt; [39] fansi_0.4.2        </span></span>
<span><span class="co">#&gt; [40] maelstRom_1.0.0         </span></span>
<span><span class="co">#&gt; [41] crayon_1.4.1       </span></span>
<span><span class="co">#&gt; [42] dplyr_1.0.4        </span></span>
<span><span class="co">#&gt; [43] MASS_7.3-53        </span></span>
<span><span class="co">#&gt; [44] rbibutils_2.0      </span></span>
<span><span class="co">#&gt; [45] grid_4.0.2         </span></span>
<span><span class="co">#&gt; [46] jsonlite_1.7.2     </span></span>
<span><span class="co">#&gt; [47] gtable_0.3.0       </span></span>
<span><span class="co">#&gt; [48] lifecycle_1.0.0    </span></span>
<span><span class="co">#&gt; [49] DBI_1.1.1          </span></span>
<span><span class="co">#&gt; [50] magrittr_2.0.1     </span></span>
<span><span class="co">#&gt; [51] scales_1.1.1       </span></span>
<span><span class="co">#&gt; [52] cli_2.3.0          </span></span>
<span><span class="co">#&gt; [53] stringi_1.5.3      </span></span>
<span><span class="co">#&gt; [54] cachem_1.0.4       </span></span>
<span><span class="co">#&gt; [55] farver_2.0.3       </span></span>
<span><span class="co">#&gt; [56] fs_1.5.0           </span></span>
<span><span class="co">#&gt; [57] bslib_0.4.0        </span></span>
<span><span class="co">#&gt; [58] ellipsis_0.3.2     </span></span>
<span><span class="co">#&gt; [59] ragg_1.2.2         </span></span>
<span><span class="co">#&gt; [60] generics_0.1.0     </span></span>
<span><span class="co">#&gt; [61] vctrs_0.3.8        </span></span>
<span><span class="co">#&gt; [62] tools_4.0.2        </span></span>
<span><span class="co">#&gt; [63] glue_1.4.2         </span></span>
<span><span class="co">#&gt; [64] purrr_0.3.4        </span></span>
<span><span class="co">#&gt; [65] parallel_4.0.2     </span></span>
<span><span class="co">#&gt; [66] fastmap_1.1.0      </span></span>
<span><span class="co">#&gt; [67] yaml_2.2.1         </span></span>
<span><span class="co">#&gt; [68] colorspace_1.4-1   </span></span>
<span><span class="co">#&gt; [69] memoise_2.0.0      </span></span>
<span><span class="co">#&gt; [70] knitr_1.31         </span></span>
<span><span class="co">#&gt; [71] patchwork_1.1.1    </span></span>
<span><span class="co">#&gt; [72] sass_0.4.2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-goovaerts2018" class="csl-entry">
Goovaerts, Tine, Sandra Steyaert, Chari A Vandenbussche, Jeroen Galle,
Olivier Thas, Wim Van Criekinge, and Tim De Meyer. 2018. <span>“A
Comprehensive Overview of Genomic Imprinting in Breast and Its
Deregulation in Cancer.”</span> <em>Nature Communications</em> 9 (1):
1–14.
</div>
<div id="ref-rohlfs" class="csl-entry">
Rohlfs, Rori V, and Bruce S Weir. 2008. <span>“Distributions of
Hardy–Weinberg Equilibrium Test Statistics.”</span> <em>Genetics</em>
180 (3): 1609–16.
</div>
<div id="ref-sha" class="csl-entry">
Sha, Qiuying, and Shuanglin Zhang. 2011. <span>“A Test of Hardy-Weinberg
Equilibrium in Structured Populations.”</span> <em>Genetic
Epidemiology</em> 35 (7): 671–78.
</div>
<div id="ref-teo" class="csl-entry">
Teo, Yik Y, Andrew E Fry, Taane G Clark, ES Tai, and Mark Seielstad.
2007. <span>“On the Usage of HWE for Identifying Genotyping
Errors.”</span> <em>Annals of Human Genetics</em> 71 (5): 701–3.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
