<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Detects relationships between beta-binomial parameters (pi, theta) and independent variables — AdvancedFitter • maelstRom</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Detects relationships between beta-binomial parameters (pi, theta) and independent variables — AdvancedFitter"><meta property="og:description" content='AdvancedFitter attempts to detect relationships between beta-binomial parameters (pi, theta) and any independent variable(s) suspected to play a role,
e.g. increased theta (overdispersion) in function of age.
It is currently, however, more of an exploratory function, as it does so by fitting these relationships assuming the supplied "WeightCol"
reflects the "true" probability that a certain observation belongs to the genotype under study (heterozygotes) as this is kept invariant;
there is of course no such thing (in reality, an observation has a one true underlying genotype, not "a probability) and, ideally,
the parameter-variable(s) relationships here studied are taken into account during the complete beta-binomial mixture model EM fit.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detects relationships between beta-binomial parameters (pi, theta) and independent variables</h1>
    
    <div class="hidden name"><code>AdvancedFitter.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>AdvancedFitter</code> attempts to detect relationships between beta-binomial parameters (pi, theta) and any independent variable(s) suspected to play a role,
e.g. increased theta (overdispersion) in function of age.
It is currently, however, more of an exploratory function, as it does so by fitting these relationships assuming the supplied "WeightCol"
reflects the "true" probability that a certain observation belongs to the genotype under study (heterozygotes) as this is kept invariant;
there is of course no such thing (in reality, an observation has a one true underlying genotype, not "a probability) and, ideally,
the parameter-variable(s) relationships here studied are taken into account during the complete beta-binomial mixture model EM fit.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AdvancedFitter</span><span class="op">(</span></span>
<span>  ThetaForm <span class="op">=</span> <span class="va">Theta</span> <span class="op">~</span> <span class="fl">1</span>,</span>
<span>  PiForm <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">SNPdata</span>,</span>
<span>  RefCol <span class="op">=</span> <span class="st">"ref_count"</span>,</span>
<span>  VarCol <span class="op">=</span> <span class="st">"var_count"</span>,</span>
<span>  WeightCol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">Pi_start</span>,</span>
<span>  <span class="va">Theta_start</span>,</span>
<span>  PiLink <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>  ThetaLink <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ResetThetaMin <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">10</span>,</span>
<span>  ResetThetaMax <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>ThetaForm</dt>
<dd><p>An object describing the (linear) relationship between theta and any independent variables present in `SNPdata`, like `Theta ~ Var1 + Var2`;
though the "ThetaLink" input (see later) allows for general linear relationships.</p></dd>


<dt>PiForm</dt>
<dd><p>An object describing the (linear) relationship between pi and any independent variables present in `SNPdata`, like `Pi ~ Var1 + Var2`;
though the "PiLink" input (see later) allows for general linear relationships.</p></dd>


<dt>SNPdata</dt>
<dd><p>Dataframe. A dataframe containing reference- and variant allele counts, in columns with names as given by the "RefCol" and "VarCol" inputs.
This dataframe also has to contain any independent variables passed to the "ThetaForm" and "PiForm" input arguments.</p></dd>


<dt>RefCol</dt>
<dd><p>String. Name of the column in SNPdata containing reference allele counts.</p></dd>


<dt>VarCol</dt>
<dd><p>String. Name of the column in SNPdata containing variant allele counts.</p></dd>


<dt>WeightCol</dt>
<dd><p>String. Optional name of a column of SNPdata containinig per-sample weights, that are - if specified - are used in a weighted maximum likelihood fit
(maximizing sum(sample-weights * sample-log-likelihoods))</p></dd>


<dt>Pi_start</dt>
<dd><p>Number. Starting pi value for numerical optimization (when fitting PiForm, this starting value is used as the intercept and all independent variables
start as having a regression coefficient of zero)</p></dd>


<dt>Theta_start</dt>
<dd><p>Number. Starting theta value for numerical optimization (when fitting ThetaForm, this starting value is used as the intercept and all independent variables
start as having a regression coefficient of zero)</p></dd>


<dt>PiLink</dt>
<dd><p>String. One of "identity", "log" or "sqrt"; the linear relationship specified by PiForm is fit as PiLink(pi) ~ PiForm</p></dd>


<dt>ThetaLink</dt>
<dd><p>String. One of "identity", "log" or "sqrt"; the linear relationship specified by ThetaForm is fit as ThetaLink(theta) ~ ThetaForm</p></dd>


<dt>center</dt>
<dd><p>Logical. If TRUE, centers all exploratory variables (relative to their mean) before performing the maximum likelihood fit.
This is recommended when the supplied Pi_start is an expected/mean pi-value across all samples (e.g. the result of a single pi-fit on these samples)</p></dd>


<dt>ResetThetaMin</dt>
<dd><p>Number. When the supplied Theta_start value is lower than this input, it is reset to this input. Default 10^-10;
it is not recommended to change this value.</p></dd>


<dt>ResetThetaMax</dt>
<dd><p>Number. When the supplied Theta_start value is higher than this input, it is reset to this input. Default 10^-1;
it is not recommended to change this value.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing the following components:</p>
<dl><dt>first list-item</dt>
<dd><p>The negative of the maximized log-likelihood.</p></dd>

<dt>second list-item</dt>
<dd><p>Optimized parameter values, given in the order (1) pi-intercept, (2) theta-intercept, (3) pi regression coefficients, (4) theta regression coefficients.</p></dd>

<dt>optional third list-item</dt>
<dd><p>If theta depends on only one independent variable, the mean of that variable across samples.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

