<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Estimate sequencing error rate and inbreeding using an Expectation-Maximisation algorithm — AllelicMeta_est • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate sequencing error rate and inbreeding using an Expectation-Maximisation algorithm — AllelicMeta_est"><meta name="description" content="AllelicMeta_est calculates sequencing error rates, inbreeding
    coefficients, allele frequencies, genotypes and genotype probabilities from ref_counts and var_counts
    for a specific locus. To this end it fits a (regular) binomial mixture model with no allelic bias in heterozygotes (both alleles equally likely).
    This somewhat simplistic model (no overdispersion parameter, no allelic bias) makes for an inferior but very fast fit,
    which is ideal to get rough estimates of the allelic population metaparameters (sequencing error rate and inbreeding coefficient) by e.g. taking their mean or median across a large amount of loci.
    The per-sample allele frequencies and genotype (probabilities) are less reliable and, as such, not this function's main use."><meta property="og:description" content="AllelicMeta_est calculates sequencing error rates, inbreeding
    coefficients, allele frequencies, genotypes and genotype probabilities from ref_counts and var_counts
    for a specific locus. To this end it fits a (regular) binomial mixture model with no allelic bias in heterozygotes (both alleles equally likely).
    This somewhat simplistic model (no overdispersion parameter, no allelic bias) makes for an inferior but very fast fit,
    which is ideal to get rough estimates of the allelic population metaparameters (sequencing error rate and inbreeding coefficient) by e.g. taking their mean or median across a large amount of loci.
    The per-sample allele frequencies and genotype (probabilities) are less reliable and, as such, not this function's main use."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Estimate sequencing error rate and inbreeding using an Expectation-Maximisation algorithm</h1>

      <div class="d-none name"><code>AllelicMeta_est.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>AllelicMeta_est</code> calculates sequencing error rates, inbreeding
    coefficients, allele frequencies, genotypes and genotype probabilities from <code>ref_counts</code> and <code>var_counts</code>
    for a specific locus. To this end it fits a (regular) binomial mixture model with no allelic bias in heterozygotes (both alleles equally likely).
    This somewhat simplistic model (no overdispersion parameter, no allelic bias) makes for an inferior but very fast fit,
    which is ideal to get rough estimates of the allelic population metaparameters (sequencing error rate and inbreeding coefficient) by e.g. taking their mean or median across a large amount of loci.
    The per-sample allele frequencies and genotype (probabilities) are less reliable and, as such, not this function's main use.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AllelicMeta_est</span><span class="op">(</span></span>
<span>  <span class="va">ref_counts</span>,</span>
<span>  <span class="va">var_counts</span>,</span>
<span>  deltaF <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span>,</span>
<span>  maxIT <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  SE_prior <span class="op">=</span> <span class="fl">0.002</span>,</span>
<span>  F_inbr_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  HetProb <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ref-counts">ref_counts<a class="anchor" aria-label="anchor" href="#arg-ref-counts"></a></dt>
<dd><p>Numeric vector. Reference counts.</p></dd>


<dt id="arg-var-counts">var_counts<a class="anchor" aria-label="anchor" href="#arg-var-counts"></a></dt>
<dd><p>Numeric vector. Variant counts.</p></dd>


<dt id="arg-deltaf">deltaF<a class="anchor" aria-label="anchor" href="#arg-deltaf"></a></dt>
<dd><p>Number. Expectation-Maximisation threshold, minimal difference between
two consecutive iterations (default is 1e-08).</p></dd>


<dt id="arg-maxit">maxIT<a class="anchor" aria-label="anchor" href="#arg-maxit"></a></dt>
<dd><p>Number. Maximum number of iterations of the Expectation-Maximisation algorithm (default is 100).</p></dd>


<dt id="arg-se-prior">SE_prior<a class="anchor" aria-label="anchor" href="#arg-se-prior"></a></dt>
<dd><p>Number. Initial estimate of the sequencing error rate (default is 0.002).</p></dd>


<dt id="arg-f-inbr-prior">F_inbr_prior<a class="anchor" aria-label="anchor" href="#arg-f-inbr-prior"></a></dt>
<dd><p>Number. Initial estimate of the inbreeding coefficient used for calculating initial genotype frequencies (default is NULL, in which case the initial genotype frequences all get set to 1/3).</p></dd>


<dt id="arg-hetprob">HetProb<a class="anchor" aria-label="anchor" href="#arg-hetprob"></a></dt>
<dd><p>Number. Allelic bias in heterozygotes (expected reference over total allele count in RNAseq data; default is 0.5</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>allelefreq</dt>
<dd><p>The estimated allele frequency.</p></dd>

<dt>SE</dt>
<dd><p>The estimated sequencing error rate.</p></dd>

<dt>F_inbr</dt>
<dd><p>The estimated inbreeding coefficient.</p></dd>

<dt>genotypes</dt>
<dd><p>The most likely genotype (rr, rv, vv) of each sample.</p></dd>

<dt>genoprobs</dt>
<dd><p>The genotype probabilties (p(rr), p(rv), p(vv)) for each sample.</p></dd>

<dt>nrep</dt>
<dd><p>The number of iterations</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">AllelicMeta_est</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">8</span>, <span class="fl">10</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">23</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span>, <span class="fl">6</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">10</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $allelefreq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.5714285</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $SE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $F_inbr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] -0.7500002</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "rv" "rv" "rv" "rv" "rv" "rv" "rr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $genoprobs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       p.rr.        p.rv. p.vv.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 0.0000000 1.000000e+00     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 0.0000000 1.000000e+00     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 0.0000000 1.000000e+00     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 0.0000000 1.000000e+00     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 0.0000000 1.000000e+00     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 0.0000000 1.000000e+00     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7 0.9999993 7.152558e-07     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nrep</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">AllelicMeta_est</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">1</span>, <span class="fl">23</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span>, <span class="fl">6</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">10</span>, <span class="fl">0</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    SE_prior <span class="op">=</span> <span class="fl">0.2</span>, F_inbr_prior <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $allelefreq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.4931604</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $SE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03413917</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $F_inbr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.678992</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "rr" "vv" "vv" "rv" "rr" "vv" "rr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $genoprobs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          p.rr.        p.rv.        p.vv.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 8.253493e-01 1.746494e-01 1.330900e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 2.353486e-12 1.936346e-03 9.980637e-01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 1.873889e-09 7.187552e-03 9.928124e-01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 7.935723e-02 9.177449e-01 2.897870e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 9.857576e-01 1.424238e-02 5.618446e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 8.271749e-14 7.557152e-03 9.924428e-01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7 9.999999e-01 1.029936e-07 4.226079e-34</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nrep</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

