<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>multi-locus wrapper function of AllelicMeta_est — AllelicMeta_est_par • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="multi-locus wrapper function of AllelicMeta_est — AllelicMeta_est_par"><meta name="description" content="AllelicMeta_est_par is a wrapper function of AllelicMeta_est allowing multiple loci to be given as input (as lists)
and focussing solely on obtaining a reliable sequencing error rate and inbreeding coefficient estimate (the population metaparameters),
this via extra inputs specifying filter criteria so as to only retain high-quality loci."><meta property="og:description" content="AllelicMeta_est_par is a wrapper function of AllelicMeta_est allowing multiple loci to be given as input (as lists)
and focussing solely on obtaining a reliable sequencing error rate and inbreeding coefficient estimate (the population metaparameters),
this via extra inputs specifying filter criteria so as to only retain high-quality loci."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>multi-locus wrapper function of AllelicMeta_est</h1>

      <div class="d-none name"><code>AllelicMeta_est_par.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>AllelicMeta_est_par</code> is a wrapper function of <code>AllelicMeta_est</code> allowing multiple loci to be given as input (as lists)
and focussing solely on obtaining a reliable sequencing error rate and inbreeding coefficient estimate (the population metaparameters),
this via extra inputs specifying filter criteria so as to only retain high-quality loci.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AllelicMeta_est_par</span><span class="op">(</span></span>
<span>  <span class="va">DataList</span>,</span>
<span>  deltaF <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span>,</span>
<span>  maxIT <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  SE_prior <span class="op">=</span> <span class="fl">0.002</span>,</span>
<span>  F_inbr_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  HetProb <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  MinAllele_filt <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  SE_filt <span class="op">=</span> <span class="fl">0.035</span>,</span>
<span>  NumSamp_filt <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  MedianCov_filt <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-datalist">DataList<a class="anchor" aria-label="anchor" href="#arg-datalist"></a></dt>
<dd><p>List of dataframes. Each dataframes should at least contain a column of reference- and variant-allele counts (named "ref_count" and "var_count" respectively).</p></dd>


<dt id="arg-deltaf">deltaF<a class="anchor" aria-label="anchor" href="#arg-deltaf"></a></dt>
<dd><p>Number. Expectation-Maximisation threshold, minimal difference between
two consecutive iterations (default is 1e-08).</p></dd>


<dt id="arg-maxit">maxIT<a class="anchor" aria-label="anchor" href="#arg-maxit"></a></dt>
<dd><p>Number. Maximum number of iterations of the Expectation-Maximisation algorithm (default is 100).</p></dd>


<dt id="arg-se-prior">SE_prior<a class="anchor" aria-label="anchor" href="#arg-se-prior"></a></dt>
<dd><p>Number. Initial estimate of the sequencing error rate (default is 0.002).</p></dd>


<dt id="arg-f-inbr-prior">F_inbr_prior<a class="anchor" aria-label="anchor" href="#arg-f-inbr-prior"></a></dt>
<dd><p>Number. Initial estimate of the inbreeding coefficient used for calculating initial genotype frequencies (default is NULL, in which case the initial genotype frequences all get set to 1/3).</p></dd>


<dt id="arg-hetprob">HetProb<a class="anchor" aria-label="anchor" href="#arg-hetprob"></a></dt>
<dd><p>Number. Allelic bias in heterozygotes (expected reference over total allele count in RNAseq data; default is 0.5</p></dd>


<dt id="arg-minallele-filt">MinAllele_filt<a class="anchor" aria-label="anchor" href="#arg-minallele-filt"></a></dt>
<dd><p>Number. Minimum allowed estimated minor allele frequency for a locus to be considered in the metaparameter estimates.</p></dd>


<dt id="arg-se-filt">SE_filt<a class="anchor" aria-label="anchor" href="#arg-se-filt"></a></dt>
<dd><p>Number. Maximum allowed estimated sequencing error rate for a locus to be considered in the metaparameter estimates.</p></dd>


<dt id="arg-numsamp-filt">NumSamp_filt<a class="anchor" aria-label="anchor" href="#arg-numsamp-filt"></a></dt>
<dd><p>Number. Minimum allowed number of samples for a locus to be considered in the metaparameter estimates.</p></dd>


<dt id="arg-mediancov-filt">MedianCov_filt<a class="anchor" aria-label="anchor" href="#arg-mediancov-filt"></a></dt>
<dd><p>Number. Minimum allowed median coverage across samples (reference plus variant allele count) for a locus to be considered in the metaparameter estimates.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>DataList_out</dt>
<dd><p>An updated DataList, now containing each locus' sequencing error estimate (est_SE), inbreeding coefficient estimate (est_inbr), reference allele frequency estimate (allelefreq_prel), preliminary genotype probabilities (prr_prel, prv_prel, pvv_prel) and preliminary genotype (genotype_prel)</p></dd>

<dt>SE_vec</dt>
<dd><p>Vector of estimated sequencing error rates of reliable loci.</p></dd>

<dt>F_vec</dt>
<dd><p>Vector of estimated inbreeding coefficients of reliable loci.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

