<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>multi-locus wrapper function of AllelicMeta_est — AllelicMeta_est_par • MAGE</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="multi-locus wrapper function of AllelicMeta_est — AllelicMeta_est_par"><meta property="og:description" content="AllelicMeta_est_par is a wrapper function of AllelicMeta_est allowing multiple loci to be given as input (as lists)
and focussing solely on obtaining a reliable sequencing error rate and inbreeding coefficient estimate (the population metaparameters),
this via extra inputs specifying filter criteria so as to only retain high-quality loci."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MAGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MAGE_expanded_tutorial.html">MAGE expanded tutorial</a>
    </li>
    <li>
      <a href="../articles/MAGE_tutorial.html">MAGE tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>multi-locus wrapper function of AllelicMeta_est</h1>
    
    <div class="hidden name"><code>AllelicMeta_est_par.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>AllelicMeta_est_par</code> is a wrapper function of <code>AllelicMeta_est</code> allowing multiple loci to be given as input (as lists)
and focussing solely on obtaining a reliable sequencing error rate and inbreeding coefficient estimate (the population metaparameters),
this via extra inputs specifying filter criteria so as to only retain high-quality loci.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AllelicMeta_est_par</span><span class="op">(</span></span>
<span>  <span class="va">DataList</span>,</span>
<span>  deltaF <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span>,</span>
<span>  maxIT <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  SE_prior <span class="op">=</span> <span class="fl">0.002</span>,</span>
<span>  F_inbr_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  HetProb <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  pA_filt <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  SE_filt <span class="op">=</span> <span class="fl">0.035</span>,</span>
<span>  NumSamp_filt <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  MedianCov_filt <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>DataList</dt>
<dd><p>List of dataframes. Each dataframes should at least contain a column of reference- and variant-allele counts (named "ref_count" and "var_count" respectively).</p></dd>


<dt>deltaF</dt>
<dd><p>Number. Expectation-Maximisation threshold, minimal difference between
two consecutive iterations (default is 1e-08).</p></dd>


<dt>maxIT</dt>
<dd><p>Number. Maximum number of iterations of the Expectation-Maximisation algorithm (default is 100).</p></dd>


<dt>SE_prior</dt>
<dd><p>Number. Initial estimate of the sequencing error rate (default is 0.002).</p></dd>


<dt>F_inbr_prior</dt>
<dd><p>Number. Initial estimate of the inbreeding coefficient used for calculating initial genotype frequencies (default is NULL, in which case the initial genotype frequences all get set to 1/3).</p></dd>


<dt>HetProb</dt>
<dd><p>Number. Allelic bias in heterozygotes (expected reference over total allele count in RNAseq data; default is 0.5</p></dd>


<dt>pA_filt</dt>
<dd><p>Number. Minimum allowed estimated minor allele frequency for a locus to be considered in the metaparameter estimates.</p></dd>


<dt>SE_filt</dt>
<dd><p>Number. Maximum allowed estimated sequencing error rate for a locus to be considered in the metaparameter estimates.</p></dd>


<dt>NumSamp_filt</dt>
<dd><p>Number. Minimum allowed number of samples for a locus to be considered in the metaparameter estimates.</p></dd>


<dt>MedianCov_filt</dt>
<dd><p>Number. Minimum allowed median coverage across samples (reference plus variant allele count) for a locus to be considered in the metaparameter estimates.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing the following components:</p>
<dl><dt>DataList_out</dt>
<dd><p>An updated DataList, now containing each locus' sequencing error estimate (est_SE), inbreeding coefficient estimate (est_inbr), reference allele frequency estimate (allelefreq_prel), preliminary genotype probabilities (prr_prel, prv_prel, pvv_prel) and preliminary genotype (genotype_prel)</p></dd>

<dt>SE_vec</dt>
<dd><p>Vector of estimated sequencing error rates of reliable loci.</p></dd>

<dt>F_vec</dt>
<dd><p>Vector of estimated inbreeding coefficients of reliable loci.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

