<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping — BetaBinomGenotyping • MAGE</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping — BetaBinomGenotyping"><meta property="og:description" content="BetaBinomGenotyping is a wrapper function of EMfit_betabinom_robust allowing multiple loci to be given as input (as lists).
Besides calling the latter to perform a EM-fit and update the inputted per-locus dataframes as listed in EMfit_betabinom_robust output,
it also returns a results-dataframe with one row for every locus, containing genotyping- and Allelic Bias-detection results.
This function is mainly useful when following MAGE's vignette step-by-step, as its input requirement are rather stictly dependent on performing all previous steps in that pipeline.
See the help page of EMfit_betabinom_robust for more information on these outputs"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MAGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MAGE_expanded_tutorial.html">MAGE expanded tutorial</a>
    </li>
    <li>
      <a href="../articles/MAGE_tutorial.html">MAGE tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping</h1>
    
    <div class="hidden name"><code>BetaBinomGenotyping.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>BetaBinomGenotyping</code> is a wrapper function of <code>EMfit_betabinom_robust</code> allowing multiple loci to be given as input (as lists).
Besides calling the latter to perform a EM-fit and update the inputted per-locus dataframes as listed in <code>EMfit_betabinom_robust</code> output,
it also returns a results-dataframe with one row for every locus, containing genotyping- and Allelic Bias-detection results.
This function is mainly useful when following MAGE's vignette step-by-step, as its input requirement are rather stictly dependent on performing all previous steps in that pipeline.
See the help page of <code>EMfit_betabinom_robust</code> for more information on these outputs</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">BetaBinomGenotyping</span><span class="op">(</span></span>
<span>  <span class="va">DataList</span>,</span>
<span>  allelefreq <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  inbr <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  dltaco <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">6</span>,</span>
<span>  HWE <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_InitEst <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ThetaInits <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  ReEstThetas <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  NoSplitHom <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  NoSplitHet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ResetThetaMin <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">10</span>,</span>
<span>  ResetThetaMax <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>,</span>
<span>  DistRob <span class="op">=</span> <span class="st">"Cook"</span>,</span>
<span>  CookMargin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  LikEmpNum <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  LikMargin <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  NumHetMin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  MaxOutFrac <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  thetaTRY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">7</span><span class="op">)</span>,</span>
<span>  fitH0 <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>DataList</dt>
<dd><p>List of dataframes. Each dataframes should at least contain the columns "ref", ref_count", "var", "var_count" and "est_SE",
which should be the case if following the vignette up to this function's appearance.</p></dd>


<dt>allelefreq, SE, inbr, dltaco, HWE, p_InitEst, ThetaInits, ReEstThetas, NoSplitHom, NoSplitHet, ResetThetaMin, ResetThetaMax, DistRob, CookMargin, LikEmpNum, LikMargin, NumHetMin, MaxOutFrac, thetaTRY, fitH0</dt>
<dd><p>All remaining parameters of <code>EMfit_betabinom_robust</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing the following components:</p>
<dl><dt>DataList_out</dt>
<dd><p>The updated DataList, corresponding to the <code>data_hash</code> output of <code>EMfit_betabinom_robust</code></p></dd>

<dt>Geno_AB_res</dt>
<dd><p>A dataframe containing per-locus results of <code>EMfit_betabinom_robust</code> EM-fit as well as some other metrics, namely:</p></dd>

   </dl><ul><li><p>positionThe locus' name, according to <code>names(DataList)</code></p></li>
<li><p>probshiftFitted reference allele fraction in RNAseq reads, indicating allelic bias when different from 0.5</p></li>
<li><p>LRTThe likelihood ratio test statistic, testing for significant allelic bias</p></li>
<li><p>pThe likelihood ratio test p-value, testing for significant allelic bias</p></li>
<li><p>qualityEquals "!" if the sample contains no fitted heterozygotes, otherwise ""</p></li>
<li><p>allele_frequencyestimated reference allele frequency in the population</p></li>
<li><p>referencereference allele nucleotide</p></li>
<li><p>variantvariant allele nucleotide</p></li>
<li><p>est_SEper-locus sequencing error rate estimate, as outputted earlier by <code>AllelicMeta_est</code></p></li>
<li><p>coveragemedian coverage across samples</p></li>
<li><p>nr_samplesnumber of samples covering this locus with at least one reference- or variant-count</p></li>
<li><p>median_ABmedian allelic bias, as outputted by <code>median_AB</code></p></li>
<li><p>rho_rrfitted reference homozygous fraction in the population</p></li>
<li><p>rho_rvfitted heterozygous fraction in the population</p></li>
<li><p>rho_vvfitted variant homozygous fraction in the population</p></li>
<li><p>theta_homfitted overdispersion parameter for the homozygous PMFs</p></li>
<li><p>theta_hetfitted overdispersion parameter for the heterozygous PMF</p></li>
<li><p>theta_hom_NoShiftfitted overdispersion parameter for the homozygous PMFs assuming no allelic bias</p></li>
<li><p>theta_het_NoShiftfitted overdispersion parameter for the heterozygous PMF assuming no allelic bias</p></li>
<li><p>Chi2PVALp-value of a chi square test assessing Hardy-Weinberg-Equilibrium on the locus given the inbreeding coefficient metaparameter; see <code>HWE_chisquared</code></p></li>
<li><p>Chi2STATtest statistic of a chi square test assessing Hardy-Weinberg-Equilibrium on the locus given the inbreeding coefficient metaparameter; see <code>HWE_chisquared</code></p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

