<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping — BetaBinomGenotyping • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping — BetaBinomGenotyping"><meta name="description" content="BetaBinomGenotyping is a wrapper function of EMfit_betabinom_robust allowing multiple loci to be given as input (as lists).
Besides calling the latter to perform a EM-fit and update the inputted per-locus dataframes as listed in EMfit_betabinom_robust output,
it also returns a results-dataframe with one row for every locus, containing genotyping- and Allelic Bias-detection results.
This function is mainly useful when following maelstRom's vignette step-by-step, as its input requirement are rather stictly dependent on performing all previous steps in that pipeline.
See the help page of EMfit_betabinom_robust for more information on these outputs"><meta property="og:description" content="BetaBinomGenotyping is a wrapper function of EMfit_betabinom_robust allowing multiple loci to be given as input (as lists).
Besides calling the latter to perform a EM-fit and update the inputted per-locus dataframes as listed in EMfit_betabinom_robust output,
it also returns a results-dataframe with one row for every locus, containing genotyping- and Allelic Bias-detection results.
This function is mainly useful when following maelstRom's vignette step-by-step, as its input requirement are rather stictly dependent on performing all previous steps in that pipeline.
See the help page of EMfit_betabinom_robust for more information on these outputs"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping</h1>

      <div class="d-none name"><code>BetaBinomGenotyping.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>BetaBinomGenotyping</code> is a wrapper function of <code>EMfit_betabinom_robust</code> allowing multiple loci to be given as input (as lists).
Besides calling the latter to perform a EM-fit and update the inputted per-locus dataframes as listed in <code>EMfit_betabinom_robust</code> output,
it also returns a results-dataframe with one row for every locus, containing genotyping- and Allelic Bias-detection results.
This function is mainly useful when following maelstRom's vignette step-by-step, as its input requirement are rather stictly dependent on performing all previous steps in that pipeline.
See the help page of <code>EMfit_betabinom_robust</code> for more information on these outputs</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">BetaBinomGenotyping</span><span class="op">(</span></span>
<span>  <span class="va">DataList</span>,</span>
<span>  allelefreq <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  inbr <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  dltaco <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">6</span>,</span>
<span>  HWE <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_InitEst <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ThetaInits <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  ReEstThetas <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  NoSplitHom <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  NoSplitHet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ResetThetaMin <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">10</span>,</span>
<span>  ResetThetaMax <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>,</span>
<span>  DistRob <span class="op">=</span> <span class="st">"Cook"</span>,</span>
<span>  CookMargin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  LikEmpNum <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  LikMargin <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  NumHetMin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  MaxOutFrac <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  thetaTRY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">7</span><span class="op">)</span>,</span>
<span>  fitH0 <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-datalist">DataList<a class="anchor" aria-label="anchor" href="#arg-datalist"></a></dt>
<dd><p>List of dataframes. Each dataframes should at least contain the columns "ref", ref_count", "var", "var_count" and "est_SE",
which should be the case if following the vignette up to this function's appearance.</p></dd>


<dt id="arg-allelefreq-se-inbr-dltaco-hwe-p-initest-thetainits-reestthetas-nosplithom-nosplithet-resetthetamin-resetthetamax-distrob-cookmargin-likempnum-likmargin-numhetmin-maxoutfrac-thetatry-fith-">allelefreq, SE, inbr, dltaco, HWE, p_InitEst, ThetaInits, ReEstThetas, NoSplitHom, NoSplitHet, ResetThetaMin, ResetThetaMax, DistRob, CookMargin, LikEmpNum, LikMargin, NumHetMin, MaxOutFrac, thetaTRY, fitH0<a class="anchor" aria-label="anchor" href="#arg-allelefreq-se-inbr-dltaco-hwe-p-initest-thetainits-reestthetas-nosplithom-nosplithet-resetthetamin-resetthetamax-distrob-cookmargin-likempnum-likmargin-numhetmin-maxoutfrac-thetatry-fith-"></a></dt>
<dd><p>All remaining parameters of <code>EMfit_betabinom_robust</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>DataList_out</dt>
<dd><p>The updated DataList, corresponding to the <code>data_hash</code> output of <code>EMfit_betabinom_robust</code></p></dd>

<dt>Geno_AB_res</dt>
<dd><p>A dataframe containing per-locus results of <code>EMfit_betabinom_robust</code> EM-fit as well as some other metrics, namely:</p></dd>

   </dl><ul><li><p>positionThe locus' name, according to <code>names(DataList)</code></p></li>
<li><p>probshiftFitted reference allele fraction in RNAseq reads, indicating allelic bias when different from 0.5</p></li>
<li><p>LRTThe likelihood ratio test statistic, testing for significant allelic bias</p></li>
<li><p>pThe likelihood ratio test p-value, testing for significant allelic bias</p></li>
<li><p>qualityEquals "!" if the sample contains no fitted heterozygotes, otherwise ""</p></li>
<li><p>allele_frequencyestimated reference allele frequency in the population</p></li>
<li><p>referencereference allele nucleotide</p></li>
<li><p>variantvariant allele nucleotide</p></li>
<li><p>est_SEper-locus sequencing error rate estimate, as outputted earlier by <code>AllelicMeta_est</code></p></li>
<li><p>coveragemedian coverage across samples</p></li>
<li><p>nr_samplesnumber of samples covering this locus with at least one reference- or variant-count</p></li>
<li><p>median_ABmedian allelic bias, as outputted by <code>median_AB</code></p></li>
<li><p>rho_rrfitted reference homozygous fraction in the population</p></li>
<li><p>rho_rvfitted heterozygous fraction in the population</p></li>
<li><p>rho_vvfitted variant homozygous fraction in the population</p></li>
<li><p>theta_homfitted overdispersion parameter for the homozygous PMFs</p></li>
<li><p>theta_hetfitted overdispersion parameter for the heterozygous PMF</p></li>
<li><p>theta_hom_NoShiftfitted overdispersion parameter for the homozygous PMFs assuming no allelic bias</p></li>
<li><p>theta_het_NoShiftfitted overdispersion parameter for the heterozygous PMF assuming no allelic bias</p></li>
<li><p>Chi2PVALp-value of a chi square test assessing Hardy-Weinberg-Equilibrium on the locus given the inbreeding coefficient metaparameter; see <code>HWE_chisquared</code></p></li>
<li><p>Chi2STATtest statistic of a chi square test assessing Hardy-Weinberg-Equilibrium on the locus given the inbreeding coefficient metaparameter; see <code>HWE_chisquared</code></p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

