<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Performs a differential Allelic Divergence analysis on a control- and case-dataset — dAD_analysis • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Performs a differential Allelic Divergence analysis on a control- and case-dataset — dAD_analysis"><meta name="description" content="dAD_analysis is a wrapper function bundling EMfit_betabinom_robust, EMfit_betabinom, EMfit_betabinom_popcomb,
pmf_betabinomMix, and HWE_chisquared to run the entire differential Allelic Divergence detection analysis as discussed in the package vignette.
To this end, it goes over the following steps:
Calling EMfit_betabinom_robust on both the control- and case-dataset separately for the sole purpose of outlier detection;
  detected outliers in both datasets play no part in future Expectation-Maximization-fits.
Perform one fit on the joint (case+control) data using EMfit_betabinom, i.e. assuming all population parameters are shared including
  the heterozygous overdispersion parameter, i.e. assuming no differential Allelic Divergence.
Perform a fit on the control- and case-data using EMfit_betabinom_popcomb, which assumes all population parameters are shared
  except for the heteroyzous overdispersion parameter, i.e. a fit accomodating differential Allelic Divergence
Statistically test for differential Allelic Divergence using a likelihood ratio test with one degree of freedom comparing the
  fits from the two previous points.
Filling out the results dataframe dAD_res.

"><meta property="og:description" content="dAD_analysis is a wrapper function bundling EMfit_betabinom_robust, EMfit_betabinom, EMfit_betabinom_popcomb,
pmf_betabinomMix, and HWE_chisquared to run the entire differential Allelic Divergence detection analysis as discussed in the package vignette.
To this end, it goes over the following steps:
Calling EMfit_betabinom_robust on both the control- and case-dataset separately for the sole purpose of outlier detection;
  detected outliers in both datasets play no part in future Expectation-Maximization-fits.
Perform one fit on the joint (case+control) data using EMfit_betabinom, i.e. assuming all population parameters are shared including
  the heterozygous overdispersion parameter, i.e. assuming no differential Allelic Divergence.
Perform a fit on the control- and case-data using EMfit_betabinom_popcomb, which assumes all population parameters are shared
  except for the heteroyzous overdispersion parameter, i.e. a fit accomodating differential Allelic Divergence
Statistically test for differential Allelic Divergence using a likelihood ratio test with one degree of freedom comparing the
  fits from the two previous points.
Filling out the results dataframe dAD_res.

"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Performs a differential Allelic Divergence analysis on a control- and case-dataset</h1>

      <div class="d-none name"><code>dAD_analysis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>dAD_analysis</code> is a wrapper function bundling <code>EMfit_betabinom_robust</code>, <code>EMfit_betabinom</code>, <code>EMfit_betabinom_popcomb</code>,
<code>pmf_betabinomMix</code>, and <code>HWE_chisquared</code> to run the entire differential Allelic Divergence detection analysis as discussed in the package vignette.
To this end, it goes over the following steps:</p><ol><li><p>Calling <code>EMfit_betabinom_robust</code> on both the control- and case-dataset separately for the sole purpose of outlier detection;
  detected outliers in both datasets play no part in future Expectation-Maximization-fits.</p></li>
<li><p>Perform one fit on the joint (case+control) data using <code>EMfit_betabinom</code>, i.e. assuming all population parameters are shared including
  the heterozygous overdispersion parameter, i.e. assuming no differential Allelic Divergence.</p></li>
<li><p>Perform a fit on the control- and case-data using <code>EMfit_betabinom_popcomb</code>, which assumes all population parameters are shared
  except for the heteroyzous overdispersion parameter, i.e. a fit accomodating differential Allelic Divergence</p></li>
<li><p>Statistically test for differential Allelic Divergence using a likelihood ratio test with one degree of freedom comparing the
  fits from the two previous points.</p></li>
<li><p>Filling out the results dataframe <code>dAD_res</code>.</p></li>
</ol></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">dAD_analysis</span><span class="op">(</span></span>
<span>  <span class="va">datas</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  <span class="va">inbr</span>,</span>
<span>  allelefreq <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  dltaco <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">6</span>,</span>
<span>  HWE <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_InitEst <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_inits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  ThetaInits <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  ReEstThetas <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  NoSplitHom <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  NoSplitHet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ResetThetaMin <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">10</span>,</span>
<span>  ResetThetaMax <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>,</span>
<span>  DistRob <span class="op">=</span> <span class="st">"Cook"</span>,</span>
<span>  CookMargin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  LikEmpNum <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  LikMargin <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  NumHetMin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  MaxOutFrac <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  thetaTRY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">7</span><span class="op">)</span>,</span>
<span>  probshift_init <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  FirstFewFixed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  MemLim <span class="op">=</span> <span class="fl">2048</span>,</span>
<span>  Xtra <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  epsabs <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  MaxIt <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-se">SE<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Number. Sequencing error rate population metaparameter.</p></dd>


<dt id="arg-inbr">inbr<a class="anchor" aria-label="anchor" href="#arg-inbr"></a></dt>
<dd><p>Number. Inbreeding coefficient population metaparameter.</p></dd>


<dt id="arg-allelefreq-dltaco-hwe-p-initest-thetainits-reestthetas-nosplithom-nosplithet-resetthetamin-resetthetamax-distrob-cookmargin-likempnum-likmargin-numhetmin-maxoutfrac-thetatry">allelefreq, dltaco, HWE, p_InitEst, ThetaInits, ReEstThetas, NoSplitHom, NoSplitHet, ResetThetaMin, ResetThetaMax, DistRob, CookMargin, LikEmpNum, LikMargin, NumHetMin, MaxOutFrac, thetaTRY<a class="anchor" aria-label="anchor" href="#arg-allelefreq-dltaco-hwe-p-initest-thetainits-reestthetas-nosplithom-nosplithet-resetthetamin-resetthetamax-distrob-cookmargin-likempnum-likmargin-numhetmin-maxoutfrac-thetatry"></a></dt>
<dd><p>Remaining
control parameters to be passed to <code>EMfit_betabinom_robust</code>, <code>EMfit_betabinom</code>, and <code>EMfit_betabinom_popcomb</code>. See their respective documentation
for more information. Usually, these parameter's default values are fine.</p></dd>


<dt id="arg-datalist">Datalist<a class="anchor" aria-label="anchor" href="#arg-datalist"></a></dt>
<dd><p>A list containing two lists of dataframes, which in turn contain control- and case-data (in that order). Subsequent entries in both lists
correspond to subsequent loci (lists must be the same size and named per-locus). Each of the dataframes should at least contain both a <code>ref_count</code> and
<code>var_count</code> column.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dataframe with one row per locus, containing for that locus in its columns:</p>
<dl><dt>LocName</dt>
<dd><p>Locus name.</p></dd>

<dt>PiFitH0</dt>
<dd><p>Heterozygous pi parameter (reference allele fraction) for the joint fit sharing all parameters.</p></dd>

<dt>PiFitH1</dt>
<dd><p>Heterozygous pi parameter (reference allele fraction) for the fit allowing differential Allelic Divergence.</p></dd>

<dt>ThetaHetH0</dt>
<dd><p>Heterozygous theta parameter (overdispersion) for the joint fit sharing all parameters.</p></dd>

<dt>ThetaHetCTRL</dt>
<dd><p>Heterozygous theta parameter (overdispersion) of controls for the fit allowing differential Allelic Divergence.</p></dd>

<dt>ThetaHetCASE</dt>
<dd><p>Heterozygous theta parameter (overdispersion) of cases for the fit allowing differential Allelic Divergence.</p></dd>

<dt>RhoHetH0</dt>
<dd><p>Heterozygous rho parameter (overdispersion, ranging from 0 to 1) for the joint fit sharing all parameters.</p></dd>

<dt>RhoHetCTRL</dt>
<dd><p>Heterozygous rho parameter (overdispersion, ranging from 0 to 1) of controls for the fit allowing differential Allelic Divergence.</p></dd>

<dt>RhoHetCASE</dt>
<dd><p>Heterozygous rho parameter (overdispersion, ranging from 0 to 1) of cases for the fit allowing differential Allelic Divergence.</p></dd>

<dt>NumHetCTRL</dt>
<dd><p>Estimate number of heterozygotes in controls, according to its separate outlier-detection fit from step 1 above. Outliers are themselves genotyped and counted towards this number, just not used in the actual fitting procedure.</p></dd>

<dt>NumHetCASE</dt>
<dd><p>Estimate number of heterozygotes in cases, according to its separate outlier-detection fit from step 1 above. Outliers are themselves genotyped and counted towards this number, just not used in the actual fitting procedure.</p></dd>

<dt>RobFlagCTRL</dt>
<dd><p>The <code>RobFlag</code> output of the separate outlier-detection fit from step 1 on controls, see <code>EMfit_betabinom_robust</code>'s documentation.</p></dd>

<dt>RobFlagCASE</dt>
<dd><p>The <code>RobFlag</code> output of the separate outlier-detection fit from step 1 on cases, see <code>EMfit_betabinom_robust</code>'s documentation.</p></dd>

<dt>HWECTRL</dt>
<dd><p>HWE chi squared p-value for controls, see <code>HWE_chisquared</code>. Performed on genotyping data according to the outlier-detection fit of step 1.</p></dd>

<dt>HWECASE</dt>
<dd><p>HWE chi squared p-value for cases, see <code>HWE_chisquared</code>. Performed on genotyping data according to the outlier-detection fit of step 1.</p></dd>

<dt>CovCTRL_mean</dt>
<dd><p>Mean coverage (reference + variant count) across control data.</p></dd>

<dt>CovCASE_mean</dt>
<dd><p>Mean coverage (reference + variant count) across case data.</p></dd>

<dt>CovCTRL_med</dt>
<dd><p>Median coverage (reference + variant count) across control data.</p></dd>

<dt>CovCASE_med</dt>
<dd><p>Median coverage (reference + variant count) across case data.</p></dd>

<dt>NumOutCTRL</dt>
<dd><p>Number of outliers detected in control data.</p></dd>

<dt>NumOutCASE</dt>
<dd><p>Number of outliers detected in case data.</p></dd>

<dt>QualityCTRL</dt>
<dd><p>Quality flag returned by <code>EMfit_betabinom_robust</code> (see documentation) of the outlier-detection fit on control data (step 1 above).</p></dd>

<dt>QualityCASE</dt>
<dd><p>Quality flag returned by <code>EMfit_betabinom_robust</code> (see documentation) of the outlier-detection fit on case data (step 1 above).</p></dd>

<dt>pr</dt>
<dd><p>Reference homozygote fraction according to the differential Allelic Divergence accomodating fit (different heterozygous overdispersion parameters).</p></dd>

<dt>prv</dt>
<dd><p>Heterozygote fraction according to the differential Allelic Divergence accomodating fit (different heterozygous overdispersion parameters).</p></dd>

<dt>pv</dt>
<dd><p>Variant homozygote fraction according to the differential Allelic Divergence accomodating fit (different heterozygous overdispersion parameters).</p></dd>

<dt>ThetaHomH0</dt>
<dd><p>Homozygous theta parameter (overdispersion) for the joint fit sharing all parameters.</p></dd>

<dt>ThetaHomH1</dt>
<dd><p>Homozygous theta parameter (overdispersion) for the fit allowing differential Allelic Divergence.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

