<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Performs a differential Allelic Dispersion analysis on a control- and case-dataset — dAD_analysis • maelstRom</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Performs a differential Allelic Dispersion analysis on a control- and case-dataset — dAD_analysis"><meta property="og:description" content="dAD_analysis is a wrapper function bundling EMfit_betabinom_robust, EMfit_betabinom, EMfit_betabinom_popcomb,
pmf_betabinomMix, and HWE_chisquared to run the entire differential Allelic Dispersion detection analysis as discussed in the package vignette.
To this end, it goes over the following steps:
Calling EMfit_betabinom_robust on both the control- and case-dataset separately for the sole purpose of outlier detection;
  detected outliers in both datasets play no part in future Expectation-Maximization-fits.
Perform one fit on the joint (case+control) data using EMfit_betabinom, i.e. assuming all population parameters are shared including
  the heterozygous overdispersion parameter, i.e. assuming no differential Allelic Dispersion.
Perform a fit on the control- and case-data using EMfit_betabinom_popcomb, which assumes all population parameters are shared
  except for the heteroyzous overdispersion parameter, i.e. a fit accomodating differential Allelic Dispersion
Statistically test for differential Allelic Dispersion using a likelihood ratio test with one degree of freedom comparing the
  fits from the two previous points.
Filling out the results dataframe dAD_res.

"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Performs a differential Allelic Dispersion analysis on a control- and case-dataset</h1>
    
    <div class="hidden name"><code>dAD_analysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>dAD_analysis</code> is a wrapper function bundling <code>EMfit_betabinom_robust</code>, <code>EMfit_betabinom</code>, <code>EMfit_betabinom_popcomb</code>,
<code>pmf_betabinomMix</code>, and <code>HWE_chisquared</code> to run the entire differential Allelic Dispersion detection analysis as discussed in the package vignette.
To this end, it goes over the following steps:</p><ol><li><p>Calling <code>EMfit_betabinom_robust</code> on both the control- and case-dataset separately for the sole purpose of outlier detection;
  detected outliers in both datasets play no part in future Expectation-Maximization-fits.</p></li>
<li><p>Perform one fit on the joint (case+control) data using <code>EMfit_betabinom</code>, i.e. assuming all population parameters are shared including
  the heterozygous overdispersion parameter, i.e. assuming no differential Allelic Dispersion.</p></li>
<li><p>Perform a fit on the control- and case-data using <code>EMfit_betabinom_popcomb</code>, which assumes all population parameters are shared
  except for the heteroyzous overdispersion parameter, i.e. a fit accomodating differential Allelic Dispersion</p></li>
<li><p>Statistically test for differential Allelic Dispersion using a likelihood ratio test with one degree of freedom comparing the
  fits from the two previous points.</p></li>
<li><p>Filling out the results dataframe <code>dAD_res</code>.</p></li>
</ol></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">dAD_analysis</span><span class="op">(</span></span>
<span>  <span class="va">datas</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  <span class="va">inbr</span>,</span>
<span>  allelefreq <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  dltaco <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">6</span>,</span>
<span>  HWE <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_InitEst <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_inits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  ThetaInits <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  ReEstThetas <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  NoSplitHom <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  NoSplitHet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ResetThetaMin <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">10</span>,</span>
<span>  ResetThetaMax <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>,</span>
<span>  DistRob <span class="op">=</span> <span class="st">"Cook"</span>,</span>
<span>  CookMargin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  LikEmpNum <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  LikMargin <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  NumHetMin <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  MaxOutFrac <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  thetaTRY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">7</span><span class="op">)</span>,</span>
<span>  probshift_init <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  FirstFewFixed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  MemLim <span class="op">=</span> <span class="fl">2048</span>,</span>
<span>  Xtra <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  epsabs <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  MaxIt <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>datas</dt>
<dd><p>A list containing two lists of dataframes, which in turn contain control- and case-data (in that order). Subsequent entries in both lists
correspond to subsequent loci (lists must be the same size and named per-locus). Each of the dataframes should at least contain both a <code>ref_count</code> and
<code>var_count</code> column.</p></dd>


<dt>SE</dt>
<dd><p>Number. Sequencing error rate population metaparameter.</p></dd>


<dt>inbr</dt>
<dd><p>Number. Inbreeding coefficient population metaparameter.</p></dd>


<dt>allelefreq, dltaco, HWE, p_InitEst, ThetaInits, ReEstThetas, NoSplitHom, NoSplitHet, ResetThetaMin, ResetThetaMax, DistRob, CookMargin, LikEmpNum, LikMargin, NumHetMin, MaxOutFrac, thetaTRY</dt>
<dd><p>Remaining
control parameters to be passed to <code>EMfit_betabinom_robust</code>, <code>EMfit_betabinom</code>, and <code>EMfit_betabinom_popcomb</code>. See their respective documentation
for more information. Usually, these parameter's default values are fine.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A dataframe with one row per locus, containing for that locus in its columns:</p>
<dl><dt>Locus</dt>
<dd><p>Locus name.</p></dd>

<dt>Gene</dt>
<dd><p>Gene name.</p></dd>

<dt>dAD_pval</dt>
<dd><p>Raw likelihood ratio test p-value testing for differential allelic dispersion between controls and cases (i.e. a difference between ThetaHetC and ThetaHetT).</p></dd>

<dt>phi_rr</dt>
<dd><p>Estimated reference homozygote mixture component in the beta-binomial mixture dAD fit.</p></dd>

<dt>phi_rv</dt>
<dd><p>Estimated heterozygote mixture component in the beta-binomial mixture dAD fit.</p></dd>

<dt>phi_vv</dt>
<dd><p>Estimated variant homozygote mixture component in the beta-binomial mixture dAD fit.</p></dd>

<dt>Pi</dt>
<dd><p>Estimated heterozygous pi parameter (assumed equal in controls and cases) in the beta-binomial mixture dAD fit.</p></dd>

<dt>ThetaHetC</dt>
<dd><p>Estimated control heterozygous theta parameter in the beta-binomial mixture dAD fit.</p></dd>

<dt>ThetaHetT</dt>
<dd><p>Estimated case heterozygous theta parameter in the beta-binomial mixture dAD fit.</p></dd>

<dt>RhoC</dt>
<dd><p>Estimated control heterozygous rho parameter (0-to-1 rescaling of theta) in the beta-binomial mixture dAD fit.</p></dd>

<dt>RhoT</dt>
<dd><p>Estimated case heterozygous rho parameter (0-to-1 rescaling of theta) in the beta-binomial mixture dAD fit.</p></dd>

<dt>ThetaHom</dt>
<dd><p>Estimated homozygous theta parameter in the beta-binomial mixture dAD fit.</p></dd>

<dt>phi_rr_H0</dt>
<dd><p>Estimated reference homozygote mixture component in the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>phi_rv_H0</dt>
<dd><p>Estimated heterozygote mixture component in the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>phi_vv_H0</dt>
<dd><p>Estimated variant homozygote mixture component in the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>PiH0</dt>
<dd><p>Estimated heterozygous pi parameter (assumed equal in controls and cases) in the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>ThetaHetH0</dt>
<dd><p>Estimated heterozygous theta parameter in the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>RhoH0</dt>
<dd><p>Estimated heterozygous rho parameter (0-to-1 rescaling of theta) in the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>ThetaHomH0</dt>
<dd><p>Estimated homozygous theta parameter in the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>phi_rr_OnlyC</dt>
<dd><p>Estimated reference homozygote mixture component in the beta-binomial mixture fit on control samples alone.</p></dd>

<dt>phi_rv_OnlyC</dt>
<dd><p>Estimated heterozygote mixture component in the beta-binomial mixture fit on control samples alone.</p></dd>

<dt>phi_vv_OnlyC</dt>
<dd><p>Estimated variant homozygote mixture component in the beta-binomial mixture fit on control samples alone.</p></dd>

<dt>PiOnlyC</dt>
<dd><p>Estimated heterozygous pi parameter in the beta-binomial mixture fit on control samples alone.</p></dd>

<dt>ThetaHetOnlyC</dt>
<dd><p>Estimated heterozygous theta parameter in the beta-binomial mixture fit on control samples alone.</p></dd>

<dt>ThetaHomOnlyC</dt>
<dd><p>Estimated homozygous theta parameter in the beta-binomial mixture fit on control samples alone.</p></dd>

<dt>phi_rr_OnlyT</dt>
<dd><p>Estimated reference homozygote mixture component in the beta-binomial mixture fit on case samples alone.</p></dd>

<dt>phi_rv_OnlyT</dt>
<dd><p>Estimated heterozygote mixture component in the beta-binomial mixture fit on case samples alone.</p></dd>

<dt>phi_vv_OnlyT</dt>
<dd><p>Estimated variant homozygote mixture component in the beta-binomial mixture fit on case samples alone.</p></dd>

<dt>PiOnlyT</dt>
<dd><p>Estimated heterozygous pi parameter in the beta-binomial mixture fit on case samples alone.</p></dd>

<dt>ThetaHetOnlyT</dt>
<dd><p>Estimated heterozygous theta parameter in the beta-binomial mixture fit on case samples alone.</p></dd>

<dt>ThetaHomOnlyT</dt>
<dd><p>Estimated homozygous theta parameter in the beta-binomial mixture fit on case samples alone.</p></dd>

<dt>NumHetC</dt>
<dd><p>Estimated number of heterozygotes in controls, according to a fit on controls alone. Outliers, which are detected at the time of this fit, are counted towards this number, but not used in any following fitting procedures.</p></dd>

<dt>NumHetT</dt>
<dd><p>Estimated number of heterozygotes in cases, according to a fit on cases alone. Outliers, which are detected at the time of this fit, are counted towards this number, but not used in any following fitting procedures.</p></dd>

<dt>RobFlagC</dt>
<dd><p>The <code>RobFlag</code> output of the separate outlier-detection fit on controls, see <code>EMfit_betabinom_robust</code>'s documentation.</p></dd>

<dt>RobFlagT</dt>
<dd><p>The <code>RobFlag</code> output of the separate outlier-detection fit on cases, see <code>EMfit_betabinom_robust</code>'s documentation.</p></dd>

<dt>HWEC</dt>
<dd><p>HWE chi squared p-value for controls, see <code>HWE_chisquared</code>. Performed on the outlier-detection fit on controls alone.</p></dd>

<dt>HWET</dt>
<dd><p>HWE chi squared p-value for cases, see <code>HWE_chisquared</code>. Performed on the outlier-detection fit on cases alone.</p></dd>

<dt>CovC</dt>
<dd><p>Mean coverage (reference + variant count) across control data.</p></dd>

<dt>CovT</dt>
<dd><p>Mean coverage (reference + variant count) across case data.</p></dd>

<dt>CovC_Med</dt>
<dd><p>Median coverage (reference + variant count) across control data.</p></dd>

<dt>CovT_Med</dt>
<dd><p>Median coverage (reference + variant count) across case data.</p></dd>

<dt>NumOutC</dt>
<dd><p>Number of outliers detected in control data.</p></dd>

<dt>NumOutT</dt>
<dd><p>Number of outliers detected in case data.</p></dd>

<dt>nrep_H0</dt>
<dd><p>Number of expectation-maximization iterations run during the beta-binomial mixture fit assuming no dAD.</p></dd>

<dt>nrep_H1</dt>
<dd><p>Number of expectation-maximization iterations run during the beta-binomial dAD mixture fit.</p></dd>

<dt>QualityC</dt>
<dd><p>Quality flag returned by <code>EMfit_betabinom_robust</code> (see documentation) of the fit on control data alone.</p></dd>

<dt>QualityT</dt>
<dd><p>Quality flag returned by <code>EMfit_betabinom_robust</code> (see documentation) of the fit on case data alone.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

