<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Models allelic RNAseq counts using an expectation-maximization fit of a binomial mixture distribution — EMfit_binom • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Models allelic RNAseq counts using an expectation-maximization fit of a binomial mixture distribution — EMfit_binom"><meta name="description" content="EMfit_binom estimates, per locus, parameters of an assumed binomial mixture model via expectation maximization (genotype
frequencies, allelic bias i.e. heterozygous p-parameter).
Through this fit, per-samples genotype probabilities are obtained as well. A fit assuming no allelic bias (heterozygous p-parameter = 0.5) is
performed as well, for the purpose of significant allelic bias detection via likelihood ratio test."><meta property="og:description" content="EMfit_binom estimates, per locus, parameters of an assumed binomial mixture model via expectation maximization (genotype
frequencies, allelic bias i.e. heterozygous p-parameter).
Through this fit, per-samples genotype probabilities are obtained as well. A fit assuming no allelic bias (heterozygous p-parameter = 0.5) is
performed as well, for the purpose of significant allelic bias detection via likelihood ratio test."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Models allelic RNAseq counts using an expectation-maximization fit of a binomial mixture distribution</h1>

      <div class="d-none name"><code>EMfit_binom.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>EMfit_binom</code> estimates, per locus, parameters of an assumed binomial mixture model via expectation maximization (genotype
frequencies, allelic bias i.e. heterozygous p-parameter).
Through this fit, per-samples genotype probabilities are obtained as well. A fit assuming no allelic bias (heterozygous p-parameter = 0.5) is
performed as well, for the purpose of significant allelic bias detection via likelihood ratio test.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">EMfit_binom</span><span class="op">(</span></span>
<span>  <span class="va">data_counts</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  allelefreq <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  inbr <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  dltaco <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  HWE <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_InitEst <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-counts">data_counts<a class="anchor" aria-label="anchor" href="#arg-data-counts"></a></dt>
<dd><p>Data frame. Data frame of a SNP with reference and variant counts ("ref_count" and
"var_count", respectively) for each sample ("sample").</p></dd>


<dt id="arg-se">SE<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Number. Sequencing error rate.</p></dd>


<dt id="arg-allelefreq">allelefreq<a class="anchor" aria-label="anchor" href="#arg-allelefreq"></a></dt>
<dd><p>Number. Allele frequency. Only used when pInitEst is TRUE (default = 0.5)</p></dd>


<dt id="arg-inbr">inbr<a class="anchor" aria-label="anchor" href="#arg-inbr"></a></dt>
<dd><p>Number. Degree of inbreeding (default = 0).</p></dd>


<dt id="arg-dltaco">dltaco<a class="anchor" aria-label="anchor" href="#arg-dltaco"></a></dt>
<dd><p>Number. Minimal difference between 2 iterations (default = 0.001).</p></dd>


<dt id="arg-hwe">HWE<a class="anchor" aria-label="anchor" href="#arg-hwe"></a></dt>
<dd><p>Logical. Should HWE be used for allele frequency estimation, not recommended (default = FALSE).</p></dd>


<dt id="arg-p-initest">p_InitEst<a class="anchor" aria-label="anchor" href="#arg-p-initest"></a></dt>
<dd><p>Logical. Calculate initial estimates of pr, pv and prv from allelefreq, not recommended (default = FALSE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>AB</dt>
<dd><p>The estimated allelic shift.</p></dd>

<dt>AB_lrt</dt>
<dd><p>The test statistic of the likelihood ratio test.</p></dd>

<dt>AB_p</dt>
<dd><p>The p-value of the likelihood ratio test.</p></dd>

<dt>GOF</dt>
<dd><p>The goodness-of-fit value based on the corrected likelihood.</p></dd>

<dt>nrep</dt>
<dd><p>The number of iterations.</p></dd>

<dt>quality</dt>
<dd><p>Indicates the quality of a locus. An "!" indicates bad data or a bad fit due to no apparent heterozygosity.</p></dd>

<dt>rho_vv</dt>
<dd><p>The variant (homozygote) genotype probability for the SNP.</p></dd>

<dt>rho_rr</dt>
<dd><p>The reference (homozygote) genotype probability for the SNP.</p></dd>

<dt>rho_rv</dt>
<dd><p>The heterozyous genotype probability for the SNP.</p></dd>

<dt>rho_vv_H0</dt>
<dd><p>The variant (homozygote) genotype probability with Allelic Bias = 0.5.</p></dd>

<dt>rho_rr_H0</dt>
<dd><p>The reference (homozygote) genotype probability with Allelic Bias = 0.5.</p></dd>

<dt>rho_rv_H0</dt>
<dd><p>The heterozyous genotype probability with Allelic Bias = 0.5.</p></dd>

<dt>data_hash</dt>
<dd><p>Data frame. Input data frame with extra columns: allelefreq, genotype and genotype probabilities (prr, prv, pvv) per sample.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

