<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Models and estimates the allelic shift for cis-eQTL data, using beta-binomial models. — EMfit_betabinom_popcomb • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Models and estimates the allelic shift for cis-eQTL data, using beta-binomial models. — EMfit_betabinom_popcomb"><meta name="description" content="EMfit_betabinom_popcomb, similar to EMfit_betabinom, fits a beta-binomial mixture model to input reference- and allele RNAseq counts,
but with those counts originating from two different populations (a control- and case-population). All distribution models are jointly fitted except
for the heterozygous overdispersion parameters, which is allowed to differ between the populations. The main purpose of this fit is getting this model's
likelihood to be used in a likelihood ratio test, checking for statistical evidence that this heterozygous overdispersion parameter
truly differs between the populations, which would indicate differential allelic divergence."><meta property="og:description" content="EMfit_betabinom_popcomb, similar to EMfit_betabinom, fits a beta-binomial mixture model to input reference- and allele RNAseq counts,
but with those counts originating from two different populations (a control- and case-population). All distribution models are jointly fitted except
for the heterozygous overdispersion parameters, which is allowed to differ between the populations. The main purpose of this fit is getting this model's
likelihood to be used in a likelihood ratio test, checking for statistical evidence that this heterozygous overdispersion parameter
truly differs between the populations, which would indicate differential allelic divergence."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Models and estimates the allelic shift for cis-eQTL data, using beta-binomial models.</h1>

      <div class="d-none name"><code>EMfit_betabinom_popcomb.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>EMfit_betabinom_popcomb</code>, similar to <code>EMfit_betabinom</code>, fits a beta-binomial mixture model to input reference- and allele RNAseq counts,
but with those counts originating from two different populations (a control- and case-population). All distribution models are jointly fitted except
for the heterozygous overdispersion parameters, which is allowed to differ between the populations. The main purpose of this fit is getting this model's
likelihood to be used in a likelihood ratio test, checking for statistical evidence that this heterozygous overdispersion parameter
truly differs between the populations, which would indicate differential allelic divergence.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">EMfit_betabinom_popcomb</span><span class="op">(</span></span>
<span>  <span class="va">data_counts</span>,</span>
<span>  allelefreq <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  inbr <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  dltaco <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">6</span>,</span>
<span>  HWE <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_InitEst <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p_inits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  ThetaInits <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  ReEstThetas <span class="op">=</span> <span class="st">"moment"</span>,</span>
<span>  NoSplitHom <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  NoSplitHet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ResetThetaMin <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">10</span>,</span>
<span>  ResetThetaMax <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>,</span>
<span>  thetaTRY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">7</span><span class="op">)</span>,</span>
<span>  probshift_init <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  FirstFewFixed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  MemLim <span class="op">=</span> <span class="fl">2048</span>,</span>
<span>  Xtra <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  epsabs <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  MaxIt <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-counts">data_counts<a class="anchor" aria-label="anchor" href="#arg-data-counts"></a></dt>
<dd><p>Data frame. Data frame of a SNP with reference and variant counts ("ref_count" and
"var_count", respectively) for each sample ("sample"). Also required is the "isCase" column, equaling 1 for case-samples and 0 for control-samples</p></dd>


<dt id="arg-allelefreq">allelefreq<a class="anchor" aria-label="anchor" href="#arg-allelefreq"></a></dt>
<dd><p>Number. Allele frequency. Only used when pInitEst is TRUE (default = 0.5)</p></dd>


<dt id="arg-se">SE<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Number. Sequencing error rate.</p></dd>


<dt id="arg-inbr">inbr<a class="anchor" aria-label="anchor" href="#arg-inbr"></a></dt>
<dd><p>Number. Degree of inbreeding (default = 0).</p></dd>


<dt id="arg-dltaco">dltaco<a class="anchor" aria-label="anchor" href="#arg-dltaco"></a></dt>
<dd><p>Number. Minimal difference between 2 iterations (default = 0.001).</p></dd>


<dt id="arg-hwe">HWE<a class="anchor" aria-label="anchor" href="#arg-hwe"></a></dt>
<dd><p>Logical. Should HWE be used for allele frequency estimation, not recommended (default = FALSE).</p></dd>


<dt id="arg-p-initest">p_InitEst<a class="anchor" aria-label="anchor" href="#arg-p-initest"></a></dt>
<dd><p>Logical. Calculate initial estimates of pr, pv and prv from allelefreq, not recommended (default = FALSE).</p></dd>


<dt id="arg-thetainits">ThetaInits<a class="anchor" aria-label="anchor" href="#arg-thetainits"></a></dt>
<dd><p>Numeric vector, "moment" or "TryThree". A vector of length two containing the initial dispersion estimate for the homozygous peaks, followed by the one for the heterozygous peak,
or "moment", in which case moment initial estimates are generated. In case of "TryThree", three initial values given by the <code>thetaTRY</code> input argument are tried out and the one yielding the best fit is retained.</p></dd>


<dt id="arg-reestthetas">ReEstThetas<a class="anchor" aria-label="anchor" href="#arg-reestthetas"></a></dt>
<dd><p>String. Accepts the methods "moment" or "simple" as arguments, in which case dispersions are re-estimated during every EM-step using a moment estimate or a simple custom procedure respectively, to be used as a starting</p></dd>


<dt id="arg-nosplithom">NoSplitHom<a class="anchor" aria-label="anchor" href="#arg-nosplithom"></a></dt>
<dd><p>Logical. If TRUE, don't allow the beta-binomial fits for homozygotes to be bimodal</p></dd>


<dt id="arg-nosplithet">NoSplitHet<a class="anchor" aria-label="anchor" href="#arg-nosplithet"></a></dt>
<dd><p>Logical. If TRUE, don't allow the beta-binomial fit for heterozygotes to be bimodal</p></dd>


<dt id="arg-resetthetamin">ResetThetaMin<a class="anchor" aria-label="anchor" href="#arg-resetthetamin"></a></dt>
<dd><p>Number. Initial theta values in numeric optimization get capped at this minimum (e.g. in case the moment estimate is even lower)</p></dd>


<dt id="arg-resetthetamax">ResetThetaMax<a class="anchor" aria-label="anchor" href="#arg-resetthetamax"></a></dt>
<dd><p>Number. Initial theta values in numeric optimization get capped at this maximum (e.g. in case the moment estimate is even higher)</p></dd>


<dt id="arg-thetatry">thetaTRY<a class="anchor" aria-label="anchor" href="#arg-thetatry"></a></dt>
<dd><p>Numeric vector. See <code>ReEstThetas</code>, in case this argument equals "TryThree".</p></dd>


<dt id="arg-probshift-init">probshift_init<a class="anchor" aria-label="anchor" href="#arg-probshift-init"></a></dt>
<dd><p>Number. Initial estimate position heterozygous peak (allelic bias; expected fraction of reference reads)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>ParamVec</dt>
<dd><p>A vector containing fitted model parameters: genotype frequencies, heterozygous peak position (allelic bias), and overdispersion of the homozygous peaks and of the separate heterozygous control- and case-peaks.</p></dd>

<dt>GenoDF</dt>
<dd><p>A dataframe containing per-samples genotypes likelihoods according to the fitted model.</p></dd>

<dt>genotypeN</dt>
<dd><p>A character vector containing the most likely per-sample genotypes according to the fitted model.</p></dd>

<dt>nrep</dt>
<dd><p>The number of iterations.</p></dd>

<dt>quality</dt>
<dd><p>An "!" indicates a bad quality locus or fit (no apparent heterozygotes)</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

