<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Final filtering of maelstRom analysis results and writing to output files — final_filter • maelstRom</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Final filtering of maelstRom analysis results and writing to output files — final_filter"><meta property="og:description" content="final_filter only retains significantly imprinted SNPs (after adjusting for multiple testing) and
SNPs of interest (with suitable GOF and degree of (median) imprinting) over all chromosomes. Results and
allelic count files are generated. When both file_all_counts and file_impr_counts are set to FALSE,
this function can be used to simply filter the results_df input."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Final filtering of maelstRom analysis results and writing to output files</h1>
    
    <div class="hidden name"><code>final_filter.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>final_filter</code> only retains significantly imprinted SNPs (after adjusting for multiple testing) and
SNPs of interest (with suitable GOF and degree of (median) imprinting) over all chromosomes. Results and
allelic count files are generated. When both file_all_counts and file_impr_counts are set to FALSE,
this function can be used to simply filter the results_df input.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">final_filter</span><span class="op">(</span></span>
<span>  <span class="va">data_hash</span>,</span>
<span>  <span class="va">results_df</span>,</span>
<span>  <span class="va">results_wd</span>,</span>
<span>  gof_filt <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>  adj_p_filt <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  med_impr_filt <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  i_filt <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  file_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  file_impr <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  file_all_counts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  file_impr_counts <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data_hash</dt>
<dd><p>Hash. Hash of SNP positions with a data frame for every SNP position.</p></dd>


<dt>results_df</dt>
<dd><p>Data frame. Results data frame with columns: "position", "gene", "LRT", "p", "estimated.i",
"allele.frequency", "dbSNP", "reference", "variant", "est_SE", "coverage", "nr_samples", "GOF", "symmetry",
"med_impr", est_inbreeding", "tot_inbreeding".</p></dd>


<dt>results_wd</dt>
<dd><p>String. Directory where results files are written to.</p></dd>


<dt>gof_filt</dt>
<dd><p>Number. Minimal Goodness of Fit, which is the mean(log(sample likelihood under imprinted model * sample coverage + 1)) across samples of a locus. A good (and default) cutoff is 0.8.</p></dd>


<dt>adj_p_filt</dt>
<dd><p>Number. The FDR adjusted singnificance level filter (default is 0.05).</p></dd>


<dt>med_impr_filt</dt>
<dd><p>Number. Minimal median imprinting (default is 0.8).</p></dd>


<dt>i_filt</dt>
<dd><p>Number. Minimal degree of imprinting (default is 0.6).</p></dd>


<dt>file_all</dt>
<dd><p>Logical. Should a file with all SNP information (imprinted and non-imprinted SNPs) be made (default is TRUE).</p></dd>


<dt>file_impr</dt>
<dd><p>Logical. Should a file with imprinted SNP information be made (default is TRUE).</p></dd>


<dt>file_all_counts</dt>
<dd><p>Logical. Should a file with all SNP counts (imprinted and non-imprinted SNPs) be made (default is FALSE).</p></dd>


<dt>file_impr_counts</dt>
<dd><p>Logical. Should a file with imprinted SNP counts be made (default is TRUE).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Data frame with results filtered on adjusted p-value, GOF, median imprinting and degree of imprinting.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

