<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>multi-locus wrapper function for imprinting detection and loss-of-imprinting analysis — impr_LOI_analysis • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="multi-locus wrapper function for imprinting detection and loss-of-imprinting analysis — impr_LOI_analysis"><meta name="description" content="impr_LOI_analysis is a wrapper function of symmetry_gof, imprinting_est, median_imprinting, final_filter, and
LOItest_logreg performing maelstRom's entire (loss of) imprinting pipeline. It goes over the following steps:
Prior filtering on basis of the allelefreq_prel column that should be present in the control dataframe,
  which it should be after running AllelicMeta_est.
Prior filtering via symmetry_gof, see its documentation for more details.
Running imprinting_est to detect imprinted loci and write results to impr_res; this includes a measure of median
  imprinting as returned by the function median_imprinting as well.
Filter impr_res down to significantly and sufficiently imprinted loci using final_filter, then perform loss-of-imprinting
  detection using LOItest_logreg. LOI-detection results are written to LOI_res.

"><meta property="og:description" content="impr_LOI_analysis is a wrapper function of symmetry_gof, imprinting_est, median_imprinting, final_filter, and
LOItest_logreg performing maelstRom's entire (loss of) imprinting pipeline. It goes over the following steps:
Prior filtering on basis of the allelefreq_prel column that should be present in the control dataframe,
  which it should be after running AllelicMeta_est.
Prior filtering via symmetry_gof, see its documentation for more details.
Running imprinting_est to detect imprinted loci and write results to impr_res; this includes a measure of median
  imprinting as returned by the function median_imprinting as well.
Filter impr_res down to significantly and sufficiently imprinted loci using final_filter, then perform loss-of-imprinting
  detection using LOItest_logreg. LOI-detection results are written to LOI_res.

"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>multi-locus wrapper function for imprinting detection and loss-of-imprinting analysis</h1>

      <div class="d-none name"><code>impr_LOI_analysis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>impr_LOI_analysis</code> is a wrapper function of <code>symmetry_gof</code>, <code>imprinting_est</code>, <code>median_imprinting</code>, <code>final_filter</code>, and
<code>LOItest_logreg</code> performing maelstRom's entire (loss of) imprinting pipeline. It goes over the following steps:</p><ol><li><p>Prior filtering on basis of the <code>allelefreq_prel</code> column that should be present in the control dataframe,
  which it should be after running <code>AllelicMeta_est</code>.</p></li>
<li><p>Prior filtering via <code>symmetry_gof</code>, see its documentation for more details.</p></li>
<li><p>Running <code>imprinting_est</code> to detect imprinted loci and write results to <code>impr_res</code>; this includes a measure of median
  imprinting as returned by the function <code>median_imprinting</code> as well.</p></li>
<li><p>Filter <code>impr_res</code> down to significantly and sufficiently imprinted loci using <code>final_filter</code>, then perform loss-of-imprinting
  detection using <code>LOItest_logreg</code>. LOI-detection results are written to <code>LOI_res</code>.</p></li>
</ol></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">impr_LOI_analysis</span><span class="op">(</span></span>
<span>  <span class="va">DataList</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  <span class="va">inbr</span>,</span>
<span>  MinMinorAllelefreq <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  sym_filt <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  adj_p_filt <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  gof_filt <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  med_i_filt <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  i_filt <span class="op">=</span> <span class="fl">0.6</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-datalist">DataList<a class="anchor" aria-label="anchor" href="#arg-datalist"></a></dt>
<dd><p>A list containing two lists of dataframes, which in turn contain control- and case-data (in that order). Subsequent entries in both lists
correspond to subsequent loci (lists must be the same size and named per-locus). Each of the dataframes should at least contain both a <code>ref_count</code> and
<code>var_count</code> column; control dataframes should also contain the columns <code>allelefreq_prel</code> (from running <code>AllelicMeta_est</code>), <code>locus_id</code>,
<code>ref</code>, <code>var</code>, and <code>coverage</code> (from running <code>prior_filter</code>).</p></dd>


<dt id="arg-se">SE<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Number. Sequencing error rate, population metaparameter.</p></dd>


<dt id="arg-inbr">inbr<a class="anchor" aria-label="anchor" href="#arg-inbr"></a></dt>
<dd><p>Number. Inbreeding coefficient, population metaparameter.</p></dd>


<dt id="arg-minminorallelefreq">MinMinorAllelefreq<a class="anchor" aria-label="anchor" href="#arg-minminorallelefreq"></a></dt>
<dd><p>Number. Minimal allowed minor allele frequency as determinde via the estimated reference allele frequency by <code>AllelicMeta_est</code> to consider the locus for imprinting analysis.</p></dd>


<dt id="arg-sym-filt">sym_filt<a class="anchor" aria-label="anchor" href="#arg-sym-filt"></a></dt>
<dd><p>Number. Minimum required symmetry statistic to consider the locus for imprinting analysis, see <code>symmetry_gof</code>.</p></dd>


<dt id="arg-adj-p-filt">adj_p_filt<a class="anchor" aria-label="anchor" href="#arg-adj-p-filt"></a></dt>
<dd><p>Number. Required FDR-corrected significance level of imprinting for loci to be considered for loss-of-imprinting detection, see the function <code>final_filter</code>.</p></dd>


<dt id="arg-gof-filt">gof_filt<a class="anchor" aria-label="anchor" href="#arg-gof-filt"></a></dt>
<dd><p>Number. Required minimum goodness-of-fit heuristic of imprinted loci to be considered for loss-of-imprinting detection, see the function <code>final_filter</code>.</p></dd>


<dt id="arg-med-i-filt">med_i_filt<a class="anchor" aria-label="anchor" href="#arg-med-i-filt"></a></dt>
<dd><p>Number. Required minimum median imprinting of imprinted loci to be considered for loss-of-imprinting detection, see the functions <code>median_imprinting</code> and <code>final_filter</code>.</p></dd>


<dt id="arg-i-filt">i_filt<a class="anchor" aria-label="anchor" href="#arg-i-filt"></a></dt>
<dd><p>Number. Required minimum estimated imprinting of imprinted loci to be considered for loss-of-imprinting detection, see the functions <code>imprinting_est</code> and <code>final_filter</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>impr_res</dt>
<dd><p>Results of the imprinting detection analysis, containing, for every locus:</p></dd>

</dl><ul><li><p>positionThe locus' name, according to <code>names(DataList)</code></p></li>
<li><p>LRTThe test statistic of the likelihood ratio test against no imprinting (see <code>imprinting_est</code>).</p></li>
<li><p>pThe p-value of the likelihood ratio test against no imprinting (see <code>imprinting_est</code>).</p></li>
<li><p>estimated.iThe estimated degree of imprinting (see <code>imprinting_est</code>).</p></li>
<li><p>allele.frequencyReference allele frequency as estimated by <code>AllelicMeta_est</code>, already given as input.</p></li>
<li><p>referenceReference allele nucleotide.</p></li>
<li><p>variantVariant allele nucleotide.</p></li>
<li><p>med_covMedian coverage (reference + variant) across the locus.</p></li>
<li><p>nr_samplesNumber of samples (with a total reference + variant read count of at least 1) covering the locus.</p></li>
<li><p>GOFGoodness-of-fit heuristic as determined by <code>imprinting_est</code>.</p></li>
<li><p>symmetrySymmetry statistic as determined by <code>symmetry_gof</code>.</p></li>
<li><p>med_imprMedian imprinting as determined by <code>median_imprinting.</code></p></li>
</ul><dl><dt>LOI_res</dt>
<dd><p>A dataframe containing, only for loci fitting well to maelstRom's imprinting model (see <code>imprinting_est</code>, determined by <code>gof_filt</code>)
that are significantly (determined by the <code>adj_p_filt input</code>) and suffiently (determined by <code>i_filt</code> and <code>med_i_filt</code>) imprinted,
the results of the differential imprinting analysis (which equals loss of imprinting if there's less imprinting in cases). This contains the same
columns as <code>impr_res</code>, but supplemented with:</p></dd>

   </dl><ul><li><p>DI_pvalThe p-value testing for differential imprinting using logistic regression, see the documentation for <code>LOItest_logreg</code> for more details.</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

