<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="AllelicMeta_est.html">AllelicMeta_est()</a></code>

        </dt>
        <dd>Estimate sequencing error rate and inbreeding using an Expectation-Maximisation algorithm</dd>
      </dl><dl><dt>

          <code><a href="AllelicMeta_est_par.html">AllelicMeta_est_par()</a></code>

        </dt>
        <dd>multi-locus wrapper function of AllelicMeta_est</dd>
      </dl><dl><dt>

          <code><a href="BetaBinomGenotyping.html">BetaBinomGenotyping()</a></code>

        </dt>
        <dd>multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping</dd>
      </dl><dl><dt>

          <code><a href="BetaBinomMix_LLDeriv.html">BetaBinomMix_LLDeriv()</a></code>

        </dt>
        <dd>Beta-binomial mixture model log-likelihood derivatives to its parameters involved in optimization</dd>
      </dl><dl><dt>

          <code><a href="BetaBinomMix_LLDeriv_transfo.html">BetaBinomMix_LLDeriv_transfo()</a></code>

        </dt>
        <dd>Beta-binomial mixture model log-likelihood derivatives to its parameters involved in optimization</dd>
      </dl><dl><dt>

          <code><a href="BetaBinom_test_pvalTS.html">BetaBinom_test_pvalTS()</a></code>

        </dt>
        <dd>Performs an exact beta-binomial p-test</dd>
      </dl><dl><dt>

          <code><a href="BrolDB.html">BrolDB()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="combine_p_gene.html">combine_p_gene()</a></code>

        </dt>
        <dd>Combine p-values of SNPs per gene.</dd>
      </dl><dl><dt>

          <code><a href="CppCnT_Optim.html">CppCnT_Optim()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="CppHetH0_Optim.html">CppHetH0_Optim()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="CppHet_Optim.html">CppHet_Optim()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="CppHom_Optim.html">CppHom_Optim()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="dAD_analysis.html">dAD_analysis()</a></code>

        </dt>
        <dd>Performs a differential Allelic Divergence analysis on a control- and case-dataset</dd>
      </dl><dl><dt>

          <code><a href="dBetaBinom_cpp_old.html">dBetaBinom_cpp_old()</a></code>

        </dt>
        <dd>Exact beta-binomial density using sums.</dd>
      </dl><dl><dt>

          <code><a href="dBetaBinom_MP.html">dBetaBinom_MP()</a></code>

        </dt>
        <dd>Exact beta-binomial density using a multiprecision library.</dd>
      </dl><dl><dt>

          <code><a href="dBetaBin_cppi.html">dBetaBin_cppi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="dBetaBin_cppi_MP.html">dBetaBin_cppi_MP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="dpqrBetaBinom.html">dBetaBinom()</a></code> <code><a href="dpqrBetaBinom.html">pBetaBinom()</a></code> <code><a href="dpqrBetaBinom.html">qBetaBinom()</a></code> <code><a href="dpqrBetaBinom.html">rBetaBinom()</a></code>

        </dt>
        <dd>The Beta-binomial distribution</dd>
      </dl><dl><dt>

          <code><a href="EMfit_betabinom.html">EMfit_betabinom()</a></code>

        </dt>
        <dd>Models allelic RNAseq counts using an expectation-maximization fit of a beta-binomial mixture distribution</dd>
      </dl><dl><dt>

          <code><a href="EMfit_betabinom_popcomb.html">EMfit_betabinom_popcomb()</a></code>

        </dt>
        <dd>Models and estimates the allelic shift for cis-eQTL data, using beta-binomial models.</dd>
      </dl><dl><dt>

          <code><a href="EMfit_betabinom_robust.html">EMfit_betabinom_robust()</a></code>

        </dt>
        <dd>Models allelic RNAseq counts using a robust expectation-maximization fit of a beta-binomial mixture distribution</dd>
      </dl><dl><dt>

          <code><a href="EMfit_betabinom_SEslack.html">EMfit_betabinom_SEslack()</a></code>

        </dt>
        <dd>This function is a work in progress. As such, it is not exported yet; but it allows for some leniency on the sequencing error parameter, hence "SESlack"</dd>
      </dl><dl><dt>

          <code><a href="EMfit_binom.html">EMfit_binom()</a></code>

        </dt>
        <dd>Models allelic RNAseq counts using an expectation-maximization fit of a binomial mixture distribution</dd>
      </dl><dl><dt>

          <code><a href="final_filter.html">final_filter()</a></code>

        </dt>
        <dd>Final filtering of maelstRom analysis results and writing to output files</dd>
      </dl><dl><dt>

          <code><a href="GOF_uniform.html">GOF_uniform()</a></code>

        </dt>
        <dd>Goodness-of-fit test comparing data following a uniform distribution to a Hardy-Weinberg distribution.</dd>
      </dl><dl><dt>

          <code><a href="GradPiPi_cppi.html">GradPiPi_cppi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradPiPi_cppi_MP.html">GradPiPi_cppi_MP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradPiTheta_cppi.html">GradPiTheta_cppi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradPiTheta_cppi_MP.html">GradPiTheta_cppi_MP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradPi_cppi.html">GradPi_cppi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradPi_cppi_MP.html">GradPi_cppi_MP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradThetaTheta_cppi.html">GradThetaTheta_cppi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradThetaTheta_cppi_MP.html">GradThetaTheta_cppi_MP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradTheta_cppi.html">GradTheta_cppi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="GradTheta_cppi_MP.html">GradTheta_cppi_MP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="grad_pi.html">grad_pi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="grad_pi_MP.html">grad_pi_MP()</a></code>

        </dt>
        <dd>Exact gradient of the beta-binomial log-likelihood function for pi using a multiprecision library.</dd>
      </dl><dl><dt>

          <code><a href="grad_pi_old.html">grad_pi_old()</a></code>

        </dt>
        <dd>Exact gradient of the beta-binomial log-likelihood function for pi using sums.</dd>
      </dl><dl><dt>

          <code><a href="grad_pi_pi.html">grad_pi_pi()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="grad_pi_theta.html">grad_pi_theta()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="grad_theta.html">grad_theta()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="grad_theta_MP.html">grad_theta_MP()</a></code>

        </dt>
        <dd>Exact gradient of the beta-binomial log-likelihood function for theta using a multiprecision library.</dd>
      </dl><dl><dt>

          <code><a href="grad_theta_old.html">grad_theta_old()</a></code>

        </dt>
        <dd>Exact gradient of the beta-binomial log-likelihood function for theta using sums.</dd>
      </dl><dl><dt>

          <code><a href="grad_theta_theta.html">grad_theta_theta()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="HWE_chisquared.html">HWE_chisquared()</a></code>

        </dt>
        <dd>Chi-squared test assessing the Hardy-Weinberg assumption.</dd>
      </dl><dl><dt>

          <code><a href="imprinting_est.html">imprinting_est()</a></code>

        </dt>
        <dd>Estimate the degree of imprinting.</dd>
      </dl><dl><dt>

          <code><a href="impr_LOI_analysis.html">impr_LOI_analysis()</a></code>

        </dt>
        <dd>multi-locus wrapper function for imprinting detection and loss-of-imprinting analysis</dd>
      </dl><dl><dt>

          <code><a href="LikelyDistsHet.html">LikelyDistsHet()</a></code>

        </dt>
        <dd>Returns parameter- and likelihood-distances based on leave-on-out (case-deletion) MLE's of <code>EMfit_betabinom_robust</code>'s fitted model</dd>
      </dl><dl><dt>

          <code><a href="LogGaussianQuad_CPP.html">LogGaussianQuad_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="LogNewtonCotes_CPP.html">LogNewtonCotes_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="logSums_MaxMethodSigned_CPP.html">logSums_MaxMethodSigned_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="logSums_MaxMethod_CPP.html">logSums_MaxMethod_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="logSums_MaxMethod_CPP_R.html">logSums_MaxMethod_CPP_R()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="LogTanhSinhQuad_CPP.html">LogTanhSinhQuad_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="LogTrapezoidalInt_CPP.html">LogTrapezoidalInt_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="LOItest_logreg.html">LOItest_logreg()</a></code>

        </dt>
        <dd>Logistic regression on the degree of heterozygosity between cases and controls</dd>
      </dl><dl><dt>

          <code><a href="LOI_calling.html">LOI_calling()</a></code>

        </dt>
        <dd>Call LOI samples over al imprinted loci.</dd>
      </dl><dl><dt>

          <code><a href="maelstRom.html">maelstRom-package</a></code> <code><a href="maelstRom.html">maelstRom</a></code>

        </dt>
        <dd>maelstRom: A package for RNAseq-based allelic analyses</dd>
      </dl><dl><dt>

          <code><a href="maelstRom_ADChromplot.html">maelstRom_ADChromplot()</a></code>

        </dt>
        <dd>Make a plot summarizing allelic divergence (-related phenomena) across a chromosome</dd>
      </dl><dl><dt>

          <code><a href="maelstRom_EMfitplot.html">maelstRom_EMfitplot()</a></code>

        </dt>
        <dd>Plot maelstRom's EM-fit results.</dd>
      </dl><dl><dt>

          <code><a href="maelstRom_imprintplot.html">maelstRom_imprintplot()</a></code>

        </dt>
        <dd>Plot maelstRom's imprinting detection results.</dd>
      </dl><dl><dt>

          <code><a href="median_AB.html">median_AB()</a></code>

        </dt>
        <dd>Calculates the median allelic bias</dd>
      </dl><dl><dt>

          <code><a href="median_imprinting.html">median_imprinting()</a></code>

        </dt>
        <dd>Calculate median imprinting.</dd>
      </dl><dl><dt>

          <code><a href="MomentEst_MixedBetaBinom.html">MomentEst_MixedBetaBinom()</a></code>

        </dt>
        <dd>Calculates moment estimates for the beta-binomial parameters given observed data</dd>
      </dl><dl><dt>

          <code><a href="MomentEst_MixedBetaBinom_SESlack.html">MomentEst_MixedBetaBinom_SESlack()</a></code>

        </dt>
        <dd>This function is a work in progress. As such, it is not exported yet; but it allows for some leniency on the sequencing error parameter, hence "SESlack"</dd>
      </dl><dl><dt>

          <code><a href="MPhelper_GradThetaTheta.html">MPhelper_GradThetaTheta()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="multinomial_coeff.html">multinomial_coeff()</a></code>

        </dt>
        <dd>Calculate multinomial coefficient(s).</dd>
      </dl><dl><dt>

          <code><a href="MyOptTest.html">MyOptTest()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="pmf_betabinomMix.html">pmf_betabinomMix()</a></code>

        </dt>
        <dd>Probability Mass Function of the beta-binomial mixture distribution modeling population-level RNAseq data</dd>
      </dl><dl><dt>

          <code><a href="pmf_betabinomMix_SEslack.html">pmf_betabinomMix_SEslack()</a></code>

        </dt>
        <dd>This function is a work in progress. As such, it is not exported yet; but it allows for some leniency on the sequencing error parameter, hence "SESlack"</dd>
      </dl><dl><dt>

          <code><a href="pmf_binomMix.html">pmf_binomMix()</a></code>

        </dt>
        <dd>Probability Mass Function of the binomial mixture distribution modeling population-level RNAseq data</dd>
      </dl><dl><dt>

          <code><a href="pmf_impr.html">pmf_impr()</a></code>

        </dt>
        <dd>Probability Mass Function for an imprinted SNP.</dd>
      </dl><dl><dt>

          <code><a href="prior_filter.html">prior_filter()</a></code>

        </dt>
        <dd>Prior filtering of loci.</dd>
      </dl><dl><dt>

          <code><a href="qbeta_C.html">qbeta_C()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="qbeta_C3.html">qbeta_C3()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="standard_alleles.html">standard_alleles()</a></code>

        </dt>
        <dd>Determine standard alleles of SNP positions.</dd>
      </dl><dl><dt>

          <code><a href="symmetry_gof.html">symmetry_gof()</a></code>

        </dt>
        <dd>Goodness-of-fit test for symmetry.</dd>
      </dl><dl><dt>

          <code><a href="TumPurHelpFun_CPP.html">TumPurHelpFun_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPurHelpFun_CPP_R.html">TumPurHelpFun_CPP_R()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP.html">TumPur_LogLik_CPP()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP2.html">TumPur_LogLik_CPP2()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP2X.html">TumPur_LogLik_CPP2X()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP2Y.html">TumPur_LogLik_CPP2Y()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP2_10.html">TumPur_LogLik_CPP2_10()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP2_100.html">TumPur_LogLik_CPP2_100()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP2_1000.html">TumPur_LogLik_CPP2_1000()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP3.html">TumPur_LogLik_CPP3()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP3X.html">TumPur_LogLik_CPP3X()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP4.html">TumPur_LogLik_CPP4()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP_DB1.html">TumPur_LogLik_CPP_DB1()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP_DB2.html">TumPur_LogLik_CPP_DB2()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl><dl><dt>

          <code><a href="TumPur_LogLik_CPP_DB3.html">TumPur_LogLik_CPP_DB3()</a></code>

        </dt>
        <dd>DO SOMETHING</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

