<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ MAGE</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MAGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MAGE_expanded_tutorial.html">MAGE expanded tutorial</a>
    </li>
    <li>
      <a href="../articles/MAGE_tutorial.html">MAGE tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="AllelicMeta_est.html">AllelicMeta_est()</a></code> </p>
        </td>
        <td><p>Estimate sequencing error rate and inbreeding using an Expectation-Maximisation algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="AllelicMeta_est_par.html">AllelicMeta_est_par()</a></code> </p>
        </td>
        <td><p>multi-locus wrapper function of AllelicMeta_est</p></td>
      </tr><tr><td>
          <p><code><a href="BetaBinomGenotyping.html">BetaBinomGenotyping()</a></code> </p>
        </td>
        <td><p>multi-locus wrapper function of EMfit_betabinom_robust for the purpose of genotyping</p></td>
      </tr><tr><td>
          <p><code><a href="BetaBinom_test_pvalTS.html">BetaBinom_test_pvalTS()</a></code> </p>
        </td>
        <td><p>Performs an exact beta-binomial p-test</p></td>
      </tr><tr><td>
          <p><code><a href="combine_p_gene.html">combine_p_gene()</a></code> </p>
        </td>
        <td><p>Combine p-values of SNPs per gene.</p></td>
      </tr><tr><td>
          <p><code><a href="dAD_analysis.html">dAD_analysis()</a></code> </p>
        </td>
        <td><p>Performs a differential Allelic Divergence analysis on a control- and case-dataset</p></td>
      </tr><tr><td>
          <p><code><a href="dBetaBinom_cpp_old.html">dBetaBinom_cpp_old()</a></code> </p>
        </td>
        <td><p>Exact beta-binomial density using sums.</p></td>
      </tr><tr><td>
          <p><code><a href="dBetaBinom_MP.html">dBetaBinom_MP()</a></code> </p>
        </td>
        <td><p>Exact beta-binomial density using a multiprecision library.</p></td>
      </tr><tr><td>
          <p><code><a href="dpqrBetaBinom.html">dBetaBinom()</a></code> <code><a href="dpqrBetaBinom.html">pBetaBinom()</a></code> <code><a href="dpqrBetaBinom.html">qBetaBinom()</a></code> <code><a href="dpqrBetaBinom.html">rBetaBinom()</a></code> </p>
        </td>
        <td><p>The Beta-binomial distribution</p></td>
      </tr><tr><td>
          <p><code><a href="EMfit_betabinom.html">EMfit_betabinom()</a></code> </p>
        </td>
        <td><p>Models allelic RNAseq counts using an expectation-maximization fit of a beta-binomial mixture distribution</p></td>
      </tr><tr><td>
          <p><code><a href="EMfit_betabinom_popcomb.html">EMfit_betabinom_popcomb()</a></code> </p>
        </td>
        <td><p>Models and estimates the allelic shift for cis-eQTL data, using beta-binomial models.</p></td>
      </tr><tr><td>
          <p><code><a href="EMfit_betabinom_robust.html">EMfit_betabinom_robust()</a></code> </p>
        </td>
        <td><p>Models allelic RNAseq counts using a robust expectation-maximization fit of a beta-binomial mixture distribution</p></td>
      </tr><tr><td>
          <p><code><a href="EMfit_betabinom_SEslack.html">EMfit_betabinom_SEslack()</a></code> </p>
        </td>
        <td><p>This function is a work in progress. As such, it is not exported yet; but it allows for some leniency on the sequencing error parameter, hence "SESlack"</p></td>
      </tr><tr><td>
          <p><code><a href="EMfit_binom.html">EMfit_binom()</a></code> </p>
        </td>
        <td><p>Models allelic RNAseq counts using an expectation-maximization fit of a binomial mixture distribution</p></td>
      </tr><tr><td>
          <p><code><a href="final_filter.html">final_filter()</a></code> </p>
        </td>
        <td><p>Final filtering of MAGE analysis results and writing to output files</p></td>
      </tr><tr><td>
          <p><code><a href="GOF_uniform.html">GOF_uniform()</a></code> </p>
        </td>
        <td><p>Goodness-of-fit test comparing data following a uniform distribution to a Hardy-Weinberg distribution.</p></td>
      </tr><tr><td>
          <p><code><a href="grad_pi.html">grad_pi()</a></code> </p>
        </td>
        <td><p>gradient of the beta-binomial log-likelihood function with respect to the pi-parameter</p></td>
      </tr><tr><td>
          <p><code><a href="grad_pi_MP.html">grad_pi_MP()</a></code> </p>
        </td>
        <td><p>Exact gradient of the beta-binomial log-likelihood function for pi using a multiprecision library.</p></td>
      </tr><tr><td>
          <p><code><a href="grad_pi_old.html">grad_pi_old()</a></code> </p>
        </td>
        <td><p>Exact gradient of the beta-binomial log-likelihood function for pi using sums.</p></td>
      </tr><tr><td>
          <p><code><a href="grad_theta.html">grad_theta()</a></code> </p>
        </td>
        <td><p>gradient of the beta-binomial log-likelihood function with respect to the theta-parameter</p></td>
      </tr><tr><td>
          <p><code><a href="grad_theta_MP.html">grad_theta_MP()</a></code> </p>
        </td>
        <td><p>Exact gradient of the beta-binomial log-likelihood function for theta using a multiprecision library.</p></td>
      </tr><tr><td>
          <p><code><a href="grad_theta_old.html">grad_theta_old()</a></code> </p>
        </td>
        <td><p>Exact gradient of the beta-binomial log-likelihood function for theta using sums.</p></td>
      </tr><tr><td>
          <p><code><a href="HWE_chisquared.html">HWE_chisquared()</a></code> </p>
        </td>
        <td><p>Chi-squared test assessing the Hardy-Weinberg assumption.</p></td>
      </tr><tr><td>
          <p><code><a href="imprinting_est.html">imprinting_est()</a></code> </p>
        </td>
        <td><p>Estimate the degree of imprinting.</p></td>
      </tr><tr><td>
          <p><code><a href="impr_LOI_analysis.html">impr_LOI_analysis()</a></code> </p>
        </td>
        <td><p>multi-locus wrapper function for imprinting detection and loss-of-imprinting analysis</p></td>
      </tr><tr><td>
          <p><code><a href="LikelyDistsHet.html">LikelyDistsHet()</a></code> </p>
        </td>
        <td><p>Returns parameter- and likelihood-distances based on leave-on-out (case-deletion) MLE's of <code>EMfit_betabinom_robust</code>'s fitted model</p></td>
      </tr><tr><td>
          <p><code><a href="LOItest_logreg.html">LOItest_logreg()</a></code> </p>
        </td>
        <td><p>Logistic regression on the degree of heterozygosity between cases and controls</p></td>
      </tr><tr><td>
          <p><code><a href="LOI_calling.html">LOI_calling()</a></code> </p>
        </td>
        <td><p>Call LOI samples over al imprinted loci.</p></td>
      </tr><tr><td>
          <p><code><a href="MAGE-package.html">MAGE-package</a></code> <code><a href="MAGE-package.html">MAGE</a></code> </p>
        </td>
        <td><p>A short title line describing what the package does</p></td>
      </tr><tr><td>
          <p><code><a href="MAGE.html">MAGE</a></code> </p>
        </td>
        <td><p>MAGE: A package for RNAseq-based allelic analyses</p></td>
      </tr><tr><td>
          <p><code><a href="MAGE_ADChromplot.html">MAGE_ADChromplot()</a></code> </p>
        </td>
        <td><p>Make a plot summarizing allelic divergence (-related phenomena) across a chromosome</p></td>
      </tr><tr><td>
          <p><code><a href="MAGE_EMfitplot.html">MAGE_EMfitplot()</a></code> </p>
        </td>
        <td><p>Plot MAGE's EM-fit results.</p></td>
      </tr><tr><td>
          <p><code><a href="MAGE_imprintplot.html">MAGE_imprintplot()</a></code> </p>
        </td>
        <td><p>Plot MAGE's imprinting detection results.</p></td>
      </tr><tr><td>
          <p><code><a href="median_AB.html">median_AB()</a></code> </p>
        </td>
        <td><p>Calculates the median allelic bias</p></td>
      </tr><tr><td>
          <p><code><a href="median_imprinting.html">median_imprinting()</a></code> </p>
        </td>
        <td><p>Calculate median imprinting.</p></td>
      </tr><tr><td>
          <p><code><a href="MomentEst_MixedBetaBinom.html">MomentEst_MixedBetaBinom()</a></code> </p>
        </td>
        <td><p>Calculates moment estimates for the beta-binomial parameters given observed data</p></td>
      </tr><tr><td>
          <p><code><a href="MomentEst_MixedBetaBinom_SESlack.html">MomentEst_MixedBetaBinom_SESlack()</a></code> </p>
        </td>
        <td><p>This function is a work in progress. As such, it is not exported yet; but it allows for some leniency on the sequencing error parameter, hence "SESlack"</p></td>
      </tr><tr><td>
          <p><code><a href="multinomial_coeff.html">multinomial_coeff()</a></code> </p>
        </td>
        <td><p>Calculate multinomial coefficient(s).</p></td>
      </tr><tr><td>
          <p><code><a href="pmf_betabinomMix.html">pmf_betabinomMix()</a></code> </p>
        </td>
        <td><p>Probability Mass Function of the beta-binomial mixture distribution modeling population-level RNAseq data</p></td>
      </tr><tr><td>
          <p><code><a href="pmf_betabinomMix_SEslack.html">pmf_betabinomMix_SEslack()</a></code> </p>
        </td>
        <td><p>This function is a work in progress. As such, it is not exported yet; but it allows for some leniency on the sequencing error parameter, hence "SESlack"</p></td>
      </tr><tr><td>
          <p><code><a href="pmf_binomMix.html">pmf_binomMix()</a></code> </p>
        </td>
        <td><p>Probability Mass Function of the binomial mixture distribution modeling population-level RNAseq data</p></td>
      </tr><tr><td>
          <p><code><a href="pmf_impr.html">pmf_impr()</a></code> </p>
        </td>
        <td><p>Probability Mass Function for an imprinted SNP.</p></td>
      </tr><tr><td>
          <p><code><a href="prior_filter.html">prior_filter()</a></code> </p>
        </td>
        <td><p>Prior filtering of loci.</p></td>
      </tr><tr><td>
          <p><code><a href="standard_alleles.html">standard_alleles()</a></code> </p>
        </td>
        <td><p>Determine standard alleles of SNP positions.</p></td>
      </tr><tr><td>
          <p><code><a href="symmetry_gof.html">symmetry_gof()</a></code> </p>
        </td>
        <td><p>Goodness-of-fit test for symmetry.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

