<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Returns parameter- and likelihood-distances based on leave-on-out (case-deletion) MLE's of EMfit_betabinom_robust's fitted model — LikelyDistsHet • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Returns parameter- and likelihood-distances based on leave-on-out (case-deletion) MLE's of EMfit_betabinom_robust's fitted model — LikelyDistsHet"><meta name="description" content="LikelyDistsHet is mainly for internal use by EMfit_betabinom_robust. It assists in robustifying the function's EM-fit by iteratively re-fitting
the model on the entire input dataset except for one point, after which the difference in heterozygous pi- and theta-estimates and likelihoods is logged.
The difference with their respective full-data fit counterparts is a measure for the left-out data point's influence on the model fit;
if either one is sufficiently high the data point could be considered an outlier."><meta property="og:description" content="LikelyDistsHet is mainly for internal use by EMfit_betabinom_robust. It assists in robustifying the function's EM-fit by iteratively re-fitting
the model on the entire input dataset except for one point, after which the difference in heterozygous pi- and theta-estimates and likelihoods is logged.
The difference with their respective full-data fit counterparts is a measure for the left-out data point's influence on the model fit;
if either one is sufficiently high the data point could be considered an outlier."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Returns parameter- and likelihood-distances based on leave-on-out (case-deletion) MLE's of <code>EMfit_betabinom_robust</code>'s fitted model</h1>

      <div class="d-none name"><code>LikelyDistsHet.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>LikelyDistsHet</code> is mainly for internal use by <code>EMfit_betabinom_robust</code>. It assists in robustifying the function's EM-fit by iteratively re-fitting
the model on the entire input dataset except for one point, after which the difference in heterozygous pi- and theta-estimates and likelihoods is logged.
The difference with their respective full-data fit counterparts is a measure for the left-out data point's influence on the model fit;
if either one is sufficiently high the data point could be considered an outlier.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">LikelyDistsHet</span><span class="op">(</span></span>
<span>  <span class="va">ref_counts</span>,</span>
<span>  <span class="va">var_counts</span>,</span>
<span>  <span class="va">sprv</span>,</span>
<span>  <span class="va">parvec_cur</span>,</span>
<span>  <span class="va">NoSplitHet</span>,</span>
<span>  <span class="va">ResetThetaMin</span>,</span>
<span>  <span class="va">ResetThetaMax</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  ReEstPars <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ref-counts">ref_counts<a class="anchor" aria-label="anchor" href="#arg-ref-counts"></a></dt>
<dd><p>Numeric vector. reference counts.</p></dd>


<dt id="arg-var-counts">var_counts<a class="anchor" aria-label="anchor" href="#arg-var-counts"></a></dt>
<dd><p>Numeric vector. variant counts.</p></dd>


<dt id="arg-sprv">sprv<a class="anchor" aria-label="anchor" href="#arg-sprv"></a></dt>
<dd><p>Numeric vector. Each sample's EM-weight reflecting its likelihood to be part of the heterozygous population.</p></dd>


<dt id="arg-parvec-cur">parvec_cur<a class="anchor" aria-label="anchor" href="#arg-parvec-cur"></a></dt>
<dd><p>Numeric vector. Pi and theta (in that order) of the heterozyous peak of the full-data fit.</p></dd>


<dt id="arg-nosplithet">NoSplitHet<a class="anchor" aria-label="anchor" href="#arg-nosplithet"></a></dt>
<dd><p>Logical. If TRUE, don't allow the beta-binomial fit for heterozygotes to be bimodal</p></dd>


<dt id="arg-resetthetamin">ResetThetaMin<a class="anchor" aria-label="anchor" href="#arg-resetthetamin"></a></dt>
<dd><p>Number. Initial theta values in numeric optimization get capped at this minimum (e.g. in case the moment estimate is even lower)</p></dd>


<dt id="arg-resetthetamax">ResetThetaMax<a class="anchor" aria-label="anchor" href="#arg-resetthetamax"></a></dt>
<dd><p>Number. Initial theta values in numeric optimization get capped at this maximum (e.g. in case the moment estimate is even higher)</p></dd>


<dt id="arg-se">SE<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Number. Sequencing error rate.</p></dd>


<dt id="arg-reestpars">ReEstPars<a class="anchor" aria-label="anchor" href="#arg-reestpars"></a></dt>
<dd><p>Logical. If TRUE, re-estimates <code>parvec_cur</code> given <code>ref_counts</code> and <code>var_counts</code>. This is useless if these are the actual
counts of the full dataset, but are useful for an emperical approach in which "expected" parameter- and likelihood-distances if the assumed model is 100<!-- % -->
correct are simulated by drawing <code>ref_counts</code> and <code>var_counts</code> from this assumed model (see <code>EMfit_betabinom_robust</code>)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>LikDists</dt>
<dd><p>A vector containing likelihood distances per sample (2 times full-data log-likelihood minus re-fitted log-likelihood leaving out the sample).</p></dd>

<dt>PiDists</dt>
<dd><p>A vector containing pi distances per sample (leave-sample-out refitted pi minus full-data pi).</p></dd>

<dt>ThetaDists</dt>
<dd><p>A vector containing theta distances per sample (leave-sample-out refitted theta minus full-data theta).</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

