<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot maelstRom's imprinting detection results. — maelstRom_imprintplot • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot maelstRom's imprinting detection results. — maelstRom_imprintplot"><meta name="description" content="maelstRom_imprintplot plots the results of maelstRom's imprinting detection analysis. More specifically, it plots a histogram of the observed reference allele fraction,
als well as the PMF of the fitted imprinted model, which is a binomial mixture model using a heterozygous peak with no variant bias (p-parameter = 0.5) but that is
split up according to a fitted degree of imprinting (see imprinting_est for this fitting procedure and pmf_impr for the imprinted pmf). Optionally,
the non-imprinted fit (so just a binomial mixture model without variant bias) is plotted as well, if plot_NoImpr is TRUE."><meta property="og:description" content="maelstRom_imprintplot plots the results of maelstRom's imprinting detection analysis. More specifically, it plots a histogram of the observed reference allele fraction,
als well as the PMF of the fitted imprinted model, which is a binomial mixture model using a heterozygous peak with no variant bias (p-parameter = 0.5) but that is
split up according to a fitted degree of imprinting (see imprinting_est for this fitting procedure and pmf_impr for the imprinted pmf). Optionally,
the non-imprinted fit (so just a binomial mixture model without variant bias) is plotted as well, if plot_NoImpr is TRUE."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot maelstRom's imprinting detection results.</h1>

      <div class="d-none name"><code>maelstRom_imprintplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>maelstRom_imprintplot</code> plots the results of maelstRom's imprinting detection analysis. More specifically, it plots a histogram of the observed reference allele fraction,
als well as the PMF of the fitted imprinted model, which is a binomial mixture model using a heterozygous peak with no variant bias (p-parameter = 0.5) but that is
split up according to a fitted degree of imprinting (see <code>imprinting_est</code> for this fitting procedure and <code>pmf_impr</code> for the imprinted pmf). Optionally,
the non-imprinted fit (so just a binomial mixture model without variant bias) is plotted as well, if <code>plot_NoImpr</code> is TRUE.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">maelstRom_imprintplot</span><span class="op">(</span></span>
<span>  <span class="va">ref_counts</span>,</span>
<span>  <span class="va">var_counts</span>,</span>
<span>  <span class="va">allelefreq</span>,</span>
<span>  <span class="va">impr</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  inbr <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  PlotCov <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  plot_NoImpr <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  SplitPeaks <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ImprCols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  NoImprCol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  wd_res <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">""</span>,</span>
<span>  position <span class="op">=</span> <span class="st">""</span>,</span>
<span>  gene <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ref-counts">ref_counts<a class="anchor" aria-label="anchor" href="#arg-ref-counts"></a></dt>
<dd><p>Numeric vector. Reference counts of the locus.</p></dd>


<dt id="arg-var-counts">var_counts<a class="anchor" aria-label="anchor" href="#arg-var-counts"></a></dt>
<dd><p>Numeric vector. Variant counts of the locus.</p></dd>


<dt id="arg-allelefreq">allelefreq<a class="anchor" aria-label="anchor" href="#arg-allelefreq"></a></dt>
<dd><p>Number. Allele frequency.</p></dd>


<dt id="arg-impr">impr<a class="anchor" aria-label="anchor" href="#arg-impr"></a></dt>
<dd><p>Number. Degree of imprinting.</p></dd>


<dt id="arg-se">SE<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Number. Sequencing error rate.</p></dd>


<dt id="arg-inbr">inbr<a class="anchor" aria-label="anchor" href="#arg-inbr"></a></dt>
<dd><p>Number. Degree of inbreeding (default = 0).</p></dd>


<dt id="arg-plotcov">PlotCov<a class="anchor" aria-label="anchor" href="#arg-plotcov"></a></dt>
<dd><p>Number. Even though the observed number of reads can vary per sample, we have to assume a single number of observed reads to plot imprinting PMFs.
One can opt to pick a set value for this (the default is 50) or e.g. use the locus' median or mean coverage here, so as to best correspond to the observed data.</p></dd>


<dt id="arg-splitpeaks">SplitPeaks<a class="anchor" aria-label="anchor" href="#arg-splitpeaks"></a></dt>
<dd><p>Logical.  If TRUE, the PMFs of both homozygotes and the heterozygotes are plotted seperately and using different colors. If FALSE, one total imprinted PMF is plotted.</p></dd>


<dt id="arg-noimprcol">NoImprCol<a class="anchor" aria-label="anchor" href="#arg-noimprcol"></a></dt>
<dd><p>String. Color for the non-imprinted PMF (also no allelic bias; heterozygous binomial p-parameter = 0.5). There's no option to split the peaks of this PMF as it would make
the final figure way too clustered, so only one color value is required. A default color is used in case non is given as input.</p></dd>


<dt id="arg-wd-res">wd_res<a class="anchor" aria-label="anchor" href="#arg-wd-res"></a></dt>
<dd><p>String. Working directory where plots are saved; if non is given, the plot itself (as a <code>ggplot2</code> object) is returned instead.</p></dd>


<dt id="arg-chr">chr<a class="anchor" aria-label="anchor" href="#arg-chr"></a></dt>
<dd><p>Number. Chromosome of the locus, to be used in naming the outpud file if a <code>wd_res</code> is given.</p></dd>


<dt id="arg-position">position<a class="anchor" aria-label="anchor" href="#arg-position"></a></dt>
<dd><p>Number. Position of the locus, to be used in naming the outpud file if a <code>wd_res</code> is given.</p></dd>


<dt id="arg-gene">gene<a class="anchor" aria-label="anchor" href="#arg-gene"></a></dt>
<dd><p>String. Gene the locus is part of, to be used in naming the outpud file if a <code>wd_res</code> is given; optional.</p></dd>


<dt id="arg-plot-noimpr">Plot_NoImpr<a class="anchor" aria-label="anchor" href="#arg-plot-noimpr"></a></dt>
<dd><p>Logical. If TRUE, also plot the non-imprintend PMF (a binomial mixture model without variant bias i.e. binomial p-parameter of heterozygotes = 0.5)</p></dd>


<dt id="arg-imrcols">ImrCols<a class="anchor" aria-label="anchor" href="#arg-imrcols"></a></dt>
<dd><p>String (vector). Colors for the plotted PMFs of the imprinted fit. Three values are required if <code>SplitPeaks</code> is TRUE, otherwise just one value is required.
Default colors are used in case none are given as input.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

