<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Make a plot summarizing allelic divergence (-related phenomena) across a chromosome — maelstRom_ADChromplot • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Make a plot summarizing allelic divergence (-related phenomena) across a chromosome — maelstRom_ADChromplot"><meta name="description" content="maelstRom_ADChromplot returns various ggplot-objects visualizing maelstRom's differential allelic divergence results across a chromosome,
together with differential expression logFC which can be determined using the same RNAseq data, as well as (optionally) methylation- and copy-number-alteration-data if available.
All of this data should be supplied as separate dataframes, which enables plotting even when some of this data is missing for some of the loci or genes."><meta property="og:description" content="maelstRom_ADChromplot returns various ggplot-objects visualizing maelstRom's differential allelic divergence results across a chromosome,
together with differential expression logFC which can be determined using the same RNAseq data, as well as (optionally) methylation- and copy-number-alteration-data if available.
All of this data should be supplied as separate dataframes, which enables plotting even when some of this data is missing for some of the loci or genes."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Make a plot summarizing allelic divergence (-related phenomena) across a chromosome</h1>

      <div class="d-none name"><code>maelstRom_ADChromplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>maelstRom_ADChromplot</code> returns various ggplot-objects visualizing maelstRom's differential allelic divergence results across a chromosome,
together with differential expression logFC which can be determined using the same RNAseq data, as well as (optionally) methylation- and copy-number-alteration-data if available.
All of this data should be supplied as separate dataframes, which enables plotting even when some of this data is missing for some of the loci or genes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">maelstRom_ADChromplot</span><span class="op">(</span></span>
<span>  <span class="va">AD_Data</span>,</span>
<span>  <span class="va">DE_Data</span>,</span>
<span>  Meth_Data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  CNAgain_Data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  CNAloss_Data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pvalSIG <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  roll_median <span class="op">=</span> <span class="fl">15</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ad-data">AD_Data<a class="anchor" aria-label="anchor" href="#arg-ad-data"></a></dt>
<dd><p>Dataframe. Should at least contain chromosomal position ("pos"), overdispersion in controls and cases expressed as rho which lies between
0 and 1 ("rho_control", "rho_case") and the p-value testing for differential allelic divergence (difference in overdispersion parameter between controls and cases; "pval")</p></dd>


<dt id="arg-de-data">DE_Data<a class="anchor" aria-label="anchor" href="#arg-de-data"></a></dt>
<dd><p>Dataframe. Should at least contain chromosomal position ("pos") and the log(2)FC of expression between cases and controls ("logFC")</p></dd>


<dt id="arg-meth-data">Meth_Data<a class="anchor" aria-label="anchor" href="#arg-meth-data"></a></dt>
<dd><p>Dataframe. Optional; should at least contain chromosomal position ("pos") and the p-value testing for case hypermethylation ("logFC")</p></dd>


<dt id="arg-cnagain-data">CNAgain_Data<a class="anchor" aria-label="anchor" href="#arg-cnagain-data"></a></dt>
<dd><p>Dataframe. Optional; should at least contain chromosomal position ("pos") and the average gain in cases ("CNAgain")</p></dd>


<dt id="arg-cnaloss-data">CNAloss_Data<a class="anchor" aria-label="anchor" href="#arg-cnaloss-data"></a></dt>
<dd><p>Dataframe. Optional; should at least contain chromosomal position ("pos") and the average loss in cases ("CNAloss")</p></dd>


<dt id="arg-pvalsig">pvalSIG<a class="anchor" aria-label="anchor" href="#arg-pvalsig"></a></dt>
<dd><p>Number. Significance cutoff of provided p-values (these should be FDR-corrected in advance)</p></dd>


<dt id="arg-roll-median">roll_median<a class="anchor" aria-label="anchor" href="#arg-roll-median"></a></dt>
<dd><p>integer. All measures are plotted as rolling medians across <code>roll_median</code> data points.</p></dd>


<dt id="arg-allelefreq">allelefreq<a class="anchor" aria-label="anchor" href="#arg-allelefreq"></a></dt>
<dd><p>Number. Allele frequency.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<dl><dt>ADDE_plot</dt>
<dd><p>A plot vizualizing Allelic Divergence and Differential Expression results.</p></dd>

<dt>MethCNA_plot</dt>
<dd><p>A plot vizualizing hypermethylation and copy-number-alteration results.</p></dd>

<dt>LEG1, LEG2, LEG3</dt>
<dd><p>Various legend items, to be combined with <code>ADDE_plot</code> and <code>MethCNA_plot</code> if the user so desires, e.g. using the <code>patchwork</code> package.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

