<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Make a plot summarizing allelic divergence (-related phenomena) across a chromosome — maelstRom_ADChromplot • maelstRom</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Make a plot summarizing allelic divergence (-related phenomena) across a chromosome — maelstRom_ADChromplot"><meta property="og:description" content="maelstRom_ADChromplot returns various ggplot-objects visualizing maelstRom's differential allelic divergence results across a chromosome,
together with differential expression logFC which can be determined using the same RNAseq data, as well as (optionally) methylation- and copy-number-alteration-data if available.
All of this data should be supplied as separate dataframes, which enables plotting even when some of this data is missing for some of the loci or genes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make a plot summarizing allelic divergence (-related phenomena) across a chromosome</h1>
    
    <div class="hidden name"><code>maelstRom_ADChromplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>maelstRom_ADChromplot</code> returns various ggplot-objects visualizing maelstRom's differential allelic divergence results across a chromosome,
together with differential expression logFC which can be determined using the same RNAseq data, as well as (optionally) methylation- and copy-number-alteration-data if available.
All of this data should be supplied as separate dataframes, which enables plotting even when some of this data is missing for some of the loci or genes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">maelstRom_ADChromplot</span><span class="op">(</span></span>
<span>  <span class="va">AD_Data</span>,</span>
<span>  <span class="va">DE_Data</span>,</span>
<span>  Meth_Data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  CNAgain_Data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  CNAloss_Data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pvalSIG <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  roll_median <span class="op">=</span> <span class="fl">15</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>AD_Data</dt>
<dd><p>Dataframe. Should at least contain chromosomal position ("pos"), overdispersion in controls and cases expressed as rho which lies between
0 and 1 ("rho_control", "rho_case") and the p-value testing for differential allelic divergence (difference in overdispersion parameter between controls and cases; "pval")</p></dd>


<dt>DE_Data</dt>
<dd><p>Dataframe. Should at least contain chromosomal position ("pos") and the log(2)FC of expression between cases and controls ("logFC")</p></dd>


<dt>Meth_Data</dt>
<dd><p>Dataframe. Optional; should at least contain chromosomal position ("pos") and the p-value testing for case hypermethylation ("logFC")</p></dd>


<dt>CNAgain_Data</dt>
<dd><p>Dataframe. Optional; should at least contain chromosomal position ("pos") and the average gain in cases ("CNAgain")</p></dd>


<dt>CNAloss_Data</dt>
<dd><p>Dataframe. Optional; should at least contain chromosomal position ("pos") and the average loss in cases ("CNAloss")</p></dd>


<dt>pvalSIG</dt>
<dd><p>Number. Significance cutoff of provided p-values (these should be FDR-corrected in advance)</p></dd>


<dt>roll_median</dt>
<dd><p>integer. All measures are plotted as rolling medians across <code>roll_median</code> data points.</p></dd>


<dt>allelefreq</dt>
<dd><p>Number. Allele frequency.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing the following components:</p>
<dl><dt>ADDE_plot</dt>
<dd><p>A plot vizualizing Allelic Divergence and Differential Expression results.</p></dd>

<dt>MethCNA_plot</dt>
<dd><p>A plot vizualizing hypermethylation and copy-number-alteration results.</p></dd>

<dt>LEG1, LEG2, LEG3</dt>
<dd><p>Various legend items, to be combined with <code>ADDE_plot</code> and <code>MethCNA_plot</code> if the user so desires, e.g. using the <code>patchwork</code> package.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

