<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot maelstRom's imprinting detection results. — maelstRom_imprintplot • maelstRom</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot maelstRom's imprinting detection results. — maelstRom_imprintplot"><meta property="og:description" content="maelstRom_imprintplot plots the results of maelstRom's imprinting detection analysis. More specifically, it plots a histogram of the observed reference allele fraction,
als well as the PMF of the fitted imprinted model, which is a binomial mixture model using a heterozygous peak with no variant bias (p-parameter = 0.5) but that is
split up according to a fitted degree of imprinting (see imprinting_est for this fitting procedure and pmf_impr for the imprinted pmf). Optionally,
the non-imprinted fit (so just a binomial mixture model without variant bias) is plotted as well, if plot_NoImpr is TRUE."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maelstRom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/maelstRom_expanded_tutorial.html">maelstRom expanded tutorial</a>
    </li>
    <li>
      <a href="../articles/maelstRom_tutorial.html">maelstRom tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot maelstRom's imprinting detection results.</h1>
    
    <div class="hidden name"><code>maelstRom_imprintplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>maelstRom_imprintplot</code> plots the results of maelstRom's imprinting detection analysis. More specifically, it plots a histogram of the observed reference allele fraction,
als well as the PMF of the fitted imprinted model, which is a binomial mixture model using a heterozygous peak with no variant bias (p-parameter = 0.5) but that is
split up according to a fitted degree of imprinting (see <code>imprinting_est</code> for this fitting procedure and <code>pmf_impr</code> for the imprinted pmf). Optionally,
the non-imprinted fit (so just a binomial mixture model without variant bias) is plotted as well, if <code>plot_NoImpr</code> is TRUE.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">maelstRom_imprintplot</span><span class="op">(</span></span>
<span>  <span class="va">ref_counts</span>,</span>
<span>  <span class="va">var_counts</span>,</span>
<span>  <span class="va">allelefreq</span>,</span>
<span>  <span class="va">impr</span>,</span>
<span>  <span class="va">SE</span>,</span>
<span>  inbr <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  PlotCov <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  plot_NoImpr <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  SplitPeaks <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ImprCols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  NoImprCol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  wd_res <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">""</span>,</span>
<span>  position <span class="op">=</span> <span class="st">""</span>,</span>
<span>  gene <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>ref_counts</dt>
<dd><p>Numeric vector. Reference counts of the locus.</p></dd>


<dt>var_counts</dt>
<dd><p>Numeric vector. Variant counts of the locus.</p></dd>


<dt>allelefreq</dt>
<dd><p>Number. Allele frequency.</p></dd>


<dt>impr</dt>
<dd><p>Number. Degree of imprinting.</p></dd>


<dt>SE</dt>
<dd><p>Number. Sequencing error rate.</p></dd>


<dt>inbr</dt>
<dd><p>Number. Degree of inbreeding (default = 0).</p></dd>


<dt>PlotCov</dt>
<dd><p>Number. Even though the observed number of reads can vary per sample, we have to assume a single number of observed reads to plot imprinting PMFs.
One can opt to pick a set value for this (the default is 50) or e.g. use the locus' median or mean coverage here, so as to best correspond to the observed data.</p></dd>


<dt>SplitPeaks</dt>
<dd><p>Logical.  If TRUE, the PMFs of both homozygotes and the heterozygotes are plotted seperately and using different colors. If FALSE, one total imprinted PMF is plotted.</p></dd>


<dt>NoImprCol</dt>
<dd><p>String. Color for the non-imprinted PMF (also no allelic bias; heterozygous binomial p-parameter = 0.5). There's no option to split the peaks of this PMF as it would make
the final figure way too clustered, so only one color value is required. A default color is used in case non is given as input.</p></dd>


<dt>wd_res</dt>
<dd><p>String. Working directory where plots are saved; if non is given, the plot itself (as a <code>ggplot2</code> object) is returned instead.</p></dd>


<dt>chr</dt>
<dd><p>Number. Chromosome of the locus, to be used in naming the outpud file if a <code>wd_res</code> is given.</p></dd>


<dt>position</dt>
<dd><p>Number. Position of the locus, to be used in naming the outpud file if a <code>wd_res</code> is given.</p></dd>


<dt>gene</dt>
<dd><p>String. Gene the locus is part of, to be used in naming the outpud file if a <code>wd_res</code> is given; optional.</p></dd>


<dt>Plot_NoImpr</dt>
<dd><p>Logical. If TRUE, also plot the non-imprintend PMF (a binomial mixture model without variant bias i.e. binomial p-parameter of heterozygotes = 0.5)</p></dd>


<dt>ImrCols</dt>
<dd><p>String (vector). Colors for the plotted PMFs of the imprinted fit. Three values are required if <code>SplitPeaks</code> is TRUE, otherwise just one value is required.
Default colors are used in case none are given as input.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

