<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Probability Mass Function for an imprinted SNP. — pmf_impr • maelstRom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Probability Mass Function for an imprinted SNP. — pmf_impr"><meta name="description" content="pmf_impr calculates the probability of observing population-level RNAseq data (reference- and variant allele counts per sample)
assuming an imprinted binomial mixture model with the mean reference allele fraction in heterozygotes fixed at 0.5 (no allelic bias).
More specifically, the formula used is (though under-the-hood not using R's built-in dbinom function but our own mathematical implementation):
pr * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = 1-SE) +
0.5 * prv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = (0.5-(impr/2)) / (1-(impr/2)) * (1-SE) + (0.5/1-(impr/2)) * SE) +
0.5 * prv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = (0.5-(impr/2)) / (1-(impr/2)) * SE + (0.5/1-(impr/2)) * (1-SE)) +
pv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = SE)
pr, pv and prv are calculated assuming Hardy-Weinberg-Equilibrium with given input parameters (inbreeding coefficient inbr and reference allele frequency allelefreq)"><meta property="og:description" content="pmf_impr calculates the probability of observing population-level RNAseq data (reference- and variant allele counts per sample)
assuming an imprinted binomial mixture model with the mean reference allele fraction in heterozygotes fixed at 0.5 (no allelic bias).
More specifically, the formula used is (though under-the-hood not using R's built-in dbinom function but our own mathematical implementation):
pr * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = 1-SE) +
0.5 * prv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = (0.5-(impr/2)) / (1-(impr/2)) * (1-SE) + (0.5/1-(impr/2)) * SE) +
0.5 * prv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = (0.5-(impr/2)) / (1-(impr/2)) * SE + (0.5/1-(impr/2)) * (1-SE)) +
pv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = SE)
pr, pv and prv are calculated assuming Hardy-Weinberg-Equilibrium with given input parameters (inbreeding coefficient inbr and reference allele frequency allelefreq)"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">maelstRom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/maelstRom_Allelic_Dispersion_tutorial.html">maelstRom Allelic Dispersion tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_imprinting_tutorial.html">maelstRom imprinting tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/maelstRom_results_exploration.html">maelstRom results exploration</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Probability Mass Function for an imprinted SNP.</h1>

      <div class="d-none name"><code>pmf_impr.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>pmf_impr</code> calculates the probability of observing population-level RNAseq data (reference- and variant allele counts per sample)
assuming an imprinted binomial mixture model with the mean reference allele fraction in heterozygotes fixed at 0.5 (no allelic bias).
More specifically, the formula used is (though under-the-hood not using R's built-in dbinom function but our own mathematical implementation):</p>
<p>pr * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = 1-SE) +</p>
<p>0.5 * prv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = (0.5-(impr/2)) / (1-(impr/2)) * (1-SE) + (0.5/1-(impr/2)) * SE) +</p>
<p>0.5 * prv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = (0.5-(impr/2)) / (1-(impr/2)) * SE + (0.5/1-(impr/2)) * (1-SE)) +</p>
<p>pv * dbinom(x = ref_counts, size = ref_counts + var_counts, prob = SE)</p>
<p>pr, pv and prv are calculated assuming Hardy-Weinberg-Equilibrium with given input parameters (inbreeding coefficient <code>inbr</code> and reference allele frequency <code>allelefreq</code>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pmf_impr</span><span class="op">(</span><span class="va">ref_counts</span>, <span class="va">var_counts</span>, <span class="va">allelefreq</span>, <span class="va">impr</span>, <span class="va">SE</span>, inbr <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ref-counts">ref_counts<a class="anchor" aria-label="anchor" href="#arg-ref-counts"></a></dt>
<dd><p>Number or Numeric vector Reference count(s).</p></dd>


<dt id="arg-var-counts">var_counts<a class="anchor" aria-label="anchor" href="#arg-var-counts"></a></dt>
<dd><p>Number or Numeric vector. Variant count(s).</p></dd>


<dt id="arg-allelefreq">allelefreq<a class="anchor" aria-label="anchor" href="#arg-allelefreq"></a></dt>
<dd><p>Number. Allele frequency.</p></dd>


<dt id="arg-impr">impr<a class="anchor" aria-label="anchor" href="#arg-impr"></a></dt>
<dd><p>Number. Degree of imprinting.</p></dd>


<dt id="arg-se">SE<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>Number. Sequencing error rate.</p></dd>


<dt id="arg-inbr">inbr<a class="anchor" aria-label="anchor" href="#arg-inbr"></a></dt>
<dd><p>Number. Degree of inbreeding (default = 0).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Probability of observing <code>ref_counts</code> and <code>var_counts</code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">pmf_impr</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span>, <span class="fl">0.5</span>, <span class="fl">0</span>, <span class="fl">0.002</span>, <span class="fl">0.12</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0775267</span>
<span class="r-in"><span><span class="fu">pmf_impr</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">0.002</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1.854402e-22</span>
<span class="r-in"><span><span class="fu">pmf_impr</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">0.002</span>, <span class="fl">0.12</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.4900895</span>
<span class="r-in"><span><span class="fu">pmf_impr</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">8</span>, <span class="fl">10</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">23</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span>, <span class="fl">6</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">10</span>, <span class="fl">0</span><span class="op">)</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">0.002</span>, <span class="fl">0.12</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 2.026483e-11 3.242372e-18 2.511768e-13 1.391617e-07 9.999602e-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6] 2.511768e-13 4.774990e-01</span>
<span class="r-in"><span><span class="fu">pmf_impr</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">1</span>, <span class="fl">23</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span>, <span class="fl">6</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">10</span>, <span class="fl">0</span><span class="op">)</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">0.002</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 5.940240e-03 4.920558e-01 4.940299e-01 1.992008e-05 4.950200e-01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6] 1.078197e-02 4.774990e-01</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Cedric Stroobandt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

